

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Advanced: Decorating the example workflow &mdash; Snakemake 5.6.0+8.g89bc383.dirty documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sphinx-argparse.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Additional features" href="additional_features.html" />
    <link rel="prev" title="Basics: An example workflow" href="basics.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Snakemake
          

          
          </a>

          
            
            
              <div class="version">
                5.6.0+8.g89bc383.dirty
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started/installation.html">安装</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="tutorial.html">Snakemake Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="basics.html">Basics: An example workflow</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Advanced: Decorating the example workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#step-1-specifying-the-number-of-used-threads">Step 1: Specifying the number of used threads</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#step-2-config-files">Step 2: Config files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-3-input-functions">Step 3: Input functions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#step-4-rule-parameters">Step 4: Rule parameters</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id2">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#step-5-logging">Step 5: Logging</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id3">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#step-6-temporary-and-protected-files">Step 6: Temporary and protected files</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id4">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#summary">Summary</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="additional_features.html">Additional features</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="short.html">Short tutorial</a></li>
</ul>
<p class="caption"><span class="caption-text">Executing workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../executable.html">Executing Snakemake</a></li>
</ul>
<p class="caption"><span class="caption-text">Defining workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/writing_snakefiles.html">Writing Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/rules.html">Rules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/modularization.html">Modularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/remote_files.html">Remote files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/utils.html">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/deployment.html">Distribution and Reproducibility</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/reporting.html">Reports</a></li>
</ul>
<p class="caption"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/snakemake.html">The Snakemake API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/snakemake_utils.html">Additional utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/internal/modules.html">Internal API</a></li>
</ul>
<p class="caption"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../project_info/citations.html">Citing and Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/more_resources.html">More Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/history.html">Change Log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/license.html">License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Snakemake</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="tutorial.html">Snakemake Tutorial</a> &raquo;</li>
        
      <li>Advanced: Decorating the example workflow</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/tutorial/advanced.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="advanced-decorating-the-example-workflow">
<h1>Advanced: Decorating the example workflow<a class="headerlink" href="#advanced-decorating-the-example-workflow" title="Permalink to this headline">¶</a></h1>
<p>Now that the basic concepts of Snakemake have been illustrated, we can introduce advanced topics.</p>
<div class="section" id="step-1-specifying-the-number-of-used-threads">
<h2>Step 1: Specifying the number of used threads<a class="headerlink" href="#step-1-specifying-the-number-of-used-threads" title="Permalink to this headline">¶</a></h2>
<p>For some tools, it is advisable to use more than one thread in order to speed up the computation.
<strong>Snakemake can be made aware of the threads a rule needs</strong> with the <code class="docutils literal notranslate"><span class="pre">threads</span></code> directive.
In our example workflow, it makes sense to use multiple threads for the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;data/samples/{sample}.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">8</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem -t {threads} {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
<p>The number of threads can be propagated to the shell command with the familiar braces notation (i.e. <code class="docutils literal notranslate"><span class="pre">{threads}</span></code>).
If no <code class="docutils literal notranslate"><span class="pre">threads</span></code> directive is given, a rule is assumed to need 1 thread.</p>
<p>When a workflow is executed, <strong>the number of threads the jobs need is considered by the Snakemake scheduler</strong>.
In particular, the scheduler ensures that the sum of the threads of all running jobs does not exceed a given number of available CPU cores.
This number can be given with the <code class="docutils literal notranslate"><span class="pre">--cores</span></code> command line argument (per default, Snakemake uses only 1 CPU core).
For example</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake --cores <span class="m">10</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">Apart from the very common thread resource, Snakemake provides a <code class="docutils literal notranslate"><span class="pre">resources</span></code> directive that can be used to <strong>specify arbitrary resources</strong>, e.g., memory usage or auxiliary computing devices like GPUs.
Similar to threads, these can be considered by the scheduler when an available amount of that resource is given with the command line argument <code class="docutils literal notranslate"><span class="pre">--resources</span></code> (see <a class="reference internal" href="../snakefiles/rules.html#snakefiles-resources"><span class="std std-ref">Resources</span></a>).</p>
</div>
<p>would execute the workflow with 10 cores.
Since the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> needs 8 threads, only one job of the rule can run at a time, and the Snakemake scheduler will try to saturate the remaining cores with other jobs like, e.g., <code class="docutils literal notranslate"><span class="pre">samtools_sort</span></code>.
The threads directive in a rule is interpreted as a maximum: when <strong>less cores than threads</strong> are provided, the number of threads a rule uses will be <strong>reduced to the number of given cores</strong>.</p>
<div class="section" id="exercise">
<h3>Exercise<a class="headerlink" href="#exercise" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>With the flag <code class="docutils literal notranslate"><span class="pre">--forceall</span></code> you can enforce a complete re-execution of the workflow. Combine this flag with different values for <code class="docutils literal notranslate"><span class="pre">--cores</span></code> and examine how the scheduler selects jobs to run in parallel.</li>
</ul>
</div>
</div>
<div class="section" id="step-2-config-files">
<h2>Step 2: Config files<a class="headerlink" href="#step-2-config-files" title="Permalink to this headline">¶</a></h2>
<p>So far, we specified the samples to consider in a Python list within the Snakefile.
However, often you want your workflow to be customizable, so that it can be easily adapted to new data.
For this purpose, Snakemake provides a config file mechanism.
Config files can be written in <a class="reference external" href="http://json.org">JSON</a> or <a class="reference external" href="http://yaml.org">YAML</a>, and loaded with the <code class="docutils literal notranslate"><span class="pre">configfile</span></code> directive.
In our example workflow, we add the line</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">configfile</span><span class="p">:</span> <span class="s2">&quot;config.yaml&quot;</span>
</pre></div>
</div>
<p>to the top of the Snakefile.
Snakemake will load the config file and store its contents into a globally available dictionary named <code class="docutils literal notranslate"><span class="pre">config</span></code>.
In our case, it makes sense to specify the samples in <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> as</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">samples</span><span class="p">:</span>
    <span class="nt">A</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">data/samples/A.fastq</span>
    <span class="nt">B</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">data/samples/B.fastq</span>
</pre></div>
</div>
<p>Now, we can remove the statement defining <code class="docutils literal notranslate"><span class="pre">SAMPLES</span></code> from the Snakefile and change the rule <code class="docutils literal notranslate"><span class="pre">bcftools_call</span></code> to</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bcftools_call</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">fa</span><span class="o">=</span><span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="n">bam</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">]),</span>
        <span class="n">bai</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam.bai&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">])</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools mpileup -g -f {input.fa} {input.bam} | &quot;</span>
        <span class="s2">&quot;bcftools call -mv - &gt; {output}&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-3-input-functions">
<span id="tutorial-input-functions"></span><h2>Step 3: Input functions<a class="headerlink" href="#step-3-input-functions" title="Permalink to this headline">¶</a></h2>
<p>Since we have stored the path to the FASTQ files in the config file, we can also generalize the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> to use these paths.
This case is different to the rule <code class="docutils literal notranslate"><span class="pre">bcftools_call</span></code> we modified above.
To understand this, it is important to know that Snakemake workflows are executed in three phases.</p>
<ul class="simple">
<li>In the <strong>initialization</strong> phase, the workflow is parsed and all rules are instantiated.</li>
<li>In the <strong>DAG</strong> phase, the DAG of jobs is built by filling wildcards and matching input files to output files.</li>
<li>In the <strong>scheduling</strong> phase, the DAG of jobs is executed.</li>
</ul>
<p>The expand functions in the list of input files of the rule <code class="docutils literal notranslate"><span class="pre">bcftools_call</span></code> are executed during the initialization phase.
In this phase, we don’t know about jobs, wildcard values and rule dependencies.
Hence, we cannot determine the FASTQ paths for rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> from the config file in this phase, because we don’t even know which jobs will be generated from that rule.
Instead, we need to defer the determination of input files to the DAG phase.
This can be achieved by specifying an <strong>input function</strong> instead of a string as inside of the input directive.
For the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> this works as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">sample</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">8</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem -t {threads} {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p>Snakemake does not automatically rerun jobs when new input files are added as
in the excercise below. However, you can get a list of output files that
are affected by such changes with <code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">--list-input-changes</span></code>.
To trigger a rerun, some bash magic helps:</p>
<div class="last highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">snakemake -n --forcerun $(snakemake --list-input-changes)</span>
</pre></div>
</div>
</div>
<p>Here, we use an anonymous function, also called <a class="reference external" href="https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions">lambda expression</a>.
Any normal function would work as well.
Input functions take as <strong>single argument</strong> a <code class="docutils literal notranslate"><span class="pre">wildcards</span></code> object, that allows to access the wildcards values via attributes (here <code class="docutils literal notranslate"><span class="pre">wildcards.sample</span></code>).
They have to <strong>return a string or a list of strings</strong>, that are interpreted as paths to input files (here, we return the path that is stored for the sample in the config file).
Input functions are evaluated once the wildcard values of a job are determined.</p>
<div class="section" id="id1">
<h3>Exercise<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>In the <code class="docutils literal notranslate"><span class="pre">data/samples</span></code> folder, there is an additional sample <code class="docutils literal notranslate"><span class="pre">C.fastq</span></code>. Add that sample to the config file and see how Snakemake wants to recompute the part of the workflow belonging to the new sample, when invoking with <code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">-n</span> <span class="pre">--reason</span> <span class="pre">--forcerun</span> <span class="pre">bcftools_call</span></code>.</li>
</ul>
</div>
</div>
<div class="section" id="step-4-rule-parameters">
<h2>Step 4: Rule parameters<a class="headerlink" href="#step-4-rule-parameters" title="Permalink to this headline">¶</a></h2>
<p>Sometimes, shell commands are not only composed of input and output files and some static flags.
In particular, it can happen that additional parameters need to be set depending on the wildcard values of the job.
For this, Snakemake allows to <strong>define arbitrary parameters</strong> for rules with the <code class="docutils literal notranslate"><span class="pre">params</span></code> directive.
In our workflow, it is reasonable to annotate aligned reads with so-called read groups, that contain metadata like the sample name.
We modify the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> accordingly:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">sample</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">rg</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;@RG\tID:{sample}\tSM:{sample}&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">8</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem -R &#39;{params.rg}&#39; -t {threads} {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">The <code class="docutils literal notranslate"><span class="pre">params</span></code> directive can also take functions like in Step 3 to defer
initialization to the DAG phase. In contrast to input functions, these can
optionally take additional arguments <code class="docutils literal notranslate"><span class="pre">input</span></code>, <code class="docutils literal notranslate"><span class="pre">output</span></code>, <code class="docutils literal notranslate"><span class="pre">threads</span></code>, and <code class="docutils literal notranslate"><span class="pre">resources</span></code>.</p>
</div>
<p>Similar to input and output files, <code class="docutils literal notranslate"><span class="pre">params</span></code> can be accessed from the shell command the Python based <code class="docutils literal notranslate"><span class="pre">run</span></code> block, or the script directive (see <a class="reference internal" href="basics.html#tutorial-script"><span class="std std-ref">Step 6: Using custom scripts</span></a>).</p>
<div class="section" id="id2">
<h3>Exercise<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Variant calling can consider a lot of parameters. A particularly important one is the prior mutation rate (1e-3 per default). It is set via the flag <code class="docutils literal notranslate"><span class="pre">-P</span></code> of the <code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">call</span></code> command. Consider making this flag configurable via adding a new key to the config file and using the <code class="docutils literal notranslate"><span class="pre">params</span></code> directive in the rule <code class="docutils literal notranslate"><span class="pre">bcftools_call</span></code> to propagate it to the shell command.</li>
</ul>
</div>
</div>
<div class="section" id="step-5-logging">
<h2>Step 5: Logging<a class="headerlink" href="#step-5-logging" title="Permalink to this headline">¶</a></h2>
<p>When executing a large workflow, it is usually desirable to store the output of each job persistently in files instead of just printing it to the terminal.
For this purpose, Snakemake allows to <strong>specify log files</strong> for rules.
Log files are defined via the <code class="docutils literal notranslate"><span class="pre">log</span></code> directive and handled similarly to output files, but they are not subject of rule matching and are not cleaned up when a job fails.
We modify our rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">sample</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">rg</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;@RG\tID:{sample}\tSM:{sample}&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/bwa_mem/{sample}.log&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">8</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;(bwa mem -R &#39;{params.rg}&#39; -t {threads} {input} | &quot;</span>
        <span class="s2">&quot;samtools view -Sb - &gt; {output}) 2&gt; {log}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">It is best practice to store all log files in a subdirectory <code class="docutils literal notranslate"><span class="pre">logs/</span></code>, prefixed by the rule or tool name.</p>
</div>
<p>The shell command is modified to collect STDERR output of both <code class="docutils literal notranslate"><span class="pre">bwa</span></code> and <code class="docutils literal notranslate"><span class="pre">samtools</span></code> and pipe it into the file referred by <code class="docutils literal notranslate"><span class="pre">{log}</span></code>.
Log files must contain exactly the same wildcards as the output files to avoid clashes.</p>
<div class="section" id="id3">
<h3>Exercise<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Add a log directive to the <code class="docutils literal notranslate"><span class="pre">bcftools_call</span></code> rule as well.</li>
<li>Time to re-run the whole workflow (remember the command line flags to force re-execution). See how log files are created for variant calling and read mapping.</li>
<li>The ability to track the provenance of each generated result is an important step towards reproducible analyses. Apart from the <code class="docutils literal notranslate"><span class="pre">report</span></code> functionality discussed before, Snakemake can summarize various provenance information for all output files of the workflow. The flag <code class="docutils literal notranslate"><span class="pre">--summary</span></code> prints a table associating each output file with the rule used to generate it, the creation date and optionally the version of the tool used for creation is provided. Further, the table informs about updated input files and changes to the source code of the rule after creation of the output file. Invoke Snakemake with <code class="docutils literal notranslate"><span class="pre">--summary</span></code> to examine the information for our example.</li>
</ul>
</div>
</div>
<div class="section" id="step-6-temporary-and-protected-files">
<span id="tutorial-temp-and-protected-files"></span><h2>Step 6: Temporary and protected files<a class="headerlink" href="#step-6-temporary-and-protected-files" title="Permalink to this headline">¶</a></h2>
<p>In our workflow, we create two BAM files for each sample, namely
the output of the rules <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> and <code class="docutils literal notranslate"><span class="pre">samtools_sort</span></code>.
When not dealing with examples, the underlying data is usually huge.
Hence, the resulting BAM files need a lot of disk space and their creation takes some time.
Snakemake allows to <strong>mark output files as temporary</strong>, such that they are deleted once every consuming job has been executed, in order to save disk space.
We use this mechanism for the output file of the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">sample</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">temp</span><span class="p">(</span><span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span><span class="p">)</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">rg</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;@RG\tID:{sample}\tSM:{sample}&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/bwa_mem/{sample}.log&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">8</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;(bwa mem -R &#39;{params.rg}&#39; -t {threads} {input} | &quot;</span>
        <span class="s2">&quot;samtools view -Sb - &gt; {output}) 2&gt; {log}&quot;</span>
</pre></div>
</div>
<p>This results in the deletion of the BAM file once the corresponding <code class="docutils literal notranslate"><span class="pre">samtools_sort</span></code> job has been executed.
Since the creation of BAM files via read mapping and sorting is computationally expensive, it is reasonable to <strong>protect</strong> the final BAM file <strong>from accidental deletion or modification</strong>.
We modify the rule <code class="docutils literal notranslate"><span class="pre">samtools_sort</span></code> by marking its output file as <code class="docutils literal notranslate"><span class="pre">protected</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">samtools_sort</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">protected</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span><span class="p">)</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools sort -T sorted_reads/{wildcards.sample} &quot;</span>
        <span class="s2">&quot;-O bam {input} &gt; {output}&quot;</span>
</pre></div>
</div>
<p>After execution of the job, Snakemake will write-protect the output file in the filesystem, so that it can’t be overwritten or deleted accidentally.</p>
<div class="section" id="id4">
<h3>Exercise<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Re-execute the whole workflow and observe how Snakemake handles the temporary and protected files.</li>
<li>Run Snakemake with the target <code class="docutils literal notranslate"><span class="pre">mapped_reads/A.bam</span></code>. Although the file is marked as temporary, you will see that Snakemake does not delete it because it is specified as a target file.</li>
<li>Try to re-execute the whole workflow again with the dry-run option. You will see that it fails (as intended) because Snakemake cannot overwrite the protected output files.</li>
</ul>
</div>
</div>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h2>
<p>The final version of our workflow looks like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">configfile</span><span class="p">:</span> <span class="s2">&quot;config.yaml&quot;</span>


<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;plots/quals.svg&quot;</span>


<span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">sample</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">temp</span><span class="p">(</span><span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span><span class="p">)</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">rg</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;@RG\tID:{sample}\tSM:{sample}&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/bwa_mem/{sample}.log&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">8</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;(bwa mem -R &#39;{params.rg}&#39; -t {threads} {input} | &quot;</span>
        <span class="s2">&quot;samtools view -Sb - &gt; {output}) 2&gt; {log}&quot;</span>


<span class="n">rule</span> <span class="n">samtools_sort</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">protected</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span><span class="p">)</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools sort -T sorted_reads/{wildcards.sample} &quot;</span>
        <span class="s2">&quot;-O bam {input} &gt; {output}&quot;</span>


<span class="n">rule</span> <span class="n">samtools_index</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;sorted_reads/{sample}.bam.bai&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools index {input}&quot;</span>


<span class="n">rule</span> <span class="n">bcftools_call</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">fa</span><span class="o">=</span><span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="n">bam</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">]),</span>
        <span class="n">bai</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam.bai&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">])</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools mpileup -g -f {input.fa} {input.bam} | &quot;</span>
        <span class="s2">&quot;bcftools call -mv - &gt; {output}&quot;</span>


<span class="n">rule</span> <span class="n">plot_quals</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;plots/quals.svg&quot;</span>
    <span class="n">script</span><span class="p">:</span>
        <span class="s2">&quot;scripts/plot-quals.py&quot;</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="additional_features.html" class="btn btn-neutral float-right" title="Additional features" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="basics.html" class="btn btn-neutral float-left" title="Basics: An example workflow" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2016, Johannes Koester

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>