

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Basics: An example workflow &mdash; Snakemake 5.6.0+8.g89bc383.dirty documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sphinx-argparse.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Advanced: Decorating the example workflow" href="advanced.html" />
    <link rel="prev" title="Setup" href="setup.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Snakemake
          

          
          </a>

          
            
            
              <div class="version">
                5.6.0+8.g89bc383.dirty
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started/installation.html">安装</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="tutorial.html">Snakemake Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="setup.html">Setup</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Basics: An example workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#background">Background</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-1-mapping-reads">Step 1: Mapping reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-2-generalizing-the-read-mapping-rule">Step 2: Generalizing the read mapping rule</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-3-sorting-read-alignments">Step 3: Sorting read alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-4-indexing-read-alignments-and-visualizing-the-dag-of-jobs">Step 4: Indexing read alignments and visualizing the DAG of jobs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#step-5-calling-genomic-variants">Step 5: Calling genomic variants</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#step-6-using-custom-scripts">Step 6: Using custom scripts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-7-adding-a-target-rule">Step 7: Adding a target rule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id4">Exercise</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#summary">Summary</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="advanced.html">Advanced: Decorating the example workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="additional_features.html">Additional features</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="short.html">Short tutorial</a></li>
</ul>
<p class="caption"><span class="caption-text">Executing workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../executable.html">Executing Snakemake</a></li>
</ul>
<p class="caption"><span class="caption-text">Defining workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/writing_snakefiles.html">Writing Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/rules.html">Rules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/modularization.html">Modularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/remote_files.html">Remote files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/utils.html">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/deployment.html">Distribution and Reproducibility</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/reporting.html">Reports</a></li>
</ul>
<p class="caption"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/snakemake.html">The Snakemake API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/snakemake_utils.html">Additional utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/internal/modules.html">Internal API</a></li>
</ul>
<p class="caption"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../project_info/citations.html">Citing and Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/more_resources.html">More Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/history.html">Change Log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/license.html">License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Snakemake</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="tutorial.html">Snakemake Tutorial</a> &raquo;</li>
        
      <li>Basics: An example workflow</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/tutorial/basics.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="basics-an-example-workflow">
<span id="tutorial-basics"></span><h1>Basics: An example workflow<a class="headerlink" href="#basics-an-example-workflow" title="Permalink to this headline">¶</a></h1>
<p>Please make sure that you have <strong>activated</strong> the environment we created before, and that you have an open terminal in the working directory you have created.</p>
<p><strong>A Snakemake workflow is defined by specifying rules in a Snakefile</strong>.
<strong>Rules decompose the workflow into small steps</strong> (e.g., the application of a single tool) by specifying how to create sets of <strong>output files</strong> from sets of <strong>input files</strong>.
Snakemake automatically <strong>determines the dependencies</strong> between the rules by matching file names.</p>
<p>The Snakemake language extends the Python language, adding syntactic structures for rule definition and additional controls.
All added syntactic structures begin with a keyword followed by a code block that is either in the same line or indented and consisting of multiple lines.
The resulting syntax resembles that of original Python constructs.</p>
<p>In the following, we will introduce the Snakemake syntax by creating an example workflow.
The workflow comes from the domain of genome analysis.
It maps sequencing reads to a reference genome and call variants on the mapped reads.
The tutorial does not require you to know what this is about.
Nevertheless, we provide some background in the following paragraph.</p>
<div class="section" id="background">
<span id="tutorial-background"></span><h2>Background<a class="headerlink" href="#background" title="Permalink to this headline">¶</a></h2>
<p>The genome of a living organism encodes its hereditary information.
It serves as a blueprint for proteins, which form living cells, carry information
and drive chemical reactions. Differences between populations, species, cancer
cells and healthy tissue, as well as syndromes or diseases can be reflected and
sometimes caused by changes in the genome.
This makes the genome a major target of biological and medical research.
Today, it is often analyzed with DNA sequencing, producing gigabytes of data from
a single biological sample (e.g. a biopsy of some tissue).
For technical reasons, DNA sequencing cuts the DNA of a sample into millions
of small pieces, called <strong>reads</strong>.
In order to recover the genome of the sample, one has to map these reads against
a known <strong>reference genome</strong> (e.g., the human one obtained during the famous
<a class="reference external" href="https://en.wikipedia.org/wiki/Human_Genome_Project">human genome project</a>).
This task is called <strong>read mapping</strong>.
Often, it is of interest where an individual genome is different from the species-wide consensus
represented with the reference genome.
Such differences are called <strong>variants</strong>. They are responsible for harmless individual
differences (like eye color), but can also cause diseases like cancer.
By investigating the differences between the all mapped reads
and the reference sequence at one position, variants can be detected.
This is a statistical challenge, because they have
to be distinguished from artifacts generated by the sequencing process.</p>
</div>
<div class="section" id="step-1-mapping-reads">
<h2>Step 1: Mapping reads<a class="headerlink" href="#step-1-mapping-reads" title="Permalink to this headline">¶</a></h2>
<p>Our first Snakemake rule maps reads of a given sample to a given reference genome (see <a class="reference internal" href="#tutorial-background"><span class="std std-ref">Background</span></a>).
For this, we will use the tool <a class="reference external" href="http://bio-bwa.sourceforge.net">bwa</a>, specifically the subcommand <code class="docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">mem</span></code>.
In the working directory, <strong>create a new file</strong> called <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> with an editor of your choice.
We propose to use the <a class="reference external" href="https://atom.io">Atom</a> editor, since it provides out-of-the-box syntax highlighting for Snakemake.
In the Snakefile, define the following rule:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;data/samples/A.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/A.bam&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">A common error is to forget the comma between the input or output items.
Since Python concatenates subsequent strings, this can lead to unexpected behavior.</p>
</div>
<p>A Snakemake rule has a name (here <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code>) and a number of directives, here <code class="docutils literal notranslate"><span class="pre">input</span></code>, <code class="docutils literal notranslate"><span class="pre">output</span></code> and <code class="docutils literal notranslate"><span class="pre">shell</span></code>.
The <code class="docutils literal notranslate"><span class="pre">input</span></code> and <code class="docutils literal notranslate"><span class="pre">output</span></code> directives are followed by lists of files that are expected to be used or created by the rule.
In the simplest case, these are just explicit Python strings.
The <code class="docutils literal notranslate"><span class="pre">shell</span></code> directive is followed by a Python string containing the shell command to execute.
In the shell command string, we can refer to elements of the rule via braces notation (similar to the Python format function).
Here, we refer to the output file by specifying <code class="docutils literal notranslate"><span class="pre">{output}</span></code> and to the input files by specifying <code class="docutils literal notranslate"><span class="pre">{input}</span></code>.
Since the rule has multiple input files, Snakemake will concatenate them separated by a whitespace.
In other words, Snakemake will replace <code class="docutils literal notranslate"><span class="pre">{input}</span></code> with <code class="docutils literal notranslate"><span class="pre">data/genome.fa</span> <span class="pre">data/samples/A.fastq</span></code> before executing the command.
The shell command invokes <code class="docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">mem</span></code> with reference genome and reads, and pipes the output into <code class="docutils literal notranslate"><span class="pre">samtools</span></code> which creates a compressed <a class="reference external" href="https://en.wikipedia.org/wiki/Binary_Alignment_Map">BAM</a> file containing the alignments.
The output of <code class="docutils literal notranslate"><span class="pre">samtools</span></code> is piped into the output file defined by the rule.</p>
<p>When a workflow is executed, Snakemake tries to generate given <strong>target</strong> files.
Target files can be specified via the command line.
By executing</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake -np mapped_reads/A.bam
</pre></div>
</div>
<p>in the working directory containing the Snakefile, we tell Snakemake to generate the target file <code class="docutils literal notranslate"><span class="pre">mapped_reads/A.bam</span></code>.
Since we used the <code class="docutils literal notranslate"><span class="pre">-n</span></code> (or <code class="docutils literal notranslate"><span class="pre">--dry-run</span></code>) flag, Snakemake will only show the execution plan instead of actually perform the steps.
The <code class="docutils literal notranslate"><span class="pre">-p</span></code> flag instructs Snakemake to also print the resulting shell command for illustration.
To generate the target files, <strong>Snakemake applies the rules given in the Snakefile in a top-down way</strong>.
The application of a rule to generate a set of output files is called <strong>job</strong>.
For each input file of a job, Snakemake again (i.e. recursively) determines rules that can be applied to generate it.
This yields a <a class="reference external" href="https://en.wikipedia.org/wiki/Directed_acyclic_graph">directed acyclic graph (DAG)</a> of jobs where the edges represent dependencies.
So far, we only have a single rule, and the DAG of jobs consists of a single node.
Nevertheless, we can <strong>execute our workflow</strong> with</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake mapped_reads/A.bam
</pre></div>
</div>
<p>Note that, after completion of above command, Snakemake will not try to create <code class="docutils literal notranslate"><span class="pre">mapped_reads/A.bam</span></code> again, because it is already present in the file system.
Snakemake <strong>only re-runs jobs if one of the input files is newer than one of the output files or one of the input files will be updated by another job</strong>.</p>
</div>
<div class="section" id="step-2-generalizing-the-read-mapping-rule">
<h2>Step 2: Generalizing the read mapping rule<a class="headerlink" href="#step-2-generalizing-the-read-mapping-rule" title="Permalink to this headline">¶</a></h2>
<p>Obviously, the rule will only work for a single sample with reads in the file <code class="docutils literal notranslate"><span class="pre">data/samples/A.fastq</span></code>.
However, Snakemake allows to <strong>generalize rules by using named wildcards</strong>.
Simply replace the <code class="docutils literal notranslate"><span class="pre">A</span></code> in the second input file and in the output file with the wildcard <code class="docutils literal notranslate"><span class="pre">{sample}</span></code>, leading to</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa_map</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;data/samples/{sample}.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">Note that if a rule has multiple output files, Snakemake requires them to all
have exactly the same wildcards. Otherwise, it could happen
that two jobs from the same rule want to write the same file.</p>
</div>
<p>When Snakemake determines that this rule can be applied to generate a target file by replacing the wildcard <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> in the output file with an appropriate value, it will propagate that value to all occurrences of <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> in the input files and thereby determine the necessary input for the resulting job.
Note that you can have multiple wildcards in your file paths, however, to avoid conflicts with other jobs of the same rule, <strong>all output files</strong> of a rule have to <strong>contain exactly the same wildcards</strong>.</p>
<p>When executing</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake -np mapped_reads/B.bam
</pre></div>
</div>
<p>Snakemake will determine that the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> can be applied to generate the target file by replacing the wildcard <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> with the value <code class="docutils literal notranslate"><span class="pre">B</span></code>.
In the output of the dry-run, you will see how the wildcard value is propagated to the input files and all filenames in the shell command.
You can also <strong>specify multiple targets</strong>, e.g.:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake -np mapped_reads/A.bam mapped_reads/B.bam
</pre></div>
</div>
<p>Some <a class="reference external" href="http://www.tldp.org/LDP/Bash-Beginners-Guide/html">Bash</a> magic can make this particularly handy. For example, you can alternatively compose our multiple targets in a single pass via</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake -np mapped_reads/<span class="o">{</span>A,B<span class="o">}</span>.bam
</pre></div>
</div>
<p>Note that this is not a special Snakemake syntax. Bash is just expanding the given path into two, one for each element of the set <code class="docutils literal notranslate"><span class="pre">{A,B}</span></code>.</p>
<p>In both cases, you will see that Snakemake only proposes to create the output file <code class="docutils literal notranslate"><span class="pre">mapped_reads/B.bam</span></code>.
This is because you already executed the workflow before (see the previous step) and no input file is newer than the output file <code class="docutils literal notranslate"><span class="pre">mapped_reads/A.bam</span></code>.
You can update the file modification date of the input file
<code class="docutils literal notranslate"><span class="pre">data/samples/A.fastq</span></code> via</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> touch data/samples/A.fastq
</pre></div>
</div>
<p>and see how Snakemake wants to re-run the job to create the file <code class="docutils literal notranslate"><span class="pre">mapped_reads/A.bam</span></code> by executing</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake -np mapped_reads/A.bam mapped_reads/B.bam
</pre></div>
</div>
</div>
<div class="section" id="step-3-sorting-read-alignments">
<h2>Step 3: Sorting read alignments<a class="headerlink" href="#step-3-sorting-read-alignments" title="Permalink to this headline">¶</a></h2>
<p>For later steps, we need the read alignments in the BAM files to be sorted.
This can be achieved with the <a class="reference external" href="http://www.htslib.org">samtools</a> command.
We add the following rule beneath the <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> rule:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">samtools_sort</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;mapped_reads/{sample}.bam&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools sort -T sorted_reads/{wildcards.sample} &quot;</span>
        <span class="s2">&quot;-O bam {input} &gt; {output}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">It is best practice to have subsequent steps of a workflow in separate, unique, output folders. This keeps the working directory structured. Further, such unique prefixes allow Snakemake to prune the search space for dependencies.</p>
</div>
<p>This rule will take the input file from the <code class="docutils literal notranslate"><span class="pre">mapped_reads</span></code> directory and store a sorted version in the <code class="docutils literal notranslate"><span class="pre">sorted_reads</span></code> directory.
Note that Snakemake <strong>automatically creates missing directories</strong> before jobs are executed.
For sorting, <code class="docutils literal notranslate"><span class="pre">samtools</span></code> requires a prefix specified with the flag <code class="docutils literal notranslate"><span class="pre">-T</span></code>.
Here, we need the value of the wildcard <code class="docutils literal notranslate"><span class="pre">sample</span></code>.
Snakemake allows to access wildcards in the shell command via the <code class="docutils literal notranslate"><span class="pre">wildcards</span></code> object that has an attribute with the value for each wildcard.</p>
<p>When issuing</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake -np sorted_reads/B.bam
</pre></div>
</div>
<p>you will see how Snakemake wants to run first the rule <code class="docutils literal notranslate"><span class="pre">bwa_map</span></code> and then the rule <code class="docutils literal notranslate"><span class="pre">samtools_sort</span></code> to create the desired target file:
as mentioned before, the dependencies are resolved automatically by matching file names.</p>
</div>
<div class="section" id="step-4-indexing-read-alignments-and-visualizing-the-dag-of-jobs">
<h2>Step 4: Indexing read alignments and visualizing the DAG of jobs<a class="headerlink" href="#step-4-indexing-read-alignments-and-visualizing-the-dag-of-jobs" title="Permalink to this headline">¶</a></h2>
<p>Next, we need to use <a class="reference external" href="http://www.htslib.org">samtools</a> again to index the sorted read alignments for random access.
This can be done with the following rule:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">samtools_index</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;sorted_reads/{sample}.bam.bai&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools index {input}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">Snakemake uses the Python format mini language to format shell commands.
Sometimes you have to use braces for something else in a shell command.
In that case, you have to escape them by doubling, e.g.,
<code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">{{A,B}}.txt</span></code>.</p>
</div>
<p>Having three steps already, it is a good time to take a closer look at the resulting DAG of jobs.
By executing</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake --dag sorted_reads/<span class="o">{</span>A,B<span class="o">}</span>.bam.bai <span class="p">|</span> dot -Tsvg &gt; dag.svg
</pre></div>
</div>
<p>we create a <strong>visualization of the DAG</strong> using the <code class="docutils literal notranslate"><span class="pre">dot</span></code> command provided by <a class="reference external" href="http://www.graphviz.org">Graphviz</a>.
For the given target files, Snakemake specifies the DAG in the dot language and pipes it into the <code class="docutils literal notranslate"><span class="pre">dot</span></code> command, which renders the definition into SVG format.
The rendered DAG is piped into the file <code class="docutils literal notranslate"><span class="pre">dag.svg</span></code> and will look similar to this:</p>
<img alt="../_images/dag_index.png" class="align-center" src="../_images/dag_index.png" />
<p>The DAG contains a node for each job and edges representing the dependencies.
Jobs that don’t need to be run because their output is up-to-date are dashed.
For rules with wildcards, the value of the wildcard for the particular job is displayed in the job node.</p>
<div class="section" id="exercise">
<h3>Exercise<a class="headerlink" href="#exercise" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Run parts of the workflow using different targets. Recreate the DAG and see how different rules become dashed because their output is present and up-to-date.</li>
</ul>
</div>
</div>
<div class="section" id="step-5-calling-genomic-variants">
<h2>Step 5: Calling genomic variants<a class="headerlink" href="#step-5-calling-genomic-variants" title="Permalink to this headline">¶</a></h2>
<p>The next step in our workflow will aggregate the mapped reads from all samples and jointly call genomic variants on them (see <a class="reference internal" href="#tutorial-background"><span class="std std-ref">Background</span></a>).
For the variant calling, we will combine the two utilities <a class="reference external" href="http://www.htslib.org">samtools</a> and <a class="reference external" href="http://www.htslib.org">bcftools</a>.
Snakemake provides a <strong>helper function for collecting input files</strong> that helps us to describe the aggregation in this step.
With</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLES</span><span class="p">)</span>
</pre></div>
</div>
<p>we obtain a list of files where the given pattern <code class="docutils literal notranslate"><span class="pre">&quot;sorted_reads/{sample}.bam&quot;</span></code> was formatted with the values in a given list of samples <code class="docutils literal notranslate"><span class="pre">SAMPLES</span></code>, i.e.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s2">&quot;sorted_reads/A.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;sorted_reads/B.bam&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>The function is particularly useful when the pattern contains multiple wildcards.
For example,</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.{replicate}.bam&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLES</span><span class="p">,</span> <span class="n">replicate</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
<p>would create the product of all elements of <code class="docutils literal notranslate"><span class="pre">SAMPLES</span></code> and the list <code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">1]</span></code>, yielding</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s2">&quot;sorted_reads/A.0.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;sorted_reads/A.1.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;sorted_reads/B.0.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;sorted_reads/B.1.bam&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Here, we use only the simple case of <code class="docutils literal notranslate"><span class="pre">expand</span></code>.
We first let Snakemake know which samples we want to consider.
Remember that Snakemake works top-down, it does not automatically infer this from, e.g., the fastq files in the data folder.
Also remember that Snakefiles are in principle Python code enhanced by some declarative statements to define workflows.
Hence, we can define the list of samples ad-hoc in plain Python at the top of the Snakefile:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SAMPLES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Later, we will learn about more sophisticated ways like <strong>config files</strong>.
Now, we can add the following rule to our Snakefile:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bcftools_call</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">fa</span><span class="o">=</span><span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="n">bam</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLES</span><span class="p">),</span>
        <span class="n">bai</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;sorted_reads/{sample}.bam.bai&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLES</span><span class="p">)</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools mpileup -g -f {input.fa} {input.bam} | &quot;</span>
        <span class="s2">&quot;bcftools call -mv - &gt; {output}&quot;</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">If you name input or output files like above, their order won’t be preserved when referring them as <code class="docutils literal notranslate"><span class="pre">{input}</span></code>.
Further, note that named and not named (i.e., positional) input and output files can be combined, but the positional ones must come first, equivalent to Python functions with keyword arguments.</p>
</div>
<p>With multiple input or output files, it is sometimes handy to refer them separately in the shell command.
This can be done by <strong>specifying names for input or output files</strong> (here, e.g., <code class="docutils literal notranslate"><span class="pre">fa=...</span></code>).
The files can then be referred in the shell command via, e.g., <code class="docutils literal notranslate"><span class="pre">{input.fa}</span></code>.
For <strong>long shell commands</strong> like this one, it is advisable to <strong>split the string over multiple indented lines</strong>.
Python will automatically merge it into one.
Further, you will notice that the <strong>input or output file lists can contain arbitrary Python statements</strong>, as long as it returns a string, or a list of strings.
Here, we invoke our <code class="docutils literal notranslate"><span class="pre">expand</span></code> function to aggregate over the aligned reads of all samples.</p>
<div class="section" id="id1">
<h3>Exercise<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>obtain the updated DAG of jobs for the target file <code class="docutils literal notranslate"><span class="pre">calls/all.vcf</span></code>, it should look like this:</li>
</ul>
<img alt="../_images/dag_call.png" class="align-center" src="../_images/dag_call.png" />
</div>
</div>
<div class="section" id="step-6-using-custom-scripts">
<span id="tutorial-script"></span><h2>Step 6: Using custom scripts<a class="headerlink" href="#step-6-using-custom-scripts" title="Permalink to this headline">¶</a></h2>
<p>Usually, a workflow not only consists of invoking various tools, but also contains custom code to e.g. calculate summary statistics or create plots.
While Snakemake also allows you to directly <a href="#id2"><span class="problematic" id="id3">:ref:`write Python code inside a rule &lt;.. _snakefiles-rules&gt;`_</span></a>, it is usually reasonable to move such logic into separate scripts.
For this purpose, Snakemake offers the <code class="docutils literal notranslate"><span class="pre">script</span></code> directive.
Add the following rule to your Snakefile:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">plot_quals</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;plots/quals.svg&quot;</span>
    <span class="n">script</span><span class="p">:</span>
        <span class="s2">&quot;scripts/plot-quals.py&quot;</span>
</pre></div>
</div>
<p>This rule shall generate a histogram of the quality scores that have been assigned to the variant calls in the file <code class="docutils literal notranslate"><span class="pre">calls/all.vcf</span></code>.
The actual Python code to generate the plot is hidden in the script <code class="docutils literal notranslate"><span class="pre">scripts/plot-quals.py</span></code>.
Script paths are always relative to the referring Snakefile.
In the script, all properties of the rule like <code class="docutils literal notranslate"><span class="pre">input</span></code>, <code class="docutils literal notranslate"><span class="pre">output</span></code>, <code class="docutils literal notranslate"><span class="pre">wildcards</span></code>, etc. are available as attributes of a global <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> object.
Create the file <code class="docutils literal notranslate"><span class="pre">scripts/plot-quals.py</span></code>, with the following content:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s2">&quot;Agg&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">pysam</span> <span class="kn">import</span> <span class="n">VariantFile</span>

<span class="n">quals</span> <span class="o">=</span> <span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">qual</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">VariantFile</span><span class="p">(</span><span class="n">snakemake</span><span class="o">.</span><span class="n">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">quals</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">snakemake</span><span class="o">.</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">It is best practice to use the script directive whenever an inline code block would have
more than a few lines of code.</p>
</div>
<p>Although there are other strategies to invoke separate scripts from your workflow
(e.g., invoking them via shell commands), the benefit of this is obvious:
the script logic is separated from the workflow logic (and can be even shared between workflows),
but <strong>boilerplate code like the parsing of command line arguments is unnecessary</strong>.</p>
<p>Apart from Python scripts, it is also possible to use R scripts. In R scripts,
an S4 object named <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> analog to the Python case above is available and
allows access to input and output files and other parameters. Here the syntax
follows that of S4 classes with attributes that are R lists, e.g. we can access
the first input file with <code class="docutils literal notranslate"><span class="pre">snakemake&#64;input[[1]]</span></code> (note that the first file does
not have index 0 here, because R starts counting from 1). Named input and output
files can be accessed in the same way, by just providing the name instead of an
index, e.g. <code class="docutils literal notranslate"><span class="pre">snakemake&#64;input[[&quot;myfile&quot;]]</span></code>.</p>
<p>For details and examples, see the <a class="reference internal" href="../snakefiles/rules.html#snakefiles-external-scripts"><span class="std std-ref">External scripts</span></a> section in the Documentation.</p>
</div>
<div class="section" id="step-7-adding-a-target-rule">
<h2>Step 7: Adding a target rule<a class="headerlink" href="#step-7-adding-a-target-rule" title="Permalink to this headline">¶</a></h2>
<p>So far, we always executed the workflow by specifying a target file at the command line.
Apart from filenames, Snakemake <strong>also accepts rule names as targets</strong> if the referred rule does not have wildcards.
Hence, it is possible to write target rules collecting particular subsets of the desired results or all results.
Moreover, if no target is given at the command line, Snakemake will define the <strong>first rule</strong> of the Snakefile as the target.
Hence, it is best practice to have a rule <code class="docutils literal notranslate"><span class="pre">all</span></code> at the top of the workflow which has all typically desired target files as input files.</p>
<p>Here, this means that we add a rule</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;plots/quals.svg&quot;</span>
</pre></div>
</div>
<p>to the top of our workflow.
When executing Snakemake with</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> snakemake -n
</pre></div>
</div>
<div class="sidebar">
<p class="first sidebar-title">Note</p>
<p class="last">In case you have mutliple reasonable sets of target files,
you can add multiple target rules at the top of the Snakefile. While
Snakemake will execute the first per default, you can target any of them via
the command line (e.g., <code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">-n</span> <span class="pre">mytarget</span></code>).</p>
</div>
<p>the execution plan for creating the file <code class="docutils literal notranslate"><span class="pre">plots/quals.svg</span></code> which contains and summarizes all our results will be shown.
Note that, apart from Snakemake considering the first rule of the workflow as default target, <strong>the appearance of rules in the Snakefile is arbitrary and does not influence the DAG of jobs</strong>.</p>
<div class="section" id="id4">
<h3>Exercise<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Create the DAG of jobs for the complete workflow.</li>
<li>Execute the complete workflow and have a look at the resulting <code class="docutils literal notranslate"><span class="pre">plots/quals.svg</span></code>.</li>
<li>Snakemake provides handy flags for forcing re-execution of parts of the workflow. Have a look at the command line help with <code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">--help</span></code> and search for the flag <code class="docutils literal notranslate"><span class="pre">--forcerun</span></code>. Then, use this flag to re-execute the rule <code class="docutils literal notranslate"><span class="pre">samtools_sort</span></code> and see what happens.</li>
<li>With <code class="docutils literal notranslate"><span class="pre">--reason</span></code> it is possible to display the execution reason for each job. Try this flag together with a dry-run and the <code class="docutils literal notranslate"><span class="pre">--forcerun</span></code> flag to understand the decisions of Snakemake.</li>
</ul>
</div>
</div>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h2>
<p>In total, the resulting workflow looks like this:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">SAMPLES = [&quot;A&quot;, &quot;B&quot;]</span>


<span class="go">rule all:</span>
<span class="go">    input:</span>
<span class="go">        &quot;plots/quals.svg&quot;</span>


<span class="go">rule bwa_map:</span>
<span class="go">    input:</span>
<span class="go">        &quot;data/genome.fa&quot;,</span>
<span class="go">        &quot;data/samples/{sample}.fastq&quot;</span>
<span class="go">    output:</span>
<span class="go">        &quot;mapped_reads/{sample}.bam&quot;</span>
<span class="go">    shell:</span>
<span class="go">        &quot;bwa mem {input} | samtools view -Sb - &gt; {output}&quot;</span>


<span class="go">rule samtools_sort:</span>
<span class="go">    input:</span>
<span class="go">        &quot;mapped_reads/{sample}.bam&quot;</span>
<span class="go">    output:</span>
<span class="go">        &quot;sorted_reads/{sample}.bam&quot;</span>
<span class="go">    shell:</span>
<span class="go">        &quot;samtools sort -T sorted_reads/{wildcards.sample} &quot;</span>
<span class="go">        &quot;-O bam {input} &gt; {output}&quot;</span>


<span class="go">rule samtools_index:</span>
<span class="go">    input:</span>
<span class="go">        &quot;sorted_reads/{sample}.bam&quot;</span>
<span class="go">    output:</span>
<span class="go">        &quot;sorted_reads/{sample}.bam.bai&quot;</span>
<span class="go">    shell:</span>
<span class="go">        &quot;samtools index {input}&quot;</span>


<span class="go">rule bcftools_call:</span>
<span class="go">    input:</span>
<span class="go">        fa=&quot;data/genome.fa&quot;,</span>
<span class="go">        bam=expand(&quot;sorted_reads/{sample}.bam&quot;, sample=SAMPLES),</span>
<span class="go">        bai=expand(&quot;sorted_reads/{sample}.bam.bai&quot;, sample=SAMPLES)</span>
<span class="go">    output:</span>
<span class="go">        &quot;calls/all.vcf&quot;</span>
<span class="go">    shell:</span>
<span class="go">        &quot;samtools mpileup -g -f {input.fa} {input.bam} | &quot;</span>
<span class="go">        &quot;bcftools call -mv - &gt; {output}&quot;</span>


<span class="go">rule plot_quals:</span>
<span class="go">    input:</span>
<span class="go">        &quot;calls/all.vcf&quot;</span>
<span class="go">    output:</span>
<span class="go">        &quot;plots/quals.svg&quot;</span>
<span class="go">    script:</span>
<span class="go">        &quot;scripts/plot-quals.py&quot;</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="advanced.html" class="btn btn-neutral float-right" title="Advanced: Decorating the example workflow" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="setup.html" class="btn btn-neutral float-left" title="Setup" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2016, Johannes Koester

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>