

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Short tutorial &mdash; Snakemake 5.6.0+8.g89bc383.dirty documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sphinx-argparse.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Executing Snakemake" href="../executable.html" />
    <link rel="prev" title="Additional features" href="additional_features.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Snakemake
          

          
          </a>

          
            
            
              <div class="version">
                5.6.0+8.g89bc383.dirty
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started/installation.html">安装</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Snakemake Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Short tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prerequisites">Prerequisites</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-1">Step 1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-2">Step 2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-3">Step 3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-4">Step 4</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-5">Step 5</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-6">Step 6</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-7">Step 7</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-8">Step 8</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#automatic-reports">Automatic reports</a></li>
<li class="toctree-l3"><a class="reference internal" href="#threads">Threads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#temporary-files">Temporary files</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#solutions">Solutions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#step-2-1">Step 2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-3-1">Step 3</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-4-1">Step 4</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-5-1">Step 5</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-6-1">Step 6</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-7-1">Step 7</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-8-1">Step 8</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Executing workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../executable.html">Executing Snakemake</a></li>
</ul>
<p class="caption"><span class="caption-text">Defining workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/writing_snakefiles.html">Writing Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/rules.html">Rules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/modularization.html">Modularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/remote_files.html">Remote files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/utils.html">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/deployment.html">Distribution and Reproducibility</a></li>
<li class="toctree-l1"><a class="reference internal" href="../snakefiles/reporting.html">Reports</a></li>
</ul>
<p class="caption"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/snakemake.html">The Snakemake API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/snakemake_utils.html">Additional utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_reference/internal/modules.html">Internal API</a></li>
</ul>
<p class="caption"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../project_info/citations.html">Citing and Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/more_resources.html">More Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/history.html">Change Log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_info/license.html">License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Snakemake</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Short tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/tutorial/short.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="short-tutorial">
<h1>Short tutorial<a class="headerlink" href="#short-tutorial" title="Permalink to this headline">¶</a></h1>
<p>Here we provide a short tutorial that guides you through the main features of Snakemake.
Note that this is not suited to learn Snakemake from scratch, rather to give a first impression.
To really learn Snakemake (starting from something simple, and extending towards advanced features), use the main <a class="reference internal" href="tutorial.html#tutorial"><span class="std std-ref">Snakemake Tutorial</span></a>.</p>
<p>This document shows all steps performed in the official <a class="reference external" href="https://youtu.be/hPrXcUUp70Y">Snakemake live demo</a>,
such that it becomes possible to follow them at your own pace.
Solutions to each step can be found at the bottom of this document.</p>
<p>The examples presented in this tutorial come from Bioinformatics.
However, Snakemake is a general-purpose workflow management system for any discipline.
For an explanation of the steps you will perform here, have a look at <a class="reference internal" href="basics.html#tutorial-background"><span class="std std-ref">Background</span></a>.
More thorough explanations are provided in the full <a class="reference internal" href="tutorial.html#tutorial"><span class="std std-ref">Snakemake Tutorial</span></a>.</p>
<div class="section" id="prerequisites">
<h2>Prerequisites<a class="headerlink" href="#prerequisites" title="Permalink to this headline">¶</a></h2>
<p>First, install Snakemake via Conda, as outlined in <a class="reference internal" href="../getting_started/installation.html#conda-install"><span class="std std-ref">Conda安装</span></a>.
The minimal version of Snakemake is sufficient for this demo.</p>
<p>Second, download and unpack the test data needed for this example from
<a class="reference external" href="https://bitbucket.org/snakemake/snakemake-tutorial/get/v5.2.3.tar.bz2">here</a>,
e.g., via</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">snakemake</span><span class="o">-</span><span class="n">demo</span>
<span class="n">cd</span> <span class="n">snakemake</span><span class="o">-</span><span class="n">demo</span>
<span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">bitbucket</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">snakemake</span><span class="o">/</span><span class="n">snakemake</span><span class="o">-</span><span class="n">tutorial</span><span class="o">/</span><span class="n">get</span><span class="o">/</span><span class="n">v5</span><span class="o">.</span><span class="mf">2.3</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">bz2</span>
<span class="n">tar</span> <span class="o">--</span><span class="n">wildcards</span> <span class="o">-</span><span class="n">xf</span> <span class="n">v5</span><span class="o">.</span><span class="mf">2.3</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">bz2</span> <span class="o">--</span><span class="n">strip</span> <span class="mi">1</span> <span class="s2">&quot;*/data&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-1">
<h2>Step 1<a class="headerlink" href="#step-1" title="Permalink to this headline">¶</a></h2>
<p>First, create an empty workflow in the current directory with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">touch</span> <span class="n">Snakefile</span>
</pre></div>
</div>
<p>Once a Snakefile is present, you can perform a dry run of Snakemake
with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">-</span><span class="n">n</span>
</pre></div>
</div>
<p>Since the Snakefile is empty, it will report that nothing has to be
done. In the next steps, we will gradually fill the Snakefile with an
example analysis workflow.</p>
</div>
<div class="section" id="step-2">
<h2>Step 2<a class="headerlink" href="#step-2" title="Permalink to this headline">¶</a></h2>
<p>The data folder in your working directory looks as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>data
├── genome.fa
├── genome.fa.amb
├── genome.fa.ann
├── genome.fa.bwt
├── genome.fa.fai
├── genome.fa.pac
├── genome.fa.sa
└── samples
    ├── A.fastq
    ├── B.fastq
    └── C.fastq
</pre></div>
</div>
<p>You will create a workflow that maps the sequencing samples in the
<code class="docutils literal notranslate"><span class="pre">data/samples</span></code> folder to the reference genome <code class="docutils literal notranslate"><span class="pre">data/genome.fa</span></code>.
Then, you will call genomic variants over the mapped samples, and create
an example plot.</p>
<p>First, create a rule called <code class="docutils literal notranslate"><span class="pre">bwa</span></code>, with input files</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">data/genome.fa</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">data/samples/A.fastq</span></code></li>
</ul>
<p>and output file</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">mapped/A.bam</span></code></li>
</ul>
<p>To generate output from input, use the shell command</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s2">&quot;bwa mem {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
<p>Providing a shell command is not enough to run your workflow on an
unprepared system. For reproducibility, you also have to provide the
required software stack and define the desired version. This can be done
with the <a class="reference external" href="https://conda.io">Conda package manager</a>, which is directly
integrated with Snakemake: add a directive
<code class="docutils literal notranslate"><span class="pre">conda:</span> <span class="pre">&quot;envs/mapping.yaml&quot;</span></code> that points to a <a class="reference external" href="https://conda.io/docs/user-guide/tasks/manage-environments.html?highlight=environment#creating-an-environment-file-manually">Conda environment
definition</a>,
with the following content</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">channels</span><span class="p">:</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">bioconda</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="nt">dependencies</span><span class="p">:</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">bwa =0.7.17</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">samtools =1.9</span>
</pre></div>
</div>
<p>Upon execution, Snakemake will automatically create that environment,
and execute the shell command within.</p>
<p>Now, test your workflow by simulating the creation of the file
<code class="docutils literal notranslate"><span class="pre">mapped/A.bam</span></code> via</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="o">-</span><span class="n">n</span> <span class="n">mapped</span><span class="o">/</span><span class="n">A</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>to perform a dry-run and</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="n">mapped</span><span class="o">/</span><span class="n">A</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>to perform the actual execution.</p>
</div>
<div class="section" id="step-3">
<h2>Step 3<a class="headerlink" href="#step-3" title="Permalink to this headline">¶</a></h2>
<p>Now, generalize the rule <code class="docutils literal notranslate"><span class="pre">bwa</span></code> by replacing the concrete sample name
<code class="docutils literal notranslate"><span class="pre">A</span></code> with a wildcard <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> in input and output file the rule
<code class="docutils literal notranslate"><span class="pre">bwa</span></code>. This way, Snakemake can apply the rule to map any of the three
available samples to the reference genome.</p>
<p>Test this by creating the file <code class="docutils literal notranslate"><span class="pre">mapped/B.bam</span></code>.</p>
</div>
<div class="section" id="step-4">
<h2>Step 4<a class="headerlink" href="#step-4" title="Permalink to this headline">¶</a></h2>
<p>Next, create a rule <code class="docutils literal notranslate"><span class="pre">sort</span></code> that sorts the obtained <code class="docutils literal notranslate"><span class="pre">.bam</span></code> file by
genomic coordinate. The rule should have the input file</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">mapped/{sample}.bam</span></code></li>
</ul>
<p>and the output file</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">mapped/{sample}.sorted.bam</span></code></li>
</ul>
<p>and uses the shell command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">sort</span> <span class="o">-</span><span class="n">o</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span> <span class="p">{</span><span class="nb">input</span><span class="p">}</span>
</pre></div>
</div>
<p>to perform the sorting. Moreover, use the same <code class="docutils literal notranslate"><span class="pre">conda:</span></code> directive as
for the previous rule.</p>
<p>Test your workflow with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="o">-</span><span class="n">n</span> <span class="n">mapped</span><span class="o">/</span><span class="n">A</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>and</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="n">mapped</span><span class="o">/</span><span class="n">A</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</div>
<div class="section" id="step-5">
<h2>Step 5<a class="headerlink" href="#step-5" title="Permalink to this headline">¶</a></h2>
<p>Now, we aggregate over all samples to perform a joint calling of genomic
variants. First, we define a variable</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>at the top of the <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code>. This serves as a definition of the
samples over which we would want to aggregate. In real life, you would
want to use an external sample sheet or a <a class="reference external" href="http://snakemake.readthedocs.io/en/stable/tutorial/advanced.html#step-2-config-files">config
file</a>
for things like this.</p>
<p>For aggregation over many files, Snakemake provides the helper function
<code class="docutils literal notranslate"><span class="pre">expand</span></code> (see <a class="reference external" href="http://snakemake.readthedocs.io/en/stable/tutorial/basics.html#step-5-calling-genomic-variants">the
docs</a>).
Create a rule <code class="docutils literal notranslate"><span class="pre">call</span></code> with input files</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">fa=&quot;data/genome.fa&quot;</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">bam=expand(&quot;mapped/{sample}.sorted.bam&quot;,</span> <span class="pre">sample=samples)</span></code></li>
</ul>
<p>output file</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">&quot;calls/all.vcf&quot;</span></code></li>
</ul>
<p>and shell command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">mpileup</span> <span class="o">-</span><span class="n">g</span> <span class="o">-</span><span class="n">f</span> <span class="p">{</span><span class="nb">input</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="p">{</span><span class="nb">input</span><span class="o">.</span><span class="n">bam</span><span class="p">}</span> <span class="o">|</span> <span class="n">bcftools</span> <span class="n">call</span> <span class="o">-</span><span class="n">mv</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span>
</pre></div>
</div>
<p>Further, define a new conda environment file with the following content:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">channels</span><span class="p">:</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">bioconda</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="nt">dependencies</span><span class="p">:</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">bcftools =1.9</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">samtools =1.9</span>
</pre></div>
</div>
</div>
<div class="section" id="step-6">
<h2>Step 6<a class="headerlink" href="#step-6" title="Permalink to this headline">¶</a></h2>
<p>Finally, we strive to calculate some exemplary statistics. This time, we
don’t use a shell command, but rather employ Snakemake’s ability to
integrate with scripting languages like R and Python.</p>
<p>First, we create a rule <code class="docutils literal notranslate"><span class="pre">stats</span></code> with input file</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">&quot;calls/all.vcf&quot;</span></code></li>
</ul>
<p>and output file</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">&quot;plots/quals.svg&quot;</span></code>.</li>
</ul>
<p>Instead of a shell command, we write</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">script</span><span class="p">:</span>
    <span class="s2">&quot;scripts/plot-quals.py&quot;</span>
</pre></div>
</div>
<p>and create the corresponding script and its containing folder in our
working directory with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">scripts</span>
<span class="n">touch</span> <span class="n">scripts</span><span class="o">/</span><span class="n">plot</span><span class="o">-</span><span class="n">quals</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<p>We open the script in the editor and add the following content</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s2">&quot;Agg&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">pysam</span> <span class="kn">import</span> <span class="n">VariantFile</span>

<span class="n">quals</span> <span class="o">=</span> <span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">qual</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">VariantFile</span><span class="p">(</span><span class="n">snakemake</span><span class="o">.</span><span class="n">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">quals</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">snakemake</span><span class="o">.</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
<p>As you can see, instead of writing a command line parser for passing
parameters like input and output files, you have direct access to the
properties of the rule via a magic <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> object, that Snakemake
automatically inserts into the script before executing the rule.</p>
<p>Finally, we have to define a conda environment for the rule, say
<code class="docutils literal notranslate"><span class="pre">envs/stats.yaml</span></code>, that provides the required Python packages to
execute the script:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">channels</span><span class="p">:</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">bioconda</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="nt">dependencies</span><span class="p">:</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">pysam =0.15</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">matplotlib =3.1</span>
  <span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">python =3.7</span>
</pre></div>
</div>
<p>Make sure to test your workflow with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="n">plots</span><span class="o">/</span><span class="n">quals</span><span class="o">.</span><span class="n">svg</span>
</pre></div>
</div>
</div>
<div class="section" id="step-7">
<h2>Step 7<a class="headerlink" href="#step-7" title="Permalink to this headline">¶</a></h2>
<p>So far, we have always specified a target file at the command line when
invoking Snakemake. When no target file is specified, Snakemake tries to
execute the first rule in the <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code>. We can use this property to
define default target files.</p>
<p>At the top of your <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> define a rule <code class="docutils literal notranslate"><span class="pre">all</span></code>, with input files</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">&quot;calls/all.vcf&quot;</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">&quot;plots/quals.svg&quot;</span></code></li>
</ul>
<p>and neither a shell command nor output files. This rule simply serves as
an indicator of what shall be collected as results.</p>
</div>
<div class="section" id="step-8">
<h2>Step 8<a class="headerlink" href="#step-8" title="Permalink to this headline">¶</a></h2>
<p>As a last step, we strive to annotate our workflow with some additional
information.</p>
<div class="section" id="automatic-reports">
<h3>Automatic reports<a class="headerlink" href="#automatic-reports" title="Permalink to this headline">¶</a></h3>
<p>Snakemake can automatically create HTML reports with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">--</span><span class="n">report</span> <span class="n">report</span><span class="o">.</span><span class="n">html</span>
</pre></div>
</div>
<p>Such a report contains runtime statistics, a visualization of the
workflow topology, used software and data provenance information.</p>
<p>In addition, you can mark any output file generated in your workflow for
inclusion into the report. It will be encoded directly into the report,
such that it can be, e.g., emailed as a self-contained document. The
reader (e.g., a collaborator of yours) can at any time download the
enclosed results from the report for further use, e.g., in a manuscript
you write together. In this example, please mark the output file
<code class="docutils literal notranslate"><span class="pre">&quot;plots/quals.svg&quot;</span></code> for inclusion by replacing it with
<code class="docutils literal notranslate"><span class="pre">report(&quot;plots/quals.svg&quot;,</span> <span class="pre">caption=&quot;report/calling.rst&quot;)</span></code> and adding a
file <code class="docutils literal notranslate"><span class="pre">report/calling.rst</span></code>, containing some description of the output
file. This description will be presented as caption in the resulting
report.</p>
</div>
<div class="section" id="threads">
<h3>Threads<a class="headerlink" href="#threads" title="Permalink to this headline">¶</a></h3>
<p>The first rule <code class="docutils literal notranslate"><span class="pre">bwa</span></code> can in theory use multiple threads. You can make
Snakemake aware of this, such that the information can be used for
scheduling. Add a directive <code class="docutils literal notranslate"><span class="pre">threads:</span> <span class="pre">8</span></code> to the rule and alter the
shell command to</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bwa</span> <span class="n">mem</span> <span class="o">-</span><span class="n">t</span> <span class="p">{</span><span class="n">threads</span><span class="p">}</span> <span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">|</span> <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">Sb</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span>
</pre></div>
</div>
<p>This passes the threads defined in the rule as a command line argument
to the <code class="docutils literal notranslate"><span class="pre">bwa</span></code> process.</p>
</div>
<div class="section" id="temporary-files">
<h3>Temporary files<a class="headerlink" href="#temporary-files" title="Permalink to this headline">¶</a></h3>
<p>The output of the <code class="docutils literal notranslate"><span class="pre">bwa</span></code> rule becomes superfluous once the sorted
version of the <code class="docutils literal notranslate"><span class="pre">.bam</span></code> file is generated by the rule <code class="docutils literal notranslate"><span class="pre">sort</span></code>.
Snakemake can automatically delete the superfluous output once it is not
needed anymore. For this, mark the output as temporary by replacing
<code class="docutils literal notranslate"><span class="pre">&quot;mapped/{sample}.bam&quot;</span></code> in the rule <code class="docutils literal notranslate"><span class="pre">bwa</span></code> with
<code class="docutils literal notranslate"><span class="pre">temp(&quot;mapped/{sample}.bam&quot;)</span></code>.</p>
</div>
</div>
<div class="section" id="solutions">
<h2>Solutions<a class="headerlink" href="#solutions" title="Permalink to this headline">¶</a></h2>
<p>Only read this if you have a problem with one of the steps.</p>
<div class="section" id="step-2-1">
<span id="id1"></span><h3>Step 2<a class="headerlink" href="#step-2-1" title="Permalink to this headline">¶</a></h3>
<p>The rule should look like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;data/samples/A.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped/A.bam&quot;</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/mapping.yaml&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-3-1">
<span id="id2"></span><h3>Step 3<a class="headerlink" href="#step-3-1" title="Permalink to this headline">¶</a></h3>
<p>The rule should look like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">bwa</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;data/samples/{sample}.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped/{sample}.bam&quot;</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/mapping.yaml&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem {input} | samtools view -Sb - &gt; {output}&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-4-1">
<span id="id3"></span><h3>Step 4<a class="headerlink" href="#step-4-1" title="Permalink to this headline">¶</a></h3>
<p>The rule should look like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">sort</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;mapped/{sample}.bam&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped/{sample}.sorted.bam&quot;</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/mapping.yaml&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools sort -o {output} {input}&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-5-1">
<span id="id4"></span><h3>Step 5<a class="headerlink" href="#step-5-1" title="Permalink to this headline">¶</a></h3>
<p>The rule should look like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>

<span class="n">rule</span> <span class="n">call</span><span class="p">:</span>
  <span class="nb">input</span><span class="p">:</span>
      <span class="n">fa</span><span class="o">=</span><span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
      <span class="n">bam</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;mapped/{sample}.sorted.bam&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">samples</span><span class="p">)</span>
  <span class="n">output</span><span class="p">:</span>
      <span class="s2">&quot;calls/all.vcf&quot;</span>
  <span class="n">conda</span><span class="p">:</span>
      <span class="s2">&quot;envs/calling.yaml&quot;</span>
  <span class="n">shell</span><span class="p">:</span>
      <span class="s2">&quot;samtools mpileup -g -f {input.fa} {input.bam} | &quot;</span>
      <span class="s2">&quot;bcftools call -mv - &gt; {output}&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-6-1">
<span id="id5"></span><h3>Step 6<a class="headerlink" href="#step-6-1" title="Permalink to this headline">¶</a></h3>
<p>The rule should look like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">stats</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;plots/quals.svg&quot;</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/stats.yaml&quot;</span>
    <span class="n">script</span><span class="p">:</span>
        <span class="s2">&quot;scripts/plot-quals.py&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-7-1">
<span id="id6"></span><h3>Step 7<a class="headerlink" href="#step-7-1" title="Permalink to this headline">¶</a></h3>
<p>The rule should look like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span><span class="p">,</span>
        <span class="s2">&quot;plots/quals.svg&quot;</span>
</pre></div>
</div>
<p>It has to appear as first rule in the <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code>.</p>
</div>
<div class="section" id="step-8-1">
<span id="id7"></span><h3>Step 8<a class="headerlink" href="#step-8-1" title="Permalink to this headline">¶</a></h3>
<p>The complete workflow should look like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">]</span>


<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span><span class="p">,</span>
        <span class="s2">&quot;plots/quals.svg&quot;</span>


<span class="n">rule</span> <span class="n">bwa</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;data/samples/{sample}.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">temp</span><span class="p">(</span><span class="s2">&quot;mapped/{sample}.bam&quot;</span><span class="p">)</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/mapping.yaml&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">8</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;bwa mem -t {threads} {input} | samtools view -Sb - &gt; {output}&quot;</span>


<span class="n">rule</span> <span class="n">sort</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;mapped/{sample}.bam&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;mapped/{sample}.sorted.bam&quot;</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/mapping.yaml&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools sort -o {output} {input}&quot;</span>



<span class="n">rule</span> <span class="n">call</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">fa</span><span class="o">=</span><span class="s2">&quot;data/genome.fa&quot;</span><span class="p">,</span>
        <span class="n">bam</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span><span class="s2">&quot;mapped/{sample}.sorted.bam&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">samples</span><span class="p">)</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/calling.yaml&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;samtools mpileup -g -f {input.fa} {input.bam} | &quot;</span>
        <span class="s2">&quot;bcftools call -mv - &gt; {output}&quot;</span>

<span class="n">rule</span> <span class="n">stats</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;calls/all.vcf&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">report</span><span class="p">(</span><span class="s2">&quot;plots/quals.svg&quot;</span><span class="p">,</span> <span class="n">caption</span><span class="o">=</span><span class="s2">&quot;report/calling.rst&quot;</span><span class="p">)</span>
    <span class="n">conda</span><span class="p">:</span>
        <span class="s2">&quot;envs/stats.yaml&quot;</span>
    <span class="n">script</span><span class="p">:</span>
        <span class="s2">&quot;scripts/plot-quals.py&quot;</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../executable.html" class="btn btn-neutral float-right" title="Executing Snakemake" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="additional_features.html" class="btn btn-neutral float-left" title="Additional features" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2016, Johannes Koester

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>