

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Snakemake &mdash; Snakemake 5.6.0+8.g89bc383.dirty documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/sphinx-argparse.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="安装" href="getting_started/installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> Snakemake
          

          
          </a>

          
            
            
              <div class="version">
                5.6.0+8.g89bc383.dirty
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started/installation.html">安装</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started/examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial.html">Snakemake Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/short.html">Short tutorial</a></li>
</ul>
<p class="caption"><span class="caption-text">Executing workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="executable.html">Executing Snakemake</a></li>
</ul>
<p class="caption"><span class="caption-text">Defining workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/writing_snakefiles.html">Writing Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/rules.html">Rules</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/modularization.html">Modularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/remote_files.html">Remote files</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/utils.html">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/deployment.html">Distribution and Reproducibility</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakefiles/reporting.html">Reports</a></li>
</ul>
<p class="caption"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api_reference/snakemake.html">The Snakemake API</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference/snakemake_utils.html">Additional utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference/internal/modules.html">Internal API</a></li>
</ul>
<p class="caption"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="project_info/citations.html">Citing and Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="project_info/more_resources.html">More Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="project_info/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="project_info/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="project_info/authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="project_info/history.html">Change Log</a></li>
<li class="toctree-l1"><a class="reference internal" href="project_info/license.html">License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Snakemake</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>Snakemake</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="snakemake">
<span id="manual-main"></span><h1>Snakemake<a class="headerlink" href="#snakemake" title="Permalink to this headline">¶</a></h1>
<a class="reference external image-reference" href="https://bioconda.github.io/recipes/snakemake/README.html"><img alt="https://img.shields.io/conda/dn/bioconda/snakemake.svg?label=Bioconda" src="https://img.shields.io/conda/dn/bioconda/snakemake.svg?label=Bioconda" /></a>
<a class="reference external image-reference" href="https://www.python.org"><img alt="https://img.shields.io/pypi/pyversions/snakemake.svg" src="https://img.shields.io/pypi/pyversions/snakemake.svg" /></a>
<a class="reference external image-reference" href="https://pypi.python.org/pypi/snakemake"><img alt="https://img.shields.io/pypi/v/snakemake.svg" src="https://img.shields.io/pypi/v/snakemake.svg" /></a>
<a class="reference external image-reference" href="https://quay.io/repository/snakemake/snakemake"><img alt="https://quay.io/repository/snakemake/snakemake/status" src="https://quay.io/repository/snakemake/snakemake/status" /></a>
<a class="reference external image-reference" href="https://circleci.com/bb/snakemake/snakemake/tree/master"><img alt="https://circleci.com/bb/snakemake/snakemake/tree/master.svg?style=shield" src="https://circleci.com/bb/snakemake/snakemake/tree/master.svg?style=shield" /></a>
<a class="reference external image-reference" href="http://stackoverflow.com/questions/tagged/snakemake"><img alt="https://img.shields.io/badge/stack-overflow-orange.svg" src="https://img.shields.io/badge/stack-overflow-orange.svg" /></a>
<a class="reference external image-reference" href="https://twitter.com/search?l=&amp;q=%23snakemake%20from%3Ajohanneskoester"><img alt="https://img.shields.io/twitter/follow/johanneskoester.svg?style=social&amp;label=Follow" src="https://img.shields.io/twitter/follow/johanneskoester.svg?style=social&amp;label=Follow" /></a>
<p>Snakemake工作流程管理系统是为创建 <strong>可重复和可扩展的</strong> 数据分析的工具。工作流程使用人类可读的基于Python的语言描述。 它们可以无缝扩展到服务器，集群，网格和云环境，而无需修改工作流程定义。 最后，Snakemake工作流程可能包含所需软件的说明，这些软件将自动部署到任何执行环境。</p>
<div class="section" id="manual-quick-example">
<span id="id1"></span><h2>快速上手<a class="headerlink" href="#manual-quick-example" title="Permalink to this headline">¶</a></h2>
<p>Snakemake工作流程本质上是由定义 <strong>规则</strong> 的声明性代码扩展而来的Python脚本。规则描述了如何从 <strong>输入文件</strong> 创建 <strong>输出文件</strong> 。</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">targets</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;plots/dataset1.pdf&quot;</span><span class="p">,</span>
        <span class="s2">&quot;plots/dataset2.pdf&quot;</span>

<span class="n">rule</span> <span class="n">plot</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;raw/{dataset}.csv&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;plots/{dataset}.pdf&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;somecommand {input} {output}&quot;</span>
</pre></div>
</div>
<ul class="simple">
<li>与GNU Make类似，可以根据顶部的伪规则指定目标;</li>
<li>对于每个目标文件和中间文件，可以创建规则定义如何从输入文件创建它们;</li>
<li>Snakemake通过匹配文件名来确定规则依赖;</li>
<li>输入和输出文件可以包含多个命名通配符。</li>
<li>规则可以使用shell命令，Python代码或外部Python或R脚本，从输入文件创建输出文件;</li>
<li>Snakemake工作流程可以在 <strong>工作站</strong>, <strong>集群</strong>, <strong>网格</strong> 和 <strong>云环境</strong> 而无需修改。作业调度可以限制任意资源，例如，可用的CPU内核，内存或GPU;</li>
<li>Snakemake可以使用 <a class="reference external" href="https://conda.io">Conda</a> 或 <a class="reference external" href="http://singularity.lbl.gov/">Singularity</a> 自动部署工作流城所需的软件依赖;</li>
<li>Snakemake可以使用Amazon S3，Google Storage，Dropbox，FTP，WebDAV，SFTP和iRODS访问输入或输出文件，也可通过HTTP和HTTPS访问输入文件。</li>
</ul>
</div>
<div class="section" id="main-getting-started">
<span id="id2"></span><h2>新手入门<a class="headerlink" href="#main-getting-started" title="Permalink to this headline">¶</a></h2>
<p>获得关于Snakemake初步印象，可参阅 <a class="reference external" href="https://slides.com/johanneskoester/snakemake-short">简介幻灯片</a> 或观看 <a class="reference external" href="https://youtu.be/hPrXcUUp70Y">演示视频</a> 。
有关Snakemake的新闻通过 <a class="reference external" href="https://twitter.com/search?l=&amp;q=%23snakemake%20from%3Ajohanneskoester">Twitter</a> 发布。
学习Snakemake，请参照 <a class="reference internal" href="tutorial/tutorial.html#tutorial"><span class="std std-ref">Snakemake Tutorial</span></a> ，然后参阅 <a class="reference internal" href="project_info/faq.html#project-info-faq"><span class="std std-ref">FAQ</span></a> 。</p>
</div>
<div class="section" id="main-support">
<span id="id5"></span><h2>支持<a class="headerlink" href="#main-support" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>有关版本，请参阅 <a class="reference internal" href="project_info/history.html#changelog"><span class="std std-ref">Changelog</span></a> 。</li>
<li>查看 <a class="reference internal" href="project_info/faq.html#project-info-faq"><span class="std std-ref">常见问题解答（FAQ）</span></a> 。</li>
<li>如有疑问，请发贴在 <a class="reference external" href="http://stackoverflow.com/questions/tagged/snakemake">stack overflow</a> 。</li>
<li>使用 <a class="reference external" href="https://groups.google.com/forum/#!forum/snakemake">mailing list</a> 与其他Snakemake用户讨论。 <strong>请不要在那里发布问题。使用stack overflow提问题。</strong></li>
<li>对于错误和新功能请求，请使用 <a class="reference external" href="https://bitbucket.org/snakemake/snakemake/issues">issue tracker</a> 。</li>
<li>有关贡献，请访问 <a class="reference external" href="https://bitbucket.org/snakemake/snakemake">bitbucket</a>，并阅读 <a class="reference internal" href="project_info/contributing.html#project-info-contributing"><span class="std std-ref">guidelines</span></a>。</li>
</ul>
</div>
<div class="section" id="id6">
<h2>引用<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/28/19/2520">Köster, Johannes and Rahmann, Sven. “Snakemake - A scalable bioinformatics workflow engine”. Bioinformatics 2012.</a></p>
<p>查看 <a class="reference internal" href="project_info/citations.html"><span class="doc">引用</span></a> 获取更多信息.</p>
</div>
<div class="section" id="id7">
<h2>相关资源<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><a class="reference external" href="https://snakemake-wrappers.readthedocs.org">Snakemake Wrappers Repository</a></dt><dd>Snakemake Wrapper Repository是一个可重复使用的包装器集合，可以快速使用Snakemake规则和工作流程中的流行工具。</dd>
<dt><a class="reference external" href="https://github.com/snakemake-workflows/docs">Snakemake Workflows Project</a></dt><dd>该项目提供了一系列高质量的模块化和可重复使用的工作流程。
提供的代码还应作为如何使用Snakemake构建生产工作流程的最佳实践。
邀请每位用户contribute。</dd>
<dt><a class="reference external" href="https://github.com/snakemake-profiles/doc">Snakemake Profiles Project</a></dt><dd>该项目为各种执行环境提供Snakemake配置文件。
如果找不到，请考虑contribute。</dd>
<dt><a class="reference external" href="https://bioconda.github.io/">Bioconda</a></dt><dd>通过定义使用的软件版本和提供二进制文件，Snakemake可以使用Bioconda创建完全可重现的工作流程。</dd>
</dl>
</div>
<div class="section" id="id8">
<h2>使用Snakemake的文献<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h2>
<p>下文是使用Snakemake进行分析的 <strong>不完整列表</strong>。请考虑添加自己的相应文献。</p>
<ul class="simple">
<li>Doris et al. 2018. <a class="reference external" href="https://doi.org/10.1016/j.molcel.2018.09.005">Spt6 is required for the fidelity of promoter selection</a>. Molecular Cell.</li>
<li>Karlsson et al. 2018. <a class="reference external" href="https://www.nature.com/articles/s41588-018-0131-y">Four evolutionary trajectories underlie genetic intratumoral variation in childhood cancer</a>. Nature Genetics.</li>
<li>Planchard et al. 2018. <a class="reference external" href="https://academic.oup.com/nar/advance-article/doi/10.1093/nar/gky489/5033161">The translational landscape of Arabidopsis mitochondria</a>. Nucleic acids research.</li>
<li>Schult et al. 2018. <a class="reference external" href="https://academic.oup.com/nar/article/46/14/7179/5047281">Effect of UV irradiation on Sulfolobus acidocaldarius and involvement of the general transcription factor TFB3 in the early UV response</a>. Nucleic acids research.</li>
<li>Goormaghtigh et al. 2018. <a class="reference external" href="https://mbio.asm.org/content/mbio/9/3/e00640-18.full.pdf">Reassessing the Role of Type II Toxin-Antitoxin Systems in Formation of Escherichia coli Type II Persister Cells</a>. mBio.</li>
<li>Ramirez et al. 2018. <a class="reference external" href="https://www.nature.com/articles/s41564-017-0062-x">Detecting macroecological patterns in bacterial communities across independent studies of global soils</a>. Nature microbiology.</li>
<li>Amato et al. 2018. <a class="reference external" href="https://www.nature.com/articles/s41396-018-0175-0">Evolutionary trends in host physiology outweigh dietary niche in structuring primate gut microbiomes</a>. The ISME journal.</li>
<li>Uhlitz et al. 2017. <a class="reference external" href="http://msb.embopress.org/content/13/5/928">An immediate–late gene expression module decodes ERK signal duration</a>. Molecular Systems Biology.</li>
<li>Akkouche et al. 2017. <a class="reference external" href="http://www.sciencedirect.com/science/article/pii/S1097276517302071">Piwi Is Required during Drosophila Embryogenesis to License Dual-Strand piRNA Clusters for Transposon Repression in Adult Ovaries</a>. Molecular Cell.</li>
<li>Beatty et al. 2017. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28237889">Giardia duodenalis induces pathogenic dysbiosis of human intestinal microbiota biofilms</a>. International Journal for Parasitology.</li>
<li>Meyer et al. 2017. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5400397/">Differential Gene Expression in the Human Brain Is Associated with Conserved, but Not Accelerated, Noncoding Sequences</a>. Molecular Biology and Evolution.</li>
<li>Lonardo et al. 2017. <a class="reference external" href="http://www.sciencedirect.com/science/article/pii/S0038071716304539">Priming of soil organic matter: Chemical structure of added compounds is more important than the energy content</a>. Soil Biology and Biochemistry.</li>
<li>Beisser et al. 2017. <a class="reference external" href="https://peerj.com/articles/2832/">Comprehensive transcriptome analysis provides new insights into nutritional strategies and phylogenetic relationships of chrysophytes</a>. PeerJ.</li>
<li>Piro et al 2017. <a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-017-0318-y">MetaMeta: integrating metagenome analysis tools to improve taxonomic profiling</a>. Microbiome.</li>
<li>Dimitrov et al 2017. <a class="reference external" href="https://peerj.com/articles/2915/">Successive DNA extractions improve characterization of soil microbial communities</a>. PeerJ.</li>
<li>de Bourcy et al. 2016. <a class="reference external" href="http://www.pnas.org/content/114/5/1105.short">Phylogenetic analysis of the human antibody repertoire reveals quantitative signatures of immune senescence and aging</a>. PNAS.</li>
<li>Bray et al. 2016. <a class="reference external" href="http://www.nature.com/nbt/journal/v34/n5/abs/nbt.3519.html">Near-optimal probabilistic RNA-seq quantification</a>. Nature Biotechnology.</li>
<li>Etournay et al. 2016. <a class="reference external" href="https://elifesciences.org/content/5/e14334">TissueMiner: a multiscale analysis toolkit to quantify how cellular processes create tissue dynamics</a>. eLife Sciences.</li>
<li>Townsend et al. 2016. <a class="reference external" href="http://www.cell.com/cancer-cell/abstract/S1535-6108%2816%2930090-3">The Public Repository of Xenografts Enables Discovery and Randomized Phase II-like Trials in Mice</a>. Cancer Cell.</li>
<li>Burrows et al. 2016. <a class="reference external" href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005793">Genetic Variation, Not Cell Type of Origin, Underlies the Majority of Identifiable Regulatory Differences in iPSCs</a>. PLOS Genetics.</li>
<li>Ziller et al. 2015. <a class="reference external" href="http://www.nature.com/nmeth/journal/v12/n3/full/nmeth.3152.html">Coverage recommendations for methylation analysis by whole-genome bisulfite sequencing</a>. Nature Methods.</li>
<li>Li et al. 2015. <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0843-6">Quality control, modeling, and visualization of CRISPR screens with MAGeCK-VISPR</a>. Genome Biology.</li>
<li>Schmied et al. 2015. <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/32/7/1112">An automated workflow for parallel processing of large multiview SPIM recordings</a>. Bioinformatics.</li>
<li>Chung et al. 2015. <a class="reference external" href="http://onlinelibrary.wiley.com/doi/10.1002/humu.22814/full">Whole-Genome Sequencing and Integrative Genomic Analysis Approach on Two 22q11.2 Deletion Syndrome Family Trios for Genotype to Phenotype Correlations</a>. Human Mutation.</li>
<li>Kim et al. 2015. <a class="reference external" href="http://emboj.embopress.org/content/34/13/1801.long">TUT7 controls the fate of precursor microRNAs by using three different uridylation mechanisms</a>. The EMBO Journal.</li>
<li>Park et al. 2015. <a class="reference external" href="http://doi.org/10.1016/j.cell.2015.06.007">Ebola Virus Epidemiology, Transmission, and Evolution during Seven Months in Sierra Leone</a>. Cell.</li>
<li>Břinda et al. 2015. <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/early/2015/09/30/bioinformatics.btv524">RNF: a general framework to evaluate NGS read mappers</a>. Bioinformatics.</li>
<li>Břinda et al. 2015. <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/early/2015/08/10/bioinformatics.btv419">Spaced seeds improve k-mer-based metagenomic classification</a>. Bioinformatics.</li>
<li>Spjuth et al. 2015. <a class="reference external" href="http://www.biologydirect.com/content/10/1/43">Experiences with workflows for automating data-intensive bioinformatics</a>. Biology Direct.</li>
<li>Schramm et al. 2015. <a class="reference external" href="http://www.nature.com/ng/journal/v47/n8/full/ng.3349.html">Mutational dynamics between primary and relapse neuroblastomas</a>. Nature Genetics.</li>
<li>Berulava et al. 2015. <a class="reference external" href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0118438">N6-Adenosine Methylation in MiRNAs</a>. PLOS ONE.</li>
<li>The Genome of the Netherlands Consortium 2014. <a class="reference external" href="http://www.nature.com/ng/journal/v46/n8/full/ng.3021.html">Whole-genome sequence variation, population structure and demographic history of the Dutch population</a>. Nature Genetics.</li>
<li>Patterson et al. 2014. <a class="reference external" href="http://online.liebertpub.com/doi/10.1089/cmb.2014.0157">WhatsHap: Haplotype Assembly for Future-Generation Sequencing Reads</a>. Journal of Computational Biology.</li>
<li>Fernández et al. 2014. <a class="reference external" href="http://genome.cshlp.org/content/25/1/27.long">H3K4me1 marks DNA regions hypomethylated during aging in human stem and differentiated cells</a>. Genome Research.</li>
<li>Köster et al. 2014. <a class="reference external" href="https://peerj.com/articles/606/">Massively parallel read mapping on GPUs with the q-group index and PEANUT</a>. PeerJ.</li>
<li>Chang et al. 2014. <a class="reference external" href="http://www.cell.com/molecular-cell/abstract/S1097-2765(14)00121-X">TAIL-seq: Genome-wide Determination of Poly(A) Tail Length and 3′ End Modifications</a>. Molecular Cell.</li>
<li>Althoff et al. 2013. <a class="reference external" href="http://onlinelibrary.wiley.com/doi/10.1002/ijc.28091/abstract;jsessionid=33613A834E2A2FDCCA49246C23DF777E.f04t02">MiR-137 functions as a tumor suppressor in neuroblastoma by downregulating KDM1A</a>. International Journal of Cancer.</li>
<li>Marschall et al. 2013. <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/29/24/3143.long">MATE-CLEVER: Mendelian-Inheritance-Aware Discovery and Genotyping of Midsize and Long Indels</a>. Bioinformatics.</li>
<li>Rahmann et al. 2013. <a class="reference external" href="http://www.sciencedirect.com/science/article/pii/S1046202312002605">Identifying transcriptional miRNA biomarkers by integrating high-throughput sequencing and real-time PCR data</a>. Methods.</li>
<li>Martin et al. 2013. <a class="reference external" href="http://www.nature.com/ng/journal/v45/n8/full/ng.2674.html">Exome sequencing identifies recurrent somatic mutations in EIF1AX and SF3B1 in uveal melanoma with disomy 3</a>. Nature Genetics.</li>
<li>Czeschik et al. 2013. <a class="reference external" href="http://link.springer.com/article/10.1007%2Fs00439-013-1295-2">Clinical and mutation data in 12 patients with the clinical diagnosis of Nager syndrome</a>. Human Genetics.</li>
<li>Marschall et al. 2012. <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/28/22/2875.long">CLEVER: Clique-Enumerating Variant Finder</a>. Bioinformatics.</li>
</ul>
<div class="toctree-wrapper compound" id="getting-started">
</div>
<div class="toctree-wrapper compound" id="execution">
</div>
<div class="toctree-wrapper compound" id="snakefiles">
</div>
<div class="toctree-wrapper compound" id="api-reference">
</div>
<div class="toctree-wrapper compound" id="project-info">
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="getting_started/installation.html" class="btn btn-neutral float-right" title="安装" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2016, Johannes Koester

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>