<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 生物学分析 | scRNA-seq数据分析</title>
  <meta name="description" content="8 生物学分析 | scRNA-seq数据分析" />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="8 生物学分析 | scRNA-seq数据分析" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 生物学分析 | scRNA-seq数据分析" />
  
  
  

<meta name="author" content="碱基吃柠檬" />


<meta name="date" content="2019-07-19" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="cleaning-the-expression-matrix.html">
<link rel="next" href="seurat-chapter.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.46.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.46.1/plotly-latest.min.js"></script>
<!-- for Facebook -->  
<meta property="og:url" content="https://github.com/sdbiodog" />
<meta property="og:description" content="本项目是将github上经典的scRNA-seq分析教程进行汉化，源项目参见hemberg-lab/scRNA.seq.course" />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="8 生物学分析 | scRNA-seq数据分析" />
<meta name="twitter:description" content="本项目是将github上经典的scRNA-seq分析教程进行汉化，源项目参见hemberg-lab/scRNA.seq.course" />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-71525309-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> 说在前面的话</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#section-1"><i class="fa fa-check"></i><b>1.1</b> 关于本项目</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#section-2"><i class="fa fa-check"></i><b>1.2</b> 视频</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#docker-"><i class="fa fa-check"></i><b>1.4</b> Docker 镜像</a><ul>
<li class="chapter" data-level="1.4.1" data-path="index.html"><a href="index.html#section-3"><i class="fa fa-check"></i><b>1.4.1</b> 运行镜像</a></li>
<li class="chapter" data-level="1.4.2" data-path="index.html"><a href="index.html#section-4"><i class="fa fa-check"></i><b>1.4.2</b> 下载数据或其它文件</a></li>
<li class="chapter" data-level="1.4.3" data-path="index.html"><a href="index.html#rstudio"><i class="fa fa-check"></i><b>1.4.3</b> RStudio</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#section-5"><i class="fa fa-check"></i><b>1.5</b> 手动安装</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#section-6"><i class="fa fa-check"></i><b>1.6</b> 许可</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#section-7"><i class="fa fa-check"></i><b>1.7</b> 准备知识</a></li>
<li class="chapter" data-level="1.8" data-path="index.html"><a href="index.html#section-8"><i class="fa fa-check"></i><b>1.8</b> 联系我们</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> scRNA-seq介绍</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-9"><i class="fa fa-check"></i><b>2.3</b> 工作流程</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-10"><i class="fa fa-check"></i><b>2.4</b> 计算分析</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-11"><i class="fa fa-check"></i><b>2.5</b> 挑战</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-12"><i class="fa fa-check"></i><b>2.6</b> 实验方法</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-13"><i class="fa fa-check"></i><b>2.7</b> 如何选择合适的平台</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html"><i class="fa fa-check"></i><b>3</b> scRNA-seq原始数据处理</a><ul>
<li class="chapter" data-level="3.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#fastqc"><i class="fa fa-check"></i><b>3.1</b> FastQC</a><ul>
<li class="chapter" data-level="3.1.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-14"><i class="fa fa-check"></i><b>3.1.1</b> 解决方案并下载报告</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-15"><i class="fa fa-check"></i><b>3.2</b> 移除接头和低质量碱基</a><ul>
<li class="chapter" data-level="3.2.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#solution"><i class="fa fa-check"></i><b>3.2.1</b> Solution</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-16"><i class="fa fa-check"></i><b>3.3</b> 文件格式</a><ul>
<li class="chapter" data-level="3.3.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#fastq"><i class="fa fa-check"></i><b>3.3.1</b> FastQ</a></li>
<li class="chapter" data-level="3.3.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#bam"><i class="fa fa-check"></i><b>3.3.2</b> BAM</a></li>
<li class="chapter" data-level="3.3.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#cram"><i class="fa fa-check"></i><b>3.3.3</b> CRAM</a></li>
<li class="chapter" data-level="3.3.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-17"><i class="fa fa-check"></i><b>3.3.4</b> 手动查看文件</a></li>
<li class="chapter" data-level="3.3.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#genome-fasta-gtf"><i class="fa fa-check"></i><b>3.3.5</b> Genome (FASTA, GTF)</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-18"><i class="fa fa-check"></i><b>3.4</b> 测序文库拆分</a><ul>
<li class="chapter" data-level="3.4.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-19"><i class="fa fa-check"></i><b>3.4.1</b> 鉴定含有细胞的液滴/微孔</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#starread"><i class="fa fa-check"></i><b>3.5</b> 使用STAR比对read</a><ul>
<li class="chapter" data-level="3.5.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#star"><i class="fa fa-check"></i><b>3.5.1</b> STAR比对命令</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallistopseudo-alignment"><i class="fa fa-check"></i><b>3.6</b> Kallisto和Pseudo-Alignment</a><ul>
<li class="chapter" data-level="3.6.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#k-mer"><i class="fa fa-check"></i><b>3.6.1</b> k-mer是什么?</a></li>
<li class="chapter" data-level="3.6.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#k-merreads"><i class="fa fa-check"></i><b>3.6.2</b> 为什么比对k-mer而不是reads?</a></li>
<li class="chapter" data-level="3.6.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto"><i class="fa fa-check"></i><b>3.6.3</b> Kallisto伪比对模式</a></li>
<li class="chapter" data-level="3.6.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto-pseudo-alignment"><i class="fa fa-check"></i><b>3.6.4</b> Kallisto Pseudo-Alignment命令</a></li>
<li class="chapter" data-level="3.6.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto-1"><i class="fa fa-check"></i><b>3.6.5</b> 理解Kallisto输出结果</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>4</b> 构建表达矩阵</a><ul>
<li class="chapter" data-level="4.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-20"><i class="fa fa-check"></i><b>4.1</b> 质量控制</a></li>
<li class="chapter" data-level="4.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-"><i class="fa fa-check"></i><b>4.2</b> Reads 比对</a></li>
<li class="chapter" data-level="4.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-21"><i class="fa fa-check"></i><b>4.3</b> 比对示例</a></li>
<li class="chapter" data-level="4.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-qc"><i class="fa fa-check"></i><b>4.4</b> Mapping QC</a></li>
<li class="chapter" data-level="4.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads"><i class="fa fa-check"></i><b>4.5</b> Reads定量</a></li>
<li class="chapter" data-level="4.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umichapter"><i class="fa fa-check"></i><b>4.6</b> 唯一分子标识符</a><ul>
<li class="chapter" data-level="4.6.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umi"><i class="fa fa-check"></i><b>4.6.1</b> UMI介绍</a></li>
<li class="chapter" data-level="4.6.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-22"><i class="fa fa-check"></i><b>4.6.2</b> 比对条形码序列</a></li>
<li class="chapter" data-level="4.6.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#barcodes"><i class="fa fa-check"></i><b>4.6.3</b> Barcodes计数</a></li>
<li class="chapter" data-level="4.6.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-23"><i class="fa fa-check"></i><b>4.6.4</b> 错误校正</a></li>
<li class="chapter" data-level="4.6.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-24"><i class="fa fa-check"></i><b>4.6.5</b> 下游分析</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html"><i class="fa fa-check"></i><b>5</b> R/Bioconductor介绍</a><ul>
<li class="chapter" data-level="5.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#r"><i class="fa fa-check"></i><b>5.1</b> 安装R包</a><ul>
<li class="chapter" data-level="5.1.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#cran"><i class="fa fa-check"></i><b>5.1.1</b> CRAN</a></li>
<li class="chapter" data-level="5.1.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#github-1"><i class="fa fa-check"></i><b>5.1.2</b> Github</a></li>
<li class="chapter" data-level="5.1.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor"><i class="fa fa-check"></i><b>5.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="5.1.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-25"><i class="fa fa-check"></i><b>5.1.4</b> 源码安装</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-26"><i class="fa fa-check"></i><b>5.2</b> 安装说明:</a></li>
<li class="chapter" data-level="5.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-27"><i class="fa fa-check"></i><b>5.3</b> 数据类型/类</a><ul>
<li class="chapter" data-level="5.3.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-28"><i class="fa fa-check"></i><b>5.3.1</b> 数值</a></li>
<li class="chapter" data-level="5.3.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-29"><i class="fa fa-check"></i><b>5.3.2</b> 字符/字符串</a></li>
<li class="chapter" data-level="5.3.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-30"><i class="fa fa-check"></i><b>5.3.3</b> 逻辑值</a></li>
<li class="chapter" data-level="5.3.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-31"><i class="fa fa-check"></i><b>5.3.4</b> 因子</a></li>
<li class="chapter" data-level="5.3.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-32"><i class="fa fa-check"></i><b>5.3.5</b> 检查数据类型</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-33"><i class="fa fa-check"></i><b>5.4</b> 基本数据结构</a></li>
<li class="chapter" data-level="5.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-34"><i class="fa fa-check"></i><b>5.5</b> 更多信息</a></li>
<li class="chapter" data-level="5.6" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-35"><i class="fa fa-check"></i><b>5.6</b> 数据类型</a><ul>
<li class="chapter" data-level="5.6.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-36"><i class="fa fa-check"></i><b>5.6.1</b> 什么是整洁的数据</a></li>
<li class="chapter" data-level="5.6.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#rich-data"><i class="fa fa-check"></i><b>5.6.2</b> 什么是Rich data?</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor-singlecellexperiment--scater"><i class="fa fa-check"></i><b>5.7</b> Bioconductor, <code>SingleCellExperiment</code> 和 <code>scater</code></a><ul>
<li class="chapter" data-level="5.7.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor-1"><i class="fa fa-check"></i><b>5.7.1</b> Bioconductor</a></li>
<li class="chapter" data-level="5.7.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#singlecellexperiment-"><i class="fa fa-check"></i><b>5.7.2</b> <code>SingleCellExperiment</code> 类</a></li>
<li class="chapter" data-level="5.7.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#scater-"><i class="fa fa-check"></i><b>5.7.3</b> <code>scater</code> 包</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2"><i class="fa fa-check"></i><b>5.8</b> ggplot2介绍</a><ul>
<li class="chapter" data-level="5.8.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2-1"><i class="fa fa-check"></i><b>5.8.1</b> ggplot2是什么?</a></li>
<li class="chapter" data-level="5.8.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2j"><i class="fa fa-check"></i><b>5.8.2</b> ggplot2j原则</a></li>
<li class="chapter" data-level="5.8.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#aes"><i class="fa fa-check"></i><b>5.8.3</b> 使用<code>aes</code>映射函数</a></li>
<li class="chapter" data-level="5.8.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#geoms"><i class="fa fa-check"></i><b>5.8.4</b> Geoms</a></li>
<li class="chapter" data-level="5.8.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-37"><i class="fa fa-check"></i><b>5.8.5</b> 绘制2个以上细胞的数据</a></li>
<li class="chapter" data-level="5.8.6" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-38"><i class="fa fa-check"></i><b>5.8.6</b> 绘制热图</a></li>
<li class="chapter" data-level="5.8.7" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-39"><i class="fa fa-check"></i><b>5.8.7</b> 主成分分析</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tabula-muris.html"><a href="tabula-muris.html"><i class="fa fa-check"></i><b>6</b> Tabula Muris</a><ul>
<li class="chapter" data-level="6.1" data-path="tabula-muris.html"><a href="tabula-muris.html#section-40"><i class="fa fa-check"></i><b>6.1</b> 介绍</a></li>
<li class="chapter" data-level="6.2" data-path="tabula-muris.html"><a href="tabula-muris.html#section-41"><i class="fa fa-check"></i><b>6.2</b> 下载数据</a></li>
<li class="chapter" data-level="6.3" data-path="tabula-muris.html"><a href="tabula-muris.html#smartseq2"><i class="fa fa-check"></i><b>6.3</b> 读取Smartseq2数据</a></li>
<li class="chapter" data-level="6.4" data-path="tabula-muris.html"><a href="tabula-muris.html#scater"><i class="fa fa-check"></i><b>6.4</b> 构建scater对象</a></li>
<li class="chapter" data-level="6.5" data-path="tabula-muris.html"><a href="tabula-muris.html#x"><i class="fa fa-check"></i><b>6.5</b> 读取10X数据</a></li>
<li class="chapter" data-level="6.6" data-path="tabula-muris.html"><a href="tabula-muris.html#scater-1"><i class="fa fa-check"></i><b>6.6</b> 创建scater对象</a></li>
<li class="chapter" data-level="6.7" data-path="tabula-muris.html"><a href="tabula-muris.html#section-42"><i class="fa fa-check"></i><b>6.7</b> 进阶练习</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html"><i class="fa fa-check"></i><b>7</b> 表达矩阵质控</a><ul>
<li class="chapter" data-level="7.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#exprs-qc"><i class="fa fa-check"></i><b>7.1</b> UMI表达质控</a><ul>
<li class="chapter" data-level="7.1.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-43"><i class="fa fa-check"></i><b>7.1.1</b> 介绍</a></li>
<li class="chapter" data-level="7.1.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tung"><i class="fa fa-check"></i><b>7.1.2</b> Tung数据集</a></li>
<li class="chapter" data-level="7.1.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-44"><i class="fa fa-check"></i><b>7.1.3</b> 细胞质控</a></li>
<li class="chapter" data-level="7.1.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-47"><i class="fa fa-check"></i><b>7.1.4</b> 细胞过滤</a></li>
<li class="chapter" data-level="7.1.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-50"><i class="fa fa-check"></i><b>7.1.5</b> 手工过滤和自动过滤比较</a></li>
<li class="chapter" data-level="7.1.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-51"><i class="fa fa-check"></i><b>7.1.6</b> 基因分析</a></li>
<li class="chapter" data-level="7.1.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-54"><i class="fa fa-check"></i><b>7.1.7</b> 保存数据</a></li>
<li class="chapter" data-level="7.1.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-55"><i class="fa fa-check"></i><b>7.1.8</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-1"><i class="fa fa-check"></i><b>7.2</b> Reads表达质控</a></li>
<li class="chapter" data-level="7.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-56"><i class="fa fa-check"></i><b>7.3</b> 数据可视化</a><ul>
<li class="chapter" data-level="7.3.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-57"><i class="fa fa-check"></i><b>7.3.1</b> 介绍</a></li>
<li class="chapter" data-level="7.3.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#visual-pca"><i class="fa fa-check"></i><b>7.3.2</b> PCA plot</a></li>
<li class="chapter" data-level="7.3.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#visual-tsne"><i class="fa fa-check"></i><b>7.3.3</b> tSNE可视化</a></li>
<li class="chapter" data-level="7.3.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-61"><i class="fa fa-check"></i><b>7.3.4</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-2"><i class="fa fa-check"></i><b>7.4</b> Reads数据可视化</a></li>
<li class="chapter" data-level="7.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-62"><i class="fa fa-check"></i><b>7.5</b> 识别混淆因子</a><ul>
<li class="chapter" data-level="7.5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-63"><i class="fa fa-check"></i><b>7.5.1</b> 介绍</a></li>
<li class="chapter" data-level="7.5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-64"><i class="fa fa-check"></i><b>7.5.2</b> 和组成分的相关性</a></li>
<li class="chapter" data-level="7.5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-66"><i class="fa fa-check"></i><b>7.5.3</b> 解释变量</a></li>
<li class="chapter" data-level="7.5.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-67"><i class="fa fa-check"></i><b>7.5.4</b> 其它混淆因子</a></li>
<li class="chapter" data-level="7.5.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-68"><i class="fa fa-check"></i><b>7.5.5</b> 练习</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-3"><i class="fa fa-check"></i><b>7.6</b> 识别READS中混淆因子</a></li>
<li class="chapter" data-level="7.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-69"><i class="fa fa-check"></i><b>7.7</b> 标准化理论</a><ul>
<li class="chapter" data-level="7.7.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-70"><i class="fa fa-check"></i><b>7.7.1</b> 介绍</a></li>
<li class="chapter" data-level="7.7.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-71"><i class="fa fa-check"></i><b>7.7.2</b> 文库大小</a></li>
<li class="chapter" data-level="7.7.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-72"><i class="fa fa-check"></i><b>7.7.3</b> 标准化</a></li>
<li class="chapter" data-level="7.7.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#effectiveness"><i class="fa fa-check"></i><b>7.7.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#umi-1"><i class="fa fa-check"></i><b>7.8</b> UMI标准化练习</a><ul>
<li class="chapter" data-level="7.8.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#raw"><i class="fa fa-check"></i><b>7.8.1</b> Raw</a></li>
<li class="chapter" data-level="7.8.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cpm-1"><i class="fa fa-check"></i><b>7.8.2</b> CPM</a></li>
<li class="chapter" data-level="7.8.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#scran-1"><i class="fa fa-check"></i><b>7.8.3</b> scran</a></li>
<li class="chapter" data-level="7.8.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-73"><i class="fa fa-check"></i><b>7.8.4</b> 下采样</a></li>
<li class="chapter" data-level="7.8.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-74"><i class="fa fa-check"></i><b>7.8.5</b> 基因/转录本长度标准化</a></li>
<li class="chapter" data-level="7.8.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-75"><i class="fa fa-check"></i><b>7.8.6</b> 练习</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-4"><i class="fa fa-check"></i><b>7.9</b> Reads标准化练习</a></li>
<li class="chapter" data-level="7.10" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>7.10</b> 处理混淆因子</a><ul>
<li class="chapter" data-level="7.10.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-76"><i class="fa fa-check"></i><b>7.10.1</b> 介绍</a></li>
<li class="chapter" data-level="7.10.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-77"><i class="fa fa-check"></i><b>7.10.2</b> 移除不需要的变异</a></li>
<li class="chapter" data-level="7.10.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#combat"><i class="fa fa-check"></i><b>7.10.3</b> Combat</a></li>
<li class="chapter" data-level="7.10.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>7.10.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="7.10.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#glm"><i class="fa fa-check"></i><b>7.10.5</b> GLM</a></li>
<li class="chapter" data-level="7.10.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#harmony"><i class="fa fa-check"></i><b>7.10.6</b> Harmony</a></li>
<li class="chapter" data-level="7.10.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-78"><i class="fa fa-check"></i><b>7.10.7</b> 评价和比较混淆因子去除策略</a></li>
<li class="chapter" data-level="7.10.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-82"><i class="fa fa-check"></i><b>7.10.8</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.11" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-5"><i class="fa fa-check"></i><b>7.11</b> 处理Reads中混淆因子</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>8</b> 生物学分析</a><ul>
<li class="chapter" data-level="8.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-83"><i class="fa fa-check"></i><b>8.1</b> 聚类介绍</a><ul>
<li class="chapter" data-level="8.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-84"><i class="fa fa-check"></i><b>8.1.1</b> 介绍</a></li>
<li class="chapter" data-level="8.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#section-85"><i class="fa fa-check"></i><b>8.1.2</b> 降维</a></li>
<li class="chapter" data-level="8.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-86"><i class="fa fa-check"></i><b>8.1.3</b> 聚类方法</a></li>
<li class="chapter" data-level="8.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#section-89"><i class="fa fa-check"></i><b>8.1.4</b> 聚类的挑战</a></li>
<li class="chapter" data-level="8.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#scrna-seq-1"><i class="fa fa-check"></i><b>8.1.5</b> scRNA-seq数据的工具</a></li>
<li class="chapter" data-level="8.1.6" data-path="biological-analysis.html"><a href="biological-analysis.html#section-90"><i class="fa fa-check"></i><b>8.1.6</b> 比较聚类结果</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clust-methods"><i class="fa fa-check"></i><b>8.2</b> 聚类示例</a><ul>
<li class="chapter" data-level="8.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-"><i class="fa fa-check"></i><b>8.2.1</b> Deng 数据集</a></li>
<li class="chapter" data-level="8.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>8.2.2</b> SC3</a></li>
<li class="chapter" data-level="8.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>8.2.3</b> tSNE + kmeans</a></li>
<li class="chapter" data-level="8.2.4" data-path="biological-analysis.html"><a href="biological-analysis.html#sincera-1"><i class="fa fa-check"></i><b>8.2.4</b> SINCERA</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-91"><i class="fa fa-check"></i><b>8.3</b> 特征选择</a><ul>
<li class="chapter" data-level="8.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#null"><i class="fa fa-check"></i><b>8.3.1</b> 识别基因与null模型</a></li>
<li class="chapter" data-level="8.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>8.3.2</b> Correlated Expression</a></li>
<li class="chapter" data-level="8.3.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-92"><i class="fa fa-check"></i><b>8.3.3</b> 比较不同的方法</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="biological-analysis.html"><a href="biological-analysis.html#section-93"><i class="fa fa-check"></i><b>8.4</b> 伪时间分析</a><ul>
<li class="chapter" data-level="8.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng"><i class="fa fa-check"></i><b>8.4.1</b> Deng数据初探</a></li>
<li class="chapter" data-level="8.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#tscan"><i class="fa fa-check"></i><b>8.4.2</b> TSCAN</a></li>
<li class="chapter" data-level="8.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>8.4.3</b> monocle</a></li>
<li class="chapter" data-level="8.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#diffusion-maps"><i class="fa fa-check"></i><b>8.4.4</b> Diffusion maps</a></li>
<li class="chapter" data-level="8.4.5" data-path="biological-analysis.html"><a href="biological-analysis.html#slicer"><i class="fa fa-check"></i><b>8.4.5</b> SLICER</a></li>
<li class="chapter" data-level="8.4.6" data-path="biological-analysis.html"><a href="biological-analysis.html#ouija"><i class="fa fa-check"></i><b>8.4.6</b> Ouija</a></li>
<li class="chapter" data-level="8.4.7" data-path="biological-analysis.html"><a href="biological-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>8.4.7</b> 方法比较</a></li>
<li class="chapter" data-level="8.4.8" data-path="biological-analysis.html"><a href="biological-analysis.html#expression-of-genes-through-time"><i class="fa fa-check"></i><b>8.4.8</b> Expression of genes through time</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="biological-analysis.html"><a href="biological-analysis.html#imputation"><i class="fa fa-check"></i><b>8.5</b> Imputation</a><ul>
<li class="chapter" data-level="8.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#scimpute"><i class="fa fa-check"></i><b>8.5.1</b> scImpute</a></li>
<li class="chapter" data-level="8.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#drimpute"><i class="fa fa-check"></i><b>8.5.2</b> DrImpute</a></li>
<li class="chapter" data-level="8.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#magic"><i class="fa fa-check"></i><b>8.5.3</b> MAGIC</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="biological-analysis.html"><a href="biological-analysis.html#dechapter"><i class="fa fa-check"></i><b>8.6</b> 差异表达分析</a><ul>
<li class="chapter" data-level="8.6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>8.6.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="8.6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>8.6.2</b> Single cell RNA-seq</a></li>
<li class="chapter" data-level="8.6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#Differences-in-Distribution"><i class="fa fa-check"></i><b>8.6.3</b> 分布的差异</a></li>
<li class="chapter" data-level="8.6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#Models-of-single-cell-RNASeq-data"><i class="fa fa-check"></i><b>8.6.4</b> scRNA-seq数据模型</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="biological-analysis.html"><a href="biological-analysis.html#DE-in-a-real-dataset"><i class="fa fa-check"></i><b>8.7</b> 差异表达分析实战</a><ul>
<li class="chapter" data-level="8.7.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-94"><i class="fa fa-check"></i><b>8.7.1</b> 介绍</a></li>
<li class="chapter" data-level="8.7.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>8.7.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="8.7.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>8.7.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="8.7.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger"><i class="fa fa-check"></i><b>8.7.4</b> edgeR</a></li>
<li class="chapter" data-level="8.7.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle-1"><i class="fa fa-check"></i><b>8.7.5</b> Monocle</a></li>
<li class="chapter" data-level="8.7.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>8.7.6</b> MAST</a></li>
<li class="chapter" data-level="8.7.7" data-path="biological-analysis.html"><a href="biological-analysis.html#slow-method1h"><i class="fa fa-check"></i><b>8.7.7</b> Slow method(&gt;1h)</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="biological-analysis.html"><a href="biological-analysis.html#scrna-seq-2"><i class="fa fa-check"></i><b>8.8</b> 比较/组合scRNA-seq数据集</a><ul>
<li class="chapter" data-level="8.8.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-96"><i class="fa fa-check"></i><b>8.8.1</b> 介绍</a></li>
<li class="chapter" data-level="8.8.2" data-path="biological-analysis.html"><a href="biological-analysis.html#datasets"><i class="fa fa-check"></i><b>8.8.2</b> Datasets</a></li>
<li class="chapter" data-level="8.8.3" data-path="biological-analysis.html"><a href="biological-analysis.html#scmap"><i class="fa fa-check"></i><b>8.8.3</b> scmap</a></li>
<li class="chapter" data-level="8.8.4" data-path="biological-analysis.html"><a href="biological-analysis.html#section-99"><i class="fa fa-check"></i><b>8.8.4</b> 细胞间映射</a></li>
<li class="chapter" data-level="8.8.5" data-path="biological-analysis.html"><a href="biological-analysis.html#metaneighbour"><i class="fa fa-check"></i><b>8.8.5</b> Metaneighbour</a></li>
<li class="chapter" data-level="8.8.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mnncorrect-1"><i class="fa fa-check"></i><b>8.8.6</b> mnnCorrect</a></li>
<li class="chapter" data-level="8.8.7" data-path="biological-analysis.html"><a href="biological-analysis.html#seurat"><i class="fa fa-check"></i><b>8.8.7</b> 典型相关分析 (Seurat)</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="biological-analysis.html"><a href="biological-analysis.html#scrna-seq-3"><i class="fa fa-check"></i><b>8.9</b> 检索scRNA-seq数据</a><ul>
<li class="chapter" data-level="8.9.1" data-path="biological-analysis.html"><a href="biological-analysis.html#about"><i class="fa fa-check"></i><b>8.9.1</b> About</a></li>
<li class="chapter" data-level="8.9.2" data-path="biological-analysis.html"><a href="biological-analysis.html#section-101"><i class="fa fa-check"></i><b>8.9.2</b> 数据集</a></li>
<li class="chapter" data-level="8.9.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-102"><i class="fa fa-check"></i><b>8.9.3</b> 基因索引</a></li>
<li class="chapter" data-level="8.9.4" data-path="biological-analysis.html"><a href="biological-analysis.html#marker"><i class="fa fa-check"></i><b>8.9.4</b> Marker基因</a></li>
<li class="chapter" data-level="8.9.5" data-path="biological-analysis.html"><a href="biological-analysis.html#section-103"><i class="fa fa-check"></i><b>8.9.5</b> 搜索基因列表</a></li>
<li class="chapter" data-level="8.9.6" data-path="biological-analysis.html"><a href="biological-analysis.html#in-silico-gating"><i class="fa fa-check"></i><b>8.9.6</b> In-silico gating</a></li>
<li class="chapter" data-level="8.9.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo"><i class="fa fa-check"></i><b>8.9.7</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="seurat-chapter.html"><a href="seurat-chapter.html"><i class="fa fa-check"></i><b>9</b> Seurat</a><ul>
<li class="chapter" data-level="9.1" data-path="seurat-chapter.html"><a href="seurat-chapter.html#setup-the-seurat-object"><i class="fa fa-check"></i><b>9.1</b> Setup the Seurat Object</a></li>
<li class="chapter" data-level="9.2" data-path="seurat-chapter.html"><a href="seurat-chapter.html#standard-pre-processing-workflow"><i class="fa fa-check"></i><b>9.2</b> Standard pre-processing workflow</a></li>
<li class="chapter" data-level="9.3" data-path="seurat-chapter.html"><a href="seurat-chapter.html#qc-and-selecting-cells-for-further-analysis"><i class="fa fa-check"></i><b>9.3</b> QC and selecting cells for further analysis</a></li>
<li class="chapter" data-level="9.4" data-path="seurat-chapter.html"><a href="seurat-chapter.html#normalizing-the-data"><i class="fa fa-check"></i><b>9.4</b> Normalizing the data</a></li>
<li class="chapter" data-level="9.5" data-path="seurat-chapter.html"><a href="seurat-chapter.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>9.5</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="9.6" data-path="seurat-chapter.html"><a href="seurat-chapter.html#scaling-the-data-and-removing-unwanted-sources-of-variation"><i class="fa fa-check"></i><b>9.6</b> Scaling the data and removing unwanted sources of variation</a></li>
<li class="chapter" data-level="9.7" data-path="seurat-chapter.html"><a href="seurat-chapter.html#perform-linear-dimensional-reduction"><i class="fa fa-check"></i><b>9.7</b> Perform linear dimensional reduction</a></li>
<li class="chapter" data-level="9.8" data-path="seurat-chapter.html"><a href="seurat-chapter.html#determine-statistically-significant-principal-components"><i class="fa fa-check"></i><b>9.8</b> Determine statistically significant principal components</a></li>
<li class="chapter" data-level="9.9" data-path="seurat-chapter.html"><a href="seurat-chapter.html#cluster-the-cells"><i class="fa fa-check"></i><b>9.9</b> Cluster the cells</a></li>
<li class="chapter" data-level="9.10" data-path="seurat-chapter.html"><a href="seurat-chapter.html#run-non-linear-dimensional-reduction-tsne"><i class="fa fa-check"></i><b>9.10</b> Run Non-linear dimensional reduction (tSNE)</a></li>
<li class="chapter" data-level="9.11" data-path="seurat-chapter.html"><a href="seurat-chapter.html#run-umap"><i class="fa fa-check"></i><b>9.11</b> Run UMAP</a></li>
<li class="chapter" data-level="9.12" data-path="seurat-chapter.html"><a href="seurat-chapter.html#finding-differentially-expressed-genes-cluster-biomarkers"><i class="fa fa-check"></i><b>9.12</b> Finding differentially expressed genes (cluster biomarkers)</a></li>
<li class="chapter" data-level="9.13" data-path="seurat-chapter.html"><a href="seurat-chapter.html#assigning-cell-type-identity-to-clusters"><i class="fa fa-check"></i><b>9.13</b> Assigning cell type identity to clusters</a></li>
<li class="chapter" data-level="9.14" data-path="seurat-chapter.html"><a href="seurat-chapter.html#further-subdivisions-within-cell-types"><i class="fa fa-check"></i><b>9.14</b> Further subdivisions within cell types</a></li>
<li class="chapter" data-level="9.15" data-path="seurat-chapter.html"><a href="seurat-chapter.html#sessioninfo-1"><i class="fa fa-check"></i><b>9.15</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>10</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/lemonbases/scRNA.seq.course.CN" target="blank">碱基吃柠檬</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNA-seq数据分析</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biological-analysis" class="section level1">
<h1><span class="header-section-number">8</span> 生物学分析</h1>
<div id="section-83" class="section level2">
<h2><span class="header-section-number">8.1</span> 聚类介绍</h2>
<p>对数据进行了标准化并移除了混淆因子后，就可以进行与手头的生物学问题相关的分析。分析的性质取决于数据集。 然而，一些方面在广泛的背景下有用，我们将在接下来的几章中讨论其中的一些方面。从scRNA-seq数据的聚类开始。</p>
<div id="section-84" class="section level3">
<h3><span class="header-section-number">8.1.1</span> 介绍</h3>
<p>scRNA-seq最有希望的应用之一是基于转录谱的<em>de novo</em>发现和注释细胞类型。在计算上，这是一个难题，因为它相当于<strong>非监督聚类</strong>。 也就是说，我们需要根据转录组的相似性识别细胞群，不需要对细胞类型的任何先验知识。 此外，在大多数情况下，我们甚至不知道<em>先验</em>集群的数量。由于噪音（技术和生物）和高维度（比如基因），问题变得更具挑战性。</p>
</div>
<div id="section-85" class="section level3">
<h3><span class="header-section-number">8.1.2</span> 降维</h3>
<p>在处理大型数据集时，应用某种降维方法是有帮助的。将数据投影到较低维度的子空间上，能够显著减少噪声。另一个好处是，在2维或3维子空间中可视化数据要容易得多。我们已经讨论了PCA(章节 <a href="cleaning-the-expression-matrix.html#visual-pca">7.3.2</a>)和t-SNE(章节 <a href="cleaning-the-expression-matrix.html#visual-pca">7.3.2</a>)。</p>
</div>
<div id="section-86" class="section level3">
<h3><span class="header-section-number">8.1.3</span> 聚类方法</h3>
<p><strong>非监督聚类</strong> 在机器学习中广泛研究并应用于不同领域。常用的流行方法是：<strong>层次聚类</strong>，<strong>k-means聚类</strong>，<strong>基于的图聚类</strong>。</p>
<div id="section-87" class="section level4">
<h4><span class="header-section-number">8.1.3.1</span> 层次聚类</h4>
<p><a href="https://en.wikipedia.org/wiki/Hierarchical_clustering">层次聚类</a>可以使用自下而上或自上而下的方法。自下而上策略中，每个细胞最初单独为一类，随后将类进行合并形成hieararchy：</p>
<div class="figure" style="text-align: center"><span id="fig:clust-hierarch-raw"></span>
<img src="figures/hierarchical_clustering1.png" alt="原始数据" width="30%" />
<p class="caption">
Figure 8.1: 原始数据
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:clust-hierarch-dendr"></span>
<img src="figures/hierarchical_clustering2.png" alt="层次聚类树状图" width="50%" />
<p class="caption">
Figure 8.2: 层次聚类树状图
</p>
</div>
<p>使用自上而下策略，最初所有观测点为一类，递归对类进行拆分拆分形成hierarchy。该策略的一个优点是该方法是确定性的。</p>
</div>
<div id="k-means" class="section level4">
<h4><span class="header-section-number">8.1.3.2</span> k-means</h4>
<p><a href="https://en.wikipedia.org/wiki/K-means_clustering"><em>k</em>-means聚类</a>,将 _N_个细胞划分成 _k_个不同的类。以迭代的方式，分配类中心并将每个细胞分配个最近的类。</p>
<div class="figure" style="text-align: center"><span id="fig:clust-k-means"></span>
<img src="figures/k-means.png" alt="Schematic representation of the k-means clustering" width="100%" />
<p class="caption">
Figure 8.3: Schematic representation of the k-means clustering
</p>
</div>
<p>scRNA-seq分析的大多数方法在某些步骤都包括 <em>k</em>-means环节。</p>
</div>
<div id="section-88" class="section level4">
<h4><span class="header-section-number">8.1.3.3</span> 基于图的方法</h4>
<p>在过去的二十年中，人们对分析各个领域的网络产生了很大的兴趣。 一个目标是识别网络中节点的群体或模块。</p>
<div class="figure" style="text-align: center"><span id="fig:clust-graph"></span>
<img src="figures/graph_network.jpg" alt="Schematic representation of the graph network" width="100%" />
<p class="caption">
Figure 8.4: Schematic representation of the graph network
</p>
</div>
<p>通过构建图，其中每个节点代表一个细胞,一些方法可以应用于scRNA-seq数据。 然而，构建图并为边指定权重并非易事。 基于图的方法的一个优点是其中一些方法非常有效，可以应用于包含数百万节点的网络。</p>
</div>
</div>
<div id="section-89" class="section level3">
<h3><span class="header-section-number">8.1.4</span> 聚类的挑战</h3>
<ul>
<li>类个数 _k_的确定？</li>
<li>什么是细胞类型？</li>
<li><strong>可拓展性</strong>：在过去的几年中，scRNA-seq实验的细胞数量增长了几个数量级，从~<span class="math inline">\(10^2\)</span> 到 ~<span class="math inline">\(10^6\)</span></li>
<li>工具不太友好</li>
</ul>
</div>
<div id="scrna-seq-1" class="section level3">
<h3><span class="header-section-number">8.1.5</span> scRNA-seq数据的工具</h3>
<div id="sincera" class="section level4">
<h4><span class="header-section-number">8.1.5.1</span> <a href="https://research.cchmc.org/pbge/sincera.html">SINCERA</a></h4>
<ul>
<li>SINCERA <span class="citation">(Guo et al. <a href="#ref-Guo2015-ok">2015</a>)</span> 基于层次聚类</li>
<li>数据在聚类前转为 <em>z</em>-scores</li>
<li>通过寻找hierarchy中第一个单例群识别 <em>k</em></li>
</ul>
</div>
<div id="sc3" class="section level4">
<h4><span class="header-section-number">8.1.5.2</span> <a href="http://bioconductor.org/packages/SC3/">SC3</a></h4>
<div class="figure" style="text-align: center"><span id="fig:clust-sc3"></span>
<img src="figures/sc3.png" alt="SC3 pipeline" width="100%" />
<p class="caption">
Figure 8.5: SC3 pipeline
</p>
</div>
<ul>
<li>SC3 <span class="citation">(Kiselev et al. <a href="#ref-Kiselev2016-bq">2017</a>)</span> 基于PCA和谱降维</li>
<li>使用 <em>k</em>-means</li>
<li>使用一致性聚类(consensus clustering)</li>
</ul>
</div>
<div id="tsne-k-means" class="section level4">
<h4><span class="header-section-number">8.1.5.3</span> tSNE + k-means</h4>
<ul>
<li>基于 __tSNE__映射</li>
<li>使用 <em>k</em>-means</li>
</ul>
</div>
<div id="seurat-clustering" class="section level4">
<h4><span class="header-section-number">8.1.5.4</span> Seurat clustering</h4>
<p><a href="https://github.com/satijalab/seurat"><code>Seurat</code></a>聚类基于类似于<code>SNN-Cliq</code>的 _community detection_方法，以及之前提出的用于分析CyTOF数据<span class="citation">(Levine et al. <a href="#ref-Levine2015-fk">2015</a>)</span>的方法。由于<code>Seurat</code>已经成为scRNA-seq数据分析的一体化工具，我们专门用一章来更详细地讨论它(<a href="mailto:章节@@ref" class="email">章节@@ref</a>(seurat-chapter))。</p>
</div>
</div>
<div id="section-90" class="section level3">
<h3><span class="header-section-number">8.1.6</span> 比较聚类结果</h3>
<p>可以使用<a href="https://en.wikipedia.org/wiki/Rand_index">adjusted Rand index</a>比较聚类效果。 该指数衡量两个数据cluster之间的相似性。ARI指数的值位<span class="math inline">\([0;1]\)</span>区间，其中<span class="math inline">\(1\)</span>表示两个聚类相同，<span class="math inline">\(0\)</span>表示随机相似程度。</p>
</div>
</div>
<div id="clust-methods" class="section level2">
<h2><span class="header-section-number">8.2</span> 聚类示例</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(pcaMethods)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">library</span>(SC3)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb1-4" title="4"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb1-5" title="5"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">library</span>(mclust)</a>
<a class="sourceLine" id="cb1-7" title="7"><span class="kw">set.seed</span>(<span class="dv">1234567</span>)</a></code></pre></div>
<p>为说明scRNA-seq数据聚类，使用来自小鼠发育胚胎细胞<span class="citation">(Deng et al. <a href="#ref-Deng2014-mx">2014</a>)</span>的<code>Deng</code>数据集。 我们已经对数据集进行预处理并创建<code>SingleCellExperiment</code>对象。 而且利用文章中识别的细胞类型对细胞进行注释(<code>colData</code>中<code>cell_type2</code>列)。</p>
<div id="deng-" class="section level3">
<h3><span class="header-section-number">8.2.1</span> Deng 数据集</h3>
<p>加载数据:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/deng/deng-reads.rds&quot;</span>)</a>
<a class="sourceLine" id="cb2-2" title="2">deng</a>
<a class="sourceLine" id="cb2-3" title="3"><span class="co">## class: SingleCellExperiment </span></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co">## dim: 22431 268 </span></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">## metadata(0):</span></a>
<a class="sourceLine" id="cb2-6" title="6"><span class="co">## assays(2): counts logcounts</span></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="co">## rownames(22431): Hvcn1 Gbp7 ... Sox5 Alg11</span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="co">## rowData names(10): feature_symbol is_feature_control ...</span></a>
<a class="sourceLine" id="cb2-9" title="9"><span class="co">##   total_counts log10_total_counts</span></a>
<a class="sourceLine" id="cb2-10" title="10"><span class="co">## colnames(268): 16cell 16cell.1 ... zy.2 zy.3</span></a>
<a class="sourceLine" id="cb2-11" title="11"><span class="co">## colData names(30): cell_type2 cell_type1 ... pct_counts_ERCC</span></a>
<a class="sourceLine" id="cb2-12" title="12"><span class="co">##   is_cell_control</span></a>
<a class="sourceLine" id="cb2-13" title="13"><span class="co">## reducedDimNames(0):</span></a>
<a class="sourceLine" id="cb2-14" title="14"><span class="co">## spikeNames(1): ERCC</span></a></code></pre></div>
<p>查看细胞类型注释:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="kw">table</span>(<span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2)</a>
<a class="sourceLine" id="cb3-2" title="2"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-3" title="3"><span class="co">##     16cell      4cell      8cell early2cell earlyblast  late2cell </span></a>
<a class="sourceLine" id="cb3-4" title="4"><span class="co">##         50         14         37          8         43         10 </span></a>
<a class="sourceLine" id="cb3-5" title="5"><span class="co">##  lateblast   mid2cell   midblast         zy </span></a>
<a class="sourceLine" id="cb3-6" title="6"><span class="co">##         30         12         60          4</span></a></code></pre></div>
<p>简单的PCA分析已经分离出一些细胞类型，并初步了解了数据结构：</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">plotPCA</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/unnamed-chunk-5-1.png" style="display: block; margin: auto;" /></p>
<p>如上图所示，早期细胞类型分得很靠，但三个囊胚时间点难以区分。</p>
</div>
<div id="sc3-1" class="section level3">
<h3><span class="header-section-number">8.2.2</span> SC3</h3>
<p>在Deng数据上运行<code>SC3</code>聚类，<code>SC3</code>的优点是其可以直接使用<code>SingleCellExperiment</code>对象。</p>
<p>假设我们不知道类 <em>k</em>(细胞类型)的数目，<code>SC3</code>可以对cluster数目进行估计。</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">sc3_estimate_k</span>(deng)</a>
<a class="sourceLine" id="cb5-2" title="2"><span class="kw">metadata</span>(deng)<span class="op">$</span>sc3<span class="op">$</span>k_estimation</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co">## 6</span></a></code></pre></div>
<p>有趣的是，<code>SC3</code>预测的细胞类型数量小于原始数据注释。 然而，不同细胞类型的早期，中期和晚期，刚好有6种细胞类型。 将合并的细胞类型存储在<code>colData</code>的<code>cell_type1</code>列中：</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw">plotPCA</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;cell_type1&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/unnamed-chunk-7-1.png" style="display: block; margin: auto;" /></p>
<p>下面运行<code>SC3</code>(同时计算类的生物学特性)：</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">sc3</span>(deng, <span class="dt">ks =</span> <span class="dv">10</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>, <span class="dt">n_cores =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><code>SC3</code>结果由几部分组成(详见<span class="citation">(Kiselev et al. <a href="#ref-Kiselev2016-bq">2017</a>)</span>和<a href="http://bioconductor.org/packages/release/bioc/vignettes/SC3/inst/doc/my-vignette.html">SC3 vignette</a>)，这里展示一些：</p>
<p>一致性矩阵:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="kw">sc3_plot_consensus</span>(deng, <span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">show_pdata =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/unnamed-chunk-9-1.png" style="display: block; margin: auto;" />
Silhouette图:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw">sc3_plot_silhouette</span>(deng, <span class="dt">k =</span> <span class="dv">10</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/unnamed-chunk-10-1.png" style="display: block; margin: auto;" />
表达矩阵热图:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="kw">sc3_plot_expression</span>(deng, <span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">show_pdata =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/unnamed-chunk-11-1.png" style="display: block; margin: auto;" />
识别的marker基因:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="kw">sc3_plot_markers</span>(deng, <span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">show_pdata =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/unnamed-chunk-12-1.png" style="display: block; margin: auto;" />
<code>SC3</code>类PCA图:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="kw">plotPCA</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;sc3_10_clusters&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/unnamed-chunk-13-1.png" style="display: block; margin: auto;" />
<code>SC3</code>聚类结果和文献中细胞类型比较：</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="kw">adjustedRandIndex</span>(<span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2, <span class="kw">colData</span>(deng)<span class="op">$</span>sc3_<span class="dv">10</span>_clusters)</a>
<a class="sourceLine" id="cb13-2" title="2"><span class="co">## [1] 0.7721292</span></a></code></pre></div>
<p><strong>注意</strong> <code>SC3</code>可以在<code>Shiny</code>会话中以交互形式运行:</p>
<p>上述命令会在浏览器中打开<code>SC3</code>。</p>
<p><strong>注意</strong> 当细胞数量<span class="math inline">\(&gt;5000\)</span>，直接计算<code>SC3</code>距离非常慢。对于包含超过<span class="math inline">\(10^5\)</span>细胞的大型数据集，推荐使用<code>Seurat</code>(见章节 <a href="seurat-chapter.html#seurat-chapter">9</a>)。</p>
<ul>
<li><p><strong>练习1</strong>: 在浏览器中运行<code>SC3</code>,设置<span class="math inline">\(k\)</span>从8-12，查看不同的聚类结果；</p></li>
<li><p><strong>练习2</strong>: 哪种聚类结果最稳定(查看“Stability”标签)；</p></li>
<li><p><strong>练习3</strong>: 使用<span class="math inline">\(k=10\)</span>，查看cluster中差异表达基因和marker基因；</p></li>
<li><p><strong>练习4</strong>: 更改marker基因阈值(默认为0.85)。<strong>SC3</strong> 找到更多的marker基因了吗？</p></li>
</ul>
</div>
<div id="tsne-kmeans" class="section level3">
<h3><span class="header-section-number">8.2.3</span> tSNE + kmeans</h3>
<p>之前(<a href="cleaning-the-expression-matrix.html#visual-tsne">7.3.3</a>)用 __scater__包画<a href="https://lvdmaaten.github.io/tsne/">tSNE</a>图是用<a href="https://cran.r-project.org/web/packages/Rtsne/index.html">Rtsne</a> 和 <a href="https://cran.r-project.org/web/packages/ggplot2/index.html">ggplot2</a>做的。这里也同样：</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(deng, <span class="dt">rand_seed =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb14-2" title="2"><span class="kw">plotTSNE</span>(deng)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-tsne"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/clust-tsne-1.png" alt="tSNE map of the patient data"  />
<p class="caption">
Figure 8.6: tSNE map of the patient data
</p>
</div>
<p>注意，上图中的所有点都是黑色的。这与之前看到的不同，当时基于注释对细胞进行着色。 这里没有任何注释，所有细胞都来自同一批次，因此所有点都是黑色的。</p>
<p>现在将 <em>k</em>-means聚类算法应用到tSNE映射上的点云上。你在点云看到了多少组？</p>
<p>从<span class="math inline">\(k=8\)</span>开始:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="kw">colData</span>(deng)<span class="op">$</span>tSNE_kmeans &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">kmeans</span>(deng<span class="op">@</span>reducedDims<span class="op">$</span>TSNE, <span class="dt">centers =</span> <span class="dv">8</span>)<span class="op">$</span>clust)</a>
<a class="sourceLine" id="cb15-2" title="2"><span class="kw">plotTSNE</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;tSNE_kmeans&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-tsne-kmeans2"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/clust-tsne-kmeans2-1.png" alt="tSNE map of the patient data with 8 colored clusters, identified by the k-means clustering algorithm"  />
<p class="caption">
Figure 8.7: tSNE map of the patient data with 8 colored clusters, identified by the k-means clustering algorithm
</p>
</div>
<p><strong>练习7</strong>: 设置<span class="math inline">\(k=10\)</span>.</p>
<p><strong>练习8</strong>: 比较<code>tSNE+kmeans</code>结果和文献中细胞类型，当改变<code>perplexity</code>参数时，结果会提高吗？</p>
<p><strong>答案</strong>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="kw">colData</span>(deng)<span class="op">$</span>tSNE_kmeans &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">kmeans</span>(deng<span class="op">@</span>reducedDims<span class="op">$</span>TSNE, <span class="dt">centers =</span> <span class="dv">10</span>)<span class="op">$</span>clust)</a>
<a class="sourceLine" id="cb16-2" title="2"><span class="kw">adjustedRandIndex</span>(<span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2, <span class="kw">colData</span>(deng)<span class="op">$</span>tSNE_kmeans)</a></code></pre></div>
<p>您可能已经注意到，<code>tSNE+kmeans</code>j结果是随机的，每次运行时都会给出不同的结果。 为了获得更好的结果，我们需要多次运行该方法。 SC3也是随机的，但由于<code>consensus</code>步骤，更稳健，不太可能产生不同的结果。</p>
</div>
<div id="sincera-1" class="section level3">
<h3><span class="header-section-number">8.2.4</span> SINCERA</h3>
<p><a href="https://research.cchmc.org/pbge/sincera.html">SINCERA</a>基于层次聚类。注意，SINCERA在聚类之前进行基因尺度z-score转换：</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="co"># use the same gene filter as in SC3</span></a>
<a class="sourceLine" id="cb17-2" title="2">input &lt;-<span class="st"> </span><span class="kw">logcounts</span>(deng[<span class="kw">rowData</span>(deng)<span class="op">$</span>sc3_gene_filter, ])</a></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="co"># perform gene-by-gene per-sample z-score transformation</span></a>
<a class="sourceLine" id="cb18-2" title="2">dat &lt;-<span class="st"> </span><span class="kw">apply</span>(input, <span class="dv">1</span>, <span class="cf">function</span>(y) scRNA.seq.funcs<span class="op">::</span><span class="kw">z.transform.helper</span>(y))</a>
<a class="sourceLine" id="cb18-3" title="3"><span class="co"># hierarchical clustering</span></a>
<a class="sourceLine" id="cb18-4" title="4">dd &lt;-<span class="st"> </span><span class="kw">as.dist</span>((<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">cor</span>(<span class="kw">t</span>(dat), <span class="dt">method =</span> <span class="st">&quot;pearson&quot;</span>))<span class="op">/</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb18-5" title="5">hc &lt;-<span class="st"> </span><span class="kw">hclust</span>(dd, <span class="dt">method =</span> <span class="st">&quot;average&quot;</span>)</a></code></pre></div>
<p>如果cluster数量未知, <a href="https://research.cchmc.org/pbge/sincera.html">SINCERA</a> 将k识别为生成不超过指定数量的单个cluster（仅包含1个细胞的clusters）的层次树的最小高度。</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">num.singleton &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb19-2" title="2">kk &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb19-3" title="3"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="kw">dim</span>(dat)[<span class="dv">2</span>]) {</a>
<a class="sourceLine" id="cb19-4" title="4">    clusters &lt;-<span class="st"> </span><span class="kw">cutree</span>(hc, <span class="dt">k =</span> i)</a>
<a class="sourceLine" id="cb19-5" title="5">    clustersizes &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">table</span>(clusters))</a>
<a class="sourceLine" id="cb19-6" title="6">    singleton.clusters &lt;-<span class="st"> </span><span class="kw">which</span>(clustersizes<span class="op">$</span>Freq <span class="op">&lt;</span><span class="st"> </span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb19-7" title="7">    <span class="cf">if</span> (<span class="kw">length</span>(singleton.clusters) <span class="op">&lt;=</span><span class="st"> </span>num.singleton) {</a>
<a class="sourceLine" id="cb19-8" title="8">        kk &lt;-<span class="st"> </span>i</a>
<a class="sourceLine" id="cb19-9" title="9">    } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb19-10" title="10">        <span class="cf">break</span>;</a>
<a class="sourceLine" id="cb19-11" title="11">    }</a>
<a class="sourceLine" id="cb19-12" title="12">}</a>
<a class="sourceLine" id="cb19-13" title="13"><span class="kw">cat</span>(kk)</a>
<a class="sourceLine" id="cb19-14" title="14"><span class="co">## 6</span></a></code></pre></div>
<p>用热图对SINCERA结果进行可视化：</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1"><span class="kw">pheatmap</span>(</a>
<a class="sourceLine" id="cb20-2" title="2">    <span class="kw">t</span>(dat),</a>
<a class="sourceLine" id="cb20-3" title="3">    <span class="dt">cluster_cols =</span> hc,</a>
<a class="sourceLine" id="cb20-4" title="4">    <span class="dt">cutree_cols =</span> kk,</a>
<a class="sourceLine" id="cb20-5" title="5">    <span class="dt">kmeans_k =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb20-6" title="6">    <span class="dt">show_rownames =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb20-7" title="7">)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-sincera"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/clustering_files/figure-html/clust-sincera-1.png" alt="Clustering solutions of SINCERA method using found $k$"  />
<p class="caption">
Figure 8.8: Clustering solutions of SINCERA method using found <span class="math inline">\(k\)</span>
</p>
</div>
<p><strong>练习10</strong>: 比较<code>SINCERA</code>结果和文献细胞类型。</p>
<p><strong>答案</strong>:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1"><span class="kw">colData</span>(deng)<span class="op">$</span>SINCERA &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">cutree</span>(hc, <span class="dt">k =</span> kk))</a>
<a class="sourceLine" id="cb21-2" title="2"><span class="kw">adjustedRandIndex</span>(<span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2, <span class="kw">colData</span>(deng)<span class="op">$</span>SINCERA)</a>
<a class="sourceLine" id="cb21-3" title="3"><span class="co">## [1] 0.3823537</span></a></code></pre></div>
<p><strong>练习11</strong>: 使用单例cluster标准确定<strong>k</strong>效果好吗？</p>
</div>
</div>
<div id="section-91" class="section level2">
<h2><span class="header-section-number">8.3</span> 特征选择</h2>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1"><span class="kw">library</span>(scRNA.seq.funcs)</a>
<a class="sourceLine" id="cb22-2" title="2"><span class="kw">library</span>(matrixStats)</a>
<a class="sourceLine" id="cb22-3" title="3"><span class="kw">library</span>(M3Drop)</a>
<a class="sourceLine" id="cb22-4" title="4"><span class="kw">library</span>(RColorBrewer)</a>
<a class="sourceLine" id="cb22-5" title="5"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb22-6" title="6"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a></code></pre></div>
<p>单细胞RNASeq能够检测每个细胞中数千个基因的表达。 然而，在大多数情况下，只有一部分基因对感兴趣的生物学条件的做出响应，例如， 细胞类型的差异，分化的驱动因素，对环境刺激的响应。 由于技术噪音，在scRNA-Seq实验中检测到的大多数基因仅在不同水平检测到。 这导致技术噪声和批量效应会混淆感兴趣的生物信号。</p>
<p>因此，进行特征选择以去除那些仅表现出来技术噪音的基因。这不仅会增加数据中的信噪比; 还通过减少数据量来降低分析的计算复杂性。</p>
<p>对于scRNASeq数据，我们将关注于非监督的特征选择方法，这些方法不需要任何先验信息，例如细胞类型或生物组，因为对大多数实验得不到这些信息，或者不可靠的。与之相反，差异表达(章节 <a href="biological-analysis.html#dechapter">8.6</a>)可以被认为是监督特征选择的一种形式，因为它使用每个样本的已知生物标记来识别在不同组间差异表达的特征（如基因）。</p>
<p>在本节中，我们将继续使用Deng数据。</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/deng/deng-reads.rds&quot;</span>)</a>
<a class="sourceLine" id="cb23-2" title="2">celltype_labs &lt;-<span class="st"> </span><span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2</a>
<a class="sourceLine" id="cb23-3" title="3">cell_colors &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="kw">max</span>(<span class="dv">3</span>,<span class="kw">length</span>(<span class="kw">unique</span>(celltype_labs))), <span class="st">&quot;Set3&quot;</span>)</a></code></pre></div>
<p>特征选择在质控后执行，但是此数据已经质控过，因此可以跳过该步骤。 M3Drop包含两种不同的特征选择方法，一个是“M3DropFeatureSelection”，基于Michaelis-Menten曲线，为全长转录本scRNA-seq（如Smartseq2）设计；另一个是“NBumiFeatureSelectionCombinedDrop”，基于负二项模型，专为UMI count数据而设计。我们在Deng Smartseq2数据上进行演示。</p>
<p>M3Drop特征选择使用标准化（但不是对数变换）表达矩阵。 使用以下命令从SingleCellExperiment对象中提取。</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1">expr_matrix &lt;-<span class="st"> </span>M3Drop<span class="op">::</span><span class="kw">M3DropConvertData</span>(deng)</a></code></pre></div>
<p>此功能与大多数scRNA-seq分析包兼容，包括：scater，SingleCellExperiment，monocle和Seurat。 如果指定矩阵是原始count或者对数转换，它可以将其转换为正确的形式（删除未检测到的基因和标准化/delogging）。查看帮助手册以获取详细信息：</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1">?M3Drop<span class="op">::</span>M3DropConvertData</a></code></pre></div>
<p><strong>练习1</strong>: 确认转换函数已经删除未检测到的基因：</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1"><span class="kw">nrow</span>(<span class="kw">counts</span>(deng)) <span class="op">-</span><span class="st"> </span><span class="kw">nrow</span>(expr_matrix)</a>
<a class="sourceLine" id="cb26-2" title="2"><span class="kw">summary</span>( <span class="kw">rowSums</span>(<span class="kw">counts</span>(deng))[<span class="op">!</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">counts</span>(deng)) <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(expr_matrix) ] )</a></code></pre></div>
<div id="null" class="section level3">
<h3><span class="header-section-number">8.3.1</span> 识别基因与null模型</h3>
<p>无监督特征选择主要有两种方法。 第一个是识别与描述数据集中预期的技术噪声的null模型不同的基因。</p>
<p>如果数据集包含spike-in RNAs，其可以用于直接模拟技术噪音。然而，spike-ins可能不像内源转录本经历相同的技术噪音 <a href="https://www.nature.com/nmeth/journal/v14/n4/full/nmeth.4220.html">(Svensson et al., 2017)</a>。而且，scRNA-seq实验只包含一小部分spike-ins，对拟合模型参数可能不足。</p>
<div id="highly-variable-genes" class="section level4">
<h4><span class="header-section-number">8.3.1.1</span> Highly Variable Genes</h4>
<p>提出用于鉴定scRNA-Seq数据集特征的第一种方法是鉴定高度可变的基因（HVG）。 HVG假设如果基因在细胞间的表达差异很大，那么这些差异中的一部分是由于细胞之间的生物学差异而不是技术噪声。然而，由于count data的性质，基因的表达均值与细胞read count的方差之间存在正相关关系。必须纠正这种关系才能正确识别HVG。</p>
<p><strong>练习2</strong>
使用rowMeans和rowVars函数绘制此数据集中所有基因的表达的均值和方差之间的关系。 （提示：使用log =“xy”在对数刻度上绘图）。</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb27-2" title="2">    <span class="kw">rowMeans</span>(expr_matrix), </a>
<a class="sourceLine" id="cb27-3" title="3">    <span class="kw">rowVars</span>(expr_matrix), </a>
<a class="sourceLine" id="cb27-4" title="4">    <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>, </a>
<a class="sourceLine" id="cb27-5" title="5">    <span class="dt">pch=</span><span class="dv">16</span>,</a>
<a class="sourceLine" id="cb27-6" title="6">    <span class="dt">xlab=</span><span class="st">&quot;Mean Expression&quot;</span>, </a>
<a class="sourceLine" id="cb27-7" title="7">    <span class="dt">ylab=</span><span class="st">&quot;Variance&quot;</span>, </a>
<a class="sourceLine" id="cb27-8" title="8">    <span class="dt">main=</span><span class="st">&quot;&quot;</span></a>
<a class="sourceLine" id="cb27-9" title="9">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/dropouts_files/figure-html/unnamed-chunk-7-1.png" style="display: block; margin: auto;" />
<a href="http://www.nature.com/nmeth/journal/v10/n11/full/nmeth.2645.html">Brennecke et al.</a>提出了一种纠正方差和平均表达之间关系的方法。为使用Brennecke方法，首先对文库大小进行标准化，然后计算均值和变异系数（方差除以表达均值）。对于ERCC spike-in，用二次曲线拟合这两个变量之间的关系，然后使用卡方检验来找到显著高于曲线的基因。此方法作为Brennecke_getVariableGenes(counts, spikes)函数包含在M3Drop包中。 但是，此数据集不包含spike-ins，因此使用整个数据集来估计技术噪声。</p>
<p>在下图中，红色曲线是拟合的技术噪声模型，虚线是95%置信区间。 粉红点是多次测试校正后具有显著生物变异性的基因。</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1">Brennecke_HVG &lt;-<span class="st"> </span><span class="kw">BrenneckeGetVariableGenes</span>(</a>
<a class="sourceLine" id="cb28-2" title="2">    expr_matrix,</a>
<a class="sourceLine" id="cb28-3" title="3">    <span class="dt">fdr =</span> <span class="fl">0.01</span>,</a>
<a class="sourceLine" id="cb28-4" title="4">    <span class="dt">minBiolDisp =</span> <span class="fl">0.5</span></a>
<a class="sourceLine" id="cb28-5" title="5">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/dropouts_files/figure-html/unnamed-chunk-8-1.png" style="display: block; margin: auto;" /></p>
<p>该函数返回显著基因的矩阵以及估计效应大小（观察和预期的变异系数之间的差异），以及原始p值和FDR校正q值的显著性。保留显著HVG基因的名称。</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1">HVG_genes &lt;-<span class="st"> </span>Brennecke_HVG<span class="op">$</span>Gene</a></code></pre></div>
<p>_练习3__
使用BrenneckeGetVariableGenes有多少基因是显著的？</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1"><span class="kw">length</span>(HVG_genes)</a>
<a class="sourceLine" id="cb30-2" title="2"><span class="co">## 1303</span></a></code></pre></div>
</div>
<div id="high-dropout-genes" class="section level4">
<h4><span class="header-section-number">8.3.1.2</span> High Dropout Genes</h4>
<p>另外一种寻找HVGs是识别具有非常多0值表达的基因。0的频率称之为<code>dropout rate</code>,和scRNA-seq数据中表达水平密切相关。scRNA-seq数据中表达矩阵超过一半的基因表达值为0，因为单细胞中mRNA无法被逆转录<a href="http://www.biorxiv.org/content/early/2017/05/25/065094">(Andrews and Hemberg, 2016)</a>。逆转录是酶促反应，可以使用Michaelis-Menten方程模拟:</p>
<p><span class="math display">\[P_{dropout} = 1 - S/(K + S)\]</span></p>
<p>其中<span class="math inline">\(S\)</span>为细胞中mRNA浓度(使用平均表达值来估计)，<span class="math inline">\(K\)</span>是Michaelis-Menten常数。</p>
<p>由于Michaelis-Menten方差是非线性凸函数，数据集中细胞间差异表达的基因位于Michaelis-Menten模型的上/右方(见下图)。</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1">K &lt;-<span class="st"> </span><span class="dv">49</span></a>
<a class="sourceLine" id="cb31-2" title="2">S_sim &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">-3</span>, <span class="dt">to =</span> <span class="dv">4</span>, <span class="dt">by =</span> <span class="fl">0.05</span>) <span class="co"># range of expression values</span></a>
<a class="sourceLine" id="cb31-3" title="3">MM &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>S_sim <span class="op">/</span><span class="st"> </span>(K <span class="op">+</span><span class="st"> </span>S_sim)</a>
<a class="sourceLine" id="cb31-4" title="4"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb31-5" title="5">    S_sim, </a>
<a class="sourceLine" id="cb31-6" title="6">    MM, </a>
<a class="sourceLine" id="cb31-7" title="7">    <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>, </a>
<a class="sourceLine" id="cb31-8" title="8">    <span class="dt">lwd =</span> <span class="dv">3</span>, </a>
<a class="sourceLine" id="cb31-9" title="9">    <span class="dt">xlab =</span> <span class="st">&quot;Expression&quot;</span>, </a>
<a class="sourceLine" id="cb31-10" title="10">    <span class="dt">ylab =</span> <span class="st">&quot;Dropout Rate&quot;</span>, </a>
<a class="sourceLine" id="cb31-11" title="11">    <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb31-12" title="12">)</a>
<a class="sourceLine" id="cb31-13" title="13">S1 &lt;-<span class="st"> </span><span class="dv">10</span> <span class="co"># Mean expression in population 1</span></a>
<a class="sourceLine" id="cb31-14" title="14">P1 &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>S1 <span class="op">/</span><span class="st"> </span>(K <span class="op">+</span><span class="st"> </span>S1) <span class="co"># Dropouts for cells in condition 1</span></a>
<a class="sourceLine" id="cb31-15" title="15">S2 &lt;-<span class="st"> </span><span class="dv">750</span> <span class="co"># Mean expression in population 2</span></a>
<a class="sourceLine" id="cb31-16" title="16">P2 &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>S2 <span class="op">/</span><span class="st"> </span>(K <span class="op">+</span><span class="st"> </span>S2) <span class="co"># Dropouts for cells in condition 2</span></a>
<a class="sourceLine" id="cb31-17" title="17"><span class="kw">points</span>(</a>
<a class="sourceLine" id="cb31-18" title="18">    <span class="kw">c</span>(S1, S2),</a>
<a class="sourceLine" id="cb31-19" title="19">    <span class="kw">c</span>(P1, P2), </a>
<a class="sourceLine" id="cb31-20" title="20">    <span class="dt">pch =</span> <span class="dv">16</span>, </a>
<a class="sourceLine" id="cb31-21" title="21">    <span class="dt">col =</span> <span class="st">&quot;grey85&quot;</span>, </a>
<a class="sourceLine" id="cb31-22" title="22">    <span class="dt">cex =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb31-23" title="23">)</a>
<a class="sourceLine" id="cb31-24" title="24">mix &lt;-<span class="st"> </span><span class="fl">0.5</span> <span class="co"># proportion of cells in condition 1</span></a>
<a class="sourceLine" id="cb31-25" title="25"><span class="kw">points</span>(</a>
<a class="sourceLine" id="cb31-26" title="26">    S1 <span class="op">*</span><span class="st"> </span>mix <span class="op">+</span><span class="st"> </span>S2 <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>mix), </a>
<a class="sourceLine" id="cb31-27" title="27">    P1 <span class="op">*</span><span class="st"> </span>mix <span class="op">+</span><span class="st"> </span>P2 <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>mix), </a>
<a class="sourceLine" id="cb31-28" title="28">    <span class="dt">pch =</span> <span class="dv">16</span>, </a>
<a class="sourceLine" id="cb31-29" title="29">    <span class="dt">col =</span> <span class="st">&quot;grey35&quot;</span>, </a>
<a class="sourceLine" id="cb31-30" title="30">    <span class="dt">cex =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb31-31" title="31">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/dropouts_files/figure-html/unnamed-chunk-11-1.png" style="display: block; margin: auto;" />
<strong>注意</strong>: 将<code>log="x"</code>加入<code>plot</code>函数，查看log尺度下的图像。</p>
<p><strong>练习4</strong>: 使用不同表达水平(S1 &amp; S2)或混合进行作图。</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb32-2" title="2">    S_sim, </a>
<a class="sourceLine" id="cb32-3" title="3">    MM, </a>
<a class="sourceLine" id="cb32-4" title="4">    <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>, </a>
<a class="sourceLine" id="cb32-5" title="5">    <span class="dt">lwd =</span> <span class="dv">3</span>, </a>
<a class="sourceLine" id="cb32-6" title="6">    <span class="dt">xlab =</span> <span class="st">&quot;Expression&quot;</span>, </a>
<a class="sourceLine" id="cb32-7" title="7">    <span class="dt">ylab =</span> <span class="st">&quot;Dropout Rate&quot;</span>, </a>
<a class="sourceLine" id="cb32-8" title="8">    <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1000</span>), </a>
<a class="sourceLine" id="cb32-9" title="9">    <span class="dt">log =</span> <span class="st">&quot;x&quot;</span></a>
<a class="sourceLine" id="cb32-10" title="10">)</a>
<a class="sourceLine" id="cb32-11" title="11">S1 &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb32-12" title="12">P1 &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>S1 <span class="op">/</span><span class="st"> </span>(K <span class="op">+</span><span class="st"> </span>S1) <span class="co"># Expression &amp; dropouts for cells in condition 1</span></a>
<a class="sourceLine" id="cb32-13" title="13">S2 &lt;-<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb32-14" title="14">P2 &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>S2 <span class="op">/</span><span class="st"> </span>(K <span class="op">+</span><span class="st"> </span>S2) <span class="co"># Expression &amp; dropouts for cells in condition 2</span></a>
<a class="sourceLine" id="cb32-15" title="15"><span class="kw">points</span>(</a>
<a class="sourceLine" id="cb32-16" title="16">    <span class="kw">c</span>(S1, S2),</a>
<a class="sourceLine" id="cb32-17" title="17">    <span class="kw">c</span>(P1, P2), </a>
<a class="sourceLine" id="cb32-18" title="18">    <span class="dt">pch =</span> <span class="dv">16</span>, </a>
<a class="sourceLine" id="cb32-19" title="19">    <span class="dt">col =</span> <span class="st">&quot;grey85&quot;</span>, </a>
<a class="sourceLine" id="cb32-20" title="20">    <span class="dt">cex =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb32-21" title="21">)</a>
<a class="sourceLine" id="cb32-22" title="22">mix &lt;-<span class="st"> </span><span class="fl">0.75</span> <span class="co"># proportion of cells in condition 1</span></a>
<a class="sourceLine" id="cb32-23" title="23"><span class="kw">points</span>(</a>
<a class="sourceLine" id="cb32-24" title="24">    S1 <span class="op">*</span><span class="st"> </span>mix <span class="op">+</span><span class="st"> </span>S2 <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>mix), </a>
<a class="sourceLine" id="cb32-25" title="25">    P1 <span class="op">*</span><span class="st"> </span>mix <span class="op">+</span><span class="st"> </span>P2 <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>mix), </a>
<a class="sourceLine" id="cb32-26" title="26">    <span class="dt">pch =</span> <span class="dv">16</span>, </a>
<a class="sourceLine" id="cb32-27" title="27">    <span class="dt">col =</span> <span class="st">&quot;grey35&quot;</span>, </a>
<a class="sourceLine" id="cb32-28" title="28">    <span class="dt">cex =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb32-29" title="29">)</a></code></pre></div>
<p>使用M3Drop识别MM曲线右侧的显著异常值。应用1％FDR多重假设校正：</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1">M3Drop_genes &lt;-<span class="st"> </span><span class="kw">M3DropFeatureSelection</span>(</a>
<a class="sourceLine" id="cb33-2" title="2">    expr_matrix,</a>
<a class="sourceLine" id="cb33-3" title="3">    <span class="dt">mt_method =</span> <span class="st">&quot;fdr&quot;</span>,</a>
<a class="sourceLine" id="cb33-4" title="4">    <span class="dt">mt_threshold =</span> <span class="fl">0.01</span></a>
<a class="sourceLine" id="cb33-5" title="5">)</a>
<a class="sourceLine" id="cb33-6" title="6">M3Drop_genes &lt;-<span class="st"> </span>M3Drop_genes<span class="op">$</span>Gene</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/dropouts_files/figure-html/unnamed-chunk-13-1.png" style="display: block; margin: auto;" /></p>
<p>M3Drop包中包含另一种方法，法专门针对UMI-tagged数据，该数据通常包含由低序列覆盖率以及逆转录不足，产生的许多0值。该模型是Depth-Adjusted Negative Binomial (DANB)。 该方法将每个表达观测描述为负二项模型，其均值和相应基因的平均表达和相应细胞的测序深度相关,其方差和基因的平均表达相关。</p>
<p>此方法直接对数据集中的raw counts进行建模，可以使用类似于M3Drop的“NBumiConvertData”函数提取相应的矩阵。 然而，我们有一个额外的步骤来拟合模型，因为这是该方法的最慢步骤，我们目前正在研究其他方法，可以将此模型信息用于其他方面（如标准化，共表达测试，检测HVG）。</p>
<p>该方法包括对每个特征的重要性进行二项式检验，但由于Deng数据不是UMI counts，因此模型不能充分地拟合噪声，并且太多基因检测为显著的。 因此，我们将按x效应大小去Top1500。</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" title="1">deng_int &lt;-<span class="st"> </span><span class="kw">NBumiConvertData</span>(deng)</a>
<a class="sourceLine" id="cb34-2" title="2">DANB_fit &lt;-<span class="st"> </span><span class="kw">NBumiFitModel</span>(deng_int) <span class="co"># DANB is fit to the raw count matrix</span></a>
<a class="sourceLine" id="cb34-3" title="3"><span class="co"># Perform DANB feature selection</span></a>
<a class="sourceLine" id="cb34-4" title="4">DropFS &lt;-<span class="st"> </span><span class="kw">NBumiFeatureSelectionCombinedDrop</span>(DANB_fit, <span class="dt">method=</span><span class="st">&quot;fdr&quot;</span>, <span class="dt">qval.thresh=</span><span class="fl">0.01</span>, <span class="dt">suppress.plot=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb34-5" title="5">DANB_genes &lt;-<span class="st"> </span>DropFS[<span class="dv">1</span><span class="op">:</span><span class="dv">1500</span>,]<span class="op">$</span>Gene</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/dropouts_files/figure-html/unnamed-chunk-14-1.png" style="display: block; margin: auto;" />
<strong>练习5</strong>
使用NBumiFeatureSelectionCombinedDrop有多少基因是显著的？</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" title="1"><span class="kw">nrow</span>(DropFS)</a>
<a class="sourceLine" id="cb35-2" title="2"><span class="co">## 10694</span></a></code></pre></div>
</div>
</div>
<div id="correlated-expression" class="section level3">
<h3><span class="header-section-number">8.3.2</span> Correlated Expression</h3>
<p>和之前方法不同时使用基因-基因相关性。该方法基于多个基因将在不同细胞类型或细胞状态之间差异表达。在相同细胞群中表达的基因彼此正相关，其中在不同细胞群中表达的基因将彼此负相关。因此，通过基因间相关程度来鉴定重要基因。</p>
<p>这种方法的局限性在于它假设技术噪声对于每个细胞是随机且独立的，因此不产生基因-基因相关性，但是这种假设被批次效应所违背，批次效应通常在不同的实验批次之间并且产生基因-基因相关性。因此，通过基因-基因相关性排序选择前几千个基因比考虑相关性的显著性更合适。</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" title="1">cor_feat &lt;-<span class="st"> </span>M3Drop<span class="op">::</span><span class="kw">corFS</span>(expr_matrix)</a>
<a class="sourceLine" id="cb36-2" title="2">Cor_genes &lt;-<span class="st"> </span><span class="kw">names</span>(cor_feat)[<span class="dv">1</span><span class="op">:</span><span class="dv">1500</span>]</a></code></pre></div>
<p>最后，scRNA-Seq数据分析中另一种常用的特征选择方法是使用PCA loadings。具有高PCA loadings的基因可能是高度可变的并且与许多其他可变基因相关，因此可能与潜在的生物学性质相关。然而，与基因-基因相关性一样，PCA loading对批次效应敏感; 因此，建议绘制PCA结果以确定对应生物变异而不是批次效应的组成分。</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" title="1"><span class="co"># PCA is typically performed on log-transformed expression data</span></a>
<a class="sourceLine" id="cb37-2" title="2">pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">log</span>(expr_matrix <span class="op">+</span><span class="st"> </span><span class="dv">1</span>) <span class="op">/</span><span class="st"> </span><span class="kw">log</span>(<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb37-3" title="3"></a>
<a class="sourceLine" id="cb37-4" title="4"><span class="co"># plot projection</span></a>
<a class="sourceLine" id="cb37-5" title="5"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb37-6" title="6">    pca<span class="op">$</span>rotation[,<span class="dv">1</span>], </a>
<a class="sourceLine" id="cb37-7" title="7">    pca<span class="op">$</span>rotation[,<span class="dv">2</span>], </a>
<a class="sourceLine" id="cb37-8" title="8">    <span class="dt">pch =</span> <span class="dv">16</span>, </a>
<a class="sourceLine" id="cb37-9" title="9">    <span class="dt">col =</span> cell_colors[<span class="kw">as.factor</span>(celltype_labs)]</a>
<a class="sourceLine" id="cb37-10" title="10">) </a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/dropouts_files/figure-html/unnamed-chunk-17-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" title="1"><span class="co"># calculate loadings for components 1 and 2</span></a>
<a class="sourceLine" id="cb38-2" title="2">score &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">abs</span>(pca<span class="op">$</span>x[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)])) </a>
<a class="sourceLine" id="cb38-3" title="3"><span class="kw">names</span>(score) &lt;-<span class="st"> </span><span class="kw">rownames</span>(expr_matrix)</a>
<a class="sourceLine" id="cb38-4" title="4">score &lt;-<span class="st"> </span>score[<span class="kw">order</span>(<span class="op">-</span>score)]</a>
<a class="sourceLine" id="cb38-5" title="5">PCA_genes &lt;-<span class="st"> </span><span class="kw">names</span>(score[<span class="dv">1</span><span class="op">:</span><span class="dv">1500</span>])</a></code></pre></div>
<p><strong>练习6</strong>
查看前5个组成分，哪个与生物学现象最相关？如果考虑这些组成分的loadings，Top1500个特征如何变化？</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" title="1"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb39-2" title="2">    pca<span class="op">$</span>rotation[,<span class="dv">2</span>], </a>
<a class="sourceLine" id="cb39-3" title="3">    pca<span class="op">$</span>rotation[,<span class="dv">3</span>], </a>
<a class="sourceLine" id="cb39-4" title="4">    <span class="dt">pch =</span> <span class="dv">16</span>, </a>
<a class="sourceLine" id="cb39-5" title="5">    <span class="dt">col =</span> cell_colors[<span class="kw">as.factor</span>(celltype_labs)]</a>
<a class="sourceLine" id="cb39-6" title="6">)</a>
<a class="sourceLine" id="cb39-7" title="7"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb39-8" title="8">    pca<span class="op">$</span>rotation[,<span class="dv">3</span>], </a>
<a class="sourceLine" id="cb39-9" title="9">    pca<span class="op">$</span>rotation[,<span class="dv">4</span>], </a>
<a class="sourceLine" id="cb39-10" title="10">    <span class="dt">pch =</span> <span class="dv">16</span>, </a>
<a class="sourceLine" id="cb39-11" title="11">    <span class="dt">col =</span> cell_colors[<span class="kw">as.factor</span>(celltype_labs)]</a>
<a class="sourceLine" id="cb39-12" title="12">)</a>
<a class="sourceLine" id="cb39-13" title="13"><span class="co"># calculate loadings for components 1 and 2</span></a>
<a class="sourceLine" id="cb39-14" title="14">score &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">abs</span>(pca<span class="op">$</span>x[,<span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)]))</a>
<a class="sourceLine" id="cb39-15" title="15"><span class="kw">names</span>(score) &lt;-<span class="st"> </span><span class="kw">rownames</span>(expr_matrix)</a>
<a class="sourceLine" id="cb39-16" title="16">score &lt;-<span class="st"> </span>score[<span class="kw">order</span>(<span class="op">-</span>score)]</a>
<a class="sourceLine" id="cb39-17" title="17">PCA_genes2 =<span class="st"> </span><span class="kw">names</span>(score[<span class="dv">1</span><span class="op">:</span><span class="dv">1500</span>])</a></code></pre></div>
</div>
<div id="section-92" class="section level3">
<h3><span class="header-section-number">8.3.3</span> 比较不同的方法</h3>
<p>检查所识别的特征是否代表了该数据集中细胞类型之间差异表达的基因。</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" title="1"><span class="kw">M3DropExpressionHeatmap</span>(</a>
<a class="sourceLine" id="cb40-2" title="2">    M3Drop_genes,</a>
<a class="sourceLine" id="cb40-3" title="3">    expr_matrix,</a>
<a class="sourceLine" id="cb40-4" title="4">    <span class="dt">cell_labels =</span> celltype_labs</a>
<a class="sourceLine" id="cb40-5" title="5">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/dropouts_files/figure-html/unnamed-chunk-19-1.png" style="display: block; margin: auto;" />
使用Jaccard指数比较特征选择方法的一致性：</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" title="1">J &lt;-<span class="st"> </span><span class="kw">sum</span>(M3Drop_genes <span class="op">%in%</span><span class="st"> </span>HVG_genes)<span class="op">/</span><span class="kw">length</span>(<span class="kw">unique</span>(<span class="kw">c</span>(M3Drop_genes, HVG_genes)))</a></code></pre></div>
<p><strong>练习7</strong></p>
<p>绘制每种方法的基因的表达值，其中哪个是差异表达？该数据集中不同方法的一致性体现在哪里？</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" title="1"><span class="kw">M3DropExpressionHeatmap</span>(</a>
<a class="sourceLine" id="cb42-2" title="2">    HVG_genes,</a>
<a class="sourceLine" id="cb42-3" title="3">    expr_matrix,</a>
<a class="sourceLine" id="cb42-4" title="4">    <span class="dt">cell_labels =</span> celltype_labs</a>
<a class="sourceLine" id="cb42-5" title="5">)</a></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" title="1"><span class="kw">M3DropExpressionHeatmap</span>(</a>
<a class="sourceLine" id="cb43-2" title="2">    Cor_genes,</a>
<a class="sourceLine" id="cb43-3" title="3">    expr_matrix,</a>
<a class="sourceLine" id="cb43-4" title="4">    <span class="dt">cell_labels =</span> celltype_labs</a>
<a class="sourceLine" id="cb43-5" title="5">)</a></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" title="1"><span class="kw">M3DropExpressionHeatmap</span>(</a>
<a class="sourceLine" id="cb44-2" title="2">    PCA_genes,</a>
<a class="sourceLine" id="cb44-3" title="3">    expr_matrix,</a>
<a class="sourceLine" id="cb44-4" title="4">    <span class="dt">cell_labels =</span> celltype_labs</a>
<a class="sourceLine" id="cb44-5" title="5">)</a></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1"><span class="kw">M3DropExpressionHeatmap</span>(</a>
<a class="sourceLine" id="cb45-2" title="2">    PCA_genes2,</a>
<a class="sourceLine" id="cb45-3" title="3">    expr_matrix,</a>
<a class="sourceLine" id="cb45-4" title="4">    <span class="dt">cell_labels =</span> celltype_labs</a>
<a class="sourceLine" id="cb45-5" title="5">)</a></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" title="1">list_of_features &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb46-2" title="2">    M3Drop_genes, </a>
<a class="sourceLine" id="cb46-3" title="3">    HVG_genes, </a>
<a class="sourceLine" id="cb46-4" title="4">    Cor_genes, </a>
<a class="sourceLine" id="cb46-5" title="5">    PCA_genes, </a>
<a class="sourceLine" id="cb46-6" title="6">    PCA_genes2</a>
<a class="sourceLine" id="cb46-7" title="7">)</a>
<a class="sourceLine" id="cb46-8" title="8">Out &lt;-<span class="st"> </span><span class="kw">matrix</span>(</a>
<a class="sourceLine" id="cb46-9" title="9">    <span class="dv">0</span>, </a>
<a class="sourceLine" id="cb46-10" title="10">    <span class="dt">ncol =</span> <span class="kw">length</span>(list_of_features), </a>
<a class="sourceLine" id="cb46-11" title="11">    <span class="dt">nrow =</span> <span class="kw">length</span>(list_of_features)</a>
<a class="sourceLine" id="cb46-12" title="12">)</a>
<a class="sourceLine" id="cb46-13" title="13"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(list_of_features) ) {</a>
<a class="sourceLine" id="cb46-14" title="14">    <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(list_of_features) ) {</a>
<a class="sourceLine" id="cb46-15" title="15">        Out[i,j] &lt;-<span class="st"> </span><span class="kw">sum</span>(list_of_features[[i]] <span class="op">%in%</span><span class="st"> </span>list_of_features[[j]])<span class="op">/</span></a>
<a class="sourceLine" id="cb46-16" title="16"><span class="st">            </span><span class="kw">length</span>(<span class="kw">unique</span>(<span class="kw">c</span>(list_of_features[[i]], list_of_features[[j]])))</a>
<a class="sourceLine" id="cb46-17" title="17">     }</a>
<a class="sourceLine" id="cb46-18" title="18">}</a>
<a class="sourceLine" id="cb46-19" title="19"><span class="kw">colnames</span>(Out) &lt;-<span class="st"> </span><span class="kw">rownames</span>(Out) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;M3Drop&quot;</span>, <span class="st">&quot;HVG&quot;</span>, <span class="st">&quot;Cor&quot;</span>, <span class="st">&quot;PCA&quot;</span>, <span class="st">&quot;PCA2&quot;</span>)</a></code></pre></div>
</div>
</div>
<div id="section-93" class="section level2">
<h2><span class="header-section-number">8.4</span> 伪时间分析</h2>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" title="1"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb47-2" title="2"><span class="kw">library</span>(TSCAN)</a>
<a class="sourceLine" id="cb47-3" title="3"><span class="kw">library</span>(M3Drop)</a>
<a class="sourceLine" id="cb47-4" title="4"><span class="kw">library</span>(monocle)</a>
<a class="sourceLine" id="cb47-5" title="5"><span class="kw">library</span>(destiny)</a>
<a class="sourceLine" id="cb47-6" title="6"><span class="kw">library</span>(SLICER)</a>
<a class="sourceLine" id="cb47-7" title="7"><span class="kw">library</span>(ouija)</a>
<a class="sourceLine" id="cb47-8" title="8"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb47-9" title="9"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb47-10" title="10"><span class="kw">library</span>(ggthemes)</a>
<a class="sourceLine" id="cb47-11" title="11"><span class="kw">library</span>(ggbeeswarm)</a>
<a class="sourceLine" id="cb47-12" title="12"><span class="kw">library</span>(corrplot)</a>
<a class="sourceLine" id="cb47-13" title="13"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a></code></pre></div>
<p>在许多情况下，科学家正在研究细胞不断变化的过程。 这包括，例如，在发育阶段的分化过程：在刺激之后，细胞将从一种细胞类型变为另一种细胞类型。 理想情况下，我们希望监测单个细胞随时间的表达水平。然而，用scRNA-seq并不能做到，因为当提取RNA时细胞被裂解（破坏）。</p>
<p>相反，在多个时间点采样并获得基因表达谱的snapshot。 由于一些细胞将在分化过程中比其他细胞更快，因此每个snapshot包括处于不同发育状态的细胞。 使用统计方法沿着一个或多个轨迹对细胞进行排序，这些轨迹代表潜在的发育轨迹，这种排序被称为“伪时间”。</p>
<p>在本章中，我们考虑五种不同的工具：Monocle，TSCAN，destiny，SLICER和ouija，用于根据伪时间发育过程对细胞进行排序。 为了演示这些方法，我们使用小鼠胚胎发育的数据集<span class="citation">(Deng et al. <a href="#ref-Deng2014-mx">2014</a>)</span>。该数据集由早期小鼠发育的10个不同时间点的268个细胞组成。 在这种情况下，不需要伪时间对齐，因为细胞标签提供关于发育轨迹的信息。 因此，细胞标签可以建立一个基准，以便评估和比较不同的方法。</p>
<p>Cannoodt等人最近的综述提供了来自单细胞转录组学的轨迹推断的各种计算方法的详细综述<span class="citation">(Cannoodt, Saelens, and Saeys <a href="#ref-Cannoodt2016-uj">2016</a>)</span>。 他们讨论了几种工具，但不幸的是，许多这些工具维护不善，或没有在R中实现。</p>
<p>Cannoodt等人包括:</p>
<ul>
<li>SCUBA - Matlab实现</li>
<li>Wanderlust - Matlab (需要注册和下载)</li>
<li>Wishbone - Python</li>
<li>SLICER - R, 但是只能从GitHub安装</li>
<li>SCOUP - C++ 命令行工具</li>
<li>Waterfall - R, 附件中的R脚本</li>
<li>Mpath - R包, GitHub安装，文档没有vignette/workflow</li>
<li>Monocle - Bioconductor包</li>
<li>TSCAN - Bioconductor包</li>
</ul>
<p>然而只有两个讨论的工具（Monocle和TSCAN）满足开源软件黄金标准。</p>
<p>下图总结了不同工具的特征。</p>
<div class="figure" style="text-align: center"><span id="fig:pseudotime-methods-description"></span>
<img src="figures/cannoodt_pseudotime_properties.png" alt="Descriptions of trajectory inference methods for single-cell transcriptomics data (Fig. 2 from Cannoodt et al, 2016)." width="90%" />
<p class="caption">
Figure 8.9: Descriptions of trajectory inference methods for single-cell transcriptomics data (Fig. 2 from Cannoodt et al, 2016).
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:pseudotime-methods"></span>
<img src="figures/cannoodt_pseudotime_methods.png" alt="Characterization of trajectory inference methods for single-cell transcriptomics data (Fig. 3 from Cannoodt et al, 2016)." width="90%" />
<p class="caption">
Figure 8.10: Characterization of trajectory inference methods for single-cell transcriptomics data (Fig. 3 from Cannoodt et al, 2016).
</p>
</div>
<div id="deng" class="section level3">
<h3><span class="header-section-number">8.4.1</span> Deng数据初探</h3>
<p>首先看看Deng data，但还没有应用复杂的伪时间方法。如下图所示，简单的PCA可以很好地显示这些数据中的结构。PCA分离的效果不错，除了囊胚细胞类型(“早期囊胚”, “中期囊胚”, “晚期囊胚”)混杂在一起。</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" title="1">deng_SCE &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/deng/deng-reads.rds&quot;</span>)</a>
<a class="sourceLine" id="cb48-2" title="2">deng_SCE<span class="op">$</span>cell_type2 &lt;-<span class="st"> </span><span class="kw">factor</span>(</a>
<a class="sourceLine" id="cb48-3" title="3">    deng_SCE<span class="op">$</span>cell_type2,</a>
<a class="sourceLine" id="cb48-4" title="4">    <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;zy&quot;</span>, <span class="st">&quot;early2cell&quot;</span>, <span class="st">&quot;mid2cell&quot;</span>, <span class="st">&quot;late2cell&quot;</span>,</a>
<a class="sourceLine" id="cb48-5" title="5">                        <span class="st">&quot;4cell&quot;</span>, <span class="st">&quot;8cell&quot;</span>, <span class="st">&quot;16cell&quot;</span>, <span class="st">&quot;earlyblast&quot;</span>,</a>
<a class="sourceLine" id="cb48-6" title="6">                        <span class="st">&quot;midblast&quot;</span>, <span class="st">&quot;lateblast&quot;</span>)</a>
<a class="sourceLine" id="cb48-7" title="7">)</a>
<a class="sourceLine" id="cb48-8" title="8">cellLabels &lt;-<span class="st"> </span>deng_SCE<span class="op">$</span>cell_type2</a>
<a class="sourceLine" id="cb48-9" title="9">deng &lt;-<span class="st"> </span><span class="kw">counts</span>(deng_SCE)</a>
<a class="sourceLine" id="cb48-10" title="10"><span class="kw">colnames</span>(deng) &lt;-<span class="st"> </span>cellLabels</a>
<a class="sourceLine" id="cb48-11" title="11">deng_SCE &lt;-<span class="st"> </span><span class="kw">runPCA</span>(deng_SCE)</a>
<a class="sourceLine" id="cb48-12" title="12"><span class="kw">plotPCA</span>(deng_SCE, <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/data-overview-1.png" style="display: block; margin: auto;" />
PCA为评估不同的伪时间方法提供了基准。对于一个非常粗糙的伪时间分析，我们可以只取第一个主成分的坐标。</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1">deng_SCE<span class="op">$</span>PC1 &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(deng_SCE, <span class="st">&quot;PCA&quot;</span>)[,<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb49-2" title="2"><span class="kw">ggplot</span>(<span class="kw">as.data.frame</span>(<span class="kw">colData</span>(deng_SCE)), <span class="kw">aes</span>(<span class="dt">x =</span> PC1, <span class="dt">y =</span> cell_type2, </a>
<a class="sourceLine" id="cb49-3" title="3">                              <span class="dt">colour =</span> cell_type2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb49-4" title="4"><span class="st">    </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb49-5" title="5"><span class="st">    </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb49-6" title="6"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;First principal component&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Timepoint&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb49-7" title="7"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Cells ordered by first principal component&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/pca-pseudotime-1.png" style="display: block; margin: auto;" /></p>
<p>如上图所示，PC1在发育时间段的早期和晚期都难以对细胞进行正确排序，但总的来说，按照发育时间对细胞排序效果不错。</p>
<p>定制的伪时间分析方法比粗糙引用PCA效果好吗？</p>
</div>
<div id="tscan" class="section level3">
<h3><span class="header-section-number">8.4.2</span> TSCAN</h3>
<p>TSCAN将聚类与伪时间分析相结合。首先，基于混合正态分布使用<code>mclust</code>聚类细胞。构建一个最小生成树来连接cluster。 连接最大数量cluster的分支是用于确定伪时间的主分支。</p>
<p>首先使用所有基因对细胞排序。</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" title="1">procdeng &lt;-<span class="st"> </span>TSCAN<span class="op">::</span><span class="kw">preprocess</span>(deng)</a>
<a class="sourceLine" id="cb50-2" title="2"><span class="kw">colnames</span>(procdeng) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(deng)</a>
<a class="sourceLine" id="cb50-3" title="3">dengclust &lt;-<span class="st"> </span>TSCAN<span class="op">::</span><span class="kw">exprmclust</span>(procdeng, <span class="dt">clusternum =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb50-4" title="4">TSCAN<span class="op">::</span><span class="kw">plotmclust</span>(dengclust)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/tscan-all-genes-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" title="1">dengorderTSCAN &lt;-<span class="st"> </span>TSCAN<span class="op">::</span><span class="kw">TSCANorder</span>(dengclust, <span class="dt">orderonly =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb51-2" title="2">pseudotime_order_tscan &lt;-<span class="st"> </span><span class="kw">as.character</span>(dengorderTSCAN<span class="op">$</span>sample_name)</a>
<a class="sourceLine" id="cb51-3" title="3">deng_SCE<span class="op">$</span>pseudotime_order_tscan &lt;-<span class="st"> </span><span class="ot">NA</span></a>
<a class="sourceLine" id="cb51-4" title="4">deng_SCE<span class="op">$</span>pseudotime_order_tscan[<span class="kw">as.numeric</span>(dengorderTSCAN<span class="op">$</span>sample_name)] &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb51-5" title="5"><span class="st">    </span>dengorderTSCAN<span class="op">$</span>Pseudotime</a></code></pre></div>
<p>令人沮丧的是，TSCAN仅为268细胞中221个细胞提供伪时间值，静默返回未指定细胞的缺失值。</p>
<p>检查哪些时间点分配给什么状态：</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" title="1">cellLabels[dengclust<span class="op">$</span>clusterid <span class="op">==</span><span class="st"> </span><span class="dv">10</span>]</a>
<a class="sourceLine" id="cb52-2" title="2"><span class="kw">ggplot</span>(<span class="kw">as.data.frame</span>(<span class="kw">colData</span>(deng_SCE)), </a>
<a class="sourceLine" id="cb52-3" title="3">       <span class="kw">aes</span>(<span class="dt">x =</span> pseudotime_order_tscan, </a>
<a class="sourceLine" id="cb52-4" title="4">           <span class="dt">y =</span> cell_type2, <span class="dt">colour =</span> cell_type2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb52-5" title="5"><span class="st">    </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb52-6" title="6"><span class="st">    </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb52-7" title="7"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;TSCAN pseudotime&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Timepoint&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb52-8" title="8"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Cells ordered by TSCAN pseudotime&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/tscan-vs-truth-1.png" style="display: block; margin: auto;" /></p>
<p>TSCAN的发展轨迹是“错误的方式”，因为后来的伪时间值对应于早期时间点，反之亦然。 这本身并不是一个问题（通过翻转排序以获得伪时间的直观解释），但总的来说，TSCAN在此数据集上的性能优于PCA（并不奇怪，因为这是一种基于PCA的方法）</p>
<p><strong>练习1</strong> 比较不同聚类数目的结果(<code>clusternum</code>).</p>
</div>
<div id="monocle" class="section level3">
<h3><span class="header-section-number">8.4.3</span> monocle</h3>
<p>Monocle跳过TSCAN的聚类阶段，并直接在细胞的降维尺度构建最小生成树以连接所有细胞。然后Monocle识别该树中最长的路径以确定伪时间。 Monocle可以识别发散轨迹（即一种细胞类型分化为两种不同的细胞类型）。每个分叉路径定义为单独的细胞状态。</p>
<p>不幸的是，Monocle在使用所有基因时不起作用，因此必须进行特征选择。首先，使用M3Drop：</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" title="1">m3dGenes &lt;-<span class="st"> </span><span class="kw">as.character</span>(</a>
<a class="sourceLine" id="cb53-2" title="2">    <span class="kw">M3DropFeatureSelection</span>(deng)<span class="op">$</span>Gene</a>
<a class="sourceLine" id="cb53-3" title="3">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/m3d-select-genes-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" title="1">d &lt;-<span class="st"> </span>deng[<span class="kw">which</span>(<span class="kw">rownames</span>(deng) <span class="op">%in%</span><span class="st"> </span>m3dGenes), ]</a>
<a class="sourceLine" id="cb54-2" title="2">d &lt;-<span class="st"> </span>d[<span class="op">!</span><span class="kw">duplicated</span>(<span class="kw">rownames</span>(d)), ]</a></code></pre></div>
<p>运行monocle:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" title="1"><span class="kw">colnames</span>(d) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(d)</a>
<a class="sourceLine" id="cb55-2" title="2">geneNames &lt;-<span class="st"> </span><span class="kw">rownames</span>(d)</a>
<a class="sourceLine" id="cb55-3" title="3"><span class="kw">rownames</span>(d) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(d)</a>
<a class="sourceLine" id="cb55-4" title="4">pd &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">timepoint =</span> cellLabels)</a>
<a class="sourceLine" id="cb55-5" title="5">pd &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;AnnotatedDataFrame&quot;</span>, <span class="dt">data=</span>pd)</a>
<a class="sourceLine" id="cb55-6" title="6">fd &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gene_short_name =</span> geneNames)</a>
<a class="sourceLine" id="cb55-7" title="7">fd &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;AnnotatedDataFrame&quot;</span>, <span class="dt">data=</span>fd)</a>
<a class="sourceLine" id="cb55-8" title="8"></a>
<a class="sourceLine" id="cb55-9" title="9">dCellData &lt;-<span class="st"> </span><span class="kw">newCellDataSet</span>(d, <span class="dt">phenoData =</span> pd, <span class="dt">featureData =</span> fd, <span class="dt">expressionFamily =</span> <span class="kw">tobit</span>())</a>
<a class="sourceLine" id="cb55-10" title="10">dCellData &lt;-<span class="st"> </span><span class="kw">setOrderingFilter</span>(dCellData, <span class="kw">which</span>(geneNames <span class="op">%in%</span><span class="st"> </span>m3dGenes))</a>
<a class="sourceLine" id="cb55-11" title="11">dCellData &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(dCellData)</a>
<a class="sourceLine" id="cb55-12" title="12">dCellDataSet &lt;-<span class="st"> </span><span class="kw">reduceDimension</span>(dCellData, <span class="dt">pseudo_expr =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb55-13" title="13">dCellDataSet &lt;-<span class="st"> </span><span class="kw">orderCells</span>(dCellDataSet, <span class="dt">reverse =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb55-14" title="14"><span class="kw">plot_cell_trajectory</span>(dCellDataSet)</a>
<a class="sourceLine" id="cb55-15" title="15"><span class="co"># Store the ordering</span></a>
<a class="sourceLine" id="cb55-16" title="16">pseudotime_monocle &lt;-</a>
<a class="sourceLine" id="cb55-17" title="17"><span class="st">    </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb55-18" title="18">        <span class="dt">Timepoint =</span> <span class="kw">phenoData</span>(dCellDataSet)<span class="op">$</span>timepoint,</a>
<a class="sourceLine" id="cb55-19" title="19">        <span class="dt">pseudotime =</span> <span class="kw">phenoData</span>(dCellDataSet)<span class="op">$</span>Pseudotime,</a>
<a class="sourceLine" id="cb55-20" title="20">        <span class="dt">State =</span> <span class="kw">phenoData</span>(dCellDataSet)<span class="op">$</span>State</a>
<a class="sourceLine" id="cb55-21" title="21">    )</a>
<a class="sourceLine" id="cb55-22" title="22"><span class="kw">rownames</span>(pseudotime_monocle) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(d)</a>
<a class="sourceLine" id="cb55-23" title="23">pseudotime_order_monocle &lt;-</a>
<a class="sourceLine" id="cb55-24" title="24"><span class="st">    </span><span class="kw">rownames</span>(pseudotime_monocle[<span class="kw">order</span>(pseudotime_monocle<span class="op">$</span>pseudotime), ])</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/monocle-all-genes-1.png" style="display: block; margin: auto;" />
将推断的伪时间与已知的采样时间点进行比较。</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" title="1">deng_SCE<span class="op">$</span>pseudotime_monocle &lt;-<span class="st"> </span>pseudotime_monocle<span class="op">$</span>pseudotime</a>
<a class="sourceLine" id="cb56-2" title="2"><span class="kw">ggplot</span>(<span class="kw">as.data.frame</span>(<span class="kw">colData</span>(deng_SCE)), </a>
<a class="sourceLine" id="cb56-3" title="3">       <span class="kw">aes</span>(<span class="dt">x =</span> pseudotime_monocle, </a>
<a class="sourceLine" id="cb56-4" title="4">           <span class="dt">y =</span> cell_type2, <span class="dt">colour =</span> cell_type2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb56-5" title="5"><span class="st">    </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb56-6" title="6"><span class="st">    </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb56-7" title="7"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;monocle pseudotime&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Timepoint&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb56-8" title="8"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Cells ordered by monocle pseudotime&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/monocle-vs-truth-1.png" style="display: block; margin: auto;" /></p>
<p>Monocle - 至少是默认设置 - 在这些数据上表现不佳。 “late2cell”与“zy”，“early2cell”和“mid2cell”细胞完全分离（尽管这些是正确排序的），并且“4cell”，“8cell”，“16cell”及其它囊胚期细胞完全没有分离。</p>
</div>
<div id="diffusion-maps" class="section level3">
<h3><span class="header-section-number">8.4.4</span> Diffusion maps</h3>
<p><a href="https://en.wikipedia.org/wiki/Diffusion_map">Diffusion maps</a>由 <a href="http://www.sciencedirect.com/science/article/pii/S1063520306000546">Ronald Coifman和Stephane Lafon</a>引入，其基本思想是假设数据是来自扩散过程的样本。 该方法通过估计与数据相关的扩散算子的特征值和特征向量来推断低维流形。</p>
<p><a href="https://academic.oup.com/bioinformatics/article/32/8/1241/1744143">Angerer et al</a>将diffusion map概念应用于scRNA-seq数据的分析，创建 <a href="http://bioconductor.org/packages/destiny">destiny</a>的R包。</p>
<p>我们将在这里将第一扩散映射分量中的细胞的顺序作为“diffusion map pseudotime”。</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">logcounts</span>(deng_SCE)</a>
<a class="sourceLine" id="cb57-2" title="2"><span class="kw">colnames</span>(deng) &lt;-<span class="st"> </span>cellLabels</a>
<a class="sourceLine" id="cb57-3" title="3">dm &lt;-<span class="st"> </span><span class="kw">DiffusionMap</span>(<span class="kw">t</span>(deng))</a>
<a class="sourceLine" id="cb57-4" title="4"></a>
<a class="sourceLine" id="cb57-5" title="5">tmp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">DC1 =</span> <span class="kw">eigenvectors</span>(dm)[,<span class="dv">1</span>],</a>
<a class="sourceLine" id="cb57-6" title="6">                  <span class="dt">DC2 =</span> <span class="kw">eigenvectors</span>(dm)[,<span class="dv">2</span>],</a>
<a class="sourceLine" id="cb57-7" title="7">                  <span class="dt">Timepoint =</span> deng_SCE<span class="op">$</span>cell_type2)</a>
<a class="sourceLine" id="cb57-8" title="8"><span class="kw">ggplot</span>(tmp, <span class="kw">aes</span>(<span class="dt">x =</span> DC1, <span class="dt">y =</span> DC2, <span class="dt">colour =</span> Timepoint)) <span class="op">+</span></a>
<a class="sourceLine" id="cb57-9" title="9"><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb57-10" title="10"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Diffusion component 1&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb57-11" title="11"><span class="st">    </span><span class="kw">ylab</span>(<span class="st">&quot;Diffusion component 2&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb57-12" title="12"><span class="st">    </span><span class="kw">theme_classic</span>()</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/destiny-deng-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" title="1">deng_SCE<span class="op">$</span>pseudotime_diffusionmap &lt;-<span class="st"> </span><span class="kw">rank</span>(<span class="kw">eigenvectors</span>(dm)[,<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb58-2" title="2"><span class="kw">ggplot</span>(<span class="kw">as.data.frame</span>(<span class="kw">colData</span>(deng_SCE)), </a>
<a class="sourceLine" id="cb58-3" title="3">       <span class="kw">aes</span>(<span class="dt">x =</span> pseudotime_diffusionmap, </a>
<a class="sourceLine" id="cb58-4" title="4">           <span class="dt">y =</span> cell_type2, <span class="dt">colour =</span> cell_type2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-5" title="5"><span class="st">    </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-6" title="6"><span class="st">    </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb58-7" title="7"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Diffusion map pseudotime (first diffusion map component)&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-8" title="8"><span class="st">    </span><span class="kw">ylab</span>(<span class="st">&quot;Timepoint&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-9" title="9"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Cells ordered by diffusion map pseudotime&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/destiny-deng-2.png" style="display: block; margin: auto;" /></p>
<p>与其他方法一样，使用destiny中的第一个扩散映射成分作为伪时间可以很好地排序早期时间点，但它无法很好区分后者。</p>
<p><strong>练习2</strong> 通过考虑额外的特征向量，是否在较晚的时间点获得更好的分辨率？
<strong>练习3</strong> 如果只是用M3Drop识别的基因，顺序如何改变？</p>
</div>
<div id="slicer" class="section level3">
<h3><span class="header-section-number">8.4.5</span> SLICER</h3>
<p>SLICER方法是用于构建轨迹的算法，其描述在连续生物过程期间的基因表达变化，就像Monocle和TSCAN。 SLICER旨在捕获高度非线性基因表达变化，自动选择与过程相关的基因，并检测轨迹中的多个分支和环路特征<span class="citation">(Welch, Hartemink, and Prins <a href="#ref-Welch2016-jr">2016</a>)</span>。SLICER R软件包可从GitHub获得，可以使用<code>devtools</code>安装。</p>
<p>使用SLICER中的<code>select_genes</code>函数自动选择用于构建细胞轨迹的基因。 该函数使用“neighbourhood variance”来识别在一组细胞中平滑变化而不是随机波动的基因。然后确定哪个“k”值（最近邻居的数量）产生最类似于轨迹的embedding。然后估计细胞的 <a href="https://en.wikipedia.org/wiki/Nonlinear_dimensionality_reduction">locally linear embedding</a>。</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" title="1"><span class="kw">library</span>(<span class="st">&quot;lle&quot;</span>)</a>
<a class="sourceLine" id="cb59-2" title="2">slicer_genes &lt;-<span class="st"> </span><span class="kw">select_genes</span>(<span class="kw">t</span>(deng))</a>
<a class="sourceLine" id="cb59-3" title="3">k &lt;-<span class="st"> </span><span class="kw">select_k</span>(<span class="kw">t</span>(deng[slicer_genes,]), <span class="dt">kmin =</span> <span class="dv">30</span>, <span class="dt">kmax=</span><span class="dv">60</span>)</a>
<a class="sourceLine" id="cb59-4" title="4">slicer_traj_lle &lt;-<span class="st"> </span><span class="kw">lle</span>(<span class="kw">t</span>(deng[slicer_genes,]), <span class="dt">m =</span> <span class="dv">2</span>, k)<span class="op">$</span>Y</a>
<a class="sourceLine" id="cb59-5" title="5"><span class="kw">reducedDim</span>(deng_SCE, <span class="st">&quot;LLE&quot;</span>) &lt;-<span class="st"> </span>slicer_traj_lle</a>
<a class="sourceLine" id="cb59-6" title="6"><span class="kw">plotReducedDim</span>(deng_SCE, <span class="dt">use_dimred =</span> <span class="st">&quot;LLE&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb59-7" title="7"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;LLE component 1&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;LLE component 2&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb59-8" title="8"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Locally linear embedding of cells from SLICER&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/slicer-analyis-1.png" style="display: block; margin: auto;" /></p>
<p>通过计算locally linear embedding，构建全连接的k-最近邻图。每个细胞用黄色圆圈表示，细胞ID蓝字位于中间。这里显示使用10个最近邻居计算的图表。这里SLICER似乎通过一个分支检测到一个主要轨迹。</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb60-1" title="1">slicer_traj_graph &lt;-<span class="st"> </span><span class="kw">conn_knn_graph</span>(slicer_traj_lle, <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb60-2" title="2"><span class="kw">plot</span>(slicer_traj_graph, <span class="dt">main =</span> <span class="st">&quot;Fully connected kNN graph from SLICER&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/slicer-build-graph-1.png" style="display: block; margin: auto;" /></p>
<p>从该图可以congoing轨迹中起始/结束细胞的识别“极端”细胞。</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" title="1">ends &lt;-<span class="st"> </span><span class="kw">find_extreme_cells</span>(slicer_traj_graph, slicer_traj_lle)</a>
<a class="sourceLine" id="cb61-2" title="2">start &lt;-<span class="st"> </span>ends[<span class="dv">1</span>]</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/slicer-1.png" style="display: block; margin: auto;" /></p>
<p>定义起始细胞后，使用估计伪时间对细胞进行排序。</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb62-1" title="1">pseudotime_order_slicer &lt;-<span class="st"> </span><span class="kw">cell_order</span>(slicer_traj_graph, start)</a>
<a class="sourceLine" id="cb62-2" title="2">branches &lt;-<span class="st"> </span><span class="kw">assign_branches</span>(slicer_traj_graph, start)</a>
<a class="sourceLine" id="cb62-3" title="3"></a>
<a class="sourceLine" id="cb62-4" title="4">pseudotime_slicer &lt;-</a>
<a class="sourceLine" id="cb62-5" title="5"><span class="st">    </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb62-6" title="6">        <span class="dt">Timepoint =</span> cellLabels,</a>
<a class="sourceLine" id="cb62-7" title="7">        <span class="dt">pseudotime =</span> <span class="ot">NA</span>,</a>
<a class="sourceLine" id="cb62-8" title="8">        <span class="dt">State =</span> branches</a>
<a class="sourceLine" id="cb62-9" title="9">    )</a>
<a class="sourceLine" id="cb62-10" title="10">pseudotime_slicer<span class="op">$</span>pseudotime[pseudotime_order_slicer] &lt;-</a>
<a class="sourceLine" id="cb62-11" title="11"><span class="st">    </span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(pseudotime_order_slicer)</a>
<a class="sourceLine" id="cb62-12" title="12">deng_SCE<span class="op">$</span>pseudotime_slicer &lt;-<span class="st"> </span>pseudotime_slicer<span class="op">$</span>pseudotime</a></code></pre></div>
<p>将推断的伪时间与已知的采样时间点进行比较，SLICER本身不提供伪时间值，只是细胞的排序。</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" title="1"><span class="kw">ggplot</span>(<span class="kw">as.data.frame</span>(<span class="kw">colData</span>(deng_SCE)), </a>
<a class="sourceLine" id="cb63-2" title="2">       <span class="kw">aes</span>(<span class="dt">x =</span> pseudotime_slicer, </a>
<a class="sourceLine" id="cb63-3" title="3">           <span class="dt">y =</span> cell_type2, <span class="dt">colour =</span> cell_type2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb63-4" title="4"><span class="st">    </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb63-5" title="5"><span class="st">    </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb63-6" title="6"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;SLICER pseudotime (cell ordering)&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb63-7" title="7"><span class="st">    </span><span class="kw">ylab</span>(<span class="st">&quot;Timepoint&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb63-8" title="8"><span class="st">    </span><span class="kw">theme_classic</span>()</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/slicer-vs-truth-1.png" style="display: block; margin: auto;" />
与之前的方法一样，SLICER对早期时间点进行正确的排序，但是其将“16cell”期置于“8cell”期之前，但是对于囊胚细胞提供了比许多早期方法更好的排序。</p>
<p><strong>练习4</strong> 当k不同时，结果如何变化(比如k=5)，如何在调用<code>conn_knn_graph</code>时改变最近邻数目？</p>
<p><strong>练习5</strong> 当使用SLICER选择的另一组基因(比如M3Drop鉴定的基因)，轨迹如何变化？</p>
</div>
<div id="ouija" class="section level3">
<h3><span class="header-section-number">8.4.6</span> Ouija</h3>
<p><a href="http://kieranrcampbell.github.io/ouija/">Ouija</a>采用了与迄今为所研究的伪时间估计方法不同的策略。 早期的方法都是“无监督的”，也就是说，除了可能选择informative genes之外，我们不提供任何关于期望某些基因或整个轨迹表现的先验信息。</p>
<p>相比之下，Ouija是一种概率框架，仅使用一小组marker基因来解释学习single-cell pseudotimes。 这个方法：</p>
<ul>
<li>从少数标记基因中推断出pseudotimes，让您了解为什么pseudotimes从这些基因中学习出来;</li>
<li>为解释基因调控行为（如峰值时间或上调时间）提供参数估计（具有不确定性）;</li>
<li>贝叶斯假设检验发现调控基因</li>
<li>识别亚稳态，即沿连续轨迹的离散细胞类型。</li>
</ul>
<p>将向Ouija提供以下标记基因（以及高度表达的预计时间点）：</p>
<ul>
<li>早期时间点: Dazl, Rnf17, Sycp3, Nanog, Pou5f1, Fgf8, Egfr, Bmp5, Bmp15</li>
<li>中期时间点: Zscan4b, Foxa1, Prdm14, Sox21</li>
<li>晚期时间点: Creb3, Gpx4, Krt8, Elf5, Eomes, Cdx2, Tdgf1, Gdf3</li>
</ul>
<p>使用Ouija，我们可以将基因建模为表现为单调上调或下调（称为开关行为），或基因短暂达到峰值的瞬态行为。 默认情况下，Ouija假设所有基因都表现出类似开关的行为（作者保证我们不要担心，如果我们弄错了，噪声模型意味着错误地指定瞬态基因，因为类似开关的影响很小）。</p>
<p>在这里，我们可以“作弊”，并检查我们选择的标记基因确实确定了分化过程的不同时间点。</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" title="1">ouija_markers_down &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Dazl&quot;</span>, <span class="st">&quot;Rnf17&quot;</span>, <span class="st">&quot;Sycp3&quot;</span>, <span class="st">&quot;Fgf8&quot;</span>, </a>
<a class="sourceLine" id="cb64-2" title="2">                        <span class="st">&quot;Egfr&quot;</span>, <span class="st">&quot;Bmp5&quot;</span>, <span class="st">&quot;Bmp15&quot;</span>, <span class="st">&quot;Pou5f1&quot;</span>)</a>
<a class="sourceLine" id="cb64-3" title="3">ouija_markers_up &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Creb3&quot;</span>, <span class="st">&quot;Gpx4&quot;</span>, <span class="st">&quot;Krt8&quot;</span>, <span class="st">&quot;Elf5&quot;</span>, <span class="st">&quot;Cdx2&quot;</span>, </a>
<a class="sourceLine" id="cb64-4" title="4">                      <span class="st">&quot;Tdgf1&quot;</span>, <span class="st">&quot;Gdf3&quot;</span>, <span class="st">&quot;Eomes&quot;</span>)</a>
<a class="sourceLine" id="cb64-5" title="5">ouija_markers_transient &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Zscan4b&quot;</span>, <span class="st">&quot;Foxa1&quot;</span>, <span class="st">&quot;Prdm14&quot;</span>, <span class="st">&quot;Sox21&quot;</span>)</a>
<a class="sourceLine" id="cb64-6" title="6">ouija_markers &lt;-<span class="st"> </span><span class="kw">c</span>(ouija_markers_down, ouija_markers_up, </a>
<a class="sourceLine" id="cb64-7" title="7">                   ouija_markers_transient)</a>
<a class="sourceLine" id="cb64-8" title="8"><span class="kw">plotExpression</span>(deng_SCE, ouija_markers, <span class="dt">x =</span> <span class="st">&quot;cell_type2&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb64-9" title="9"><span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">60</span>, <span class="dt">hjust =</span> <span class="dv">1</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-response-type-1.png" style="display: block; margin: auto;" />
为了拟合pseudotimes，只需调用<code>ouija</code>，传递预期的响应类型。 注意，如果没有提供响应类型，则默认情况下它们都被假定为类似开关。 Ouija的输入可以是非负表达值的cell-by-gene矩阵，也可以是ExpressionSet对象，或者SingleCellExperiment对象中选择<code>logcounts</code>值。</p>
<p>我们可以应用关于基因在分化过程中是上调还是下调的先验信息，并且还提供关于何时可能发生表达转换或表达峰值的先验信息。</p>
<p>使用下列拟合Ouija模型：</p>
<ul>
<li>Hamiltonian Monte Carlo (HMC) - MCMC推断，其中log-posterior梯度信息用于“引导”随机遍历参数空间，或</li>
<li>Automatic Differentiation Variational Bayes (ADVI or simply VI) - 近似推断，其中最小化KL近似分布偏差。</li>
</ul>
<p>一般而言，HMC将为所有参数提供更准确的推理和近似正确的后验方差。 然而，VB比HMC快几个数量级，虽然它可能低估了后验方差，但是Ouija的作者认为，通常情况下，其在发现后验pseudotimes和HMC效果相同。</p>
<p>为了帮助Ouija模型，我们为它提供了关于上调和下调基因的开关强度的先验信息。 通过将下调基因的开关强度设置为-10，将上调基因的开关强度设置为10，先验强度标准差为0.5，告诉模型我们对这些基因在分化过程中的预期行为充满信心。</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" title="1"><span class="kw">options</span>(<span class="dt">mc.cores =</span> parallel<span class="op">::</span><span class="kw">detectCores</span>())</a>
<a class="sourceLine" id="cb65-2" title="2">response_type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;switch&quot;</span>, <span class="kw">length</span>(ouija_markers_down) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb65-3" title="3"><span class="st">                           </span><span class="kw">length</span>(ouija_markers_up)), </a>
<a class="sourceLine" id="cb65-4" title="4">                   <span class="kw">rep</span>(<span class="st">&quot;transient&quot;</span>, <span class="kw">length</span>(ouija_markers_transient)))</a>
<a class="sourceLine" id="cb65-5" title="5">switch_strengths &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="op">-</span><span class="dv">10</span>, <span class="kw">length</span>(ouija_markers_down)),</a>
<a class="sourceLine" id="cb65-6" title="6">                      <span class="kw">rep</span>(<span class="dv">10</span>, <span class="kw">length</span>(ouija_markers_up)))</a>
<a class="sourceLine" id="cb65-7" title="7">switch_strength_sd &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="fl">0.5</span>, <span class="kw">length</span>(ouija_markers_down)),</a>
<a class="sourceLine" id="cb65-8" title="8">                      <span class="kw">rep</span>(<span class="fl">0.5</span>, <span class="kw">length</span>(ouija_markers_up)))</a>
<a class="sourceLine" id="cb65-9" title="9">garbage &lt;-<span class="st"> </span><span class="kw">capture.output</span>(</a>
<a class="sourceLine" id="cb65-10" title="10">    oui_vb &lt;-<span class="st"> </span><span class="kw">ouija</span>(deng_SCE[ouija_markers,],</a>
<a class="sourceLine" id="cb65-11" title="11">                    <span class="dt">single_cell_experiment_assay =</span> <span class="st">&quot;logcounts&quot;</span>, </a>
<a class="sourceLine" id="cb65-12" title="12">                    <span class="dt">response_type =</span> response_type,</a>
<a class="sourceLine" id="cb65-13" title="13">                    <span class="dt">switch_strengths =</span> switch_strengths,</a>
<a class="sourceLine" id="cb65-14" title="14">                    <span class="dt">switch_strength_sd =</span> switch_strength_sd,</a>
<a class="sourceLine" id="cb65-15" title="15">                    <span class="dt">inference_type =</span> <span class="st">&quot;vb&quot;</span>)</a>
<a class="sourceLine" id="cb65-16" title="16">)</a>
<a class="sourceLine" id="cb65-17" title="17"></a>
<a class="sourceLine" id="cb65-18" title="18"><span class="kw">print</span>(oui_vb)</a></code></pre></div>
<p>使用plot_expression函数绘制伪时间基因表达以及平均函数（sigmoid或高斯瞬态函数）的最大后验（MAP）估计。</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb66-1" title="1"><span class="kw">plot_expression</span>(oui_vb)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-plot-exprs-1.png" style="display: block; margin: auto;" /></p>
<p>使用plot_switch_times和plot_transient_times函数，在轨迹基因调控行为发生时，以switch时间或峰值时间（对于类似开关或瞬态基因）的形式进行可视化：</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" title="1"><span class="kw">plot_switch_times</span>(oui_vb)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-plot-switch-times-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb68-1" title="1"><span class="kw">plot_peak_times</span>(oui_vb)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-plot-switch-times-2.png" style="display: block; margin: auto;" />
使用一致性矩阵识别亚稳态。</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" title="1">cmo &lt;-<span class="st"> </span><span class="kw">consistency_matrix</span>(oui_vb)</a>
<a class="sourceLine" id="cb69-2" title="2"><span class="kw">plot_consistency</span>(oui_vb)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-consistency-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb70-1" title="1">cell_classifications &lt;-<span class="st"> </span><span class="kw">cluster_consistency</span>(cmo)</a></code></pre></div>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb71-1" title="1">map_pst &lt;-<span class="st"> </span><span class="kw">map_pseudotime</span>(oui_vb)</a>
<a class="sourceLine" id="cb71-2" title="2">ouija_pseudotime &lt;-<span class="st"> </span><span class="kw">data.frame</span>(map_pst, cell_classifications)</a>
<a class="sourceLine" id="cb71-3" title="3"></a>
<a class="sourceLine" id="cb71-4" title="4"><span class="kw">ggplot</span>(ouija_pseudotime, <span class="kw">aes</span>(<span class="dt">x =</span> map_pst, <span class="dt">y =</span> cell_classifications)) <span class="op">+</span></a>
<a class="sourceLine" id="cb71-5" title="5"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb71-6" title="6"><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;MAP pseudotime&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb71-7" title="7"><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Cell classification&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-pseudotime-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb72-1" title="1">deng_SCE<span class="op">$</span>pseudotime_ouija &lt;-<span class="st"> </span>ouija_pseudotime<span class="op">$</span>map_pst</a>
<a class="sourceLine" id="cb72-2" title="2">deng_SCE<span class="op">$</span>ouija_cell_class &lt;-<span class="st"> </span>ouija_pseudotime<span class="op">$</span>cell_classifications</a>
<a class="sourceLine" id="cb72-3" title="3"></a>
<a class="sourceLine" id="cb72-4" title="4"><span class="kw">ggplot</span>(<span class="kw">as.data.frame</span>(<span class="kw">colData</span>(deng_SCE)), </a>
<a class="sourceLine" id="cb72-5" title="5">       <span class="kw">aes</span>(<span class="dt">x =</span> pseudotime_ouija, </a>
<a class="sourceLine" id="cb72-6" title="6">           <span class="dt">y =</span> cell_type2, <span class="dt">colour =</span> cell_type2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb72-7" title="7"><span class="st">    </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb72-8" title="8"><span class="st">    </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb72-9" title="9"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Ouija pseudotime&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb72-10" title="10"><span class="st">    </span><span class="kw">ylab</span>(<span class="st">&quot;Timepoint&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb72-11" title="11"><span class="st">    </span><span class="kw">theme_classic</span>()</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-pseudotime-2.png" style="display: block; margin: auto;" /></p>
<p>虽然Ouija对标记基因的选择和提供的先验信息敏感，但Ouija在细胞的排序方面表现相当不错。 如果选择不同的标记基因或更改先验信息，结果会如何变化？</p>
<p>Ouija识别四种亚稳态，将其注释为“zygote/2cell”, “4/8/16 cell”, “blast1” and “blast2”。</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb73-1" title="1"><span class="kw">ggplot</span>(<span class="kw">as.data.frame</span>(<span class="kw">colData</span>(deng_SCE)), </a>
<a class="sourceLine" id="cb73-2" title="2">       <span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">as.factor</span>(ouija_cell_class), </a>
<a class="sourceLine" id="cb73-3" title="3">           <span class="dt">y =</span> pseudotime_ouija, <span class="dt">colour =</span> cell_type2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb73-4" title="4"><span class="st">    </span><span class="kw">geom_boxplot</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb73-5" title="5"><span class="st">    </span><span class="kw">coord_flip</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb73-6" title="6"><span class="st">    </span><span class="kw">scale_color_tableau</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb73-7" title="7"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Ouija cell classification&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb73-8" title="8"><span class="st">    </span><span class="kw">ylab</span>(<span class="st">&quot;Ouija pseudotime&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb73-9" title="9"><span class="st">    </span><span class="kw">theme_classic</span>()</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/ouija-states-1.png" style="display: block; margin: auto;" />
一个常见的分析是基因的调节顺序。 例如，基因A在基因B之前是否上调？ 基因C在基因D下调前是否达到峰值？ Ouija根据贝叶斯假设检验来回答这些问题，该检验调控时间的差异（switch时间或峰值时间）是否和0有显著差异。</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" title="1">gene_regs &lt;-<span class="st"> </span><span class="kw">gene_regulation</span>(oui_vb)</a>
<a class="sourceLine" id="cb74-2" title="2"><span class="kw">head</span>(gene_regs)</a>
<a class="sourceLine" id="cb74-3" title="3"><span class="co">## # A tibble: 6 x 7</span></a>
<a class="sourceLine" id="cb74-4" title="4"><span class="co">## # Groups:   label, gene_A [6]</span></a>
<a class="sourceLine" id="cb74-5" title="5"><span class="co">##   label         gene_A gene_B mean_difference lower_95 upper_95 significant</span></a>
<a class="sourceLine" id="cb74-6" title="6"><span class="co">##   &lt;chr&gt;         &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;lgl&gt;      </span></a>
<a class="sourceLine" id="cb74-7" title="7"><span class="co">## 1 Bmp15 - Cdx2  Bmp15  Cdx2           -0.0658 -1.17e-1  -0.0136 TRUE       </span></a>
<a class="sourceLine" id="cb74-8" title="8"><span class="co">## 2 Bmp15 - Creb3 Bmp15  Creb3           0.266   1.96e-1   0.321  TRUE       </span></a>
<a class="sourceLine" id="cb74-9" title="9"><span class="co">## 3 Bmp15 - Elf5  Bmp15  Elf5           -0.665  -7.03e-1  -0.622  TRUE       </span></a>
<a class="sourceLine" id="cb74-10" title="10"><span class="co">## 4 Bmp15 - Eomes Bmp15  Eomes           0.0765 -1.30e-2   0.150  FALSE      </span></a>
<a class="sourceLine" id="cb74-11" title="11"><span class="co">## 5 Bmp15 - Foxa1 Bmp15  Foxa1          -0.0157 -4.95e-2   0.0128 FALSE      </span></a>
<a class="sourceLine" id="cb74-12" title="12"><span class="co">## 6 Bmp15 - Gdf3  Bmp15  Gdf3            0.0592  2.06e-4   0.114  TRUE</span></a></code></pre></div>
<p>可以从Ouija的基因调控输出中得出什么结论？</p>
<p>尝试HMC推断，查看Ouija结果是否变化？</p>
</div>
<div id="comparison-of-the-methods" class="section level3">
<h3><span class="header-section-number">8.4.7</span> 方法比较</h3>
<p>如何比较TSCAN，Monocle，Diffusion Map，SLICER和Ouija推断出的轨迹？</p>
<p>TSCAN和Diffusion Map方法轨迹“方向错误”，比对之前先对其进行调整。</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb75-1" title="1">df_pseudotime &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(</a>
<a class="sourceLine" id="cb75-2" title="2">    <span class="kw">colData</span>(deng_SCE)[, <span class="kw">grep</span>(<span class="st">&quot;pseudotime&quot;</span>, <span class="kw">colnames</span>(<span class="kw">colData</span>(deng_SCE)))]</a>
<a class="sourceLine" id="cb75-3" title="3">)</a>
<a class="sourceLine" id="cb75-4" title="4"><span class="kw">colnames</span>(df_pseudotime) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;pseudotime_&quot;</span>, <span class="st">&quot;&quot;</span>, </a>
<a class="sourceLine" id="cb75-5" title="5">                                <span class="kw">colnames</span>(df_pseudotime))</a>
<a class="sourceLine" id="cb75-6" title="6">df_pseudotime<span class="op">$</span>PC1 &lt;-<span class="st"> </span>deng_SCE<span class="op">$</span>PC1</a>
<a class="sourceLine" id="cb75-7" title="7">df_pseudotime<span class="op">$</span>order_tscan &lt;-<span class="st"> </span><span class="op">-</span>df_pseudotime<span class="op">$</span>order_tscan</a>
<a class="sourceLine" id="cb75-8" title="8">df_pseudotime<span class="op">$</span>diffusionmap &lt;-<span class="st"> </span><span class="op">-</span>df_pseudotime<span class="op">$</span>diffusionmap</a>
<a class="sourceLine" id="cb75-9" title="9"></a>
<a class="sourceLine" id="cb75-10" title="10"><span class="kw">corrplot.mixed</span>(<span class="kw">cor</span>(df_pseudotime, <span class="dt">use =</span> <span class="st">&quot;na.or.complete&quot;</span>), </a>
<a class="sourceLine" id="cb75-11" title="11">               <span class="dt">order =</span> <span class="st">&quot;hclust&quot;</span>, <span class="dt">tl.col =</span> <span class="st">&quot;black&quot;</span>,</a>
<a class="sourceLine" id="cb75-12" title="12">               <span class="dt">main =</span> <span class="st">&quot;Correlation matrix for pseudotime results&quot;</span>,</a>
<a class="sourceLine" id="cb75-13" title="13">               <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">3.1</span>, <span class="dv">0</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/compare-results-1.png" style="display: block; margin: auto;" /></p>
<p>上图所示，Ouija，TSCAN和SLICER都提供了与PC1类似且强相关的轨迹。Diffusion Map与这些方法的相关性较低，而Monocle给出了非常不同的结果。</p>
<p><strong>练习6</strong>: 将destiny和SLICER与TSCAN，Monocle和Ouija进行更深入的比较。区别在哪里？</p>
</div>
<div id="expression-of-genes-through-time" class="section level3">
<h3><span class="header-section-number">8.4.8</span> Expression of genes through time</h3>
<p>每个包可以通过伪时间显示基因表达。检查单个基因非常有助于鉴定在分化过程中起重要作用的基因。使用<code>Rhoa</code>基因进行演示。</p>
<p>我们已经将使用所有方法计算的伪时间值添加到<code>SCE</code>对象的<code>colData</code>中。<code>scater</code>软件包的绘图功能可用于研究基因表达，细胞群和pseudotime的关系。 这对于不提供绘图功能的SLICER等软件包非常有用。</p>
<p><strong>主成分</strong></p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" title="1"><span class="kw">plotExpression</span>(deng_SCE, <span class="st">&quot;Rhoa&quot;</span>, <span class="dt">x =</span> <span class="st">&quot;PC1&quot;</span>, </a>
<a class="sourceLine" id="cb76-2" title="2">               <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>, <span class="dt">show_violin =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb76-3" title="3">               <span class="dt">show_smooth =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/Rhoa-pc1-1.png" style="display: block; margin: auto;" />
<strong>TSCAN</strong></p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb77-1" title="1"><span class="kw">plotExpression</span>(deng_SCE, <span class="st">&quot;Rhoa&quot;</span>, <span class="dt">x =</span> <span class="st">&quot;pseudotime_order_tscan&quot;</span>, </a>
<a class="sourceLine" id="cb77-2" title="2">               <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>, <span class="dt">show_violin =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb77-3" title="3">               <span class="dt">show_smooth =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/Rhoa-tscan-1.png" style="display: block; margin: auto;" />
<strong>Monocle</strong></p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" title="1"><span class="kw">plotExpression</span>(deng_SCE, <span class="st">&quot;Rhoa&quot;</span>, <span class="dt">x =</span> <span class="st">&quot;pseudotime_monocle&quot;</span>, </a>
<a class="sourceLine" id="cb78-2" title="2">               <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>, <span class="dt">show_violin =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb78-3" title="3">               <span class="dt">show_smooth =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/Rhoa-monocle-1.png" style="display: block; margin: auto;" />
<strong>Diffusion Map</strong></p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb79-1" title="1"><span class="kw">plotExpression</span>(deng_SCE, <span class="st">&quot;Rhoa&quot;</span>, <span class="dt">x =</span> <span class="st">&quot;pseudotime_diffusionmap&quot;</span>, </a>
<a class="sourceLine" id="cb79-2" title="2">               <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>, <span class="dt">show_violin =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb79-3" title="3">               <span class="dt">show_smooth =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/Rhoa-diff-map-1.png" style="display: block; margin: auto;" /></p>
<p><strong>SLICER</strong></p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb80-1" title="1"><span class="kw">plotExpression</span>(deng_SCE, <span class="st">&quot;Rhoa&quot;</span>, <span class="dt">x =</span> <span class="st">&quot;pseudotime_slicer&quot;</span>, </a>
<a class="sourceLine" id="cb80-2" title="2">               <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>, <span class="dt">show_violin =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb80-3" title="3">               <span class="dt">show_smooth =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/Rhoa-slicer-1.png" style="display: block; margin: auto;" />
<strong>Ouija</strong></p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb81-1" title="1"><span class="kw">plotExpression</span>(deng_SCE, <span class="st">&quot;Rhoa&quot;</span>, <span class="dt">x =</span> <span class="st">&quot;pseudotime_ouija&quot;</span>, </a>
<a class="sourceLine" id="cb81-2" title="2">               <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>, <span class="dt">show_violin =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb81-3" title="3">               <span class="dt">show_smooth =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/pseudotime_files/figure-html/Rhoa-ouija-1.png" style="display: block; margin: auto;" /></p>
<p>这些方法中有多少优于使用第一主成分表示伪时间？</p>
<p><strong>练习7</strong>: 使用一部基因比如<code>Brennecke_getVariableGenes()</code>得到的HVG重复上面练习。</p>
</div>
</div>
<div id="imputation" class="section level2">
<h2><span class="header-section-number">8.5</span> Imputation</h2>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb82-1" title="1"><span class="kw">library</span>(scImpute)</a>
<a class="sourceLine" id="cb82-2" title="2"><span class="kw">library</span>(SC3)</a>
<a class="sourceLine" id="cb82-3" title="3"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb82-4" title="4"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb82-5" title="5"><span class="kw">library</span>(mclust)</a>
<a class="sourceLine" id="cb82-6" title="6"><span class="kw">library</span>(DrImpute)</a>
<a class="sourceLine" id="cb82-7" title="7"><span class="kw">library</span>(Rmagic)</a>
<a class="sourceLine" id="cb82-8" title="8"><span class="kw">set.seed</span>(<span class="dv">1234567</span>)</a></code></pre></div>
<p>如前所述，分析scRNA-seq数据时的主要挑战之一是存在零或dropout。Dropouts可能由于以下原因：</p>
<ul>
<li>基因不在细胞中表达，因此没有转录物</li>
<li>基因表达，但由于某些原因，转录本在测序前丢失</li>
<li>基因表达并捕获转录物并转化为cDNA，但测序深度低。</li>
</ul>
<p>因此，dropout可能是实验缺陷，如果是这种情况，希望提供计算校正。一种策略是将从表达矩阵插补缺失值，为了能够插补基因表达值，必须有一个基础模型。由于不知道哪些dropout是技术原因，哪些对应于真实缺失，因此插补是一项艰巨的挑战，并且倾向于在下游分析中产生 <a href="https://f1000research.com/articles/7-1740/v2">假阳性结果</a>。</p>
<p>目前有很多插值的方法，再次我们考虑三种运行速度快而且发表的方法：
<a href="https://github.com/pkathail/magic">MAGIC</a> <span class="citation">(Dijk et al. <a href="#ref-Van_Dijk2017-bh">2017</a>)</span>,
<a href="https://github.com/gongx030/DrImpute">DrImpute</a> and
<a href="https://github.com/Vivianstats/scImpute">scImpute</a> <span class="citation">(Li and Li <a href="#ref-Li2017-tz">2017</a>)</span>.</p>
<p>DrImpute和scImpute都使用模型来确定哪些零值是技术原因造成的，并且对这些值进行插补。二者都使用聚类来识别具有同源表达的一组细胞。 DrImpute插补cluster中细胞表达不一致为0的值。然而，scImpute使用零膨胀的正态分布拟合对数归一化的表达值并估算inflated零值。</p>
<div id="scimpute" class="section level3">
<h3><span class="header-section-number">8.5.1</span> scImpute</h3>
<p>为测试scImpute，使用默认参数应用于Deng数据集。scImpute将.csv或.txt文件作为输入：</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb83-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/deng/deng-reads.rds&quot;</span>)</a>
<a class="sourceLine" id="cb83-2" title="2"><span class="kw">write.csv</span>(<span class="kw">counts</span>(deng), <span class="st">&quot;deng.csv&quot;</span>)</a>
<a class="sourceLine" id="cb83-3" title="3"><span class="kw">scimpute</span>(</a>
<a class="sourceLine" id="cb83-4" title="4">    <span class="dt">count_path =</span> <span class="st">&quot;deng.csv&quot;</span>,</a>
<a class="sourceLine" id="cb83-5" title="5">    <span class="dt">infile =</span> <span class="st">&quot;csv&quot;</span>,</a>
<a class="sourceLine" id="cb83-6" title="6">    <span class="dt">outfile =</span> <span class="st">&quot;txt&quot;</span>, </a>
<a class="sourceLine" id="cb83-7" title="7">    <span class="dt">out_dir =</span> <span class="st">&quot;./&quot;</span>,</a>
<a class="sourceLine" id="cb83-8" title="8">    <span class="dt">Kcluster =</span> <span class="dv">10</span>,</a>
<a class="sourceLine" id="cb83-9" title="9">    <span class="dt">ncores =</span> <span class="dv">2</span></a>
<a class="sourceLine" id="cb83-10" title="10">)</a></code></pre></div>
<p>通过PCA图来比较结果和原始数据</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb84-1" title="1">res &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;scimpute_count.txt&quot;</span>)</a>
<a class="sourceLine" id="cb84-2" title="2"><span class="kw">colnames</span>(res) &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb84-3" title="3">res &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(</a>
<a class="sourceLine" id="cb84-4" title="4">    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">logcounts =</span> <span class="kw">log2</span>(<span class="kw">as.matrix</span>(res) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)), </a>
<a class="sourceLine" id="cb84-5" title="5">    <span class="dt">colData =</span> <span class="kw">colData</span>(deng)</a>
<a class="sourceLine" id="cb84-6" title="6">)</a>
<a class="sourceLine" id="cb84-7" title="7"><span class="kw">rowData</span>(res)<span class="op">$</span>feature_symbol &lt;-<span class="st"> </span><span class="kw">rowData</span>(deng)<span class="op">$</span>feature_symbol</a>
<a class="sourceLine" id="cb84-8" title="8"><span class="kw">plotPCA</span>(</a>
<a class="sourceLine" id="cb84-9" title="9">    res, </a>
<a class="sourceLine" id="cb84-10" title="10">    <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span></a>
<a class="sourceLine" id="cb84-11" title="11">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/imputation_files/figure-html/unnamed-chunk-4-1.png" style="display: block; margin: auto;" /></p>
<p>比较结果与章节<a href="biological-analysis.html#clust-methods">8.2</a>的原始数剧。最显著的差异是什么？</p>
<p>检查特定基因的表达，观察插补对表达分布的影响。</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" title="1"><span class="kw">plotExpression</span>(res, <span class="kw">c</span>(<span class="st">&quot;Sox2&quot;</span>, <span class="st">&quot;Eomes&quot;</span>, <span class="st">&quot;Zscan4d&quot;</span>, <span class="st">&quot;Fgf4&quot;</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/imputation_files/figure-html/unnamed-chunk-5-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb86-1" title="1"><span class="kw">plotExpression</span>(deng, <span class="kw">c</span>(<span class="st">&quot;Sox2&quot;</span>, <span class="st">&quot;Eomes&quot;</span>, <span class="st">&quot;Zscan4d&quot;</span>, <span class="st">&quot;Fgf4&quot;</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/imputation_files/figure-html/unnamed-chunk-5-2.png" style="display: block; margin: auto;" /></p>
<p>为评价插补的效果，使用<code>SC3</code>对插补后的矩阵进行聚类。</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb87-1" title="1">res &lt;-<span class="st"> </span><span class="kw">sc3_estimate_k</span>(res)</a>
<a class="sourceLine" id="cb87-2" title="2"><span class="kw">metadata</span>(res)<span class="op">$</span>sc3<span class="op">$</span>k_estimation</a>
<a class="sourceLine" id="cb87-3" title="3"><span class="co">## [1] 6</span></a>
<a class="sourceLine" id="cb87-4" title="4">res &lt;-<span class="st"> </span><span class="kw">sc3</span>(res, <span class="dt">ks =</span> <span class="dv">10</span>, <span class="dt">n_cores =</span> <span class="dv">1</span>, <span class="dt">gene_filter =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb87-5" title="5"><span class="kw">adjustedRandIndex</span>(<span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2, <span class="kw">colData</span>(res)<span class="op">$</span>sc3_<span class="dv">10</span>_clusters)</a>
<a class="sourceLine" id="cb87-6" title="6"><span class="co">## [1] 0.5062983</span></a>
<a class="sourceLine" id="cb87-7" title="7"><span class="kw">plotPCA</span>(</a>
<a class="sourceLine" id="cb87-8" title="8">    res, </a>
<a class="sourceLine" id="cb87-9" title="9">    <span class="dt">colour_by =</span> <span class="st">&quot;sc3_10_clusters&quot;</span></a>
<a class="sourceLine" id="cb87-10" title="10">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/imputation_files/figure-html/unnamed-chunk-6-1.png" style="display: block; margin: auto;" /></p>
<p><strong>练习:</strong> 基于PCA和聚类结果，使用<code>scImpute</code>对Deng数据集进行插补效果好吗？</p>
</div>
<div id="drimpute" class="section level3">
<h3><span class="header-section-number">8.5.2</span> DrImpute</h3>
<p>使用DrImpute进行同样的操作。DrImpute在R中直接运行对数标准化表达矩阵，使用scater生成此矩阵，然后运行DrImpute。 与scImpute不同，DrImpute使用两个不同的相关距离来考虑跨越一系列ks的一致性插补：</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">normalize</span>(deng)</a>
<a class="sourceLine" id="cb88-2" title="2">res &lt;-<span class="st"> </span><span class="kw">DrImpute</span>(deng<span class="op">@</span>assays[[<span class="st">&quot;logcounts&quot;</span>]], <span class="dt">ks=</span><span class="dv">8</span><span class="op">:</span><span class="dv">12</span>)</a>
<a class="sourceLine" id="cb88-3" title="3"><span class="kw">colnames</span>(res) &lt;-<span class="st"> </span><span class="kw">colnames</span>(deng)</a>
<a class="sourceLine" id="cb88-4" title="4"><span class="kw">rownames</span>(res) &lt;-<span class="st"> </span><span class="kw">rownames</span>(deng)</a>
<a class="sourceLine" id="cb88-5" title="5">res &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(</a>
<a class="sourceLine" id="cb88-6" title="6">    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">logcounts =</span> <span class="kw">as.matrix</span>(res)), </a>
<a class="sourceLine" id="cb88-7" title="7">    <span class="dt">colData =</span> <span class="kw">colData</span>(deng)</a>
<a class="sourceLine" id="cb88-8" title="8">)</a>
<a class="sourceLine" id="cb88-9" title="9"><span class="kw">rowData</span>(res)<span class="op">$</span>feature_symbol &lt;-<span class="st"> </span><span class="kw">rowData</span>(deng)<span class="op">$</span>feature_symbol</a>
<a class="sourceLine" id="cb88-10" title="10"><span class="kw">plotPCA</span>(</a>
<a class="sourceLine" id="cb88-11" title="11">    res, </a>
<a class="sourceLine" id="cb88-12" title="12">    <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span></a>
<a class="sourceLine" id="cb88-13" title="13">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/imputation_files/figure-html/unnamed-chunk-7-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb89-1" title="1"><span class="kw">plotExpression</span>(res, <span class="kw">c</span>(<span class="st">&quot;Sox2&quot;</span>, <span class="st">&quot;Eomes&quot;</span>, <span class="st">&quot;Zscan4d&quot;</span>, <span class="st">&quot;Fgf4&quot;</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/imputation_files/figure-html/unnamed-chunk-7-2.png" style="display: block; margin: auto;" />
<strong>练习:</strong> 检查DrImpute矩阵的sc3聚类，<code>DrImpute</code>适合对Deng数据集进行插补吗？</p>
</div>
<div id="magic" class="section level3">
<h3><span class="header-section-number">8.5.3</span> MAGIC</h3>
<p><strong>(由于技术原因，本章作为练习)</strong></p>
<p>MAGIC是一个python包，但是作者提供了一个R包封装，所以它可以从R无缝地运行。</p>
<p>与scImpute和DrImpute不同，MAGIC可以smooth整个数据集。它会对零值进行插补，但也会smooth非零值，以进一步扩大数据集中的结构。 由于它基于扩散过程，因此特别增强了数据集中类似轨迹的结构，与scImpute和DrImpute相反，后者假设数据具有类似cluster的结构。</p>
<p>比对结果与章节<a href="biological-analysis.html#clust-methods">8.2</a>的原始数据。显著差异是什么？</p>
<p>为评价插补的效果，使用<code>SC3</code>对插补后的矩阵进行聚类。</p>
<p><strong>练习:</strong> MAGIC包含2个影响插补水平的参数：<code>t</code> 和 <code>k</code>，增大或减小这些参数，结果如何？(提示：尝试t=1, t=10, k=2, k=50)</p>
<p><strong>练习:</strong> 基于PCA和聚类分析的<code>scImpute</code>和<code>MAGIC</code>有什么区别？哪一个更好用？</p>
</div>
</div>
<div id="dechapter" class="section level2">
<h2><span class="header-section-number">8.6</span> 差异表达分析</h2>
<div id="bulk-rna-seq-1" class="section level3">
<h3><span class="header-section-number">8.6.1</span> Bulk RNA-seq</h3>
<p>bulk RNA-seq数据最常见的分析类型之一是鉴定差异表达的基因。通过比较在两种条件下变化的基因，例如 突变体和野生型或刺激和未刺激的，寻找背后的分子机制。</p>
<p>用于bulk RNA-seq差异表达分析有几种不同的方法，比如 <a href="https://bioconductor.org/packages/DESeq2">DESeq2</a>和
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>,而且有大量的 <a href="http://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-9-r95">数据集</a>，其中RNA-seq数据被RT-qPCR验证。这些数据可以对差异表达分析的算法进行基准测试。现有的证据表明算法表现不错。</p>
</div>
<div id="single-cell-rna-seq" class="section level3">
<h3><span class="header-section-number">8.6.2</span> Single cell RNA-seq</h3>
<p>与bulk RNA-seq相反，在scRNA-seq中，通常没有一组确定的实验条件。如前一章(<a href="biological-analysis.html#clust-methods">8.2</a>)所示，通过无监督的聚类方法来识别细胞群体，然后通过比较各群体之间的差异（如SC3的Kruskal-Wallis检验），或成对比较cluster中基因表达，找到差异表达的基因。 在下一章中，我们将主要考虑为成对比较开发的工具。</p>
</div>
<div id="Differences-in-Distribution" class="section level3">
<h3><span class="header-section-number">8.6.3</span> 分布的差异</h3>
<p>与bulk RNA-seq不同，在单细胞实验中，每个group有大量的样品（即细胞）。 因此，利用每组中表达值的完整分布来识别组间的差异，而不是向bulk RNA-seq中仅仅比较平均表达。</p>
<p>主要有两种方法比较分布。首先，可以使用现有的统计模型/分布拟合到每个组中的表达，然后检验每个模型的参数差异，或者允许特定参数不同，模型是否拟合地更好。例如在章节<a href="cleaning-the-expression-matrix.html#dealing-with-confounders">7.10</a>，使用edgeR检验在不同批次中允许平均表达不同是否显著提高数据的负二项模型的拟合。</p>
<p>或者，使用非参数检验，该检验不假设表达值服从任何特定分布，例如， <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov-Smirnov test (KS-test)</a>。 非参数测试通常将观察到的表达值转换为rank并检验一组的rank分布是否与另一组的rank分布显著不同。然而，一些非参数方法不适用存在大量tied values，例如单细胞RNA-seq表达数据中的dropout的情况。 此外，如果参数检验的条件成立，那么它通常比非参数测试更强大。</p>
</div>
<div id="Models-of-single-cell-RNASeq-data" class="section level3">
<h3><span class="header-section-number">8.6.4</span> scRNA-seq数据模型</h3>
<p>最常用的RNA-seq数据模型是负二项分布模型：</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" title="1"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb90-2" title="2"><span class="kw">hist</span>(</a>
<a class="sourceLine" id="cb90-3" title="3">    <span class="kw">rnbinom</span>(</a>
<a class="sourceLine" id="cb90-4" title="4">        <span class="dv">1000</span>, </a>
<a class="sourceLine" id="cb90-5" title="5">        <span class="dt">mu =</span> <span class="dv">10</span>, </a>
<a class="sourceLine" id="cb90-6" title="6">        <span class="dt">size =</span> <span class="dv">100</span>), </a>
<a class="sourceLine" id="cb90-7" title="7">    <span class="dt">col =</span> <span class="st">&quot;grey50&quot;</span>, </a>
<a class="sourceLine" id="cb90-8" title="8">    <span class="dt">xlab =</span> <span class="st">&quot;Read Counts&quot;</span>, </a>
<a class="sourceLine" id="cb90-9" title="9">    <span class="dt">main =</span> <span class="st">&quot;Negative Binomial&quot;</span></a>
<a class="sourceLine" id="cb90-10" title="10">)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:nb-plot"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-intro_files/figure-html/nb-plot-1.png" alt="Negative Binomial distribution of read counts for a single gene across 1000 cells"  />
<p class="caption">
Figure 8.11: Negative Binomial distribution of read counts for a single gene across 1000 cells
</p>
</div>
<p>均值:
<span class="math inline">\(\mu = mu\)</span></p>
<p>方差:
<span class="math inline">\(\sigma^2 = mu + mu^2/size\)</span></p>
<p>其有两个参数，平均表达值(<span class="math inline">\(\mu\)</span>)和离散程度(dispersion)，离散程度与方差负相关。负二项分布模型很好地拟合了bulk RNA-seq数据，并且应用于大多数统计方法。此外，它也很好地你好了UMI数据的molecule coutns分布(<a href="http://www.nature.com/nmeth/journal/v11/n6/full/nmeth.2930.html">Grun et al. 2014</a>, <a href="http://genome.cshlp.org/content/21/7/1160">Islam et al. 2011</a>)。</p>
<p>然而由于相对于非零read counts的高dropout率，原始负二项分布模型不适合全长转录本数据。对于这种类型的数据，已经提出了各种零膨胀负二项模型，比如 <a href="https://bioconductor.org/packages/release/bioc/html/MAST.html">MAST</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scde.html">SCDE</a>。</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb91-1" title="1">d &lt;-<span class="st"> </span><span class="fl">0.5</span>;</a>
<a class="sourceLine" id="cb91-2" title="2">counts &lt;-<span class="st"> </span><span class="kw">rnbinom</span>(</a>
<a class="sourceLine" id="cb91-3" title="3">    <span class="dv">1000</span>, </a>
<a class="sourceLine" id="cb91-4" title="4">    <span class="dt">mu =</span> <span class="dv">10</span>, </a>
<a class="sourceLine" id="cb91-5" title="5">    <span class="dt">size =</span> <span class="dv">100</span></a>
<a class="sourceLine" id="cb91-6" title="6">)</a>
<a class="sourceLine" id="cb91-7" title="7">counts[<span class="kw">runif</span>(<span class="dv">1000</span>) <span class="op">&lt;</span><span class="st"> </span>d] &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb91-8" title="8"><span class="kw">hist</span>(</a>
<a class="sourceLine" id="cb91-9" title="9">    counts, </a>
<a class="sourceLine" id="cb91-10" title="10">    <span class="dt">col =</span> <span class="st">&quot;grey50&quot;</span>, </a>
<a class="sourceLine" id="cb91-11" title="11">    <span class="dt">xlab =</span> <span class="st">&quot;Read Counts&quot;</span>, </a>
<a class="sourceLine" id="cb91-12" title="12">    <span class="dt">main =</span> <span class="st">&quot;Zero-inflated NB&quot;</span></a>
<a class="sourceLine" id="cb91-13" title="13">)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:zero-inflation-plot"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-intro_files/figure-html/zero-inflation-plot-1.png" alt="Zero-inflated Negative Binomial distribution"  />
<p class="caption">
Figure 8.12: Zero-inflated Negative Binomial distribution
</p>
</div>
<p>均值:
<span class="math inline">\(\mu = mu \cdot (1 - d)\)</span></p>
<p>方差:
<span class="math inline">\(\sigma^2 = \mu \cdot (1-d) \cdot (1 + d \cdot \mu + \mu / size)\)</span></p>
<p>这个模型为dropout引入新的参数<span class="math inline">\(d\)</span>到负二项分布模型，如第19章中所看到的，基因的dropout率与基因的平均表达密切相关。 不同的零膨胀负二项模型使用mu和d之间的不同关系，并且一些可以使<span class="math inline">\(\mu\)</span>和<span class="math inline">\(d\)</span>拟合每个基因的表达。</p>
<p>最后，一些方法使用Poisson-Beta分布，该分布基于转录爆发的机制模型。 这个模型有很强的实验支持( <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-1-r7">Kim and Marioni, 2013</a>)，它较好地拟合scRNA-seq数据，但它比负二项模型不容易使用，而且现有的方法交少。</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" title="1">a &lt;-<span class="st"> </span><span class="fl">0.1</span></a>
<a class="sourceLine" id="cb92-2" title="2">b &lt;-<span class="st"> </span><span class="fl">0.1</span></a>
<a class="sourceLine" id="cb92-3" title="3">g &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb92-4" title="4">lambdas &lt;-<span class="st"> </span><span class="kw">rbeta</span>(<span class="dv">1000</span>, a, b)</a>
<a class="sourceLine" id="cb92-5" title="5">counts &lt;-<span class="st"> </span><span class="kw">sapply</span>(g<span class="op">*</span>lambdas, <span class="cf">function</span>(l) {<span class="kw">rpois</span>(<span class="dv">1</span>, <span class="dt">lambda =</span> l)})</a>
<a class="sourceLine" id="cb92-6" title="6"><span class="kw">hist</span>(</a>
<a class="sourceLine" id="cb92-7" title="7">    counts, </a>
<a class="sourceLine" id="cb92-8" title="8">    <span class="dt">col =</span> <span class="st">&quot;grey50&quot;</span>, </a>
<a class="sourceLine" id="cb92-9" title="9">    <span class="dt">xlab =</span> <span class="st">&quot;Read Counts&quot;</span>, </a>
<a class="sourceLine" id="cb92-10" title="10">    <span class="dt">main =</span> <span class="st">&quot;Poisson-Beta&quot;</span></a>
<a class="sourceLine" id="cb92-11" title="11">)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-intro_files/figure-html/pois-beta-plot-1.png" style="display: block; margin: auto;" />
均值:
<span class="math inline">\(\mu = g \cdot a / (a + b)\)</span></p>
<p>方差:
<span class="math inline">\(\sigma^2 = g^2 \cdot a \cdot b/((a + b + 1) \cdot (a + b)^2)\)</span></p>
<p>该模型使用三个参数：<span class="math inline">\(a\)</span>转录激活率; <span class="math inline">\(b\)</span>转录抑制率; 和<span class="math inline">\(g\)</span>转录激活状态时转录本产生的速率。 差异表达方法可以检验每个参数的组间差异或仅一个（通常是<span class="math inline">\(g\)</span>）。</p>
<p>所有这些模型可以进一步扩展来明确地解释基因表达差异的其他来源，例如批次效应或文库深度，这取决于特定的DE算法。</p>
<p><strong>练习</strong>: 改变每个分布的参数，探索它们如何影响基因表达的分布。Poisson-Beta和Negative Binomial模型相似处在哪里？</p>
</div>
</div>
<div id="DE-in-a-real-dataset" class="section level2">
<h2><span class="header-section-number">8.7</span> 差异表达分析实战</h2>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb93-1" title="1"><span class="kw">library</span>(scRNA.seq.funcs)</a>
<a class="sourceLine" id="cb93-2" title="2"><span class="kw">library</span>(edgeR)</a>
<a class="sourceLine" id="cb93-3" title="3"><span class="kw">library</span>(monocle)</a>
<a class="sourceLine" id="cb93-4" title="4"><span class="kw">library</span>(MAST)</a>
<a class="sourceLine" id="cb93-5" title="5"><span class="kw">library</span>(ROCR)</a>
<a class="sourceLine" id="cb93-6" title="6"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a></code></pre></div>
<div id="section-94" class="section level3">
<h3><span class="header-section-number">8.7.1</span> 介绍</h3>
<p>为了测试不同的单细胞差异表达方法，我们使用第7-17章的Blischak数据集。对于该实验，除了单细胞数据之外，还产生每个细胞系的bulk RNA-seq数据。 使用标准方法分别在bulk RNA-seq和scRNA-seq鉴定差异表达基因，并用bulk RNA-seq结果评估单细胞方法准确性。为了节省时间，我们预先计算了这些。 可以运行以下命令来加载这些数据。</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" title="1">DE &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/tung/TPs.txt&quot;</span>)</a>
<a class="sourceLine" id="cb94-2" title="2">notDE &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/tung/TNs.txt&quot;</span>)</a>
<a class="sourceLine" id="cb94-3" title="3">GroundTruth &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb94-4" title="4">    <span class="dt">DE =</span> <span class="kw">as.character</span>(<span class="kw">unlist</span>(DE)), </a>
<a class="sourceLine" id="cb94-5" title="5">    <span class="dt">notDE =</span> <span class="kw">as.character</span>(<span class="kw">unlist</span>(notDE))</a>
<a class="sourceLine" id="cb94-6" title="6">)</a></code></pre></div>
<p>比较NA19101到NA19239的基准已经产生，下面加载相应的单细胞数据：</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb95-1" title="1">molecules &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/tung/molecules.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb95-2" title="2">anno &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/tung/annotation.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb95-3" title="3">keep &lt;-<span class="st"> </span>anno[,<span class="dv">1</span>] <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span> <span class="op">|</span><span class="st"> </span>anno[,<span class="dv">1</span>] <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span></a>
<a class="sourceLine" id="cb95-4" title="4">data &lt;-<span class="st"> </span>molecules[,keep]</a>
<a class="sourceLine" id="cb95-5" title="5">group &lt;-<span class="st"> </span>anno[keep,<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb95-6" title="6">batch &lt;-<span class="st"> </span>anno[keep,<span class="dv">4</span>]</a>
<a class="sourceLine" id="cb95-7" title="7"><span class="co"># remove genes that aren&#39;t expressed in at least 6 cells</span></a>
<a class="sourceLine" id="cb95-8" title="8">gkeep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(data <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>;</a>
<a class="sourceLine" id="cb95-9" title="9">counts &lt;-<span class="st"> </span>data[gkeep,]</a>
<a class="sourceLine" id="cb95-10" title="10"><span class="co"># Library size normalization</span></a>
<a class="sourceLine" id="cb95-11" title="11">lib_size =<span class="st"> </span><span class="kw">colSums</span>(counts)</a>
<a class="sourceLine" id="cb95-12" title="12">norm &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(counts)<span class="op">/</span>lib_size <span class="op">*</span><span class="st"> </span><span class="kw">median</span>(lib_size)) </a>
<a class="sourceLine" id="cb95-13" title="13"><span class="co"># Variant of CPM for datasets with library sizes of fewer than 1 mil molecules</span></a></code></pre></div>
<p>现在比较各种单细胞差异表达分析方法。请注意，这里只运行R语言环境并且运行速度相对较快的方法。</p>
</div>
<div id="kolmogorov-smirnov-test" class="section level3">
<h3><span class="header-section-number">8.7.2</span> Kolmogorov-Smirnov test</h3>
<p>最容易使用的检验类型是非参数测试。最常用的非参数检验是 <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov-Smirnov test</a> (KS-test)，可以用它来比较两个个体中每个基因的分布。</p>
<p>KS检验量化了两个群体中每个基因表达的经验累积分布之间的距离。它对平均表达值的变化和方差的变化很敏感。 然而，它假设数据是连续的，并且当数据包含大量相同的值（例如，0）时可能表现不佳。 KS检验的另一个问题是它对于大样本量可能非常敏感，因此即使差异的幅度非常小，它也可能结果为显著。</p>
<div class="figure" style="text-align: center"><span id="fig:ks-statistic"></span>
<img src="figures/KS2_Example.png" alt="Illustration of the two-sample Kolmogorov–Smirnov statistic. Red and blue lines each correspond to an empirical distribution function, and the black arrow is the two-sample KS statistic. (taken from [here](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test))" width="60%" />
<p class="caption">
Figure 8.13: Illustration of the two-sample Kolmogorov–Smirnov statistic. Red and blue lines each correspond to an empirical distribution function, and the black arrow is the two-sample KS statistic. (taken from <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">here</a>)
</p>
</div>
<p>运行KS检验:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb96-1" title="1">pVals &lt;-<span class="st"> </span><span class="kw">apply</span>(</a>
<a class="sourceLine" id="cb96-2" title="2">    norm, <span class="dv">1</span>, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb96-3" title="3">        <span class="kw">ks.test</span>(</a>
<a class="sourceLine" id="cb96-4" title="4">            x[group <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>], </a>
<a class="sourceLine" id="cb96-5" title="5">            x[group <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>]</a>
<a class="sourceLine" id="cb96-6" title="6">        )<span class="op">$</span>p.value</a>
<a class="sourceLine" id="cb96-7" title="7">    }</a>
<a class="sourceLine" id="cb96-8" title="8">)</a>
<a class="sourceLine" id="cb96-9" title="9"><span class="co"># multiple testing correction</span></a>
<a class="sourceLine" id="cb96-10" title="10">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a></code></pre></div>
<p>此代码将函数“appy”到表达矩阵的每一行（由1指定）。函数只返回ks.test输出的p.value。 我们现在可以考虑通过KS检验检测到多少true positive和true negative DE基因.</p>
<div id="section-95" class="section level4">
<h4><span class="header-section-number">8.7.2.1</span> 评价准确率</h4>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb97-1" title="1">sigDE &lt;-<span class="st"> </span><span class="kw">names</span>(pVals)[pVals <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>]</a>
<a class="sourceLine" id="cb97-2" title="2"><span class="kw">length</span>(sigDE) </a>
<a class="sourceLine" id="cb97-3" title="3"><span class="co">## [1] 5095</span></a>
<a class="sourceLine" id="cb97-4" title="4"><span class="co"># Number of KS-DE genes</span></a>
<a class="sourceLine" id="cb97-5" title="5"><span class="kw">sum</span>(GroundTruth<span class="op">$</span>DE <span class="op">%in%</span><span class="st"> </span>sigDE) </a>
<a class="sourceLine" id="cb97-6" title="6"><span class="co">## [1] 792</span></a>
<a class="sourceLine" id="cb97-7" title="7"><span class="co"># Number of KS-DE genes that are true DE genes</span></a>
<a class="sourceLine" id="cb97-8" title="8"><span class="kw">sum</span>(GroundTruth<span class="op">$</span>notDE <span class="op">%in%</span><span class="st"> </span>sigDE)</a>
<a class="sourceLine" id="cb97-9" title="9"><span class="co">## [1] 3190</span></a>
<a class="sourceLine" id="cb97-10" title="10"><span class="co"># Number of KS-DE genes that are truly not-DE</span></a></code></pre></div>
<p>通过KS检验将基准中true negative基因识别为DE(假阳性)，但这可能是由于大量的f非DE基因。通常将这些计数归一化为真阳性率(TPR), TP/(TP + FN)和假阳性率 (FPR), FP/(FP+TP)。</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb98-1" title="1">tp &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>DE <span class="op">%in%</span><span class="st"> </span>sigDE)</a>
<a class="sourceLine" id="cb98-2" title="2">fp &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>notDE <span class="op">%in%</span><span class="st"> </span>sigDE)</a>
<a class="sourceLine" id="cb98-3" title="3">tn &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>notDE <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(pVals)[pVals <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.05</span>])</a>
<a class="sourceLine" id="cb98-4" title="4">fn &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>DE <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(pVals)[pVals <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.05</span>])</a>
<a class="sourceLine" id="cb98-5" title="5">tpr &lt;-<span class="st"> </span>tp<span class="op">/</span>(tp <span class="op">+</span><span class="st"> </span>fn)</a>
<a class="sourceLine" id="cb98-6" title="6">fpr &lt;-<span class="st"> </span>fp<span class="op">/</span>(fp <span class="op">+</span><span class="st"> </span>tn)</a>
<a class="sourceLine" id="cb98-7" title="7"><span class="kw">cat</span>(<span class="kw">c</span>(tpr, fpr))</a>
<a class="sourceLine" id="cb98-8" title="8"><span class="co">## 0.7346939 0.2944706</span></a></code></pre></div>
<p>TPR大于FPR，表明KS检验识别出差异表达基因。</p>
<p>目前为止，我们只评估了单个显著性阈值的性能。通常，可以改变阈值并评估一系列值的性能。 然后将其绘制为ROC(receiver-operating-characteristic curve)曲线，并且计算ROC曲线下的面积AUC。我们将使用ROCR包来绘图。</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb99-1" title="1"><span class="co"># Only consider genes for which we know the ground truth</span></a>
<a class="sourceLine" id="cb99-2" title="2">pVals &lt;-<span class="st"> </span>pVals[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE <span class="op">|</span><span class="st"> </span></a>
<a class="sourceLine" id="cb99-3" title="3"><span class="st">               </span><span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>notDE] </a>
<a class="sourceLine" id="cb99-4" title="4">truth &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>, <span class="dt">times =</span> <span class="kw">length</span>(pVals));</a>
<a class="sourceLine" id="cb99-5" title="5">truth[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE] =<span class="st"> </span><span class="dv">0</span>;</a>
<a class="sourceLine" id="cb99-6" title="6">pred &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">prediction</span>(pVals, truth)</a>
<a class="sourceLine" id="cb99-7" title="7">perf &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;tpr&quot;</span>, <span class="st">&quot;fpr&quot;</span>)</a>
<a class="sourceLine" id="cb99-8" title="8">ROCR<span class="op">::</span><span class="kw">plot</span>(perf)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ks-roc-plot"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-real_files/figure-html/ks-roc-plot-1.png" alt="ROC curve for KS-test."  />
<p class="caption">
Figure 8.14: ROC curve for KS-test.
</p>
</div>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb100-1" title="1">aucObj &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;auc&quot;</span>)</a>
<a class="sourceLine" id="cb100-2" title="2">aucObj<span class="op">@</span>y.values[[<span class="dv">1</span>]] <span class="co"># AUC</span></a>
<a class="sourceLine" id="cb100-3" title="3"><span class="co">## [1] 0.7954796</span></a></code></pre></div>
<p>最后，为了便于比较其他DE方法，将这段代码放入一个函数中，这样就不需要重复它:</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" title="1">DE_Quality_AUC &lt;-<span class="st"> </span><span class="cf">function</span>(pVals) {</a>
<a class="sourceLine" id="cb101-2" title="2">    pVals &lt;-<span class="st"> </span>pVals[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE <span class="op">|</span><span class="st"> </span></a>
<a class="sourceLine" id="cb101-3" title="3"><span class="st">                   </span><span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>notDE]</a>
<a class="sourceLine" id="cb101-4" title="4">    truth &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>, <span class="dt">times =</span> <span class="kw">length</span>(pVals));</a>
<a class="sourceLine" id="cb101-5" title="5">    truth[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE] =<span class="st"> </span><span class="dv">0</span>;</a>
<a class="sourceLine" id="cb101-6" title="6">    pred &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">prediction</span>(pVals, truth)</a>
<a class="sourceLine" id="cb101-7" title="7">    perf &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;tpr&quot;</span>, <span class="st">&quot;fpr&quot;</span>)</a>
<a class="sourceLine" id="cb101-8" title="8">    ROCR<span class="op">::</span><span class="kw">plot</span>(perf)</a>
<a class="sourceLine" id="cb101-9" title="9">    aucObj &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;auc&quot;</span>)</a>
<a class="sourceLine" id="cb101-10" title="10">    <span class="kw">return</span>(aucObj<span class="op">@</span>y.values[[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb101-11" title="11">}</a></code></pre></div>
</div>
</div>
<div id="wilcoxmann-whitney-u-test" class="section level3">
<h3><span class="header-section-number">8.7.3</span> Wilcox/Mann-Whitney-U Test</h3>
<p>The Wilcox-rank-sum检验是另一种非参数检验，检验一组中的值是否大于/小于另一组中的值。因此，它通常被认为是检验两组间表达值中位数差异; 而KS检验对表达值分布的任何变化都很敏感。</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb102-1" title="1">pVals &lt;-<span class="st"> </span><span class="kw">apply</span>(</a>
<a class="sourceLine" id="cb102-2" title="2">    norm, <span class="dv">1</span>, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb102-3" title="3">        <span class="kw">wilcox.test</span>(</a>
<a class="sourceLine" id="cb102-4" title="4">            x[group <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>], </a>
<a class="sourceLine" id="cb102-5" title="5">            x[group <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>]</a>
<a class="sourceLine" id="cb102-6" title="6">        )<span class="op">$</span>p.value</a>
<a class="sourceLine" id="cb102-7" title="7">    }</a>
<a class="sourceLine" id="cb102-8" title="8">)</a>
<a class="sourceLine" id="cb102-9" title="9"><span class="co"># multiple testing correction</span></a>
<a class="sourceLine" id="cb102-10" title="10">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb102-11" title="11"><span class="kw">DE_Quality_AUC</span>(pVals)</a>
<a class="sourceLine" id="cb102-12" title="12"><span class="co">## [1] 0.8320326</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:wilcox-plot"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-real_files/figure-html/wilcox-plot-1.png" alt="ROC curve for Wilcox test."  />
<p class="caption">
Figure 8.15: ROC curve for Wilcox test.
</p>
</div>
</div>
<div id="edger" class="section level3">
<h3><span class="header-section-number">8.7.4</span> edgeR</h3>
<p>我们已经在章节<a href="cleaning-the-expression-matrix.html#dealing-with-confounders">7.10</a>中使用了edgeR来计算差异表达。 edgeR基于基因表达的负二项式模型，并使用广义线性模型（GLM），使我们能够将其他因素（例如批次）包含到模型中。</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb103-1" title="1">dge &lt;-<span class="st"> </span><span class="kw">DGEList</span>(</a>
<a class="sourceLine" id="cb103-2" title="2">    <span class="dt">counts =</span> counts, </a>
<a class="sourceLine" id="cb103-3" title="3">    <span class="dt">norm.factors =</span> <span class="kw">rep</span>(<span class="dv">1</span>, <span class="kw">length</span>(counts[<span class="dv">1</span>,])), </a>
<a class="sourceLine" id="cb103-4" title="4">    <span class="dt">group =</span> group</a>
<a class="sourceLine" id="cb103-5" title="5">)</a>
<a class="sourceLine" id="cb103-6" title="6">group_edgeR &lt;-<span class="st"> </span><span class="kw">factor</span>(group)</a>
<a class="sourceLine" id="cb103-7" title="7">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>group_edgeR)</a>
<a class="sourceLine" id="cb103-8" title="8">dge &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(dge, <span class="dt">design =</span> design, <span class="dt">trend.method =</span> <span class="st">&quot;none&quot;</span>)</a>
<a class="sourceLine" id="cb103-9" title="9">fit &lt;-<span class="st"> </span><span class="kw">glmFit</span>(dge, design)</a>
<a class="sourceLine" id="cb103-10" title="10">res &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit)</a>
<a class="sourceLine" id="cb103-11" title="11">pVals &lt;-<span class="st"> </span>res<span class="op">$</span>table[,<span class="dv">4</span>]</a>
<a class="sourceLine" id="cb103-12" title="12"><span class="kw">names</span>(pVals) &lt;-<span class="st"> </span><span class="kw">rownames</span>(res<span class="op">$</span>table)</a>
<a class="sourceLine" id="cb103-13" title="13"></a>
<a class="sourceLine" id="cb103-14" title="14">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb103-15" title="15"><span class="kw">DE_Quality_AUC</span>(pVals)</a>
<a class="sourceLine" id="cb103-16" title="16"><span class="co">## [1] 0.8466764</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:edger-plot"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-real_files/figure-html/edger-plot-1.png" alt="ROC curve for edgeR."  />
<p class="caption">
Figure 8.16: ROC curve for edgeR.
</p>
</div>
</div>
<div id="monocle-1" class="section level3">
<h3><span class="header-section-number">8.7.5</span> Monocle</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/monocle.html">Monocle</a>可以使用几种不同的DE模型。 对于count数据，它建议使用负二项模型（negbinomial.size）。对于标准化数据，对其进行对数变换使用正态分布（gaussianff）。 与edgeR类似，此方法使用GLM框架，因此理论上可以考虑批次效应，但实际上，如果包含批次，模型将不适用此数据集。</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb104-1" title="1">pd &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">group =</span> group, <span class="dt">batch =</span> batch)</a>
<a class="sourceLine" id="cb104-2" title="2"><span class="kw">rownames</span>(pd) &lt;-<span class="st"> </span><span class="kw">colnames</span>(counts)</a>
<a class="sourceLine" id="cb104-3" title="3">pd &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;AnnotatedDataFrame&quot;</span>, <span class="dt">data =</span> pd)</a>
<a class="sourceLine" id="cb104-4" title="4"></a>
<a class="sourceLine" id="cb104-5" title="5">Obj &lt;-<span class="st"> </span><span class="kw">newCellDataSet</span>(</a>
<a class="sourceLine" id="cb104-6" title="6">    <span class="kw">as.matrix</span>(counts), </a>
<a class="sourceLine" id="cb104-7" title="7">    <span class="dt">phenoData =</span> pd, </a>
<a class="sourceLine" id="cb104-8" title="8">    <span class="dt">expressionFamily =</span> <span class="kw">negbinomial.size</span>()</a>
<a class="sourceLine" id="cb104-9" title="9">)</a>
<a class="sourceLine" id="cb104-10" title="10">Obj &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(Obj)</a>
<a class="sourceLine" id="cb104-11" title="11">Obj &lt;-<span class="st"> </span><span class="kw">estimateDispersions</span>(Obj)</a>
<a class="sourceLine" id="cb104-12" title="12">res &lt;-<span class="st"> </span><span class="kw">differentialGeneTest</span>(Obj, <span class="dt">fullModelFormulaStr =</span> <span class="st">&quot;~group&quot;</span>)</a>
<a class="sourceLine" id="cb104-13" title="13"></a>
<a class="sourceLine" id="cb104-14" title="14">pVals &lt;-<span class="st"> </span>res[,<span class="dv">3</span>]</a>
<a class="sourceLine" id="cb104-15" title="15"><span class="kw">names</span>(pVals) &lt;-<span class="st"> </span><span class="kw">rownames</span>(res)</a>
<a class="sourceLine" id="cb104-16" title="16">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb104-17" title="17"><span class="kw">DE_Quality_AUC</span>(pVals)</a>
<a class="sourceLine" id="cb104-18" title="18"><span class="co">## [1] 0.8252662</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:Monocle-plot"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-real_files/figure-html/Monocle-plot-1.png" alt="ROC curve for Monocle."  />
<p class="caption">
Figure 8.17: ROC curve for Monocle.
</p>
</div>
<p><strong>练习</strong>: 比较使用负二项式模型结果和对数变换使用正态/高斯模型(<code>gaussianff()</code>)模型结果。</p>
<p><strong>Answer</strong>:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb105-1" title="1">pd &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">group =</span> group, <span class="dt">batch =</span> batch)</a>
<a class="sourceLine" id="cb105-2" title="2"><span class="kw">rownames</span>(pd) &lt;-<span class="st"> </span><span class="kw">colnames</span>(norm)</a>
<a class="sourceLine" id="cb105-3" title="3">pd &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;AnnotatedDataFrame&quot;</span>, <span class="dt">data =</span> pd)</a>
<a class="sourceLine" id="cb105-4" title="4"></a>
<a class="sourceLine" id="cb105-5" title="5">Obj_log &lt;-<span class="st"> </span><span class="kw">newCellDataSet</span>(</a>
<a class="sourceLine" id="cb105-6" title="6">    <span class="kw">as.matrix</span>(<span class="kw">log</span>(norm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>) <span class="op">/</span><span class="st"> </span><span class="kw">log</span>(<span class="dv">2</span>)), </a>
<a class="sourceLine" id="cb105-7" title="7">    <span class="dt">phenoData =</span> pd, </a>
<a class="sourceLine" id="cb105-8" title="8">    <span class="dt">expressionFamily =</span> <span class="kw">gaussianff</span>()</a>
<a class="sourceLine" id="cb105-9" title="9">)</a>
<a class="sourceLine" id="cb105-10" title="10">Obj_log &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(Obj_log)</a>
<a class="sourceLine" id="cb105-11" title="11"><span class="co"># Obj_log &lt;- estimateDispersions(Obj_log)</span></a>
<a class="sourceLine" id="cb105-12" title="12">res &lt;-<span class="st"> </span><span class="kw">differentialGeneTest</span>(Obj_log, <span class="dt">fullModelFormulaStr =</span> <span class="st">&quot;~group&quot;</span>)</a>
<a class="sourceLine" id="cb105-13" title="13"></a>
<a class="sourceLine" id="cb105-14" title="14">pVals &lt;-<span class="st"> </span>res[,<span class="dv">3</span>]</a>
<a class="sourceLine" id="cb105-15" title="15"><span class="kw">names</span>(pVals) &lt;-<span class="st"> </span><span class="kw">rownames</span>(res)</a>
<a class="sourceLine" id="cb105-16" title="16">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb105-17" title="17"><span class="kw">DE_Quality_AUC</span>(pVals)</a>
<a class="sourceLine" id="cb105-18" title="18"><span class="co">## [1] 0.8294642</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:Monocle-plot2"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-real_files/figure-html/Monocle-plot2-1.png" alt="ROC curve for Monocle-gaussian."  />
<p class="caption">
Figure 8.18: ROC curve for Monocle-gaussian.
</p>
</div>
</div>
<div id="mast" class="section level3">
<h3><span class="header-section-number">8.7.6</span> MAST</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/MAST.html">MAST</a>基于零膨胀负二项模型。 它使用hurdle模型检验差异表达，组合基因表达的离散（0对非0）检验和连续（非零值）检验。同样，其使用线性建模框架来考虑复杂模型。</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb106-1" title="1">log_counts &lt;-<span class="st"> </span><span class="kw">log</span>(counts <span class="op">+</span><span class="st"> </span><span class="dv">1</span>) <span class="op">/</span><span class="st"> </span><span class="kw">log</span>(<span class="dv">2</span>)</a>
<a class="sourceLine" id="cb106-2" title="2">fData &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">names =</span> <span class="kw">rownames</span>(log_counts))</a>
<a class="sourceLine" id="cb106-3" title="3"><span class="kw">rownames</span>(fData) &lt;-<span class="st"> </span><span class="kw">rownames</span>(log_counts);</a>
<a class="sourceLine" id="cb106-4" title="4">cData &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">cond =</span> group)</a>
<a class="sourceLine" id="cb106-5" title="5"><span class="kw">rownames</span>(cData) &lt;-<span class="st"> </span><span class="kw">colnames</span>(log_counts)</a>
<a class="sourceLine" id="cb106-6" title="6"></a>
<a class="sourceLine" id="cb106-7" title="7">obj &lt;-<span class="st"> </span><span class="kw">FromMatrix</span>(<span class="kw">as.matrix</span>(log_counts), cData, fData)</a>
<a class="sourceLine" id="cb106-8" title="8"><span class="kw">colData</span>(obj)<span class="op">$</span>cngeneson &lt;-<span class="st"> </span><span class="kw">scale</span>(<span class="kw">colSums</span>(<span class="kw">assay</span>(obj) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>))</a>
<a class="sourceLine" id="cb106-9" title="9">cond &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">colData</span>(obj)<span class="op">$</span>cond)</a>
<a class="sourceLine" id="cb106-10" title="10"></a>
<a class="sourceLine" id="cb106-11" title="11"><span class="co"># Model expression as function of condition &amp; number of detected genes</span></a>
<a class="sourceLine" id="cb106-12" title="12">zlmCond &lt;-<span class="st"> </span><span class="kw">zlm.SingleCellAssay</span>(<span class="op">~</span><span class="st"> </span>cond <span class="op">+</span><span class="st"> </span>cngeneson, obj) </a>
<a class="sourceLine" id="cb106-13" title="13"></a>
<a class="sourceLine" id="cb106-14" title="14">summaryCond &lt;-<span class="st"> </span><span class="kw">summary</span>(zlmCond, <span class="dt">doLRT =</span> <span class="st">&quot;condNA19101&quot;</span>)</a>
<a class="sourceLine" id="cb106-15" title="15">summaryDt &lt;-<span class="st"> </span>summaryCond<span class="op">$</span>datatable</a>
<a class="sourceLine" id="cb106-16" title="16"></a>
<a class="sourceLine" id="cb106-17" title="17">summaryDt &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(summaryDt)</a>
<a class="sourceLine" id="cb106-18" title="18">pVals &lt;-<span class="st"> </span><span class="kw">unlist</span>(summaryDt[summaryDt<span class="op">$</span>component <span class="op">==</span><span class="st"> &quot;H&quot;</span>,<span class="dv">4</span>]) <span class="co"># H = hurdle model</span></a>
<a class="sourceLine" id="cb106-19" title="19"><span class="kw">names</span>(pVals) &lt;-<span class="st"> </span><span class="kw">unlist</span>(summaryDt[summaryDt<span class="op">$</span>component <span class="op">==</span><span class="st"> &quot;H&quot;</span>,<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb106-20" title="20">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb106-21" title="21"><span class="kw">DE_Quality_AUC</span>(pVals)</a>
<a class="sourceLine" id="cb106-22" title="22"><span class="co">## [1] 0.8284046</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:MAST-plot"></span>
<img src="https://scrnaseq-course.cog.sanger.ac.uk/website/de-real_files/figure-html/MAST-plot-1.png" alt="ROC curve for MAST."  />
<p class="caption">
Figure 8.19: ROC curve for MAST.
</p>
</div>
</div>
<div id="slow-method1h" class="section level3">
<h3><span class="header-section-number">8.7.7</span> Slow method(&gt;1h)</h3>
<p>这些方法今天运行速度太慢，但鼓励您自己尝试。</p>
<div id="bpsc" class="section level4">
<h4><span class="header-section-number">8.7.7.1</span> BPSC</h4>
<p><a href="https://academic.oup.com/bioinformatics/article/32/14/2128/2288270/Beta-Poisson-model-for-single-cell-RNA-seq-data">BPSC</a>使用单细胞基因表达的Poisson-Beta模型，并将与广义线性模型相结合。 BPSC将一个或多个组与参考组（“Control”）进行比较，并且可以包括其他因素，例如模型中的批次效应。</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb107-1" title="1"><span class="kw">library</span>(BPSC)</a>
<a class="sourceLine" id="cb107-2" title="2">bpsc_data &lt;-<span class="st"> </span>norm[,batch<span class="op">==</span><span class="st">&quot;NA19101.r1&quot;</span> <span class="op">|</span><span class="st"> </span>batch<span class="op">==</span><span class="st">&quot;NA19239.r1&quot;</span>]</a>
<a class="sourceLine" id="cb107-3" title="3">bpsc_group =<span class="st"> </span>group[batch<span class="op">==</span><span class="st">&quot;NA19101.r1&quot;</span> <span class="op">|</span><span class="st"> </span>batch<span class="op">==</span><span class="st">&quot;NA19239.r1&quot;</span>]</a>
<a class="sourceLine" id="cb107-4" title="4"></a>
<a class="sourceLine" id="cb107-5" title="5">control_cells &lt;-<span class="st"> </span><span class="kw">which</span>(bpsc_group <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>)</a>
<a class="sourceLine" id="cb107-6" title="6">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span>bpsc_group)</a>
<a class="sourceLine" id="cb107-7" title="7">coef=<span class="dv">2</span> <span class="co"># group label</span></a>
<a class="sourceLine" id="cb107-8" title="8">res=<span class="kw">BPglm</span>(<span class="dt">data=</span>bpsc_data, <span class="dt">controlIds=</span>control_cells, <span class="dt">design=</span>design, <span class="dt">coef=</span>coef, </a>
<a class="sourceLine" id="cb107-9" title="9">                <span class="dt">estIntPar=</span><span class="ot">FALSE</span>, <span class="dt">useParallel =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb107-10" title="10">pVals =<span class="st"> </span>res<span class="op">$</span>PVAL</a>
<a class="sourceLine" id="cb107-11" title="11">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb107-12" title="12"><span class="kw">DE_Quality_AUC</span>(pVals)</a></code></pre></div>
</div>
<div id="scde" class="section level4">
<h4><span class="header-section-number">8.7.7.2</span> SCDE</h4>
<p><a href="http://hms-dbmi.github.io/scde/">SCDE</a>是单细胞特异性差异表达分析方法。它使用贝叶斯统计量将零膨胀负二项式模型拟合表达数据。 以下用法检验各组间基因的平均表达差异，但最近的版本包括检验平均表达差异或一组基因的离散程度，通常代表一条通路。</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb108-1" title="1"><span class="kw">library</span>(scde)</a>
<a class="sourceLine" id="cb108-2" title="2">cnts &lt;-<span class="st"> </span><span class="kw">apply</span>(</a>
<a class="sourceLine" id="cb108-3" title="3">    counts,</a>
<a class="sourceLine" id="cb108-4" title="4">    <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb108-5" title="5">    <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb108-6" title="6">        <span class="kw">storage.mode</span>(x) &lt;-<span class="st"> &#39;integer&#39;</span></a>
<a class="sourceLine" id="cb108-7" title="7">        <span class="kw">return</span>(x)</a>
<a class="sourceLine" id="cb108-8" title="8">    }</a>
<a class="sourceLine" id="cb108-9" title="9">)</a>
<a class="sourceLine" id="cb108-10" title="10"><span class="kw">names</span>(group) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(group)</a>
<a class="sourceLine" id="cb108-11" title="11"><span class="kw">colnames</span>(cnts) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(group)</a>
<a class="sourceLine" id="cb108-12" title="12">o.ifm &lt;-<span class="st"> </span>scde<span class="op">::</span><span class="kw">scde.error.models</span>(</a>
<a class="sourceLine" id="cb108-13" title="13">    <span class="dt">counts =</span> cnts,</a>
<a class="sourceLine" id="cb108-14" title="14">    <span class="dt">groups =</span> group,</a>
<a class="sourceLine" id="cb108-15" title="15">    <span class="dt">n.cores =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb108-16" title="16">    <span class="dt">threshold.segmentation =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb108-17" title="17">    <span class="dt">save.crossfit.plots =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb108-18" title="18">    <span class="dt">save.model.plots =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb108-19" title="19">    <span class="dt">verbose =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb108-20" title="20">    <span class="dt">min.size.entries =</span> <span class="dv">2</span></a>
<a class="sourceLine" id="cb108-21" title="21">)</a>
<a class="sourceLine" id="cb108-22" title="22">priors &lt;-<span class="st"> </span>scde<span class="op">::</span><span class="kw">scde.expression.prior</span>(</a>
<a class="sourceLine" id="cb108-23" title="23">    <span class="dt">models =</span> o.ifm,</a>
<a class="sourceLine" id="cb108-24" title="24">    <span class="dt">counts =</span> cnts,</a>
<a class="sourceLine" id="cb108-25" title="25">    <span class="dt">length.out =</span> <span class="dv">400</span>,</a>
<a class="sourceLine" id="cb108-26" title="26">    <span class="dt">show.plot =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb108-27" title="27">)</a>
<a class="sourceLine" id="cb108-28" title="28">resSCDE &lt;-<span class="st"> </span>scde<span class="op">::</span><span class="kw">scde.expression.difference</span>(</a>
<a class="sourceLine" id="cb108-29" title="29">    o.ifm,</a>
<a class="sourceLine" id="cb108-30" title="30">    cnts,</a>
<a class="sourceLine" id="cb108-31" title="31">    priors,</a>
<a class="sourceLine" id="cb108-32" title="32">    <span class="dt">groups =</span> group,</a>
<a class="sourceLine" id="cb108-33" title="33">    <span class="dt">n.randomizations =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb108-34" title="34">    <span class="dt">n.cores =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb108-35" title="35">    <span class="dt">verbose =</span> <span class="dv">0</span></a>
<a class="sourceLine" id="cb108-36" title="36">)</a>
<a class="sourceLine" id="cb108-37" title="37"><span class="co"># Convert Z-scores into 2-tailed p-values</span></a>
<a class="sourceLine" id="cb108-38" title="38">pVals &lt;-<span class="st"> </span><span class="kw">pnorm</span>(<span class="kw">abs</span>(resSCDE<span class="op">$</span>cZ), <span class="dt">lower.tail =</span> <span class="ot">FALSE</span>) <span class="op">*</span><span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb108-39" title="39"><span class="kw">DE_Quality_AUC</span>(pVals)</a></code></pre></div>
</div>
</div>
</div>
<div id="scrna-seq-2" class="section level2">
<h2><span class="header-section-number">8.8</span> 比较/组合scRNA-seq数据集</h2>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb109-1" title="1"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb109-2" title="2"><span class="kw">library</span>(SingleCellExperiment)</a></code></pre></div>
<div id="section-96" class="section level3">
<h3><span class="header-section-number">8.8.1</span> 介绍</h3>
<p>随着scRNA-seq数据集越来越多，进行merged_seurat比较是关键。比较scRNA-Seq数据集有两种主要方法。 第一种方法是“label-centric”，其通过比较单个细胞或一组细胞识别数据集间相同的细胞类型/状态。另一种方法是“cross-dataset normalization”，试图去除特定实验的技术/生物效应，使得组合联合分析来自多个实验的数据。</p>
<p>label-centric方法可以与高置信度细胞注释的数据集一起使用，例如，人类细胞图谱（HCA）<span class="citation">(Regev et al. <a href="#ref-Regev2017-mw">2017</a>)</span>或Tabula Muris <span class="citation">(Consortium and others <a href="#ref-tabula2018single">2018</a>)</span>。一旦这些项目完成，将新样本的细胞或cluster投影到该参考图谱上以识别组织成分或识别细胞类型。 从概念上讲，这种预测类似于流行的BLAST方法<span class="citation">(Altschul et al. <a href="#ref-Altschul1990-ts">1990</a>)</span>，这使得可以在数据库中快速找到最接近的匹配，用于新鉴定的核苷酸或氨基酸序列。 label-centric方法还可用于比较由不同实验室收集的相似生物来源的数据集，以确保注释和分析是一致的。</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-137"></span>
<img src="figures/CourseCompareTypes.png" alt="Label-centric dataset comparison can be used to compare the annotations of two different samples." width="950" />
<p class="caption">
Figure 8.20: Label-centric dataset comparison can be used to compare the annotations of two different samples.
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-138"></span>
<img src="figures/CourseAtlasAssignment.png" alt="Label-centric dataset comparison can project cells from a new experiment onto an annotated reference." width="871" />
<p class="caption">
Figure 8.21: Label-centric dataset comparison can project cells from a new experiment onto an annotated reference.
</p>
</div>
<p>cross-dataset normalization方法也可用于比较相似生物来源的数据集，不像label-centric方法，它可以实现多个数据集的联合分析，以便于更可靠地识别稀有细胞类型，其可能在每个个体中过于稀疏地采样。然而，cross-dataset normalization不适用于非常大且不同的reference，因为它假设每个数据集中的生物变异性的重要部分与其他数据集重叠。</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-139"></span>
<img src="figures/CourseCrossNorm.png" alt="Cross-dataset normalization enables joint-analysis of 2+ scRNASeq datasets." width="822" />
<p class="caption">
Figure 8.22: Cross-dataset normalization enables joint-analysis of 2+ scRNASeq datasets.
</p>
</div>
</div>
<div id="datasets" class="section level3">
<h3><span class="header-section-number">8.8.2</span> Datasets</h3>
<p>在两个人胰腺数据集上运行这些方法：<span class="citation">(Muraro et al. <a href="#ref-Muraro2016-yk">2016</a>)</span>和<span class="citation">(Segerstolpe et al. <a href="#ref-Segerstolpe2016-wc">2016</a>)</span>。 由于胰腺已被广泛研究，这些数据集得到了很好的注释。</p>
<p>此数据已针对scmap格式化。细胞类型label必须存储在<code>colData</code>的<code>cell_type1</code>列中，并且两个数据集之间一致的基因ID必须存储在<code>rowData</code>的<code>feature_symbol</code>列中。</p>
<p>首先，检查两个数据集中的gene-id匹配情况:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb110-1" title="1"><span class="kw">sum</span>(<span class="kw">rowData</span>(muraro)<span class="op">$</span>feature_symbol <span class="op">%in%</span><span class="st"> </span><span class="kw">rowData</span>(segerstolpe)<span class="op">$</span>feature_symbol)<span class="op">/</span><span class="kw">nrow</span>(muraro)</a>
<a class="sourceLine" id="cb110-2" title="2"><span class="co">## [1] 0.9599519</span></a>
<a class="sourceLine" id="cb110-3" title="3"><span class="kw">sum</span>(<span class="kw">rowData</span>(segerstolpe)<span class="op">$</span>feature_symbol <span class="op">%in%</span><span class="st"> </span><span class="kw">rowData</span>(muraro)<span class="op">$</span>feature_symbol)<span class="op">/</span><span class="kw">nrow</span>(segerstolpe)</a>
<a class="sourceLine" id="cb110-4" title="4"><span class="co">## [1] 0.719334</span></a></code></pre></div>
<p>muraro中6％基因与segerstople中的基因匹配，而segerstolpe中72％的基因与muraro匹配。因为segerstolpe数据集比muraro数据集测序深度更深。 然而，它突出了比较scRNA-Seq数据集的一些困难。</p>
<p>通过检查这两个数据集的总体大小来确认这一点。</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb111-1" title="1"><span class="kw">dim</span>(muraro)</a>
<a class="sourceLine" id="cb111-2" title="2"><span class="co">## [1] 19127  2126</span></a>
<a class="sourceLine" id="cb111-3" title="3"><span class="kw">dim</span>(segerstolpe)</a>
<a class="sourceLine" id="cb111-4" title="4"><span class="co">## [1] 25525  3514</span></a></code></pre></div>
<p>此外，使用以下命令检查每个数据集的细胞类型注释：</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb112-1" title="1"><span class="kw">summary</span>(<span class="kw">factor</span>(<span class="kw">colData</span>(muraro)<span class="op">$</span>cell_type1))</a>
<a class="sourceLine" id="cb112-2" title="2"><span class="co">##      acinar       alpha        beta       delta      ductal endothelial </span></a>
<a class="sourceLine" id="cb112-3" title="3"><span class="co">##         219         812         448         193         245          21 </span></a>
<a class="sourceLine" id="cb112-4" title="4"><span class="co">##     epsilon       gamma mesenchymal     unclear </span></a>
<a class="sourceLine" id="cb112-5" title="5"><span class="co">##           3         101          80           4</span></a>
<a class="sourceLine" id="cb112-6" title="6"><span class="kw">summary</span>(<span class="kw">factor</span>(<span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1))</a>
<a class="sourceLine" id="cb112-7" title="7"><span class="co">##                 acinar                  alpha                   beta </span></a>
<a class="sourceLine" id="cb112-8" title="8"><span class="co">##                    185                    886                    270 </span></a>
<a class="sourceLine" id="cb112-9" title="9"><span class="co">##          co-expression                  delta                 ductal </span></a>
<a class="sourceLine" id="cb112-10" title="10"><span class="co">##                     39                    114                    386 </span></a>
<a class="sourceLine" id="cb112-11" title="11"><span class="co">##            endothelial                epsilon                  gamma </span></a>
<a class="sourceLine" id="cb112-12" title="12"><span class="co">##                     16                      7                    197 </span></a>
<a class="sourceLine" id="cb112-13" title="13"><span class="co">##                   mast           MHC class II         not applicable </span></a>
<a class="sourceLine" id="cb112-14" title="14"><span class="co">##                      7                      5                   1305 </span></a>
<a class="sourceLine" id="cb112-15" title="15"><span class="co">##                    PSC           unclassified unclassified endocrine </span></a>
<a class="sourceLine" id="cb112-16" title="16"><span class="co">##                     54                      2                     41</span></a></code></pre></div>
<p>即使两个数据集来自相同的生物组织，注释细胞类型也略有差异。如果您熟悉胰腺生物学，您可能会认识到segerstolpe中的pancreatic stellate cells (PSCs)是一种间充质干细胞，在muraro中属于"mesenchymal“类型。但是，目前尚不清楚这两个注释是否应该被视为同义词。 可以使用label-centric的比较方法来确定这两个细胞类型的注释是否确实相同。</p>
<p>或者，我们可以通过利用数据集中的形成来理解那些“unclassified endocrine”或被认为对每个数据集中的原始聚类质量太差（“not applicable”）的细胞的功能。 我们可以尝试使用label-centric的方法来推断它们最可能属于哪个现有注释，或者使用cross-dataset normalization来揭示其中的新颖细胞类型（或现有注释中的子类型）。</p>
<p>为了简化演示分析，我们将删除少量未分配类型的细胞和质量较差的细胞。但保留“unclassified endocrine”，看看这些方法中是否阐明它们属于哪种细胞类型。</p>
</div>
<div id="scmap" class="section level3">
<h3><span class="header-section-number">8.8.3</span> scmap</h3>
<p><code>scmap</code> <span class="citation">(Kiselev and Hemberg <a href="#ref-Kiselev2017-nb">2017</a>)</span>，一种将scRNA-seq实验细胞投影到其他实验中鉴定的细胞类型上的方法。 此外，云版本的scmap可以免费运行，但有merged_seurat限制， <a href="http://www.hemberg-lab.cloud/scmap">http://www.hemberg-lab.cloud/scmap</a>。</p>
<div id="section-97" class="section level4">
<h4><span class="header-section-number">8.8.3.1</span> 特征选择</h4>
<p>创建<code>SingleCellExperiment</code>对象后可以直接运行<code>scmap</code>。首先构建参考cluster的“索引”。 由于我们想知道PSC和间充质细胞是否是同义词，需要将每个数据集投射到另一个，因此为每个数据集构建索引。 这需要首先为参考数据集选择信息量最大的特征。</p>
<pre><code>## Warning in linearModel(object, n_features): Your object does not contain
## counts() slot. Dropouts were calculated using logcounts() slot...</code></pre>
<p><img src="08biological-analysis_files/figure-html/unnamed-chunk-146-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>红色高亮的基因将用于进一步分析（投影）。</p>
<p><img src="08biological-analysis_files/figure-html/unnamed-chunk-148-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>从这些图的y轴可以看出scmap使用基于dropmerged_seurat的特征选择方法。</p>
<p>计算细胞类型索引:</p>
<p>可视化索引:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb114-1" title="1"><span class="kw">heatmap</span>(<span class="kw">as.matrix</span>(<span class="kw">metadata</span>(muraro)<span class="op">$</span>scmap_cluster_index))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/projection_files/figure-html/unnamed-chunk-15-1.png" style="display: block; margin: auto;" /></p>
<p>如果特征过于集中在少数细胞类型中，可以使用<code>setFeatures</code>函数调整特征。</p>
<p><strong>练习</strong>
使用每个数据集的rowData，两个数据集中有多少基因作为特征？</p>
<p><strong>答案</strong></p>
</div>
<div id="section-98" class="section level4">
<h4><span class="header-section-number">8.8.3.2</span> 投影</h4>
<p>scmap计算从每个细胞到参考索引中每个细胞类型的距离，然后应用经验阈值来确定将哪些细胞分配给最近的参考细胞类型以及哪些细胞未分配。 为了解释测序深度的差异，使用spearman相关系数和余弦距离计算距离，并且仅当两种距离返回一致结果时，将结果分配给该细胞。</p>
<p>将<code>segerstolpe</code>投影到<code>muraro</code>数据集:</p>
<p>将<code>muraro</code>投影到<code>segerstolpe</code></p>
<p>请注意，在每种情况下，我们都会投影到单个数据集，但这可以扩展到任意数量已计算索引的数据集。</p>
<p>比较原始细胞类型label与投影label：</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb115-1" title="1"><span class="kw">table</span>(<span class="kw">colData</span>(muraro)<span class="op">$</span>cell_type1, muraro_to_seger<span class="op">$</span>scmap_cluster_labs)</a>
<a class="sourceLine" id="cb115-2" title="2"><span class="co">##              </span></a>
<a class="sourceLine" id="cb115-3" title="3"><span class="co">##               acinar alpha beta co-expression delta ductal endothelial</span></a>
<a class="sourceLine" id="cb115-4" title="4"><span class="co">##   acinar         211     0    0             0     0      0           0</span></a>
<a class="sourceLine" id="cb115-5" title="5"><span class="co">##   alpha            1   763    0            18     0      2           0</span></a>
<a class="sourceLine" id="cb115-6" title="6"><span class="co">##   beta             2     1  397             7     2      2           0</span></a>
<a class="sourceLine" id="cb115-7" title="7"><span class="co">##   delta            0     0    2             1   173      0           0</span></a>
<a class="sourceLine" id="cb115-8" title="8"><span class="co">##   ductal           7     0    0             0     0    208           0</span></a>
<a class="sourceLine" id="cb115-9" title="9"><span class="co">##   endothelial      0     0    0             0     0      0          15</span></a>
<a class="sourceLine" id="cb115-10" title="10"><span class="co">##   epsilon          0     0    0             0     0      0           0</span></a>
<a class="sourceLine" id="cb115-11" title="11"><span class="co">##   gamma            2     0    0             0     0      0           0</span></a>
<a class="sourceLine" id="cb115-12" title="12"><span class="co">##   mesenchymal      0     0    0             0     0      1           0</span></a>
<a class="sourceLine" id="cb115-13" title="13"><span class="co">##              </span></a>
<a class="sourceLine" id="cb115-14" title="14"><span class="co">##               epsilon gamma MHC class II PSC unassigned</span></a>
<a class="sourceLine" id="cb115-15" title="15"><span class="co">##   acinar            0     0            0   0          8</span></a>
<a class="sourceLine" id="cb115-16" title="16"><span class="co">##   alpha             0     2            0   0         26</span></a>
<a class="sourceLine" id="cb115-17" title="17"><span class="co">##   beta              0     5            1   2         29</span></a>
<a class="sourceLine" id="cb115-18" title="18"><span class="co">##   delta             0     0            0   0         17</span></a>
<a class="sourceLine" id="cb115-19" title="19"><span class="co">##   ductal            0     0            5   3         22</span></a>
<a class="sourceLine" id="cb115-20" title="20"><span class="co">##   endothelial       0     0            0   1          5</span></a>
<a class="sourceLine" id="cb115-21" title="21"><span class="co">##   epsilon           3     0            0   0          0</span></a>
<a class="sourceLine" id="cb115-22" title="22"><span class="co">##   gamma             0    95            0   0          4</span></a>
<a class="sourceLine" id="cb115-23" title="23"><span class="co">##   mesenchymal       0     0            0  77          2</span></a></code></pre></div>
<p>在这里，我们可以看到细胞类型映射到segerstolpe，重要的是除了一个“间充质“细胞之外的所有细胞都被分配到“PSC”类。</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb116-1" title="1"><span class="kw">table</span>(<span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1, seger_to_muraro<span class="op">$</span>scmap_cluster_labs)</a>
<a class="sourceLine" id="cb116-2" title="2"><span class="co">##                         </span></a>
<a class="sourceLine" id="cb116-3" title="3"><span class="co">##                          acinar alpha beta delta ductal endothelial</span></a>
<a class="sourceLine" id="cb116-4" title="4"><span class="co">##   acinar                    181     0    0     0      4           0</span></a>
<a class="sourceLine" id="cb116-5" title="5"><span class="co">##   alpha                       0   869    1     0      0           0</span></a>
<a class="sourceLine" id="cb116-6" title="6"><span class="co">##   beta                        0     0  260     0      0           0</span></a>
<a class="sourceLine" id="cb116-7" title="7"><span class="co">##   co-expression               0     7   31     0      0           0</span></a>
<a class="sourceLine" id="cb116-8" title="8"><span class="co">##   delta                       0     0    1   111      0           0</span></a>
<a class="sourceLine" id="cb116-9" title="9"><span class="co">##   ductal                      0     0    0     0    383           0</span></a>
<a class="sourceLine" id="cb116-10" title="10"><span class="co">##   endothelial                 0     0    0     0      0          14</span></a>
<a class="sourceLine" id="cb116-11" title="11"><span class="co">##   epsilon                     0     0    0     0      0           0</span></a>
<a class="sourceLine" id="cb116-12" title="12"><span class="co">##   gamma                       0     2    0     0      0           0</span></a>
<a class="sourceLine" id="cb116-13" title="13"><span class="co">##   mast                        0     0    0     0      0           0</span></a>
<a class="sourceLine" id="cb116-14" title="14"><span class="co">##   MHC class II                0     0    0     0      0           0</span></a>
<a class="sourceLine" id="cb116-15" title="15"><span class="co">##   PSC                         0     0    1     0      0           0</span></a>
<a class="sourceLine" id="cb116-16" title="16"><span class="co">##   unclassified endocrine      0     0    0     0      0           0</span></a>
<a class="sourceLine" id="cb116-17" title="17"><span class="co">##                         </span></a>
<a class="sourceLine" id="cb116-18" title="18"><span class="co">##                          epsilon gamma mesenchymal unassigned</span></a>
<a class="sourceLine" id="cb116-19" title="19"><span class="co">##   acinar                       0     0           0          0</span></a>
<a class="sourceLine" id="cb116-20" title="20"><span class="co">##   alpha                        0     0           0         16</span></a>
<a class="sourceLine" id="cb116-21" title="21"><span class="co">##   beta                         0     0           0         10</span></a>
<a class="sourceLine" id="cb116-22" title="22"><span class="co">##   co-expression                0     0           0          1</span></a>
<a class="sourceLine" id="cb116-23" title="23"><span class="co">##   delta                        0     0           0          2</span></a>
<a class="sourceLine" id="cb116-24" title="24"><span class="co">##   ductal                       0     0           0          3</span></a>
<a class="sourceLine" id="cb116-25" title="25"><span class="co">##   endothelial                  0     0           0          2</span></a>
<a class="sourceLine" id="cb116-26" title="26"><span class="co">##   epsilon                      6     0           0          1</span></a>
<a class="sourceLine" id="cb116-27" title="27"><span class="co">##   gamma                        0   192           0          3</span></a>
<a class="sourceLine" id="cb116-28" title="28"><span class="co">##   mast                         0     0           0          7</span></a>
<a class="sourceLine" id="cb116-29" title="29"><span class="co">##   MHC class II                 0     0           0          5</span></a>
<a class="sourceLine" id="cb116-30" title="30"><span class="co">##   PSC                          0     0          53          0</span></a>
<a class="sourceLine" id="cb116-31" title="31"><span class="co">##   unclassified endocrine       0     0           0         41</span></a></code></pre></div>
<p>再次看到细胞类型相互匹配，除了“PSC”所有细胞除一个外都与“间充质”细胞相匹配，这提供了强有力的证据证明这两个注释应该被认为是同义词。</p>
<p>可以通过 <a href="https://developers.google.com/chart/interactive/docs/gallery/sankey">桑葚图</a>对上述表格进行可视化:</p>
<!-- Sankey generated in R 3.6.0 by googleVis 0.6.4 package -->
<!-- Fri Jul 19 22:23:08 2019 -->
<!-- jsHeader -->
<script type="text/javascript">
 
// jsData 
function gvisDataSankeyID9ad86ff83a7 () {
var data = new google.visualization.DataTable();
var datajson =
[
 [
"alpha ",
" alpha",
763
],
[
"beta ",
" beta",
397
],
[
"acinar ",
" acinar",
211
],
[
"ductal ",
" ductal",
208
],
[
"delta ",
" delta",
173
],
[
"gamma ",
" gamma",
95
],
[
"mesenchymal ",
" PSC",
77
],
[
"endothelial ",
" endothelial",
15
],
[
"epsilon ",
" epsilon",
3
],
[
"acinar ",
" unassigned",
8
],
[
"alpha ",
" ductal",
2
],
[
"alpha ",
" unassigned",
26
],
[
"alpha ",
" acinar",
1
],
[
"alpha ",
" co-expression",
18
],
[
"alpha ",
" gamma",
2
],
[
"beta ",
" unassigned",
29
],
[
"beta ",
" gamma",
5
],
[
"beta ",
" MHC class II",
1
],
[
"beta ",
" alpha",
1
],
[
"beta ",
" co-expression",
7
],
[
"beta ",
" acinar",
2
],
[
"beta ",
" PSC",
2
],
[
"beta ",
" ductal",
2
],
[
"beta ",
" delta",
2
],
[
"delta ",
" beta",
2
],
[
"delta ",
" unassigned",
17
],
[
"delta ",
" co-expression",
1
],
[
"ductal ",
" acinar",
7
],
[
"ductal ",
" PSC",
3
],
[
"ductal ",
" MHC class II",
5
],
[
"ductal ",
" unassigned",
22
],
[
"endothelial ",
" PSC",
1
],
[
"endothelial ",
" unassigned",
5
],
[
"gamma ",
" acinar",
2
],
[
"gamma ",
" unassigned",
4
],
[
"mesenchymal ",
" ductal",
1
],
[
"mesenchymal ",
" unassigned",
2
] 
];
data.addColumn('string','From');
data.addColumn('string','To');
data.addColumn('number','# of cells');
data.addRows(datajson);
return(data);
}
 
// jsDrawChart
function drawChartSankeyID9ad86ff83a7() {
var data = gvisDataSankeyID9ad86ff83a7();
var options = {};
options["width"] = 400;
options["height"] = 400;
options["sankey"] = {
                node:{
                    label:{
                        fontName:'Arial',
                        fontSize:11,color:
                        '#000000',
                        bold:true,
                        italic:false
                    },
                    colors:'#FFFFFF',
                    nodePadding:12
                },iterations:0
            };

    var chart = new google.visualization.Sankey(
    document.getElementById('SankeyID9ad86ff83a7')
    );
    chart.draw(data,options);
    

}
  
 
// jsDisplayChart
(function() {
var pkgs = window.__gvisPackages = window.__gvisPackages || [];
var callbacks = window.__gvisCallbacks = window.__gvisCallbacks || [];
var chartid = "sankey";
  
// Manually see if chartid is in pkgs (not all browsers support Array.indexOf)
var i, newPackage = true;
for (i = 0; newPackage && i < pkgs.length; i++) {
if (pkgs[i] === chartid)
newPackage = false;
}
if (newPackage)
  pkgs.push(chartid);
  
// Add the drawChart function to the global list of callbacks
callbacks.push(drawChartSankeyID9ad86ff83a7);
})();
function displayChartSankeyID9ad86ff83a7() {
  var pkgs = window.__gvisPackages = window.__gvisPackages || [];
  var callbacks = window.__gvisCallbacks = window.__gvisCallbacks || [];
  window.clearTimeout(window.__gvisLoad);
  // The timeout is set to 100 because otherwise the container div we are
  // targeting might not be part of the document yet
  window.__gvisLoad = setTimeout(function() {
  var pkgCount = pkgs.length;
  google.load("visualization", "1", { packages:pkgs, callback: function() {
  if (pkgCount != pkgs.length) {
  // Race condition where another setTimeout call snuck in after us; if
  // that call added a package, we must not shift its callback
  return;
}
while (callbacks.length > 0)
callbacks.shift()();
} });
}, 100);
}
 
// jsFooter
</script>
<!-- jsChart -->
<script type="text/javascript" src="https://www.google.com/jsapi?callback=displayChartSankeyID9ad86ff83a7"></script>
<!-- divChart -->
<div id="SankeyID9ad86ff83a7" style="width: 400; height: 400;">

</div>
<p><strong>练习</strong></p>
<p>之前未分类的细胞有多少可以使用scmap分配细胞类型？</p>
<p><strong>答案</strong></p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb117-1" title="1">t2 &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1, seger_to_muraro<span class="op">$</span>scmap_cluster_labs)</a>
<a class="sourceLine" id="cb117-2" title="2">seger_assigned &lt;-<span class="st"> </span><span class="kw">sum</span>(t2[<span class="kw">c</span>(<span class="st">&quot;unclassified endocrine&quot;</span>), <span class="dv">1</span><span class="op">:</span><span class="dv">9</span>])</a></code></pre></div>
</div>
</div>
<div id="section-99" class="section level3">
<h3><span class="header-section-number">8.8.4</span> 细胞间映射</h3>
<p>scmap还可以将一个数据集中的每个细胞投影到参考数据集中的近似最近邻细胞。 这使用高度优化的搜索算法，允许它扩展到非常大的参考空间（理论上是100,000万个细胞）。 但是，这个过程是随机的，因此必须固定随机种子以确保可以重现结果。</p>
<p>我们已经为此数据集进行了特征选择，可以直接构建索引。</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb118-1" title="1"><span class="kw">set.seed</span>(<span class="dv">193047</span>)</a>
<a class="sourceLine" id="cb118-2" title="2">segerstolpe &lt;-<span class="st"> </span><span class="kw">indexCell</span>(segerstolpe)</a>
<a class="sourceLine" id="cb118-3" title="3">muraro &lt;-<span class="st"> </span><span class="kw">indexCell</span>(muraro)</a></code></pre></div>
<p>在这种情况下，索引是使用不同特征集的每个细胞的一系列聚类，参数k和M是聚类的数量和在这些子聚类中使用的特征的数量。 将新细胞分配给每个subcluster中最近的cluster，以生成唯一的聚类分配模式。 然后，在参考数据集中寻找具有相同或最相似的聚类分配模式的细胞。</p>
<p>检查参考数据集的聚类分配模式</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb119-1" title="1"><span class="kw">metadata</span>(muraro)<span class="op">$</span>scmap_cell_index<span class="op">$</span>subclusters[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb119-2" title="2"><span class="co">##      D28.1_1 D28.1_13 D28.1_15 D28.1_17 D28.1_2</span></a>
<a class="sourceLine" id="cb119-3" title="3"><span class="co">## [1,]       4        6        4       24      34</span></a>
<a class="sourceLine" id="cb119-4" title="4"><span class="co">## [2,]      16       27       14       11       3</span></a>
<a class="sourceLine" id="cb119-5" title="5"><span class="co">## [3,]      18       12       17       26      24</span></a>
<a class="sourceLine" id="cb119-6" title="6"><span class="co">## [4,]      44       33       41       44      12</span></a>
<a class="sourceLine" id="cb119-7" title="7"><span class="co">## [5,]      39       44       23       34      32</span></a></code></pre></div>
<p>投影并寻找<code>w</code>最近邻：</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb120-1" title="1">muraro_to_seger &lt;-<span class="st"> </span><span class="kw">scmapCell</span>(</a>
<a class="sourceLine" id="cb120-2" title="2">  <span class="dt">projection =</span> muraro,</a>
<a class="sourceLine" id="cb120-3" title="3">  <span class="dt">index_list =</span> <span class="kw">list</span>(</a>
<a class="sourceLine" id="cb120-4" title="4">    <span class="dt">seger =</span> <span class="kw">metadata</span>(segerstolpe)<span class="op">$</span>scmap_cell_index</a>
<a class="sourceLine" id="cb120-5" title="5">  ),</a>
<a class="sourceLine" id="cb120-6" title="6">  <span class="dt">w =</span> <span class="dv">5</span></a>
<a class="sourceLine" id="cb120-7" title="7">)</a></code></pre></div>
<p>查看结果:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb121-1" title="1">muraro_to_seger<span class="op">$</span>seger[[<span class="dv">1</span>]][,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb121-2" title="2"><span class="co">##      D28.1_1 D28.1_13 D28.1_15 D28.1_17 D28.1_2</span></a>
<a class="sourceLine" id="cb121-3" title="3"><span class="co">## [1,]    1448     1381     1125     1834    1553</span></a>
<a class="sourceLine" id="cb121-4" title="4"><span class="co">## [2,]    1731      855     1081     1731    1437</span></a>
<a class="sourceLine" id="cb121-5" title="5"><span class="co">## [3,]    1229     2113     2104     1790    1890</span></a>
<a class="sourceLine" id="cb121-6" title="6"><span class="co">## [4,]    2015     1833     1153     1882    1593</span></a>
<a class="sourceLine" id="cb121-7" title="7"><span class="co">## [5,]    1159     1854     1731     1202    1078</span></a></code></pre></div>
<p>这显示了segerstolpe到muraro中的细胞5个最近邻居的列号。然后从segerstolpe数据的colData中选择适当的数据来计算伪时间估计，分支分配或其他细胞水平数据。作为演示，查找每个细胞的最近邻居的细胞类型。</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb122-1" title="1">cell_type_NN &lt;-<span class="st"> </span><span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1[muraro_to_seger<span class="op">$</span>seger[[<span class="dv">1</span>]][<span class="dv">1</span>,]]</a>
<a class="sourceLine" id="cb122-2" title="2"><span class="kw">head</span>(cell_type_NN)</a>
<a class="sourceLine" id="cb122-3" title="3">cell_type_NN &lt;-<span class="st"> </span><span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1[muraro_to_seger<span class="op">$</span>seger[[<span class="dv">1</span>]][<span class="dv">1</span>,]]</a>
<a class="sourceLine" id="cb122-4" title="4"><span class="kw">head</span>(cell_type_NN)</a></code></pre></div>
</div>
<div id="metaneighbour" class="section level3">
<h3><span class="header-section-number">8.8.5</span> Metaneighbour</h3>
<p><a href="https://www.biorxiv.org/content/early/2017/06/16/150524">Metaneighbour</a>专门用于检查细胞类型是否在数据集之间保持一致。它有两个版本。 首先是一种完全监督的方法，它假设所有数据集中细胞类型一致，并计算这些细胞类型标记的“good”的程度。（下面将描述“good”的确切含义）。 或者，metaneighbour可以估计数据集内和数据集之间所有细胞类型彼此之间的相似程度。 我们将只使用无监督版本，因为它具有更广泛的适用性，并且更容易解释结果。</p>
<p>Metaneighbour通过构建细胞-细胞spearman相关网络来比较数据集中的细胞类型。然后通过最近邻居的加权投票来预测细胞类型。 然后将两个cluster之间的整体相似性评分为AUROC，基于加权投票将类型A的细胞分配给类型B. AUROC为1表示在任何其他细胞出现之前，所有类型A的细胞都被分配到类型B，如果细胞随机分配，则AUROC为0.5。</p>
<p>Metanighbour只是一些R函数而不是一个完整的R包，所以必须使用<code>source</code>加载它们</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" title="1"><span class="kw">source</span>(<span class="st">&quot;utils/2017-08-28-runMN-US.R&quot;</span>)</a></code></pre></div>
<div id="section-100" class="section level4">
<h4><span class="header-section-number">8.8.5.1</span> 数据准备</h4>
<p>Metaneighbour要求在运行之前将所有数据集组合成单个表达矩阵：</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb124-1" title="1">is.common &lt;-<span class="st"> </span><span class="kw">rowData</span>(muraro)<span class="op">$</span>feature_symbol <span class="op">%in%</span><span class="st"> </span><span class="kw">rowData</span>(segerstolpe)<span class="op">$</span>feature_symbol</a>
<a class="sourceLine" id="cb124-2" title="2">muraro &lt;-<span class="st"> </span>muraro[is.common,]</a>
<a class="sourceLine" id="cb124-3" title="3">segerstolpe &lt;-<span class="st"> </span>segerstolpe[<span class="kw">match</span>(<span class="kw">rowData</span>(muraro)<span class="op">$</span>feature_symbol, <span class="kw">rowData</span>(segerstolpe)<span class="op">$</span>feature_symbol),]</a>
<a class="sourceLine" id="cb124-4" title="4"><span class="kw">rownames</span>(segerstolpe) &lt;-<span class="st"> </span><span class="kw">rowData</span>(segerstolpe)<span class="op">$</span>feature_symbol</a>
<a class="sourceLine" id="cb124-5" title="5"><span class="kw">rownames</span>(muraro) &lt;-<span class="st"> </span><span class="kw">rowData</span>(muraro)<span class="op">$</span>feature_symbol</a>
<a class="sourceLine" id="cb124-6" title="6"><span class="kw">identical</span>(<span class="kw">rownames</span>(segerstolpe), <span class="kw">rownames</span>(muraro))</a>
<a class="sourceLine" id="cb124-7" title="7"></a>
<a class="sourceLine" id="cb124-8" title="8">combined_logcounts &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">logcounts</span>(muraro), <span class="kw">logcounts</span>(segerstolpe))</a>
<a class="sourceLine" id="cb124-9" title="9">dataset_labels &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;m&quot;</span>, <span class="st">&quot;s&quot;</span>), <span class="dt">times=</span><span class="kw">c</span>(<span class="kw">ncol</span>(muraro), <span class="kw">ncol</span>(segerstolpe)))</a>
<a class="sourceLine" id="cb124-10" title="10">cell_type_labels &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">colData</span>(muraro)<span class="op">$</span>cell_type1, <span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1)</a>
<a class="sourceLine" id="cb124-11" title="11"></a>
<a class="sourceLine" id="cb124-12" title="12">pheno &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Sample_ID =</span> <span class="kw">colnames</span>(combined_logcounts),</a>
<a class="sourceLine" id="cb124-13" title="13">                <span class="dt">Study_ID=</span>dataset_labels,</a>
<a class="sourceLine" id="cb124-14" title="14">                <span class="dt">Celltype=</span><span class="kw">paste</span>(cell_type_labels, dataset_labels, <span class="dt">sep=</span><span class="st">&quot;-&quot;</span>))</a>
<a class="sourceLine" id="cb124-15" title="15"><span class="kw">rownames</span>(pheno) &lt;-<span class="st"> </span><span class="kw">colnames</span>(combined_logcounts)</a></code></pre></div>
<p>Metaneighbor包括识别HVG的特征选择方法。</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb125-1" title="1">var.genes =<span class="st"> </span><span class="kw">get_variable_genes</span>(combined_logcounts, pheno)</a></code></pre></div>
<p>由于Metaneighbor比<code>scmap</code>慢得多，对这些数据集进行下采样。</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb126-1" title="1">subset &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(pheno), <span class="dv">2000</span>)</a>
<a class="sourceLine" id="cb126-2" title="2">combined_logcounts &lt;-<span class="st"> </span>combined_logcounts[,subset]</a>
<a class="sourceLine" id="cb126-3" title="3">pheno &lt;-<span class="st"> </span>pheno[subset,]</a>
<a class="sourceLine" id="cb126-4" title="4">cell_type_labels &lt;-<span class="st"> </span>cell_type_labels[subset]</a>
<a class="sourceLine" id="cb126-5" title="5">dataset_labels &lt;-<span class="st"> </span>dataset_labels[subset]</a></code></pre></div>
<p>现在准备运行Metaneighbor。首先，运行无监督版本，查看两个数据集中哪些细胞类型最相似。</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" title="1">unsup &lt;-<span class="st"> </span><span class="kw">run_MetaNeighbor_US</span>(var.genes, combined_logcounts, <span class="kw">unique</span>(pheno<span class="op">$</span>Celltype), pheno)</a>
<a class="sourceLine" id="cb127-2" title="2"><span class="kw">heatmap</span>(unsup)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/projection_files/figure-html/unnamed-chunk-32-1.png" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="mnncorrect-1" class="section level3">
<h3><span class="header-section-number">8.8.6</span> mnnCorrect</h3>
<p><a href="https://www.biorxiv.org/content/early/2017/07/18/165118">mnnCorrect</a>校正数据集以促进联合分析。 为了解释两个重复或两个不同实验之间的组成差异，首先匹配实验中的单个细胞以找到重叠的生物结构。 使用该重叠结构学习表达的哪个维度对应于生物状态以及哪个维度对应于批次/实验效应; mnnCorrect假设这些维度在高维表达空间中是彼此正交的。 最后，从整个表达矩阵中删除批次/实验效应，返回校正后的矩阵。</p>
<p>为了跨数据集匹配细胞，mnnCorrect使用余弦距离来避免文库大小效应，跨数据集识别mututal nearest neighbours（k确定邻居大小）。 只有重叠的生物群应该有mutual nearest neighbours（见下图b）。 然而，这假设k被设置为数据集中最小生物群的大小，但是k越小将识别太少的mutual nearest-neighbour对来对批次效应进行良好估计。</p>
<p>学习生物/技术效应的方法是使用奇异值分解，类似于批次校正部分中的RUV，或者使用优化的irlba包进行主成分分析，这比SVD快。 参数<code>svd.dim</code>指定应该保留多少维度来总结数据的生物结构，我们将它设置为三，因为使用上面的Metaneighbor发现有三个主要组。 可以通过平滑（sigma）和/或方差调整（var.adj）进一步调整这些估计。</p>
<p>mnnCorrect还假设您已经对表达式矩阵去子集，以便相同的基因具有相同的顺。当为Metaneighbor设置数据时，我们已经为数据集做了这些。</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-171"></span>
<img src="figures/mnnCorrectDiagramCropped.png" alt="mnnCorrect batch/dataset effect correction. From Haghverdi et al. 2017" width="688" />
<p class="caption">
Figure 8.23: mnnCorrect batch/dataset effect correction. From Haghverdi et al. 2017
</p>
</div>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb128-1" title="1"><span class="kw">require</span>(<span class="st">&quot;scran&quot;</span>)</a>
<a class="sourceLine" id="cb128-2" title="2"><span class="co"># mnnCorrect will take several minutes to run</span></a>
<a class="sourceLine" id="cb128-3" title="3">corrected &lt;-<span class="st"> </span><span class="kw">mnnCorrect</span>(<span class="kw">logcounts</span>(muraro), <span class="kw">logcounts</span>(segerstolpe), <span class="dt">k=</span><span class="dv">20</span>, <span class="dt">sigma=</span><span class="dv">1</span>, <span class="dt">pc.approx=</span><span class="ot">TRUE</span>, <span class="dt">subset.row=</span>var.genes, <span class="dt">svd.dim=</span><span class="dv">3</span>)</a></code></pre></div>
<p>首先让我们检查一下是否找到了足够数量的mnn对，mnnCorrect返回一个数据框列表，其中包含每个数据集的mnn对。</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" title="1"><span class="kw">dim</span>(corrected<span class="op">$</span>pairs[[<span class="dv">1</span>]]) <span class="co"># muraro -&gt; others</span></a>
<a class="sourceLine" id="cb129-2" title="2"><span class="co">## [1] 0 3</span></a>
<a class="sourceLine" id="cb129-3" title="3"><span class="kw">dim</span>(corrected<span class="op">$</span>pairs[[<span class="dv">2</span>]]) <span class="co"># seger -&gt; others</span></a>
<a class="sourceLine" id="cb129-4" title="4"><span class="co">## [1] 2533    3</span></a></code></pre></div>
<p>第一列和第二列包含细胞列ID，第三列数字代表第二列细胞属于哪个数据集/批次。在我们的例子中只比较两个数据集，因此所有mnn对都已分配给第二个表，第三个列只包含一个。</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb130-1" title="1"><span class="kw">head</span>(corrected<span class="op">$</span>pairs[[<span class="dv">2</span>]])</a>
<a class="sourceLine" id="cb130-2" title="2"><span class="co">## DataFrame with 6 rows and 3 columns</span></a>
<a class="sourceLine" id="cb130-3" title="3"><span class="co">##   current.cell other.cell other.batch</span></a>
<a class="sourceLine" id="cb130-4" title="4"><span class="co">##      &lt;integer&gt;      &lt;Rle&gt;       &lt;Rle&gt;</span></a>
<a class="sourceLine" id="cb130-5" title="5"><span class="co">## 1         1553          5           1</span></a>
<a class="sourceLine" id="cb130-6" title="6"><span class="co">## 2         1078          5           1</span></a>
<a class="sourceLine" id="cb130-7" title="7"><span class="co">## 3         1437          5           1</span></a>
<a class="sourceLine" id="cb130-8" title="8"><span class="co">## 4         1890          5           1</span></a>
<a class="sourceLine" id="cb130-9" title="9"><span class="co">## 5         1569          5           1</span></a>
<a class="sourceLine" id="cb130-10" title="10"><span class="co">## 6          373          5           1</span></a>
<a class="sourceLine" id="cb130-11" title="11">total_pairs &lt;-<span class="st"> </span><span class="kw">nrow</span>(corrected<span class="op">$</span>pairs[[<span class="dv">2</span>]])</a>
<a class="sourceLine" id="cb130-12" title="12">n_unique_seger &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>((corrected<span class="op">$</span>pairs[[<span class="dv">2</span>]][,<span class="dv">1</span>])))</a>
<a class="sourceLine" id="cb130-13" title="13">n_unique_muraro &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>((corrected<span class="op">$</span>pairs[[<span class="dv">2</span>]][,<span class="dv">2</span>])))</a></code></pre></div>
<p>mnnCorrect在<code>n_unique_seger</code> segerstolpe cell和<code>n_unique_muraro</code> muraro cells之间找到了2533组mutual nearest-neighbours。 这应该是足够数量的对，但每个数据集中的独特细胞数量较少表明可能没有捕获每个数据集中的完整生物信号。</p>
<p><strong>练习</strong>
哪些细胞类型在这些数据集中有mnns？应该增加/减少k吗？</p>
<p><strong>答案</strong></p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb131-1" title="1">seger_matched_cells &lt;-<span class="st"> </span><span class="kw">unique</span>((corrected<span class="op">$</span>pairs[[<span class="dv">2</span>]][,<span class="dv">1</span>]))</a>
<a class="sourceLine" id="cb131-2" title="2"><span class="kw">summary</span>(<span class="kw">factor</span>(<span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1[seger_matched_cells]))</a>
<a class="sourceLine" id="cb131-3" title="3"></a>
<a class="sourceLine" id="cb131-4" title="4">muraro_matched_cells &lt;-<span class="st"> </span><span class="kw">unique</span>((corrected<span class="op">$</span>pairs[[<span class="dv">2</span>]][,<span class="dv">2</span>]))</a>
<a class="sourceLine" id="cb131-5" title="5"><span class="kw">summary</span>(<span class="kw">factor</span>(<span class="kw">colData</span>(muraro)<span class="op">$</span>cell_type1[muraro_matched_cells]))</a></code></pre></div>
<p>现在创建一个组合数据集来联合分析这些数据。但是，校正的数据不再是counts，通常包含负值，因此某些分析工具可能不再适用。 简单起见，绘制一个联合TSNE。</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb132-1" title="1"><span class="kw">require</span>(<span class="st">&quot;Rtsne&quot;</span>)</a>
<a class="sourceLine" id="cb132-2" title="2">joint_expression_matrix &lt;-<span class="st"> </span><span class="kw">cbind</span>(corrected<span class="op">$</span>corrected[[<span class="dv">1</span>]], corrected<span class="op">$</span>corrected[[<span class="dv">2</span>]])</a>
<a class="sourceLine" id="cb132-3" title="3"></a>
<a class="sourceLine" id="cb132-4" title="4"><span class="co"># Tsne will take some time to run on the full dataset</span></a>
<a class="sourceLine" id="cb132-5" title="5">joint_tsne &lt;-<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">t</span>(joint_expression_matrix[<span class="kw">rownames</span>(joint_expression_matrix) <span class="op">%in%</span><span class="st"> </span>var.genes,]), <span class="dt">initial_dims=</span><span class="dv">10</span>, <span class="dt">theta=</span><span class="fl">0.75</span>,</a>
<a class="sourceLine" id="cb132-6" title="6">                        <span class="dt">check_duplicates=</span><span class="ot">FALSE</span>, <span class="dt">max_iter=</span><span class="dv">200</span>, <span class="dt">stop_lying_iter=</span><span class="dv">50</span>, <span class="dt">mom_switch_iter=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb132-7" title="7">dataset_labels &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;m&quot;</span>, <span class="st">&quot;s&quot;</span>), <span class="dt">times=</span><span class="kw">c</span>(<span class="kw">ncol</span>(muraro), <span class="kw">ncol</span>(segerstolpe))))</a>
<a class="sourceLine" id="cb132-8" title="8">cell_type_labels &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">c</span>(<span class="kw">colData</span>(muraro)<span class="op">$</span>cell_type1, <span class="kw">colData</span>(segerstolpe)<span class="op">$</span>cell_type1))</a>
<a class="sourceLine" id="cb132-9" title="9"><span class="kw">plot</span>(joint_tsne<span class="op">$</span>Y[,<span class="dv">1</span>], joint_tsne<span class="op">$</span>Y[,<span class="dv">2</span>], <span class="dt">pch=</span><span class="kw">c</span>(<span class="dv">16</span>,<span class="dv">1</span>)[dataset_labels], <span class="dt">col=</span><span class="kw">rainbow</span>(<span class="kw">length</span>(<span class="kw">levels</span>(cell_type_labels)))[cell_type_labels])</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/projection_files/figure-html/unnamed-chunk-38-1.png" style="display: block; margin: auto;" /></p>
</div>
<div id="seurat" class="section level3">
<h3><span class="header-section-number">8.8.7</span> 典型相关分析 (Seurat)</h3>
<p>Seurat包中包含另一种用于组合多个数据集的校正方法，称为 <a href="https://www.biorxiv.org/content/early/2017/07/18/164889">CCA</a>。 但是，与mnnCorrect不同，它不直接校正表达矩阵本身。相反，Seurat为每个数据集找到一个较低维度的子空间，然后校正这些子空间。 与mnnCorrect不同，Seurat一次只组合一对数据集。</p>
<p>Seurat通过称为典型相关分析（CCA）的方法使用基因-基因相关性识别数据集中的生物结构。Seurat学习基因-基因相关性的共享结构，然后评估每个细胞与这种结构的拟合程度。必须通过数据特异的降维方法而不是共享相关结构描述的细胞代表数据集特异的细胞类型/状态，在比对两个数据集前舍去。 最后，使用’warping’算法比对两个数据集，该算法对每个数据集的低维表示进行归一化，其对群体密度差异稳健。</p>
<p>注意，因为Seurat占用了大量的库空间，必须重新启动R-session才能加载它，并且不会在此页面上自动生成plots / merged_seuratput。</p>
<p>重新加载数据:</p>
<p>首先，将数据重新格式化为Seurat对象：</p>
<p>对每个数据集进行标准化，缩放和识别HGV：</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-181"></span>
<img src="figures/muraro_seurat_hvg.png" alt="muraro variable genes" width="1050" />
<p class="caption">
Figure 8.24: muraro variable genes
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-182"></span>
<img src="figures/seger_seurat_hvg.png" alt="segerstolpe variable genes" width="1050" />
<p class="caption">
Figure 8.25: segerstolpe variable genes
</p>
</div>
<p>尽管Seurat校正了离散程度和表达均值之间的关系，对特征进行排名时也不使用校正值。将以下命令的结果与上图中的结果进行比较：</p>
<p>按照他们的例子使用前2000个最分散的基因并用merged_seurat来校正每个数据集的表达均值。</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" title="1">gene.use &lt;-<span class="st"> </span><span class="kw">union</span>(<span class="kw">rownames</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> muraro_seurat<span class="op">@</span>hvg.info, <span class="dt">n =</span> <span class="dv">2000</span>)),</a>
<a class="sourceLine" id="cb133-2" title="2">                  <span class="kw">rownames</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> seger_seurat<span class="op">@</span>hvg.info, <span class="dt">n =</span> <span class="dv">2000</span>)))</a></code></pre></div>
<p><strong>练习</strong>
如果选择按均值缩放后top2000的特征，这些特征是什么？
<strong>答案</strong></p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb134-1" title="1">seger_hvg &lt;-<span class="st"> </span>seger_seurat<span class="op">@</span>hvg.info</a>
<a class="sourceLine" id="cb134-2" title="2">seger_hvg &lt;-<span class="st"> </span>seger_hvg[<span class="kw">order</span>(seger_hvg[,<span class="dv">3</span>], <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a>
<a class="sourceLine" id="cb134-3" title="3">muraro_hvg &lt;-<span class="st"> </span>seger_seurat<span class="op">@</span>hvg.info</a>
<a class="sourceLine" id="cb134-4" title="4">muraro_hvg &lt;-<span class="st"> </span>muraro_hvg[<span class="kw">order</span>(muraro_hvg[,<span class="dv">3</span>], <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a>
<a class="sourceLine" id="cb134-5" title="5"></a>
<a class="sourceLine" id="cb134-6" title="6">gene.use.scaled &lt;-<span class="st"> </span><span class="kw">union</span>(<span class="kw">rownames</span>(seger_hvg[<span class="dv">1</span><span class="op">:</span><span class="dv">2000</span>,]), <span class="kw">rownames</span>(muraro_hvg[<span class="dv">1</span><span class="op">:</span><span class="dv">2000</span>,]))</a></code></pre></div>
<p>运行CCA来查找这两个数据集的共享相关结构：</p>
<p>请注意，为了加快计算速度，我们将仅使用前5个维度，但理想情况下，您会考虑更多，然后使用<code>DimHeatmap</code>选择信息最丰富的维度。</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" title="1">merged_seurat &lt;-<span class="st"> </span><span class="kw">RunCCA</span>(<span class="dt">object=</span>muraro_seurat, <span class="dt">object2=</span>seger_seurat, <span class="dt">genes.use=</span>gene.use, <span class="dt">add.cell.id1=</span><span class="st">&quot;m&quot;</span>, <span class="dt">add.cell.id2=</span><span class="st">&quot;s&quot;</span>, <span class="dt">num.cc =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb135-2" title="2"><span class="kw">DimPlot</span>(<span class="dt">object =</span> merged_seurat, <span class="dt">reduction.use =</span> <span class="st">&quot;cca&quot;</span>, <span class="dt">group.by =</span> <span class="st">&quot;dataset&quot;</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>) <span class="co"># Before correcting</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-187"></span>
<img src="figures/cca_before.png" alt="Before Aligning" width="1050" />
<p class="caption">
Figure 8.26: Before Aligning
</p>
</div>
<p>为了识别特定于数据集的细胞类型，我们比较了CCA与特定数据集主成分分析“解释”细胞的程度。</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb136-1" title="1">merged_seurat &lt;-<span class="st"> </span><span class="kw">CalcVarExpRatio</span>(<span class="dt">object =</span> merged_seurat, <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">grouping.var =</span> <span class="st">&quot;dataset&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)</a>
<a class="sourceLine" id="cb136-2" title="2">merged.all &lt;-<span class="st"> </span>merged_seurat</a>
<a class="sourceLine" id="cb136-3" title="3">merged_seurat &lt;-<span class="st"> </span><span class="kw">SubsetData</span>(<span class="dt">object=</span>merged_seurat, <span class="dt">subset.name=</span><span class="st">&quot;var.ratio.pca&quot;</span>, <span class="dt">accept.low =</span> <span class="fl">0.5</span>) <span class="co"># CCA &gt; 1/2 as good as PCA</span></a>
<a class="sourceLine" id="cb136-4" title="4">merged.discard &lt;-<span class="st"> </span><span class="kw">SubsetData</span>(<span class="dt">object=</span>merged.all, <span class="dt">subset.name=</span><span class="st">&quot;var.ratio.pca&quot;</span>, <span class="dt">accept.high =</span> <span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb136-5" title="5"></a>
<a class="sourceLine" id="cb136-6" title="6"><span class="kw">summary</span>(<span class="kw">factor</span>(merged.discard<span class="op">@</span>meta.data<span class="op">$</span>celltype)) <span class="co"># check the cell-type of the discarded cells.</span></a></code></pre></div>
<p>在这里我们可以看到，尽管两个数据集都包含内皮细胞，但几乎都作为“dataset-specific”被丢弃。下面比对数据集：</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb137-1" title="1">merged_seurat &lt;-<span class="st"> </span><span class="kw">AlignSubspace</span>(<span class="dt">object =</span> merged_seurat, <span class="dt">reduction.type =</span> <span class="st">&quot;cca&quot;</span>, <span class="dt">grouping.var =</span> <span class="st">&quot;dataset&quot;</span>, <span class="dt">dims.align =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)</a>
<a class="sourceLine" id="cb137-2" title="2"><span class="kw">DimPlot</span>(<span class="dt">object =</span> merged_seurat, <span class="dt">reduction.use =</span> <span class="st">&quot;cca.aligned&quot;</span>, <span class="dt">group.by =</span> <span class="st">&quot;dataset&quot;</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>) <span class="co"># After aligning subspaces</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-190"></span>
<img src="figures/cca_after.png" alt="After Aligning" width="1050" />
<p class="caption">
Figure 8.27: After Aligning
</p>
</div>
<p><strong>练习</strong>
比较使用离散程度缩放的特征结果</p>
<p><strong>答案</strong></p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb138-1" title="1">merged_seurat &lt;-<span class="st"> </span><span class="kw">RunCCA</span>(<span class="dt">object=</span>muraro_seurat, <span class="dt">object2=</span>seger_seurat, <span class="dt">genes.use=</span>gene.use.scaled, <span class="dt">add.cell.id1=</span><span class="st">&quot;m&quot;</span>, <span class="dt">add.cell.id2=</span><span class="st">&quot;s&quot;</span>, <span class="dt">num.cc =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb138-2" title="2">merged_seurat &lt;-<span class="st"> </span><span class="kw">AlignSubspace</span>(<span class="dt">object =</span> merged_seurat, <span class="dt">reduction.type =</span> <span class="st">&quot;cca&quot;</span>, <span class="dt">grouping.var =</span> <span class="st">&quot;dataset&quot;</span>, <span class="dt">dims.align =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)</a>
<a class="sourceLine" id="cb138-3" title="3"><span class="kw">DimPlot</span>(<span class="dt">object =</span> merged_seurat, <span class="dt">reduction.use =</span> <span class="st">&quot;cca.aligned&quot;</span>, <span class="dt">group.by =</span> <span class="st">&quot;dataset&quot;</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>) <span class="co"># After aligning subspaces</span></a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-192"></span>
<img src="figures/cca_after2.png" alt="After Aligning" width="1050" />
<p class="caption">
Figure 8.28: After Aligning
</p>
</div>
<p><strong>进阶练习</strong>
使用我们之前在组合数据集上介绍的聚类方法。是否发现任何新型细胞类型？</p>
</div>
</div>
<div id="scrna-seq-3" class="section level2">
<h2><span class="header-section-number">8.9</span> 检索scRNA-seq数据</h2>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb139-1" title="1"><span class="kw">library</span>(scfind)</a>
<a class="sourceLine" id="cb139-2" title="2"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb139-3" title="3"><span class="kw">library</span>(plotly)</a>
<a class="sourceLine" id="cb139-4" title="4"><span class="kw">set.seed</span>(<span class="dv">1234567</span>)</a></code></pre></div>
<div id="about" class="section level3">
<h3><span class="header-section-number">8.9.1</span> About</h3>
<p><code>scfind</code>是可以使用基因列表搜索单细胞RNA-Seq集合（Atlas）的工具，寻找表达特定基因集表达的细胞和细胞类型。<code>scfind</code>是一个<a href="(https://github.com/hemberg-lab/scfind)">Github包</a>。 <a href="https://scfind.sanger.ac.uk/">网站</a>提供了包含大量数据集的<code>scfind</code>云版本。</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-194"></span>
<img src="figures/scfind.png" alt="scfind can be used to search large collection of scRNA-seq data by a list of gene IDs." width="80%" />
<p class="caption">
Figure 8.29: scfind can be used to search large collection of scRNA-seq data by a list of gene IDs.
</p>
</div>
</div>
<div id="section-101" class="section level3">
<h3><span class="header-section-number">8.9.2</span> 数据集</h3>
<p>在Tabula Muris 10X数据集上运行<code>scfind</code>。<code>scfind</code>还在<code>SingleCellExperiment</code>类上运行：</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb140-1" title="1">tm10x_heart &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/sce/Heart_10X.rds&quot;</span>)</a>
<a class="sourceLine" id="cb140-2" title="2">tm10x_heart</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 23341 654 
## metadata(1): log.exprs.offset
## assays(3): counts normcounts logcounts
## rownames(23341): Xkr4 Rp1 ... Tdtom_transgene zsGreen_transgene
## rowData names(1): feature_symbol
## colnames(654): 10X_P7_4_AAACCTGCACGACGAA-1
##   10X_P7_4_AAACCTGGTTCCACGG-1 ... 10X_P7_4_TTTGTCAGTCGATTGT-1
##   10X_P7_4_TTTGTCAGTGCGCTTG-1
## colData names(5): mouse cell_type1 tissue subtissue sex
## reducedDimNames(3): UMAP PCA TSNE
## spikeNames(0):</code></pre>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb142-1" title="1"><span class="kw">colData</span>(tm10x_heart)</a></code></pre></div>
<pre><code>## DataFrame with 654 rows and 5 columns
##                                mouse          cell_type1   tissue
##                             &lt;factor&gt;            &lt;factor&gt; &lt;factor&gt;
## 10X_P7_4_AAACCTGCACGACGAA-1   3-F-56    endothelial cell    Heart
## 10X_P7_4_AAACCTGGTTCCACGG-1   3-F-56          fibroblast    Heart
## 10X_P7_4_AAACCTGTCTCGATGA-1   3-F-56    endothelial cell    Heart
## 10X_P7_4_AAACGGGAGCGCTCCA-1   3-F-56 cardiac muscle cell    Heart
## 10X_P7_4_AAAGCAAAGTGAATTG-1   3-F-56          fibroblast    Heart
## ...                              ...                 ...      ...
## 10X_P7_4_TTTGGTTCAGACTCGC-1   3-F-56 cardiac muscle cell    Heart
## 10X_P7_4_TTTGTCAAGTGGAGAA-1   3-F-56 cardiac muscle cell    Heart
## 10X_P7_4_TTTGTCACAAGCCATT-1   3-F-56    endocardial cell    Heart
## 10X_P7_4_TTTGTCAGTCGATTGT-1   3-F-56    endothelial cell    Heart
## 10X_P7_4_TTTGTCAGTGCGCTTG-1   3-F-56          fibroblast    Heart
##                             subtissue      sex
##                              &lt;factor&gt; &lt;factor&gt;
## 10X_P7_4_AAACCTGCACGACGAA-1        NA        F
## 10X_P7_4_AAACCTGGTTCCACGG-1        NA        F
## 10X_P7_4_AAACCTGTCTCGATGA-1        NA        F
## 10X_P7_4_AAACGGGAGCGCTCCA-1        NA        F
## 10X_P7_4_AAAGCAAAGTGAATTG-1        NA        F
## ...                               ...      ...
## 10X_P7_4_TTTGGTTCAGACTCGC-1        NA        F
## 10X_P7_4_TTTGTCAAGTGGAGAA-1        NA        F
## 10X_P7_4_TTTGTCACAAGCCATT-1        NA        F
## 10X_P7_4_TTTGTCAGTCGATTGT-1        NA        F
## 10X_P7_4_TTTGTCAGTGCGCTTG-1        NA        F</code></pre>
</div>
<div id="section-102" class="section level3">
<h3><span class="header-section-number">8.9.3</span> 基因索引</h3>
<p>使用数据集创建基因索引：</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb144-1" title="1">heart_index &lt;-<span class="st"> </span><span class="kw">buildCellTypeIndex</span>(</a>
<a class="sourceLine" id="cb144-2" title="2">  <span class="dt">sce =</span> tm10x_heart,</a>
<a class="sourceLine" id="cb144-3" title="3">  <span class="dt">assay.name =</span> <span class="st">&quot;counts&quot;</span>,</a>
<a class="sourceLine" id="cb144-4" title="4">  <span class="dt">cell.type.label =</span> <span class="st">&quot;cell_type1&quot;</span>,</a>
<a class="sourceLine" id="cb144-5" title="5">  <span class="dt">dataset.name =</span> <span class="st">&quot;Heart&quot;</span></a>
<a class="sourceLine" id="cb144-6" title="6">)</a></code></pre></div>
<pre><code>## Generating index for Heart from &#39; counts &#39; assay</code></pre>
<pre><code>##  Indexing cardiac muscle cell as Heart.cardiac muscle cell  with  83  cells.</code></pre>
<pre><code>##  Indexing endocardial cell as Heart.endocardial cell  with  65  cells.</code></pre>
<pre><code>##  Indexing endothelial cell as Heart.endothelial cell  with  178  cells.</code></pre>
<pre><code>##  Indexing erythrocyte as Heart.erythrocyte  with  55  cells.</code></pre>
<pre><code>##  Indexing fibroblast as Heart.fibroblast  with  222  cells.</code></pre>
<pre><code>##  Indexing smooth muscle cell as Heart.smooth muscle cell  with  21  cells.</code></pre>
<p><code>scfind</code>采用两步压缩策略，有效压缩大型cell-by-gene矩阵，并允许通过基因查询快速检索数据。估计用这种方法可以达到2个数量级的压缩。</p>
<p>索引的输入矩阵是<code>SingleCellExperiment</code>类的原始count矩阵。 默认情况下，<code>SingleCellExperiment</code>对象的<code>colData</code>的<code>cell_type1</code>列用于定义细胞类型，但也可以使用<code>buildCellTypeIndex</code>的<code>cell.type.label</code>参数手动定义。对于具有多个组织的数据集，还可以使用函数<code>mergeDataset</code>将所有组织合并在一起以创建超级索引。</p>
<p>索引可以使用<code>saveObject</code>函数以.rds格式保存，并使用<code>loadObject</code>函数加载以备将来使用。</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb152-1" title="1">tm10x_thymus &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/sce/Thymus_10X.rds&quot;</span>)</a>
<a class="sourceLine" id="cb152-2" title="2">thymus_index &lt;-<span class="st"> </span><span class="kw">buildCellTypeIndex</span>(</a>
<a class="sourceLine" id="cb152-3" title="3">  <span class="dt">sce =</span> tm10x_thymus, </a>
<a class="sourceLine" id="cb152-4" title="4">  <span class="dt">assay.name =</span> <span class="st">&quot;counts&quot;</span>, </a>
<a class="sourceLine" id="cb152-5" title="5">  <span class="dt">cell.type.label =</span> <span class="st">&quot;cell_type1&quot;</span>, </a>
<a class="sourceLine" id="cb152-6" title="6">  <span class="dt">dataset.name =</span> <span class="st">&quot;Thymus&quot;</span></a>
<a class="sourceLine" id="cb152-7" title="7">)</a></code></pre></div>
<pre><code>## Generating index for Thymus from &#39; counts &#39; assay</code></pre>
<pre><code>##  Indexing stromal cell as Thymus.stromal cell  with  112  cells.</code></pre>
<pre><code>##  Indexing T cell as Thymus.T cell  with  1317  cells.</code></pre>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb156-1" title="1">scfind_index &lt;-<span class="st"> </span><span class="kw">mergeDataset</span>(heart_index, thymus_index)</a></code></pre></div>
<pre><code>## Merging Thymus</code></pre>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb158-1" title="1">scfind_index<span class="op">@</span>datasets</a></code></pre></div>
<pre><code>## [1] &quot;Heart&quot;  &quot;Thymus&quot;</code></pre>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb160-1" title="1"><span class="kw">cellTypeNames</span>(scfind_index)</a></code></pre></div>
<pre><code>## [1] &quot;Heart.cardiac muscle cell&quot; &quot;Heart.endocardial cell&quot;   
## [3] &quot;Heart.endothelial cell&quot;    &quot;Heart.erythrocyte&quot;        
## [5] &quot;Heart.fibroblast&quot;          &quot;Heart.smooth muscle cell&quot; 
## [7] &quot;Thymus.stromal cell&quot;       &quot;Thymus.T cell&quot;</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb162-1" title="1"><span class="kw">sample</span>(<span class="kw">scfindGenes</span>(scfind_index),<span class="dv">20</span>)</a></code></pre></div>
<pre><code>##  [1] &quot;Cwc15&quot;         &quot;Ttc33&quot;         &quot;Plbd2&quot;         &quot;Arhgap12&quot;     
##  [5] &quot;AA465934&quot;      &quot;Ulk3&quot;          &quot;Foxa3&quot;         &quot;Htra3&quot;        
##  [9] &quot;Gpr84&quot;         &quot;Lcn8&quot;          &quot;Zfp498&quot;        &quot;Basp1&quot;        
## [13] &quot;A430105I19Rik&quot; &quot;Dhrs2&quot;         &quot;Zmym3&quot;         &quot;Rara&quot;         
## [17] &quot;Lrrc24&quot;        &quot;Rps8&quot;          &quot;Ttc14&quot;         &quot;Igsf5&quot;</code></pre>
<p>使用交互式Shiny快速找到细胞类型，使用如下命令：
To quickly and easily find the enriched cell type using an interactive Shiny application use the following method:</p>
</div>
<div id="marker" class="section level3">
<h3><span class="header-section-number">8.9.4</span> Marker基因</h3>
<p>在数据集中找到Thymus T细胞的标记基因</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb164-1" title="1"><span class="co"># Showing the top 5 marker genes for each cell type and sort by F1 score.</span></a>
<a class="sourceLine" id="cb164-2" title="2">t_cell_markers &lt;-<span class="st"> </span><span class="kw">cellTypeMarkers</span>(scfind_index, <span class="dt">cell.types =</span> <span class="st">&quot;Thymus.T cell&quot;</span>, <span class="dt">top.k =</span> <span class="dv">5</span>, <span class="dt">sort.field =</span> <span class="st">&quot;f1&quot;</span>) </a></code></pre></div>
<pre><code>## Considering the whole DB..</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb166-1" title="1">t_cell_markers</a></code></pre></div>
<pre><code>##           cellType genes   tp  fp fn precision    recall        f1
## 2628 Thymus.T cell  Cd3d 1311 108  6 0.9238901 0.9954442 0.9583333
## 6616 Thymus.T cell   Lck 1305 106 12 0.9248759 0.9908884 0.9567448
## 2631 Thymus.T cell  Cd3g 1304 108 13 0.9235128 0.9901291 0.9556614
## 2629 Thymus.T cell  Cd3e 1279  98 38 0.9288308 0.9711465 0.9495174
## 6603 Thymus.T cell   Lat 1280 109 37 0.9215263 0.9719059 0.9460459</code></pre>
<p>评估Thymus stromal细胞中Thymus T cell的marker</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb168-1" title="1"><span class="kw">evaluateMarkers</span>(</a>
<a class="sourceLine" id="cb168-2" title="2">  scfind_index, </a>
<a class="sourceLine" id="cb168-3" title="3">  <span class="dt">gene.list =</span> <span class="kw">as.character</span>(t_cell_markers<span class="op">$</span>genes), </a>
<a class="sourceLine" id="cb168-4" title="4">  <span class="dt">cell.types =</span> <span class="st">&quot;Thymus.stromal cell&quot;</span>, </a>
<a class="sourceLine" id="cb168-5" title="5">  <span class="dt">sort.field =</span> <span class="st">&quot;f1&quot;</span></a>
<a class="sourceLine" id="cb168-6" title="6">)</a></code></pre></div>
<pre><code>## Considering the whole DB..</code></pre>
<pre><code>##              cellType genes tp   fp fn  precision    recall        f1
## 2 Thymus.stromal cell  Cd3e 90 1287 22 0.06535947 0.8035714 0.1208865
## 3 Thymus.stromal cell  Cd3g 94 1318 18 0.06657224 0.8392857 0.1233596
## 5 Thymus.stromal cell   Lck 94 1317 18 0.06661942 0.8392857 0.1234406
## 1 Thymus.stromal cell  Cd3d 98 1321 14 0.06906272 0.8749999 0.1280209
## 4 Thymus.stromal cell   Lat 99 1290 13 0.07127430 0.8839286 0.1319121</code></pre>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb171-1" title="1"><span class="co"># By default, the marker evaluation takes all cell types in the dataset as background cell type, but you can use the argument `background.cell.types` to fine tune the evaluation</span></a>
<a class="sourceLine" id="cb171-2" title="2"></a>
<a class="sourceLine" id="cb171-3" title="3">background &lt;-<span class="st"> </span><span class="kw">cellTypeNames</span>(scfind_index, <span class="dt">datasets =</span> <span class="st">&quot;Thymus&quot;</span>)</a>
<a class="sourceLine" id="cb171-4" title="4">background</a></code></pre></div>
<pre><code>## [1] &quot;Thymus.stromal cell&quot; &quot;Thymus.T cell&quot;</code></pre>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb173-1" title="1"><span class="kw">evaluateMarkers</span>(</a>
<a class="sourceLine" id="cb173-2" title="2">  scfind_index, </a>
<a class="sourceLine" id="cb173-3" title="3">  <span class="dt">gene.list =</span> <span class="kw">as.character</span>(t_cell_markers<span class="op">$</span>genes), </a>
<a class="sourceLine" id="cb173-4" title="4">  <span class="dt">cell.types =</span> <span class="st">&quot;Thymus.stromal cell&quot;</span>, </a>
<a class="sourceLine" id="cb173-5" title="5">  <span class="dt">sort.field =</span> <span class="st">&quot;f1&quot;</span>, </a>
<a class="sourceLine" id="cb173-6" title="6">  <span class="dt">background.cell.types =</span> background</a>
<a class="sourceLine" id="cb173-7" title="7">)</a></code></pre></div>
<pre><code>##              cellType genes tp   fp fn  precision    recall        f1
## 2 Thymus.stromal cell  Cd3e 90 1279 22 0.06574142 0.8035714 0.1215395
## 5 Thymus.stromal cell   Lck 94 1305 18 0.06719085 0.8392857 0.1244209
## 3 Thymus.stromal cell  Cd3g 94 1304 18 0.06723891 0.8392857 0.1245033
## 1 Thymus.stromal cell  Cd3d 98 1311 14 0.06955288 0.8749999 0.1288626
## 4 Thymus.stromal cell   Lat 99 1280 13 0.07179116 0.8839286 0.1327968</code></pre>
</div>
<div id="section-103" class="section level3">
<h3><span class="header-section-number">8.9.5</span> 搜索基因列表</h3>
<p><code>scfind</code>可以快速从大型单细胞数据集中识别出最能代表感兴趣基因的细胞类型。我们使用文献 <a href="https://www.nature.com/articles/cr201599">Yanbin et al. 2015</a>鉴定的标记基因。
用于免疫染色的心肌细胞特异性标志物如图1所示。</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb175-1" title="1">cardiomyocytes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Mef2c&quot;</span>, <span class="st">&quot;Gata4&quot;</span>, <span class="st">&quot;Nkx2.5&quot;</span>, <span class="st">&quot;Myh6&quot;</span>, <span class="st">&quot;tnnt2&quot;</span>, <span class="st">&quot;tnni3&quot;</span>, <span class="st">&quot;CDH2&quot;</span>, <span class="st">&quot;Cx43&quot;</span>, <span class="st">&quot;GJA1&quot;</span>) </a>
<a class="sourceLine" id="cb175-2" title="2">result &lt;-<span class="st"> </span><span class="kw">markerGenes</span>(</a>
<a class="sourceLine" id="cb175-3" title="3">  scfind_index, </a>
<a class="sourceLine" id="cb175-4" title="4">  <span class="dt">gene.list =</span> cardiomyocytes</a>
<a class="sourceLine" id="cb175-5" title="5">)</a></code></pre></div>
<pre><code>## Ignored Nkx2.5, Cx43. Not found in the index</code></pre>
<pre><code>## calculating tfidf for the reduced expression matrix...</code></pre>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb178-1" title="1">result</a></code></pre></div>
<pre><code>##     Genes                                  Query      tfidf Cells
## 1       2                             Cdh2,Gata4   1.075548    27
## 2       3                        Cdh2,Gata4,Gja1   2.085320    18
## 3       4                  Cdh2,Gata4,Gja1,Mef2c   2.812912    11
## 4       5             Cdh2,Gata4,Gja1,Mef2c,Myh6   9.754036     8
## 5       6       Cdh2,Gata4,Gja1,Mef2c,Myh6,Tnni3  13.983566     8
## 6       7 Cdh2,Gata4,Gja1,Mef2c,Myh6,Tnni3,Tnnt2  18.241363     8
## 7       6       Cdh2,Gata4,Gja1,Mef2c,Myh6,Tnnt2  14.011833     8
## 8       5            Cdh2,Gata4,Gja1,Mef2c,Tnni3   7.042442     8
## 9       6      Cdh2,Gata4,Gja1,Mef2c,Tnni3,Tnnt2  11.300239     8
## 10      5            Cdh2,Gata4,Gja1,Mef2c,Tnnt2   7.070709     8
## 11      4                   Cdh2,Gata4,Gja1,Myh6   9.026444    14
## 12      5             Cdh2,Gata4,Gja1,Myh6,Tnni3  13.255974    14
## 13      6       Cdh2,Gata4,Gja1,Myh6,Tnni3,Tnnt2  17.513771    14
## 14      5             Cdh2,Gata4,Gja1,Myh6,Tnnt2  13.284241    14
## 15      4                  Cdh2,Gata4,Gja1,Tnni3   6.314850    14
## 16      5            Cdh2,Gata4,Gja1,Tnni3,Tnnt2  10.572647    14
## 17      4                  Cdh2,Gata4,Gja1,Tnnt2   6.343117    15
## 18      3                       Cdh2,Gata4,Mef2c   1.803140    14
## 19      4                  Cdh2,Gata4,Mef2c,Myh6   8.744264     8
## 20      5            Cdh2,Gata4,Mef2c,Myh6,Tnni3  12.973794     8
## 21      6      Cdh2,Gata4,Mef2c,Myh6,Tnni3,Tnnt2  17.231591     8
## 22      5            Cdh2,Gata4,Mef2c,Myh6,Tnnt2  13.002062     8
## 23      4                 Cdh2,Gata4,Mef2c,Tnni3   6.032670     9
## 24      5           Cdh2,Gata4,Mef2c,Tnni3,Tnnt2  10.290467     8
## 25      4                 Cdh2,Gata4,Mef2c,Tnnt2   6.060938     8
## 26      3                        Cdh2,Gata4,Myh6   8.016672    15
## 27      4                  Cdh2,Gata4,Myh6,Tnni3  12.246202    14
## 28      5            Cdh2,Gata4,Myh6,Tnni3,Tnnt2  16.503999    14
## 29      4                  Cdh2,Gata4,Myh6,Tnnt2  12.274470    14
## 30      3                       Cdh2,Gata4,Tnni3   5.305078    17
## 31      4                 Cdh2,Gata4,Tnni3,Tnnt2   9.562876    15
## 32      3                       Cdh2,Gata4,Tnnt2   5.333346    17
## 33      2                              Cdh2,Gja1   1.864063    39
## 34      3                        Cdh2,Gja1,Mef2c   2.716969    18
## 35      4                   Cdh2,Gja1,Mef2c,Myh6  10.853567    15
## 36      5             Cdh2,Gja1,Mef2c,Myh6,Tnni3  15.811551    15
## 37      6       Cdh2,Gja1,Mef2c,Myh6,Tnni3,Tnnt2  20.802671    15
## 38      5             Cdh2,Gja1,Mef2c,Myh6,Tnnt2  15.844687    15
## 39      4                  Cdh2,Gja1,Mef2c,Tnni3   7.674953    15
## 40      5            Cdh2,Gja1,Mef2c,Tnni3,Tnnt2  12.666073    15
## 41      4                  Cdh2,Gja1,Mef2c,Tnnt2   7.708089    15
## 42      3                         Cdh2,Gja1,Myh6  10.000661    32
## 43      4                   Cdh2,Gja1,Myh6,Tnni3  14.958645    32
## 44      5             Cdh2,Gja1,Myh6,Tnni3,Tnnt2  19.949766    32
## 45      4                   Cdh2,Gja1,Myh6,Tnnt2  14.991781    32
## 46      3                        Cdh2,Gja1,Tnni3   6.822047    33
## 47      4                  Cdh2,Gja1,Tnni3,Tnnt2  11.813168    33
## 48      3                        Cdh2,Gja1,Tnnt2   6.855183    34
## 49      2                             Cdh2,Mef2c   1.533283    27
## 50      3                        Cdh2,Mef2c,Myh6   9.669881    17
## 51      4                  Cdh2,Mef2c,Myh6,Tnni3  14.627866    15
## 52      5            Cdh2,Mef2c,Myh6,Tnni3,Tnnt2  19.618986    15
## 53      4                  Cdh2,Mef2c,Myh6,Tnnt2  14.661002    16
## 54      3                       Cdh2,Mef2c,Tnni3   6.491268    18
## 55      4                 Cdh2,Mef2c,Tnni3,Tnnt2  11.482388    16
## 56      3                       Cdh2,Mef2c,Tnnt2   6.524404    18
## 57      2                              Cdh2,Myh6   8.816976    36
## 58      3                        Cdh2,Myh6,Tnni3  13.774960    32
## 59      4                  Cdh2,Myh6,Tnni3,Tnnt2  18.766080    32
## 60      3                        Cdh2,Myh6,Tnnt2  13.808096    33
## 61      2                             Cdh2,Tnni3   5.638362    38
## 62      3                       Cdh2,Tnni3,Tnnt2  10.629482    35
## 63      2                             Cdh2,Tnnt2   5.671498    43
## 64      2                             Gata4,Gja1   1.504907    41
## 65      3                       Gata4,Gja1,Mef2c   2.232499    19
## 66      4                  Gata4,Gja1,Mef2c,Myh6   9.173623     9
## 67      5            Gata4,Gja1,Mef2c,Myh6,Tnni3  13.403153     8
## 68      6      Gata4,Gja1,Mef2c,Myh6,Tnni3,Tnnt2  17.660950     8
## 69      5            Gata4,Gja1,Mef2c,Myh6,Tnnt2  13.431420     8
## 70      4                 Gata4,Gja1,Mef2c,Tnni3   6.462029    11
## 71      5           Gata4,Gja1,Mef2c,Tnni3,Tnnt2  10.719826    10
## 72      4                 Gata4,Gja1,Mef2c,Tnnt2   6.490296    11
## 73      3                        Gata4,Gja1,Myh6   8.446031    22
## 74      4                  Gata4,Gja1,Myh6,Tnni3  12.675561    19
## 75      5            Gata4,Gja1,Myh6,Tnni3,Tnnt2  16.933358    19
## 76      4                  Gata4,Gja1,Myh6,Tnnt2  12.703828    19
## 77      3                       Gata4,Gja1,Tnni3   5.734437    22
## 78      4                 Gata4,Gja1,Tnni3,Tnnt2   9.992234    21
## 79      3                       Gata4,Gja1,Tnnt2   5.762704    26
## 80      2                            Gata4,Mef2c   1.222727    65
## 81      3                       Gata4,Mef2c,Myh6   8.163851    20
## 82      4                 Gata4,Mef2c,Myh6,Tnni3  12.393381    13
## 83      5           Gata4,Mef2c,Myh6,Tnni3,Tnnt2  16.651178     9
## 84      4                 Gata4,Mef2c,Myh6,Tnnt2  12.421649    10
## 85      3                      Gata4,Mef2c,Tnni3   5.452257    25
## 86      4                Gata4,Mef2c,Tnni3,Tnnt2   9.710054    14
## 87      3                      Gata4,Mef2c,Tnnt2   5.480525    21
## 88      2                             Gata4,Myh6   7.436259    46
## 89      3                       Gata4,Myh6,Tnni3  11.665789    30
## 90      4                 Gata4,Myh6,Tnni3,Tnnt2  15.923586    24
## 91      3                       Gata4,Myh6,Tnnt2  11.694057    25
## 92      2                            Gata4,Tnni3   4.724665    57
## 93      3                      Gata4,Tnni3,Tnnt2   8.982463    34
## 94      2                            Gata4,Tnnt2   4.752933    61
## 95      2                             Gja1,Mef2c   2.553022    91
## 96      3                        Gja1,Mef2c,Myh6  12.752868    35
## 97      4                  Gja1,Mef2c,Myh6,Tnni3  18.968080    31
## 98      5            Gja1,Mef2c,Myh6,Tnni3,Tnnt2  25.224830    26
## 99      4                  Gja1,Mef2c,Myh6,Tnnt2  19.009619    27
## 100     3                       Gja1,Mef2c,Tnni3   8.768234    43
## 101     4                 Gja1,Mef2c,Tnni3,Tnnt2  15.024984    34
## 102     3                       Gja1,Mef2c,Tnnt2   8.809772    44
## 103     2                              Gja1,Myh6  16.214932    71
## 104     3                        Gja1,Myh6,Tnni3  24.840569    61
## 105     4                  Gja1,Myh6,Tnni3,Tnnt2  33.523855    54
## 106     3                        Gja1,Myh6,Tnnt2  24.898218    57
## 107     2                             Gja1,Tnni3  10.684950    85
## 108     3                       Gja1,Tnni3,Tnnt2  19.368235    67
## 109     2                             Gja1,Tnnt2  10.742598    88
## 110     2                             Mef2c,Myh6  11.269029    71
## 111     3                       Mef2c,Myh6,Tnni3  17.484240    41
## 112     4                 Mef2c,Myh6,Tnni3,Tnnt2  23.740991    30
## 113     3                       Mef2c,Myh6,Tnnt2  17.525779    43
## 114     2                            Mef2c,Tnni3   7.284394   102
## 115     3                      Mef2c,Tnni3,Tnnt2  13.541144    60
## 116     2                            Mef2c,Tnnt2   7.325932   109
## 117     2                             Myh6,Tnni3  95.421586    91
## 118     3                       Myh6,Tnni3,Tnnt2 131.792399    72
## 119     2                             Myh6,Tnnt2  96.302401    94
## 120     2                            Tnni3,Tnnt2  72.500162   118
##     Cell.Types
## 1            0
## 2            0
## 3            0
## 4            0
## 5            0
## 6            0
## 7            0
## 8            0
## 9            0
## 10           0
## 11           0
## 12           0
## 13           0
## 14           0
## 15           0
## 16           0
## 17           0
## 18           0
## 19           0
## 20           0
## 21           0
## 22           0
## 23           0
## 24           0
## 25           0
## 26           0
## 27           0
## 28           0
## 29           0
## 30           0
## 31           0
## 32           0
## 33           0
## 34           0
## 35           0
## 36           0
## 37           0
## 38           0
## 39           0
## 40           0
## 41           0
## 42           0
## 43           0
## 44           0
## 45           0
## 46           0
## 47           0
## 48           0
## 49           0
## 50           0
## 51           0
## 52           0
## 53           0
## 54           0
## 55           0
## 56           0
## 57           0
## 58           0
## 59           0
## 60           0
## 61           0
## 62           0
## 63           0
## 64           0
## 65           0
## 66           0
## 67           0
## 68           0
## 69           0
## 70           0
## 71           0
## 72           0
## 73           0
## 74           0
## 75           0
## 76           0
## 77           0
## 78           0
## 79           0
## 80           0
## 81           0
## 82           0
## 83           0
## 84           0
## 85           0
## 86           0
## 87           0
## 88           0
## 89           0
## 90           0
## 91           0
## 92           0
## 93           0
## 94           0
## 95           0
## 96           0
## 97           0
## 98           0
## 99           0
## 100          0
## 101          0
## 102          0
## 103          0
## 104          0
## 105          0
## 106          0
## 107          0
## 108          0
## 109          0
## 110          0
## 111          0
## 112          0
## 113          0
## 114          0
## 115          0
## 116          0
## 117          0
## 118          0
## 119          0
## 120          0</code></pre>
<p>为了允许从基因列表查询搜索富集的细胞类型，<code>scfind</code>使用查询优化程序。首先，函数<code>markerGenes</code>将统计在数据集中具有最高支持度的子查询。 每个基因集的TF-IDF得分允许用户识别最佳子查询以找到最相关的细胞类型。</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb180-1" title="1">best_subquery &lt;-<span class="st"> </span>result[<span class="kw">which.max</span>(result<span class="op">$</span>tfidf),] <span class="co"># get the best subquery by ranking TF-IDF score</span></a>
<a class="sourceLine" id="cb180-2" title="2">best_subquery &lt;-<span class="st"> </span><span class="kw">strsplit</span>(<span class="kw">as.character</span>(best_subquery<span class="op">$</span>Query), <span class="st">&quot;,&quot;</span>)[[<span class="dv">1</span>]] <span class="co"># obtain gene list</span></a>
<a class="sourceLine" id="cb180-3" title="3"><span class="kw">hyperQueryCellTypes</span>(</a>
<a class="sourceLine" id="cb180-4" title="4">  scfind_index,</a>
<a class="sourceLine" id="cb180-5" title="5">  <span class="dt">gene.list =</span> best_subquery</a>
<a class="sourceLine" id="cb180-6" title="6">)</a></code></pre></div>
<pre><code>##                   cell_type cell_hits total_cells pval
## 1 Heart.cardiac muscle cell        57          83    0
## 2    Heart.endocardial cell         3          65    1
## 3    Heart.endothelial cell         7         178    1
## 4          Heart.fibroblast         5         222    1</code></pre>
<p><code>hyperQueryCellTypes</code>函数返回与给定数据集中的所有细胞类型对应的p值列表。它还输出一组细胞，其基因与列表中基因共表达。</p>
<p><strong>练习1</strong></p>
<p>在Heart数据集中查找所有细胞类型的标记基因</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb182-1" title="1">heart_cell_types &lt;-<span class="st"> </span><span class="kw">cellTypeNames</span>(scfind_index, <span class="st">&quot;Heart&quot;</span>)</a>
<a class="sourceLine" id="cb182-2" title="2"></a>
<a class="sourceLine" id="cb182-3" title="3"><span class="cf">for</span>(cell_type_name <span class="cf">in</span> heart_cell_types)</a>
<a class="sourceLine" id="cb182-4" title="4">{</a>
<a class="sourceLine" id="cb182-5" title="5">  result &lt;-<span class="st"> </span><span class="kw">cellTypeMarkers</span>(</a>
<a class="sourceLine" id="cb182-6" title="6">    scfind_index, </a>
<a class="sourceLine" id="cb182-7" title="7">    <span class="dt">cell.types =</span> cell_type_name, </a>
<a class="sourceLine" id="cb182-8" title="8">    <span class="dt">sort.field =</span> <span class="st">&quot;f1&quot;</span>, </a>
<a class="sourceLine" id="cb182-9" title="9">    <span class="dt">background =</span> heart_cell_types) </a>
<a class="sourceLine" id="cb182-10" title="10">  <span class="kw">print</span>(result)</a>
<a class="sourceLine" id="cb182-11" title="11">}</a></code></pre></div>
<pre><code>##                        cellType genes tp fp fn precision    recall
## 11309 Heart.cardiac muscle cell Xirp2 51  4 32 0.9272727 0.6144578
## 6628  Heart.cardiac muscle cell Myom2 53  8 30 0.8688524 0.6385542
## 8925  Heart.cardiac muscle cell  Ryr2 51  8 32 0.8644068 0.6144578
## 5699  Heart.cardiac muscle cell  Ldb3 53 13 30 0.8030303 0.6385542
## 10657 Heart.cardiac muscle cell  Trdn 49  6 34 0.8909091 0.5903614
##              f1
## 11309 0.7391304
## 6628  0.7361111
## 8925  0.7183099
## 5699  0.7114094
## 10657 0.7101449
##                    cellType  genes tp fp fn precision    recall        f1
## 4581 Heart.endocardial cell  Gpm6a 50  5 15 0.9090909 0.7692308 0.8333333
## 1736 Heart.endocardial cell    Bmx 52 16 13 0.7647059 0.8000000 0.7819549
## 6851 Heart.endocardial cell   Npr3 61 32  4 0.6559140 0.9384615 0.7721519
## 7533 Heart.endocardial cell Plagl1 51 18 14 0.7391304 0.7846154 0.7611941
## 2552 Heart.endocardial cell   Comp 45 10 20 0.8181818 0.6923077 0.7500001
##                     cellType         genes  tp fp fn precision    recall
## 1951  Heart.endothelial cell       C1qtnf9 153 17 25 0.9000000 0.8595505
## 76    Heart.endothelial cell 1190002H23Rik 161 33 17 0.8298969 0.9044943
## 10534 Heart.endothelial cell         Tcf15 157 30 21 0.8395722 0.8820225
## 2274  Heart.endothelial cell       Cd300lg 163 50 15 0.7652582 0.9157303
## 4850  Heart.endothelial cell       Gpihbp1 168 58 10 0.7433628 0.9438202
##              f1
## 1951  0.8793103
## 76    0.8655914
## 10534 0.8602740
## 2274  0.8337595
## 4850  0.8316832
##               cellType   genes tp fp fn precision    recall        f1
## 2074 Heart.erythrocyte    Cd68 51  5  4 0.9107143 0.9272727 0.9189188
## 2346 Heart.erythrocyte Clec4a2 50  4  5 0.9259259 0.9090909 0.9174312
## 5623 Heart.erythrocyte    Lyz1 55 12  0 0.8208955 1.0000000 0.9016393
## 983  Heart.erythrocyte Alox5ap 55 13  0 0.8088235 1.0000000 0.8943089
## 3393 Heart.erythrocyte    Emr1 48  5  7 0.9056604 0.8727272 0.8888889
##               cellType         genes  tp fp fn precision    recall
## 3024  Heart.fibroblast        Col1a1 209 18 13 0.9207048 0.9414415
## 11653 Heart.fibroblast         Tcf21 193  6 29 0.9698492 0.8693694
## 8609  Heart.fibroblast        Pdgfra 194  8 28 0.9603961 0.8738739
## 3830  Heart.fibroblast           Dpt 215 40  7 0.8431373 0.9684685
## 683   Heart.fibroblast 6330406I15Rik 190 11 32 0.9452736 0.8558559
##              f1
## 3024  0.9309577
## 11653 0.9168646
## 8609  0.9150944
## 3830  0.9014676
## 683   0.8983452
##                      cellType   genes tp fp fn precision    recall
## 6968 Heart.smooth muscle cell    Rgs4 20  9  1 0.6896551 0.9523810
## 5662 Heart.smooth muscle cell   Nrip2 15  3  6 0.8333333 0.7142857
## 5779 Heart.smooth muscle cell Olfr558 16  6  5 0.7272727 0.7619048
## 890  Heart.smooth muscle cell    Ano1 15  5  6 0.7500000 0.7142857
## 3565 Heart.smooth muscle cell Gm13889 18 11  3 0.6206896 0.8571429
##             f1
## 6968 0.8000000
## 5662 0.7692308
## 5779 0.7441860
## 890  0.7317073
## 3565 0.7200000</code></pre>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb184-1" title="1">cardiac_contractility &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Ace2&quot;</span>,<span class="st">&quot;Fkbp1b&quot;</span>,<span class="st">&quot;Gh&quot;</span>,<span class="st">&quot;Cacna1c&quot;</span>,<span class="st">&quot;Cd59b&quot;</span>,<span class="st">&quot;Ppp1r1a&quot;</span>,<span class="st">&quot;Tnnt2&quot;</span>,<span class="st">&quot;Nos1&quot;</span>,<span class="st">&quot;Agtr1a&quot;</span>,<span class="st">&quot;Camk2g&quot;</span>,<span class="st">&quot;Grk2&quot;</span>,<span class="st">&quot;Ins2&quot;</span>,<span class="st">&quot;Dnah8&quot;</span>,<span class="st">&quot;Igf1&quot;</span>,<span class="st">&quot;Nos3&quot;</span>,<span class="st">&quot;Nppa&quot;</span>,<span class="st">&quot;Nppb&quot;</span>,<span class="st">&quot;Il6&quot;</span>,<span class="st">&quot;Myh6&quot;</span>,<span class="st">&quot;Ren2&quot;</span>,<span class="st">&quot;Tnni3&quot;</span>,<span class="st">&quot;Apln&quot;</span>,<span class="st">&quot;Kcnmb1&quot;</span>,<span class="st">&quot;Pik3cg&quot;</span>,<span class="st">&quot;Prkca&quot;</span>,<span class="st">&quot;Aplnr&quot;</span>,<span class="st">&quot;Slc8a1&quot;</span>,<span class="st">&quot;Ace&quot;</span>,<span class="st">&quot;Akt1&quot;</span>,<span class="st">&quot;Edn1&quot;</span>,<span class="st">&quot;Kcnmb2&quot;</span>,<span class="st">&quot;Nos2&quot;</span>,<span class="st">&quot;Tnf&quot;</span>,<span class="st">&quot;Myh14&quot;</span>,<span class="st">&quot;Adrb2&quot;</span>,<span class="st">&quot;Agt&quot;</span>,<span class="st">&quot;Adrb1&quot;</span>,<span class="st">&quot;Atp2a2&quot;</span>,<span class="st">&quot;Ryr2&quot;</span>,<span class="st">&quot;Pln&quot;</span>)</a></code></pre></div>
<p>_练习2__</p>
<p>输入与“心肌收缩力”相关的基因列表，找到具有最高支持度的最佳基因组。识别此查询富集的细胞类型。</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb185-1" title="1">result &lt;-<span class="st"> </span><span class="kw">markerGenes</span>(scfind_index, cardiac_contractility)</a></code></pre></div>
<pre><code>## Ignored Gh, Nos1, Grk2, Ins2, Dnah8, Ren2, Kcnmb2. Not found in the index</code></pre>
<pre><code>## calculating tfidf for the reduced expression matrix...</code></pre>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb188-1" title="1">best_sub_query &lt;-<span class="st"> </span><span class="kw">as.character</span>(result[<span class="kw">which.max</span>(result<span class="op">$</span>tfidf), <span class="kw">c</span>(<span class="st">&quot;Query&quot;</span>)])</a>
<a class="sourceLine" id="cb188-2" title="2">best_sub_query &lt;-<span class="st"> </span><span class="kw">strsplit</span>(best_sub_query, <span class="st">&quot;,&quot;</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb188-3" title="3"><span class="kw">hyperQueryCellTypes</span>(scfind_index, best_sub_query)</a></code></pre></div>
<pre><code>##                   cell_type cell_hits total_cells pval
## 1 Heart.cardiac muscle cell        57          83    0
## 2    Heart.endocardial cell         3          65    1
## 3    Heart.endothelial cell         7         178    1
## 4          Heart.fibroblast         4         222    1</code></pre>
</div>
<div id="in-silico-gating" class="section level3">
<h3><span class="header-section-number">8.9.6</span> In-silico gating</h3>
<p>使用<code>findCellTypes</code>函数，可以执行in-silico gating以识别细胞类型子集，类似细胞排序。 在基因名称前面分别加上逻辑运算符，包括“-”,"*“,”no“和”intermediate"表达式。 在这里，我们使用运算符将Thymus数据集的T细胞分为效应T调节细胞和效应记忆T细胞子集。</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb190-1" title="1">effector_t_reg_cells &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;*Ptprc&quot;</span>, <span class="st">&quot;-Il7r&quot;</span>, <span class="st">&quot;Ctla4&quot;</span>, <span class="st">&quot;-Il7r&quot;</span>)</a>
<a class="sourceLine" id="cb190-2" title="2">effector_memory_t_cells &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;-Il2ra&quot;</span>, <span class="st">&quot;*Ptprc&quot;</span>, <span class="st">&quot;Il7r&quot;</span>)</a>
<a class="sourceLine" id="cb190-3" title="3"></a>
<a class="sourceLine" id="cb190-4" title="4">subset_treg &lt;-<span class="st"> </span><span class="kw">findCellTypes</span>(scfind_index, effector_t_reg_cells, <span class="st">&quot;Thymus&quot;</span>) </a></code></pre></div>
<pre><code>## Found 41 cells co-expressing Ctla4 ∧ Ptprc</code></pre>
<pre><code>## Excluded 7 cells expressing Il7r</code></pre>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb193-1" title="1">subset_tmem &lt;-<span class="st"> </span><span class="kw">findCellTypes</span>(scfind_index, effector_memory_t_cells, <span class="st">&quot;Thymus&quot;</span>)</a></code></pre></div>
<pre><code>## Found 167 cells co-expressing Il7r ∧ Ptprc</code></pre>
<pre><code>## Excluded 13 cells expressing Il2ra</code></pre>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb196-1" title="1">subset_treg</a></code></pre></div>
<pre><code>## $`Thymus.T cell`
##  [1]   28   54   80  121  125  144  161  165  166  180  229  235  251  320
## [15]  366  371  476  497  663  667  713  737  739  740  786  879  888  918
## [29]  928  953 1098 1225 1232 1277</code></pre>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb198-1" title="1">subset_tmem</a></code></pre></div>
<pre><code>## $`Thymus.T cell`
##   [1]    1   18   19   34   48   96  100  101  137  138  150  171  177  186
##  [15]  198  215  237  254  267  284  292  293  312  313  323  329  362  376
##  [29]  405  416  430  455  456  500  535  546  555  561  574  580  599  610
##  [43]  634  673  681  686  687  698  701  712  718  719  722  733  735  745
##  [57]  754  755  789  790  800  808  814  830  837  846  848  850  851  861
##  [71]  893  895  896  897  902  907  908  913  916  920  921  930  931  946
##  [85]  949  956  970  979  989  997 1020 1046 1047 1054 1055 1068 1076 1087
##  [99] 1115 1124 1128 1129 1130 1140 1148 1152 1159 1168 1170 1190 1206 1213
## [113] 1226 1227 1237 1240 1251 1255 1275 1282 1284 1290 1298 1305 1307 1312
## [127] 1315
## 
## $`Thymus.stromal cell`
##  [1]   6   7   9  12  17  18  20  25  27  34  50  52  59  63  65  68  69
## [18]  71  72  79  80  84  88  98 100 104 106</code></pre>
<p>使用来自Thymus的<code>SingleCellExperiment</code>的TSNE图信息来展示gating结果。</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb200-1" title="1">map &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb200-2" title="2">  tm10x_thymus<span class="op">@</span>reducedDims[[<span class="st">&#39;TSNE&#39;</span>]], </a>
<a class="sourceLine" id="cb200-3" title="3">  <span class="dt">cell_type =</span> <span class="kw">as.character</span>(<span class="kw">colData</span>(tm10x_thymus)<span class="op">$</span>cell_type1), </a>
<a class="sourceLine" id="cb200-4" title="4">  <span class="dt">stringsAsFactors =</span> F</a>
<a class="sourceLine" id="cb200-5" title="5">  )</a>
<a class="sourceLine" id="cb200-6" title="6"></a>
<a class="sourceLine" id="cb200-7" title="7">map &lt;-<span class="st"> </span><span class="kw">subset</span>(map, cell_type <span class="op">==</span><span class="st"> &quot;T cell&quot;</span>) </a>
<a class="sourceLine" id="cb200-8" title="8"></a>
<a class="sourceLine" id="cb200-9" title="9"><span class="kw">plot_ly</span>(map, <span class="dt">x =</span> <span class="op">~</span>X1 , <span class="dt">y =</span> <span class="op">~</span>X2, <span class="dt">type=</span><span class="st">&quot;scatter&quot;</span>)</a></code></pre></div>
<pre><code>## No scatter mode specifed:
##   Setting the mode to markers
##   Read more about this attribute -&gt; https://plot.ly/r/reference/#scatter-mode</code></pre>
<div id="htmlwidget-923b1f469b1d2b4d85c0" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-923b1f469b1d2b4d85c0">{"x":{"visdat":{"9ad84a4a7516":["function () ","plotlyVisDat"]},"cur_data":"9ad84a4a7516","attrs":{"9ad84a4a7516":{"x":{},"y":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter"}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"xaxis":{"domain":[0,1],"automargin":true,"title":"X1"},"yaxis":{"domain":[0,1],"automargin":true,"title":"X2"},"hovermode":"closest","showlegend":false},"source":"A","config":{"showSendToCloud":false},"data":[{"x":[-0.64700797896146,9.82891143936811,-3.00196647225079,-16.2636878973757,5.11026001190674,-7.96327259272952,9.8278810873859,8.23099598692419,-14.1328068830468,7.55202175312812,5.58212148331875,-0.968089213823785,-0.691558447469152,4.22047401496733,9.13347093938178,-13.2389163454206,12.7275993677326,10.9829194193465,-0.683169312033188,-13.385212780602,-11.6767711905462,3.90796220999321,-10.1058977187402,-15.4974877269412,-11.7262439499746,-8.96469744886844,-3.42860593124477,9.05700339075373,7.33006942842282,-11.7174038942314,3.93814303620079,7.46224187040622,-2.52271126038625,0.121598337692018,-15.9687795641159,-8.57235426903485,-9.21562151777873,-8.26416314323915,-4.84601253223187,11.3290849681427,1.95401739442417,-12.0106767257361,-6.99277813008846,10.1993777742303,-0.925864772883279,-0.0294976427393101,-1.62499371485408,2.38673232083063,-3.89113272565772,-3.37191805685573,-6.06683634915917,7.44489014696503,11.8506979167224,1.31582770457569,-16.1165585028952,7.75200687218933,9.13599607227886,-3.9052052980904,9.42711905621988,-11.5764456740183,-2.30986632161948,-2.80101344559873,-8.47418423574407,3.60594209112652,-12.818035489478,6.12061932001222,-4.93997138252942,-10.0709655747583,-11.3080798752615,-11.7615114398318,-9.6511354498064,7.87559868152668,-1.59995583539679,-12.1076158938546,10.6185783604724,-4.27072367504397,-6.6289221469232,-3.93109620627119,4.37209016342535,7.71129866826428,7.78196798797721,-11.9993497926742,-5.40163299673636,0.844356426968267,7.25149832125925,-14.549257814265,-2.84591581148156,7.75011498084225,-2.40760105455274,11.8563972429839,-6.89775579946556,-4.43890490045562,-10.4567779527525,4.1944754345686,-12.2686664262743,-11.2077529755415,-17.9774965905025,3.33098043326833,2.25920179307337,1.95983387495633,1.355658058858,-11.6729610313437,9.81837505506589,-4.51042888834907,-14.460521305944,9.719407141031,1.16614902939685,-2.20399706923565,-9.74109013296507,2.51685241926093,-5.74660583936413,6.03341468458895,-1.48333881074084,5.14639604945107,-12.0479781225904,9.44862150446,-3.02307650368023,9.51015296043139,1.76512017529795,-2.0101180061106,9.64284253210722,3.11848686773141,-5.17226368049842,3.4528729681505,-2.93054633868947,-11.2053652896679,-3.17230186705166,5.23743507166201,-10.557885415587,5.36297744476268,-2.52871812581956,2.20963427724461,0.312493109546483,6.59956451840895,7.46740832723688,1.91010262260896,2.28442859883303,3.16351677858245,-1.98629965354965,-12.9475085909362,-0.787555785873725,6.73115074510438,-2.27636621399645,8.01488449123776,2.97792546441982,0.960353809208676,8.1304638915541,8.4836950741601,-10.7166417232331,11.2807423417454,9.60213585682176,4.05664331570355,-12.7630804328154,-0.679192258286923,11.6507763440007,-1.16253798832006,4.43155910718925,-16.6155095179111,7.21174356368151,-2.08320304186372,1.73017234445989,-12.9901107730417,-3.2495720438173,6.69384435463233,8.8512205635191,-10.6085782363227,-5.06207605418766,0.749349394627608,-4.04566337259661,-0.809704681726989,5.52011162055027,-9.59505729460747,10.4501732061606,-16.305104514804,-11.6444120095217,8.81181963063932,11.4783367949828,-1.77739152145807,-12.6968328803359,3.00191378192751,-10.7049809306755,-12.4882943828428,10.2884510063357,0.371997644469344,-15.1391489673992,-1.09254313923237,8.81894416698965,-1.46010378954201,-10.1755845010003,-1.35850404432497,-5.14819950373142,-8.41251184644206,3.4914738601816,-9.9394576287063,3.64380835425618,-4.07982903761715,3.19201618620682,7.81987178814859,-13.0337361922979,7.70603774293683,-0.533564774166623,5.50025867192459,-15.5505999207836,-17.3404843193353,-1.28459518800085,-5.08459841534678,-0.889907541794594,2.89131505992518,9.83111307936576,-13.3934892257984,-8.18246290564943,1.05871145735768,-6.83681856140259,8.72685719935613,11.3704066013725,6.6019429208772,-3.93993006200646,-7.33109921881688,-13.8246756280819,-4.89727184809966,-16.8383949538528,10.7346360732132,5.18740054083238,4.78186902636418,-10.7771865434061,9.26369926671915,-11.1981176072996,-2.2351493628046,10.2663589499106,1.54855336118919,4.2102841265481,5.14763157067476,4.89652319571709,7.09691582977028,9.60754861635615,-7.0005755478137,8.49868714453964,2.92429166154218,5.75355179555096,-1.94018351402956,9.94334428389341,9.91861370883366,6.4350254255442,-8.31775825570102,-1.10999772469839,6.80639392821921,4.72908710375811,2.39648035112287,-6.47837473384119,8.11723377952883,6.36554730667016,9.01185872985275,-3.47195785338812,6.75622092439215,1.42258535179755,-12.8995307978385,12.928799632771,-3.85517979721668,-2.64739120566826,-3.15581452052875,-10.9004231645292,6.84044166266121,-8.60348685857969,1.89243982929869,1.16177013190608,5.16078907281793,-2.17811482875469,-3.34341606459096,7.67993241909487,-11.5993243747319,7.70232886715235,7.21905228203758,-10.8316024258295,1.98342671662567,-5.24395941561369,1.42276917088958,-3.46942869890395,-0.160396590088969,-0.52973552678463,5.67038227928319,7.11647437555116,-13.447417763681,-7.67537730573923,6.44083015176533,-10.2679583593352,-1.94807156536674,-13.9876122953144,-14.8087686026045,-1.95394043338005,-5.77755944378688,9.64487625869794,4.58506236787563,4.36284980705886,-15.0491119991027,-3.23870455928025,7.01981910268452,-15.5246449826532,6.2136957990831,-8.88722627972841,-4.25695055555794,-14.8940624230048,5.04888619943999,9.00091945343663,-0.419742225070688,-9.25907816321094,-0.0389637447942507,-1.35395065230248,9.20105750422091,7.46927957871424,-3.31507008261641,0.572508261887164,3.40514142870881,10.8994791252426,2.23124919368925,6.03360917359753,11.2146005924885,-4.66441568165984,-15.2736422733279,-9.15377752041512,2.78085260040477,5.12904382335145,4.0882266665905,3.34521572800305,0.461831314799731,3.97482665515735,10.8519664180908,12.3010948523928,-1.50861297602589,-0.264389549753781,9.40180723432858,-10.1267842781459,-14.3254330087046,-11.4493864145288,1.12529842400453,7.52782449746136,8.51973455277696,3.75371496161553,3.28753576627298,3.23157223048133,2.96246901198722,6.80137599040706,7.78501145263659,-12.6570833769504,-3.1063261082921,-11.4411946732459,-12.9362281851671,-12.3555114540239,-0.28333411748761,0.755350387541639,-3.9649969760844,3.76756782740247,6.77036626743522,-11.8324950698558,-8.37539909528126,0.679085042846249,-14.9601217324881,-5.1818116282347,-0.0942536726845936,3.56548195374302,-12.2926679294128,-6.12910432435701,-2.34287276102339,-9.918482383013,7.09903538302189,2.16008718644457,1.60518134357344,-8.10410403638876,-16.7532427217557,4.41264354137519,6.36600922795468,7.08497284512453,3.39611828225795,-11.7250008361746,-9.05975337407891,5.0792954776135,-8.59900198432096,-8.26470094678932,-12.3290026309716,-12.3810163075483,-2.37752360481419,-4.45136544309075,8.14355923049767,-10.189806344882,8.35094830303059,9.26690983712412,9.70043224865771,3.83264269064231,-13.6943386504416,5.63531880052481,8.0839711172369,-9.38215899273452,10.7606295555205,10.466995043177,-12.4160028951189,-1.00825621425155,6.55730973409536,11.8634528474964,-7.16235345508303,-9.87857700705185,-11.8643616669115,-0.0655645728983408,3.95909449379202,-6.66110477582011,3.6124314740147,10.9314522568183,6.06000611014882,-6.26920877808011,8.70414236241388,-12.3708513107692,-3.07606905859856,-11.0075685844169,4.0836704198752,-12.1123681950746,-3.88244070834873,-10.3426964760793,11.000742604909,-15.8829612504354,2.54520334250744,4.67677432635109,-3.56405154453272,0.565117438569289,11.9748153156141,-10.6201297025315,-11.5121392495465,9.56972139586113,9.38493519816435,0.0498449281812405,1.98049346977713,3.33354445117847,4.09323862442843,4.49402228006215,-7.96186224336824,-12.6564613966401,-11.2428640140026,10.71085448709,5.309910826302,-4.1912008612685,-6.00060207984623,11.2320287050076,11.4479800255823,7.90371451438669,1.07785797462281,2.50537639501785,0.72359120657806,-14.4541637533114,1.39585806479195,7.00338781258431,2.50753719573538,4.84587729618694,-3.73320579678222,6.97267260261632,-1.55629665582251,-3.37079325635419,7.27811566716854,0.00106162044567765,-1.15749993922686,-13.6135104878247,-3.4482749963,-2.01995382347535,-2.14611337848434,-4.09072615262254,-6.15456234787786,11.6116225418866,3.56318558404695,-0.327441989918096,-3.3290997782366,3.08861359082291,0.20021924189669,-3.55698669208339,2.0445313741246,4.85878957851101,-10.7368285093069,6.54247510900478,10.3738336790499,0.916300211224866,-9.93701471824346,4.45474755223373,-11.0742695809191,9.6909220236751,-5.14179262305946,-9.13925256040715,2.70019371264956,-13.8722650225966,-3.06348763310777,-2.27084700771599,-0.00480882368855796,2.95134655502424,8.89090405719254,-14.8214364838395,-7.21582354574512,-1.18710806271312,-4.64800975333439,-14.4103146911509,9.18076126642862,-2.38016115252994,-0.446759142144465,-3.73724777055028,-9.55133364613346,-0.0984271813938266,6.62847044191489,3.19063506532798,-14.3846558359694,3.11361846883912,-9.59222540963671,-14.6879580193745,5.14681015456595,2.85735569271041,-6.7935056372843,-2.40100686752974,3.99181744517366,7.20908365649441,9.65398552292046,0.335664680685584,-12.3288088475584,-13.1940751569452,8.92004905480146,-15.9497816728839,-10.1828289522161,4.89453888547524,-13.4531697559213,-9.80858728741571,8.44650653952169,-2.81820794831571,8.12187068293113,-11.5662906299865,4.7455467329387,7.86012121985834,10.2731104021915,-11.6384853997638,-3.61926738524991,8.75423568781965,8.73415503739917,1.77214723493637,-13.0747956341718,4.13166737628574,0.0970966036085597,-5.28553097863298,8.72958344929029,5.01480765578096,7.13305932617364,-6.90367885550704,-1.90166662639275,-0.0785130993241402,-2.51546247466487,-11.3664861845708,2.84976094903702,1.84139901284619,-3.25399720559288,-4.82246099656796,4.26384066272578,3.59839926693447,7.59098262208168,3.22524360992331,10.6337403406005,4.83917615535764,3.46964466644503,8.20869534662265,-1.85838165159572,-0.471454226324056,-3.56385103464068,12.187307821636,-15.6470803289408,5.04855201904907,6.79406144720244,-3.43959168860988,-6.96098368540143,7.04638677988106,-8.36890024646393,-9.94027992213227,-10.0738783419843,10.9618017977764,-11.5686983386592,10.2245565289438,11.8086040828073,-0.041511442884092,0.0171818410957833,-1.4877167314915,-17.4798777156874,2.5099724083714,6.55723374404436,0.223102107868022,7.67234711509862,-3.830453305316,-15.0906908203436,-11.0437018224022,8.85760497580631,-13.793000506398,-5.42770628760829,7.58952885387356,-3.68547193382706,-15.5563004823308,-11.221441402074,-7.03066717707251,-0.423337520241322,8.841301385236,-6.90208983428567,-11.3195604595538,7.06830473483865,7.0424681444496,-10.3539460145979,-11.4771218969258,11.2932943684557,-1.35557713695463,-4.18602739352296,-10.0751465490147,1.8788847399775,-4.10020933548653,8.26116887242672,1.87384805832659,7.65009438485869,-11.6643611778711,7.49096581669704,6.89886624881443,-14.0038814165246,-3.07178865739004,-7.53119984941236,-9.23371108437883,0.129995946427896,-11.5931867703794,2.48109859300615,0.72066770057411,-4.00261296836749,0.0565345794784102,-4.98689691880957,-1.0440663695784,1.06092720810148,-4.32537949856768,5.3075877512733,-11.4554704790414,7.87989838024182,-13.4777994123988,11.2946535504296,4.41543754996927,5.30907608211705,-3.48331152847722,-1.48866093092278,-1.3940242218004,11.2775353622639,3.33562273664726,6.30098588723641,8.70658523087676,-13.2705672696049,0.766327923268357,7.55707215819368,5.19587118321291,2.31800317440879,-1.29638517874767,-1.1266492514967,-11.4429159185132,1.06045938930989,-10.5373687527947,-12.4923984070613,11.8117442771488,-0.949649808405418,-3.14918213595859,7.47680074253922,9.47983136223432,-3.51315127892931,-4.61340504968856,-9.42293560363734,3.77537212102496,-10.2695216542302,-10.9464800679939,11.1821224324796,9.62687118450968,-1.17121668558615,5.83027605241538,-3.93403756478549,-5.06708348145476,-9.80520498638217,6.74911722868136,7.15162221475909,10.3024174940888,7.27480994335103,10.5147485288601,-12.6425818934727,6.74280671691854,0.831822765126827,8.08729861824141,-3.69389188826327,12.2124896762216,-12.6707253821162,-16.2972092291776,3.77439236409139,2.7236908730811,6.40845826135662,10.6968203170107,10.3593393560673,-12.3229140351064,-12.8855862343476,7.19395444800778,-9.37433759057026,-2.61246268326568,-4.32203717191813,-7.49633456691126,-0.339933868028644,1.18671238330189,10.9361793741712,-3.8372384469258,-2.16840907514794,4.75953339124891,-8.78464730418214,10.646569505448,6.23702262641971,-2.46278095277826,-3.32608342500181,-11.9994339708017,4.78595728883319,10.1004626161565,-14.0854228157649,7.11563097701212,6.57113355494872,9.74802910763305,-3.35268081168093,11.2651997047495,-5.11895583150842,-10.9498110847956,10.058247357314,10.9208742224803,-6.78685441429928,-13.9667144729033,4.71517557420859,3.19715576239278,7.66434786434726,-14.0155111023421,-6.23364047608626,-10.5932224233214,-13.2109844382228,7.7868944659012,-10.8365314353441,7.57081927345916,10.72335241332,6.78400552552356,-3.7232025873966,11.6657448844689,2.65036000281584,7.23375632097735,3.05988412708265,10.3147756943685,2.63537008716921,-3.41732955929093,9.62208983522594,2.4041545341629,-2.82137499922696,3.85762682711509,10.8093579261054,-6.06474366432108,7.62195715175079,7.59637948642491,6.1109066943915,-16.4797181727674,-7.01528517137479,-2.15360715931152,0.494066188198416,6.5801785461178,-11.7738516352146,-8.83953071922091,3.86737733893302,-1.20069981809382,-2.715492437111,-8.21232322755488,2.13932134820078,-8.76068582757213,2.17047392431433,10.4171908216825,5.40644041323777,10.2313878323923,-1.37376226948617,0.672197857752152,-12.5284917213217,-9.72213533062268,-15.0534595981337,-2.79203323676675,-0.795172957832658,6.22536395110988,-13.622218871723,-10.7960227323767,-10.1821594286835,0.130571295687463,1.21353874847712,4.65471198806589,-17.4813038107817,-8.03636110772787,7.00714741638335,2.61557346020311,9.87969503104111,-6.49582629789168,9.98447415557156,-0.688165968947004,11.2278961065549,-12.4319361717574,4.98994819641734,-10.2922069562845,4.15105991583381,-6.13060666993313,9.7459362422876,-1.50030216202154,10.112396741618,-6.60131977511768,11.0142674390149,0.152023997005444,-8.0878030317311,-14.5141855649691,9.3041642813024,-9.61653552327437,-15.4088469327069,-1.46647589381195,-2.10098875565224,-6.10026229643796,-10.7228584739178,3.35477992160217,-8.64448573233908,0.810216909208728,0.425374376100636,-8.18835435137973,1.47545522992215,8.17655875575261,6.94351802584242,4.99121655893888,-12.2001833095526,8.23983376841392,5.08508410049415,-0.19144892564878,-2.15080996698666,3.05305762261268,3.27855265597793,2.8255953921857,-0.199305439454856,3.07838114462375,11.1926430953936,-7.95372136419564,4.48744260557881,-3.00441739893663,-2.08236999039003,-7.41668049262339,4.10275902944458,-1.102078458571,-3.6697836394086,11.9901917760387,6.56274637827125,6.90033151976202,9.83015451386409,-9.68079557349856,7.4701131100295,-11.8108389200304,-1.45661849552363,-14.8622937010971,3.92410129409312,-11.0635992246399,2.07568337915508,9.29157384710789,8.07968803706528,-13.5747680802732,3.20612307421428,9.64472845844451,-3.35022835109835,9.32532694051381,-1.58215914467997,6.81347418097091,1.06436527474142,-1.91146276115677,3.17120508230376,-3.66711732098798,-2.98244750703895,-11.0645368570523,-9.40135268997227,5.55212476112239,-6.16560493452022,3.81873019124052,-1.27030861456813,-11.1959796197472,2.96144423219601,5.20116990332279,-5.22434283445764,6.83515194188371,-11.6664765045506,-1.78128658383716,6.27810516453127,2.9947648967973,5.6509142931066,2.78277732669555,7.70356117259486,-14.1027219497052,6.58500357982688,9.65950771855894,-14.5457196925591,-13.4435243927939,0.851096695143485,6.52260281146573,6.90576457774171,6.23752392817671,6.86910783375688,10.7020156677305,-14.4532693433643,0.812936146985994,3.24280819380646,11.6066917033058,7.42649178348599,-7.20622652350954,-0.550317965567239,-0.611697750365122,2.36136333003448,-8.13406693871669,1.325187784636,-9.54157969318261,-7.09189932007124,6.83611484977029,4.67244286208184,-1.23712731787688,9.18406881853282,-14.0227312612812,9.31155063438081,4.0697905092024,-13.0109435734237,6.5098992248851,-1.43511258056262,-16.6779390834791,2.95982330040759,-7.16494029513774,11.1920356560971,2.32879903511123,-5.08923495179521,1.5963118888362,-8.97359580462543,-3.60962659757962,-0.61744387483345,-2.85701156634217,5.5357724584898,2.56357794696911,-15.3017457132328,4.58491410596088,2.98747824076449,-12.896102287132,-13.9258736873381,2.98671548638628,8.49153238398527,7.08553815402189,-15.9867652858164,-13.6709538394743,-3.65088878315723,3.11953849342822,-13.1839496917453,8.75766847081078,-3.88420442655663,9.64350275686616,-1.37408953470131,-3.59943552623882,7.89382320624922,4.58295220662718,-14.9878224817506,-3.95301432540554,7.02835741300142,-4.36001595253798,-9.12678758665595,6.21723769253142,-0.105397174020197,4.08111628731198,10.9990127174105,2.01010014786725,-0.878415037347965,10.3845028865434,3.40155729752614,-12.1418086386029,-1.71673740896422,-2.65388907888998,10.4333713655889,10.1158335067172,3.20112670404568,-1.01033115770864,0.486357037986709,-3.1756773065346,4.22578116202339,-4.40176738214793,9.12595870466929,7.96244806136765,-12.4770167606737,2.87518122474179,-4.23078435556712,1.03991644902279,-12.5771694231855,8.37341467262872,5.75138764679351,0.660279006477143,7.3323395626482,6.06139339432244,4.13183137810494,2.19975638204812,-10.5504892958463,7.70220668792144,-2.82072296151232,-14.1858394386605,9.2789493402567,-16.5799941812136,-8.90464766502288,9.21875256640506,9.20421894411271,9.11570175777651,2.10932795320815,-3.9976735942059,2.10512781362829,9.33830611652755,2.50730663695904,4.58318311997114,6.80932128478773,3.92607583300854,-1.23540773182524,-7.51064271365029,10.5357372122005,-5.5218875911005,0.00976673020773924,-9.89805005605661,-2.23483889784124,-0.832349705887485,7.88706968943287,0.96543954928539,4.97476417641664,0.244998614532728,-9.10127472597457,4.28473610982041,-0.0934993254164325,5.22274727155551,-13.0070939544239,-11.8493001484245,1.42854381602167,-12.7399866144193,7.07007002175112,1.24385949720144,-13.4335350527399,7.35756042659514,-12.1372952236963,-14.2731920527005,-8.58410173741731,-7.35729850580197,-14.2132663520284,7.77145318355852,-2.01806303375981,2.11165533581418,3.0771901181428,-8.90779277253557,-16.9509982281864,-3.55023891533676,-14.0097814184945,7.32963773682705,9.118271566155,10.3223779667283,-0.785458781544937,1.39113719941521,11.9914218289086,-7.40357547883626,-3.21503236779638,-4.90486210994078,-13.6901823001047,1.40596362647126,5.11656436238653,10.8440180659292,6.42591343768377,1.98140451255794,-0.259143915494176,1.22751045600258,-12.5335842699494,-10.1679865991152,5.11157833173426,9.92632957743197,-9.70017982570916,3.22030478152524,6.43676328908583,-0.739480456406954,7.37021089235498,-0.976495840637493,-0.568353632806302,10.2018934153319,-4.22612674752566,-7.44511756534601,5.5549194184107,0.601330692589122,1.94484055193084,-0.620886096735016,7.99427464095841,-3.93334559288339,-10.8504684645923,2.84081317661909,6.83395463308155,-3.78073461754307,3.78409018387511,-1.30541425020812,-12.3357203394803,0.48410980485091,6.38971578616204,9.65684994542042,-2.75438339932516,-14.4803163665767,-2.15501430053395,0.465994702172914,3.2285167988725,4.8884276486059,-2.29902687961026,9.16434995883591,-4.5995400684101,9.40110109511421,1.1837201930044,10.4419501548189,-14.8429672895073,-0.564433723719789,-2.05830651686259,-4.35391670968321,11.414165116044,8.5939396898019,-3.7853091020217,-8.46395013637652,-8.06935307815328,8.8385487744995,3.84313804530459,-3.05304359390583,7.08273347127422,2.72718712194514,8.40531559988768,3.97870544551576,-16.4670822395134,-7.07753069567801,-14.2384247196042,-0.126657088594808,-5.28809161792464,-12.5874350120199,4.18832517656818,-10.5913232687267,-2.32011198439004,-10.4884599256429,10.9771942267818,6.92733062346478,8.22544958377789,-4.00501773898503,-1.66675170750261,-1.44492856125172,6.59171897248965,-13.7807630629173,-3.34071056786941,-9.92595191234454,-0.871271531215126,4.84447210190695,11.5338033754264,-7.1683656823241,0.220125787243174,4.18600892926953,-3.66702122765043,-5.8709026100168,-9.61498001990717,-11.1673490949517,12.1847838412572,-7.50753628267548,-4.18769204949873,-10.2328153946908,-15.2736902813511,-1.29671711423883,9.25233922256825,11.6739934569865,0.0996158734085468,3.96691514485562,-9.12681537118177,11.2687716473871,-2.81762341296063,-7.24269479141646,-12.8997467944318,1.26425093913796,-11.9080891106839,-9.60075867733841,4.57872269421269,10.100918332724,4.46726215723052,2.93114678782632,7.58436483303276,0.971223867035387,2.88436550294712,-12.8614336085156,-8.12749823276476,-13.75374088855,-12.3035540169631,2.32327514643619,7.46189835034588,4.33386843396052,-9.17945031266364,-5.55991360900063,-4.39404665637217,7.36862763331266,4.21658567315317,8.8030100448261,5.54592921961176,11.4502358224436,-4.94818308226012,10.8379170747792,10.8374117598183,-7.90282113886091,8.68292250505573,-2.51049372240199,-10.2590582809269,0.414952157253194,5.13842097257281,7.88167915832968,9.33209008128285,9.77688304066469,-3.09939171531768,6.96408772603081,4.3295620612701,-15.332191387827,3.46344178929219,2.33018685450472,3.08052478431283,4.29597182025093,-1.47876778696363,-13.0381138315334,-3.12570613346694,-2.73059870786524,-8.63795891628221,7.08924980892202,1.69157077101361,-12.1278478231357,8.80881877372487,4.53269780406071,-2.5785420050338,-3.02785550085931,9.03179903908925,-8.30283467289226,-3.84538757753671,10.4030243228389,-10.8869128348097,10.0048414242583,3.28560379927315,9.29156576981388,-11.0365835810929,-9.80976296813444,0.672843607644832,-11.1581879938641,10.2834341290521,3.52455081731651,1.24233166113872,-12.5760171546041,10.5665549718346,11.7863449304956,10.1568243344168,-2.36676340444677,11.0706423843404,0.86883969650285,-11.6771019361012,-5.67916944004976,-13.114908550882,-16.7699164496718,-4.26972676319082,3.46789248006443,-3.71584694999347,6.49582498715062,8.51423657394286,-1.71095371214338,10.5617539414593,11.7707552627807,-4.52441167404108,10.2917018827769,5.25270585839545,0.789519676614693,-6.46361634924249,-0.561773394102143,7.10202796386736,-15.165033503964,-3.14938720237319,-3.10205777486574,-12.3142986537011,3.92461712474966,-4.35439725062635,7.27740366109258,-2.66093884507689,-2.69122531011063,1.16265003024978,2.54606165930274,-5.33757755794002,10.4707150312347,5.30053421230119,10.3161270180132,-10.6125172138372,6.64017497081236,-8.29851165258077,6.6694061446567,-3.46212287384644,-10.8633090396677,9.25038427002594,-1.98841667427302,5.30844999186754,-7.84047504370984,0.934151589473817,-3.3548279262894,-11.8910231717265,4.87870037834936,9.56295717864335,7.64330829935797,3.29250827097822,-14.3121121453113,-2.45711302943458,-1.46618076547506,4.32191941222138,-2.36503119749703,4.45190660277448,-16.8222330026763,-1.51368749211814,-9.79123124114864,-9.51879561798784,-13.9987247818322,1.27830781675901,-1.48481746792535,2.48925240981731,-2.74809653318622,5.01034774217534,-8.3226926630704,9.65498894519923,-12.4866849083789,4.37324345918158,10.3136304346567,-4.30743442868922,11.1944363981471,-11.2332928482887,6.94753258461691],"y":[-15.9673377599113,5.66975964902548,-8.77818950821644,-2.13961178993674,-19.6984183705609,-0.196920975539678,-2.42580987010058,-19.2343069267998,-2.99109571081673,9.52752074000473,17.1364426797165,-15.6402161539033,9.78267413804593,9.43263556966627,5.06757808058878,-4.47926752294106,-17.3691746558434,8.48695469183638,-13.0574160987575,3.99760455244532,1.4151029400085,8.89224469808186,-2.83992481423707,-0.492336956733319,3.28290991087231,3.85990793553297,-8.90705092873849,9.42016369662749,-17.7902663841582,1.74661216774987,-19.6411813602325,19.3742870135737,18.9811333759661,-18.4203719385579,0.765645695938311,-4.73163549436878,-0.855021596928189,-2.48653374675299,-12.6792841846805,-17.4875155763953,17.1627125138593,-2.16264078943451,2.15754423793356,-2.80383147240977,-10.7142090887233,17.9389565154667,18.8395192760639,-19.803975252085,13.7658831814431,-11.1422906848907,7.91357465907774,5.02020416822446,12.3781104233532,16.2598662505788,1.20421776039783,7.84583189043147,6.3837089411763,13.8932999344707,-2.35928201920734,0.927266973523188,19.5254608802006,-13.198523180346,3.12796600297153,-20.5434998098939,3.52704695759401,10.3820067840877,-9.02868546572871,1.67331190700338,-0.228720331597061,-1.39100558198826,-0.122412142921674,16.5671611812012,-14.4204757055845,1.81665722687687,-16.8537163561203,-11.9264435258769,2.68955297255271,-13.817486831441,7.87312132250302,6.16073009561685,8.78501077174469,-2.41065861023483,7.74209477699574,-17.875029477808,-20.3382525365537,-0.156889809164829,-8.66988290236121,7.20077879554241,-16.5236246366409,-17.5383460930866,5.4825751966243,-13.2242586383442,2.04709014195926,-14.1287282778212,3.18596852675273,-4.92644528565067,-0.0870944565277938,-21.4542252074024,21.0191128581882,-13.4694097102671,-12.5317104441329,1.60385474551617,-16.964333485658,7.53925839322867,0.806244469881588,-2.8033779330827,-17.3861728309692,17.494924040051,-4.02931917499083,20.413866737375,6.68150994018296,-10.4460495406797,20.1855077863121,15.3628640532909,5.70775721881636,12.6828090398904,15.4353103914631,11.0990568738171,17.0636916704775,-10.3337881049011,-2.40995216662941,11.8316255098366,-6.68512894750448,9.20532089676397,17.0802490611973,1.18244122540038,15.3853592458602,-20.5062040646386,-3.85040652471447,17.5423193218868,17.8169330823617,15.8112019462963,21.3387802744448,15.8797497722738,5.41224439717998,20.6249474983657,-16.1928973513659,-17.7188971162668,-8.98080173055824,-0.903288417948641,19.6846563527395,19.7363430603188,-9.34848351542492,6.0884210026157,19.2118956522669,19.8645388208863,-17.7399069662793,13.7281148329862,-2.73031759082039,-17.3027700684193,-2.81041070006992,8.1957884771828,1.03636643502591,-12.2927630262754,12.4362631504786,-14.4579071556128,17.900888083007,-3.98292286863356,19.4754664953722,-14.5664465530997,13.0324851488086,2.31148961428687,16.6299468011584,6.27441369187808,9.63182608415293,-1.28370244684747,8.01395323549763,-12.6676261670212,8.19553323534016,-8.3074271563913,-18.3370279971478,3.84597510272244,11.6956907659842,-2.76544888628933,2.29912225727876,15.5046538207351,-16.8881379685924,17.1014019053288,-2.00312377145404,8.71977138971156,4.92284235823714,-0.695932334115392,7.76873083402207,9.75931456833239,0.953908160832141,-13.0262221327575,18.0787377677213,-12.1647309934251,-2.42537512431964,-8.10786758225995,9.38031138299535,5.42039734168583,18.89016137245,3.75812702143235,18.1215022624889,-12.4198631850512,-9.68705242633241,6.63700726756897,-0.462287420125959,16.5428625688585,17.2963388559552,19.5845643045607,1.99839679438684,1.05203043609798,-9.44022344452382,-12.5098966893475,19.9883653260893,-11.4668381940495,-2.51920279720007,-1.94362589422666,1.30041141159447,15.4712630727988,-7.97105542424684,9.0885550700181,-17.3789860517129,-16.0708143259203,13.651996886796,4.11706185657858,1.21517331414382,10.8910790040793,-2.59064794134656,11.5909231940057,17.6927126224925,19.7962819482183,1.31752426671125,-2.6266716631909,-2.9914589958855,-15.4066676597385,-16.2026383984467,21.2287038403327,9.17465382615656,6.0829249596365,-21.4100486909801,15.8998983618823,5.62077466672877,0.121534486468602,5.91677917150417,16.6542024316003,-18.7837797889461,18.2713082869209,15.7583643487246,-2.47482895690965,15.6281478652042,-4.89756518440682,-13.0722727188178,-9.24452462160515,-18.549461938582,14.5802721093655,4.04565429687868,7.79383009582645,6.37329153855569,5.46632268170458,-12.5048047973784,-9.05011487704128,-20.5155466098898,1.54284473769175,-17.8643543468521,14.9389946351783,17.8152148860166,15.0874650458087,4.12253813767634,6.79536626466934,-3.16933669862998,19.2736094883604,-21.2466435802668,7.69934125545749,-12.8488573059695,15.1299277220383,6.46163062266483,-0.515585277589916,7.22444121228801,15.6311472667667,1.57422509992914,15.5932848851086,7.82841272501197,20.4393105600063,14.7704126306148,10.0610786618068,-10.0934141484493,-19.3735458271777,13.3824888085295,-3.91201560727871,-6.17802969035687,-7.00944800627085,-3.0782854722129,18.5672730871287,-0.213116993316094,-3.92745993225775,-12.1339659087895,6.24338226102606,-2.81738502775102,-17.316741005132,-19.1028167013768,-2.08228378769089,17.8136610420233,13.3151896067032,-2.6487991668168,18.2526662464085,-3.2693215964209,8.72366486429651,2.70168099469592,13.2827947914069,16.572979953092,-14.3405385514874,2.70772730819146,-13.3323497518113,16.4590349549654,-2.24230932647231,17.1724444086406,-10.994384941468,-8.60463894901688,-18.8549436183589,-16.8913154882993,8.96317709717228,-19.3901642733629,14.5168847896221,8.69169027305702,-1.53771585385232,-0.984246464803202,-16.3254510507658,13.6827072255216,-17.9832895498747,-20.9232278108037,20.8282787744234,-17.9281699803905,13.8382915891702,13.9770802016478,19.9704223768714,-16.4891007910443,-3.34561909460168,3.34767777678513,4.43979394280692,-5.74764595566222,-14.2799481430226,8.26328633034095,12.8295408942651,-19.869082967735,-9.48802658676562,16.003876005742,16.0020751450403,16.5498827131526,10.3223313012824,0.207876009194289,7.66663009032203,-3.63817244182129,0.528683552321408,4.48202850203215,19.9210461358447,20.389113945851,14.7172745793256,-18.6724228331844,-20.5754501842944,-5.45855740488608,-6.57079439419139,16.2457160493122,-3.53396956780216,-13.2141498278453,10.3884748029506,8.72763756535123,1.38116179980807,3.52742574951163,-11.4932606569854,-2.04680639205122,-14.3704430620874,-14.7898605268285,16.1117903310734,2.07214115574314,-1.24163081787943,-20.1604873612004,6.41051763781159,13.2111800958681,-17.3849461031013,0.518411375416865,2.54199777444288,-10.402762631616,-4.41454406033073,3.16938865516432,2.92341685921823,-0.205227723697595,-17.9325411357957,17.5345607378754,9.89385110045275,-0.767757027691558,17.0659670848245,5.05277920748838,11.0046103480639,19.8772905951954,2.4920515631586,7.18944438465895,5.23412624050007,2.71874127630484,12.4916241216863,11.5684429432305,0.422853835792536,18.1503072466146,-7.44621265828443,14.5780235338787,5.58378617164743,1.20760050517734,3.84678867018867,-10.3670858521612,19.6165993844413,5.71517593971534,15.6485901211095,-17.242088178452,14.8901614364899,-7.82933930400907,14.2692900046865,-3.18561111362506,10.4087048554138,3.77554647180408,15.8170369844064,2.22511779786275,14.6738227638015,-0.901465467116397,-16.7604708621207,-0.651668428603639,14.4059680995789,9.2294033772844,-12.5101697823064,17.4184586800377,-17.426280146391,5.05283338120363,1.67258858208148,-2.2352070495508,-2.49439076976666,-17.9761502107718,16.7376775395646,8.79808686415993,-11.6127127205241,9.58896387521299,0.614259375566449,1.29394408542557,-1.32625445163692,12.4464633902524,16.048594334689,9.55711796359211,-1.24845211538619,14.2485122683077,14.1230741649444,8.83774829576955,15.8414998896177,18.1796254252336,-12.0142657680521,-5.85742795066542,19.8039687681025,14.2356381590713,-12.3597768573599,-19.5095828495316,11.370482328829,16.7097120707782,-10.3481467503858,17.5443569903107,16.7838508213956,-17.4979397322253,-16.2086696675427,-1.27673547941285,16.4034788498926,-9.71142959076393,-8.25137359370254,-14.8338455410491,8.4099998140794,14.9766472579633,-17.9690356760235,16.6089944916805,-15.1259477866307,-18.4553106420749,20.138651478641,15.8731321364437,-18.7289301428151,18.6225119994419,-1.60319356130381,19.0058014396552,12.6696519109037,-19.3353047884268,-5.28165490662122,7.90722260339127,-1.64623142233308,13.3162789105471,-11.5582763606447,-3.05660753524773,8.6383702549955,-3.19176762395208,-11.1129436308296,16.7462900604341,15.2031185312269,13.5354455967921,14.5437388126716,0.360911927572727,-1.38157104266409,18.392931863343,14.5735525859071,-1.01138202350945,5.13578129156389,15.7360762391944,17.267381228412,15.5273491786175,-3.17754490026488,-9.19494389485269,-9.03417421354525,8.40233366159696,-4.58927213645346,20.8856941629589,6.25946881583429,0.762827368134978,19.6809767643837,18.0899668568342,4.72906544750177,-10.993287349382,-13.9248106263865,6.96041971373793,15.1737701506904,-14.1175660737513,0.891720192916639,0.515578437033615,12.8907920132349,2.40182642167669,-4.26564994701717,13.5109105393715,1.68161762133074,-0.102606395166109,15.7416421600608,17.0735040655409,5.37112495912424,-2.06419947190856,9.50598069363063,7.8591359210575,8.16061450027677,2.5389377327347,10.0588314859113,16.3241975179064,14.1493181368791,8.9689187411779,4.84202946466112,-19.9869450071342,-8.99565672630343,6.64267618933627,11.305339668824,16.8164087320148,10.2335370628633,1.18051060483815,-8.91174989068111,9.279526665975,-14.9626391890019,-3.54743925160024,-14.4668476657397,-15.5825932740027,9.80249637160744,11.536734485222,-18.158195997143,8.65504572529829,14.5665427077952,16.9890147914619,14.7465324701929,-19.1289619938819,9.46403491185392,-18.6675922727859,-8.69350076898722,19.0108274353206,11.2801540572842,-17.3401808479988,-3.04146782020303,7.60749853195372,7.02783548494267,-11.8650018589081,1.10819839262213,13.1853502798213,-0.81137001408533,-0.666704095578504,-1.17910134422225,-16.5468362602718,3.60282661385323,14.2132971369446,-17.683520614018,-15.3978288470131,19.4702393886969,-10.4696816694967,1.13322845796472,15.0638196519836,-12.910607967393,15.3837084614033,9.11368801560278,10.9886534986002,3.66309055326045,-1.66960029582586,-2.29142586827462,-3.76804304592304,8.20451639434588,15.2264758981547,16.6244402265399,3.87545633225048,-1.10973349003385,5.39761494610955,17.1692346844561,-1.17549876120324,4.96799123093356,-2.22814040129757,-20.3913156583966,-9.46695192809555,-2.77443811816909,-3.97399573621483,15.4875875599323,19.5626047152569,-7.97550114439175,1.93839121081026,19.4011905787773,14.4293790519012,16.4376369157374,-13.7781594894936,6.53313772857693,-3.26005927841387,8.35391776239691,16.6080078493928,-2.28365750710105,-15.8687204609801,-3.00515673429595,-2.17465273551824,-13.0225396773461,2.29235424629154,15.8058080252767,19.9674226884741,-12.1839988688698,-17.1261885940117,8.82661517280576,-8.5550739907339,17.296493292508,-11.3559590810084,-20.7727192209101,-0.429788572160804,14.217692337805,-5.19240840581989,13.0449999445833,-18.8321260165657,-11.8762651921903,17.0389882456598,20.4094249330553,15.3371273469079,11.1986964073298,-17.2008678425196,19.0797795082565,10.8855094569571,2.40168209896513,19.9168014105335,17.2553117514865,17.8522171780002,-20.2222723962623,-9.53046464406806,-16.4295558483845,-1.15391471582,9.34542881783022,3.26312313326077,-2.63711089177538,14.5961155339532,-11.8841138816897,-9.90787752099693,15.3221247031161,7.31537632606183,10.2752625177923,-11.9159128882196,0.449192868289589,17.0352734523319,2.90349567418807,0.307001981217754,14.6684309469985,8.22109747786773,13.8811626967834,6.30616274348156,13.6188142248342,-11.7549944345284,4.41789300211265,16.1180147701378,14.1288159298647,7.69904059068943,14.3379385868962,11.4336450879225,-3.81625715860506,-9.12035471944441,8.84925969324289,18.2390604000628,15.1147434875783,-17.0456362432215,3.53066784936364,-2.55730794272695,19.3574124561834,8.69597319814983,-17.2702472951602,-16.0661283779956,15.8969573833719,-4.43751700619594,-3.49781214597606,7.13597358392323,4.17892692005263,-9.83006458092438,-14.236024260158,2.40572730383288,20.0030711636918,16.2443050712752,-15.7201539039815,13.5894964716118,-12.8385728515398,-18.3570873952702,-4.57778255173067,-18.2581302472305,-21.7217644117845,17.3272879784995,18.4834366614735,-4.5668838163869,18.9812725004085,16.5279336265263,3.1536463570625,7.27988690003319,-9.53824479915129,-2.41001757364162,-10.9992408413667,15.1156199886862,6.34218200942106,-1.74759771989044,7.68908871112614,-17.0672303327482,3.82200928693169,3.32164201229202,-13.1113079713175,-9.6629663018304,15.1507528644543,2.07699543867406,7.17151566870513,2.65972548383651,-2.55222020811325,18.2618196571966,-7.06106249781447,7.36671558412857,11.5612359416949,-9.21927654693501,-12.8941489491538,-17.0460962517196,-16.8951081129048,6.83076770789913,19.0252425970689,7.79857006571157,-20.8580407225331,13.3705204996846,-1.97990681578481,-21.4403644740376,-14.5115651918593,-19.2567620295311,-17.1829381801626,5.30579219982831,5.80812981762389,16.2899141269879,20.4593958498046,3.30394972968619,3.72562259841015,-8.33421125611347,-16.6507969656835,-19.4690256282487,0.424972557420911,1.85736699791352,-21.1430359438421,19.333101311897,15.7473602132762,3.29249246741572,17.8121463279215,0.900072044157013,-20.7552528920612,10.5209344628855,-21.5553368657892,14.2816457511048,-10.8321394728403,9.55760865254239,0.729087732189361,1.26883155784994,-0.59765103640581,-8.05603760278139,19.4229032231311,19.6025498787028,2.57474606672288,-4.53930499474173,0.388110576501327,20.6560793611693,-10.4353293429493,7.65317442890324,1.43240249699367,3.27336207931473,14.2276587589568,17.3190151236262,6.24823848686261,-2.07468389769746,6.60447381677595,-16.2719502000009,-18.0745780166978,2.16800203691403,18.8098324126826,-2.15843873761192,-22.2326126603977,0.174082379051333,-2.96894671984182,19.6535331394537,-2.42572579917431,-3.69369933084895,-15.9549500573829,15.2822632686775,2.19443563309576,-0.00916421248447897,15.0330020505199,-1.71532678013858,-1.7064764118561,-11.6270438601738,-16.9034282195424,-9.26454032177326,-0.480694220675785,8.23138940430177,-1.77228094960533,-19.8024799628438,-18.9372776792985,-2.23457686316135,-12.0589796861826,14.4567313462072,10.1100635264806,18.705083449952,-1.8053801197738,14.1580909399068,12.643752485126,19.9873244032206,-16.9324782556308,10.4545531705857,-17.9187179635911,-17.3454899940387,15.2696328957819,-19.5107780160869,11.8499872722199,0.0751085382410047,20.8623405564885,-14.621980999277,-13.5699642051154,1.96253890591315,15.8398235424118,-15.0255909763955,-9.10123590695494,-9.73111640168443,8.38491059501091,15.2469748983534,12.1729888611388,-2.19371414743663,5.938134576989,-0.796182141560785,-14.6205747758983,2.69190214953539,-20.0029949159605,-0.161806611707093,-18.2861464708817,-1.78372279938135,19.6900701487769,0.57290472349682,11.8143424471561,-2.7505665508003,-13.3960900834486,15.7170308985772,-13.6680830285086,16.506365408677,6.74678884926463,-16.2877508943129,20.4588207905883,9.87846066161024,9.96469233767841,-2.51598940114582,-0.979431479890049,19.2478578288578,-10.9087608052963,19.3903194869668,-9.63609447203665,2.09166768531411,19.8408129340323,9.69023832779171,7.05596320797161,13.8621824686197,-2.28899041931767,14.9904019068918,-19.3186243401973,-18.9733055199059,9.19114174801527,-20.1638671218294,13.9522958245874,-3.12013199725636,12.9979427134782,-2.55646438981654,1.5771938578123,1.92678821446162,-18.084864146926,-7.35741387977289,-9.33568939974354,19.6389851756646,18.4335016408565,-17.73814012697,-2.02843902686987,-20.09545798639,-18.9591752914419,-17.6651736210257,18.7170765163384,4.34718714154463,-14.565310362859,18.711219812327,-14.6443703013033,6.20476625038319,8.97853137326519,4.48777763912365,-1.75226295272448,-9.51722570815222,16.4318376558537,17.0472460909636,16.6180218811418,-1.10887933425826,14.5593694794081,-16.1134538800913,-0.352991532538151,17.4120864341467,-15.6098793958951,-0.041668469963273,-20.2878324195875,-0.621690368820052,-16.6772641362291,-17.7450575961841,8.12929424612562,-10.6215285316348,0.353813545611738,16.2256897254895,19.5961743419478,17.6263672434147,9.38916201681057,14.0128251449006,1.22949770587053,-18.9741262950951,8.70010978178972,-1.61829291509501,-0.988706838742032,-10.4673041301781,16.6716515651436,13.2365923454686,0.430013306399646,0.27495061153459,13.5491427369848,-9.40492819783824,3.03575397766656,8.1308741111949,17.9385420330635,13.3393570161314,-14.3683485880362,-13.5762293354778,9.89487617730381,-18.8756870182492,-0.153984270042586,-8.96438283106515,16.6675799038749,9.53774164368478,1.47270080371427,-20.9785748877748,-17.0350385879842,-17.5296719644304,15.2483850673073,19.9798562679143,-9.5237715986463,14.4700228293009,19.9161407097141,0.959397204237898,18.8625751413655,14.7695880933912,5.63845411616412,7.74390247570908,13.2540381426418,15.3659450420157,-12.9542861529085,9.4065629238653,19.7161887993213,11.6769491717758,-1.60534684618136,18.5505570505702,-1.93733020555345,11.8261834124673,14.9931209836848,-17.1106969499494,-1.45495527104816,9.93978974911447,14.9217098534065,-19.7394980397688,7.8549693892823,7.75663425744448,20.9422658748666,20.8019699424477,0.85893574860937,-17.8729513506339,10.3764259158204,-1.78500155199551,15.4485988274013,-3.97561722291679,-1.87500774338994,5.09878374110306,5.2179290157884,-2.09639029852204,-19.1254647914821,7.46864481741188,-11.1960305529515,8.4935616554094,-14.2523828714132,-19.9761762745815,-8.60175717835101,-17.1630731728224,-16.7201291384846,-6.91464845314963,10.9192029630302,8.72416905753062,-9.38851296960255,3.2192728128208,-12.3710648628355,-13.5123755779893,8.19577274135588,-10.1895330746395,18.5368910430836,18.4951535750688,-4.05391199324103,19.4818918813363,-12.4440204571499,-20.0692261361627,0.706705250713441,-0.704858315696957,-18.6350291618185,-5.54863086101619,13.8673574724741,15.0816942906005,1.23253475253951,13.3774146163719,-0.255135073963282,4.08779531461649,0.629557551379091,6.31263917320316,0.375907984189696,17.3177989775091,-9.39244761227013,19.5266656043227,-22.2877749722028,2.9931788056219,2.49161762964059,-12.5111852310286,1.42014864874895,9.02740869929596,14.3018852512707,7.87077030986005,-14.7496723691055,-12.1066821531919,10.9729051559684,5.86241827080051,-13.5215021294372,13.0355403767206,-3.80915641184492,18.0108214051236,-13.1375062122689,-17.4298983777903,18.9157277221578,13.6831366928957,20.7278460559288,-17.2361633156262,1.84826284862035,0.432035640654324,8.59737999131398,-18.1455846950282,2.45760126095541,-15.7674047738025,18.5702610298533,-15.9598464307772,7.35768775205052,-8.75022627142851,-18.1853231679671,7.88150534900053,-11.576337948329,3.40512300979407,18.5448930772796,-12.4866655906516,-18.9156175913494,-17.0130616029137,8.17315836275974,10.3860841964863,1.9142555341536,-19.4830231764535,18.0886180483037,15.2140264091526,-19.1755068282761,17.9325299986165,2.21423069629377,-12.5964368690418,-14.3491769022764,-2.47638031500891,-9.10398558954643,-5.95706679238743,-10.1744001789963,-15.4702347090409,-10.4323103092065,16.4843193962816,17.6047038748644,13.4644105585632,9.02122351543075,-2.19592475909374,-18.8947934457866,11.8381243113958,1.7151862558127,-8.67570683478086,-11.534044300752,9.08556760702293,14.9377623573207,18.4683314773426,9.63513491095147,1.31401915324597,-1.271225224239,15.397118521621,16.1527441846277,15.4629089159033,-10.0877367905372,-19.5328497348615,8.15232419539049,8.12238751713012,-0.205833274392683,-4.84138526244564,-2.77894224395762,-13.9448332729583,10.1871378895179,-5.61887869764122,-21.3542540641275,-4.8646317704411,10.0490569905279,-3.78895015051356,-17.1938085156693,-9.6329623717877,9.43974543889997,-10.4082211497873,16.1497195428697,-17.7882167027861,7.54066807039838,-0.408065406356117,-8.40547288859975,2.64177130706061,-7.97186409424854,13.359357562061,-17.9287331070032,-3.93616855691398,-15.8960856219169,20.577686134897,13.9427721977091,6.15354020451476,-5.09228915686128,1.16633154919136,-17.2201516170362,-2.74347047940198,11.0529867657383,0.803549222062129,-4.84176821865319,-13.6239142359406,13.8178430857289,-17.4693556163821,-9.03507213606613,-15.4892150185873,-6.53126428249449,-16.7953189616291,16.34413616629,-10.0671349499422,-5.87789549981324,9.02832461999755,5.64875186058826,-3.92103396030346,-19.9757099484824,11.5119662435307,-11.0827962482735,-20.4291824270241,6.13341827723552,16.5062183118915,-11.7302757271135,-0.578599471146989,-5.05305668594502,5.81965401449357,-2.67231847003062,-14.2825822772759,7.10698253044451,-17.8438621769918,0.247612054058772,-10.8695012639889,7.84571768660693,15.0760936521291,-20.9896925695329,11.8489908805816,-20.9762946461049,11.3572062431146,6.91518816788796,13.7032299605268,-17.5751442795026,-3.33281966814274,11.3488702383265,-12.1253005943129,-1.03986794129841,19.2839456431107,17.6928660283789,-14.1593242890703,-2.29093196191276,-2.93440806063435,-11.9658361061073,6.50239323778827,18.9046768716738,-0.105132073970499,18.7564823711172,13.9330544270098,-18.1707284336896,-14.8785073119708,17.987779310703,-2.69247151637152,-8.41990274471359,17.7478670478794,-0.72170821221358,-17.9125342962248,17.3930206723904,0.839712519444847,12.1770556828123,-14.5088592050095,-9.41808655186277,15.7704811143105,17.0577410238385,-4.31006002406029,11.3179990400539,15.0064872579722,-7.00072914645733,11.8694209390078,-19.9074083254161,-2.14150761576518,0.671524174425753,-5.62430408468967,15.8281324761621,-2.54609926546576,8.23792530574266,-18.9987209218952,17.559650939614,0.0391167638735266,-17.0451463956825,-18.7873616696664,7.17743331285697,18.7830821145138,-16.5348266863965,21.1871887619488,-1.37783388861623,4.39999947594913,-1.06905161418157,0.00537903355749417,9.83466669189674,19.7337454827457,8.78045774583373,-19.7092401521396,16.9509332803418,-14.767790923637,15.2192815999903,-16.116060876017,7.55614040969096,-17.5291587337506,13.5453902298381,15.4296801221977,4.55327606829284,17.1853626664548,-10.2110878965165,-1.1557859858247,-11.1687145813132,16.8557798760247,-1.1893675458757,-18.3156070345885,6.46004018607397,6.57408912951245,-8.63285797086896,16.6770500324721,-15.3413750858785,16.3645765587824,5.54272545528201,-17.1014383596195,20.5682045889648,8.00790936893595,3.50373413625571,6.67187709635809,5.15603422399348,-18.9273024416562,11.8649366522599,5.83506239276883,-16.2228319227006,-9.46712003146065,-20.175934174087,4.8708977670112,-13.3430000735059,-16.6237041717262,0.510803400526167,19.8292226059539,7.4804232904945,-21.9601089146488,-16.4192070967549,0.715291138696762,-9.26215837807002,-16.9938542258804,-17.1268705713455,15.6054871165768,-14.4620055879497,-0.986028339001036,10.0658485168423,-3.30425427896296,-3.80908961717448,-0.577962518282289,18.6995134848471,-16.113056527219,13.8014950290531,-13.6759044528117,-20.4561685707944,3.9512791602266,5.41351864330652,-3.47403412902778,13.2022293470373,-2.60907998598067,15.8630352320114,-17.6657982103591,3.06069548794654,-20.1797265139826],"type":"scatter","mode":"markers","marker":{"color":"rgba(31,119,180,1)","line":{"color":"rgba(31,119,180,1)"}},"error_y":{"color":"rgba(31,119,180,1)"},"error_x":{"color":"rgba(31,119,180,1)"},"line":{"color":"rgba(31,119,180,1)"},"xaxis":"x","yaxis":"y","frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb202-1" title="1">map<span class="op">$</span>cell_type[subset_treg<span class="op">$</span><span class="st">`</span><span class="dt">Thymus.T cell</span><span class="st">`</span>] &lt;-<span class="st"> &quot;Effector T Regulatory Cell&quot;</span></a>
<a class="sourceLine" id="cb202-2" title="2">map<span class="op">$</span>cell_type[subset_tmem<span class="op">$</span><span class="st">`</span><span class="dt">Thymus.T cell</span><span class="st">`</span>] &lt;-<span class="st"> &quot;Effector Memory T Cell&quot;</span></a>
<a class="sourceLine" id="cb202-3" title="3"></a>
<a class="sourceLine" id="cb202-4" title="4"><span class="kw">plot_ly</span>(map, <span class="dt">x =</span> <span class="op">~</span>X1 , <span class="dt">y =</span> <span class="op">~</span>X2, <span class="dt">type=</span><span class="st">&quot;scatter&quot;</span>, <span class="dt">color =</span> <span class="op">~</span>cell_type)</a></code></pre></div>
<pre><code>## No scatter mode specifed:
##   Setting the mode to markers
##   Read more about this attribute -&gt; https://plot.ly/r/reference/#scatter-mode</code></pre>
<div id="htmlwidget-678ea4eaf6ac0e87eab0" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-678ea4eaf6ac0e87eab0">{"x":{"visdat":{"9ad86dec688a":["function () ","plotlyVisDat"]},"cur_data":"9ad86dec688a","attrs":{"9ad86dec688a":{"x":{},"y":{},"color":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter"}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"xaxis":{"domain":[0,1],"automargin":true,"title":"X1"},"yaxis":{"domain":[0,1],"automargin":true,"title":"X2"},"hovermode":"closest","showlegend":true},"source":"A","config":{"showSendToCloud":false},"data":[{"x":[-0.64700797896146,10.9829194193465,-0.683169312033188,0.121598337692018,2.38673232083063,-11.2077529755415,1.95983387495633,1.355658058858,2.28442859883303,3.16351677858245,11.2807423417454,5.52011162055027,11.4783367949828,-1.09254313923237,7.81987178814859,11.3704066013725,8.49868714453964,6.75622092439215,-2.17811482875469,6.44083015176533,4.58506236787563,4.36284980705886,3.40514142870881,10.8994791252426,3.34521572800305,-0.264389549753781,-2.34287276102339,-8.59900198432096,10.9314522568183,11.000742604909,4.09323862442843,0.00106162044567765,-1.15749993922686,6.62847044191489,4.13166737628574,2.84976094903702,4.83917615535764,12.187307821636,11.8086040828073,6.55723374404436,7.0424681444496,7.65009438485869,5.30907608211705,10.3024174940888,12.2124896762216,6.40845826135662,10.6968203170107,10.9361793741712,4.75953339124891,6.57113355494872,10.058247357314,10.9208742224803,4.71517557420859,6.78400552552356,11.6657448844689,3.85762682711509,0.494066188198416,6.5801785461178,-0.688165968947004,11.2278961065549,11.0142674390149,-2.10098875565224,0.425374376100636,11.1926430953936,-1.102078458571,-1.45661849552363,3.92410129409312,2.07568337915508,9.29157384710789,-1.91146276115677,10.7020156677305,0.812936146985994,3.24280819380646,11.6066917033058,2.36136333003448,6.83611484977029,4.67244286208184,4.0697905092024,-1.43511258056262,11.1920356560971,2.32879903511123,-15.3017457132328,4.58491410596088,-1.37408953470131,4.58295220662718,-0.105397174020197,0.486357037986709,1.03991644902279,7.70220668792144,9.11570175777651,-0.0934993254164325,-0.785458781544937,1.39113719941521,5.11656436238653,10.8440180659292,7.37021089235498,1.94484055193084,0.48410980485091,2.72718712194514,4.18832517656818,10.9771942267818,6.92733062346478,8.22544958377789,11.5338033754264,12.1847838412572,-15.2736902813511,11.2687716473871,4.46726215723052,7.58436483303276,10.8374117598183,3.08052478431283,7.08924980892202,3.28560379927315,9.29156576981388,11.7863449304956,11.0706423843404,-1.71095371214338,10.2917018827769,10.3161270180132,9.25038427002594,5.30844999186754,9.56295717864335,4.45190660277448,-1.48481746792535,-2.74809653318622,4.37324345918158,11.1944363981471],"y":[-15.9673377599113,8.48695469183638,-13.0574160987575,-18.4203719385579,-19.803975252085,-4.92644528565067,-13.4694097102671,-12.5317104441329,-16.1928973513659,-17.7188971162668,-17.3027700684193,-18.3370279971478,-16.8881379685924,-13.0262221327575,6.63700726756897,-17.3789860517129,5.91677917150417,-9.05011487704128,-12.8488573059695,-7.00944800627085,-17.316741005132,-19.1028167013768,-18.8549436183589,-16.8913154882993,-20.9232278108037,-16.4891007910443,-11.4932606569854,-4.41454406033073,-17.242088178452,-16.7604708621207,-11.6127127205241,-17.4979397322253,-16.2086696675427,-9.03417421354525,-19.9869450071342,-14.4668476657397,-19.1289619938819,-17.3401808479988,-17.683520614018,-12.910607967393,-9.46695192809555,6.53313772857693,-11.8762651921903,7.69904059068943,-17.0456362432215,-17.2702472951602,-16.0661283779956,-15.7201539039815,-18.3570873952702,-9.53824479915129,7.68908871112614,-17.0672303327482,-13.1113079713175,-9.21927654693501,-17.0460962517196,-19.2567620295311,-16.6507969656835,-19.4690256282487,-16.2719502000009,-18.0745780166978,-15.9549500573829,-16.9034282195424,-18.9372776792985,11.8499872722199,-15.0255909763955,-14.6205747758983,-20.0029949159605,-18.2861464708817,-1.78372279938135,-16.2877508943129,-17.73814012697,-20.09545798639,-18.9591752914419,-17.6651736210257,-14.6443703013033,-9.51722570815222,16.4318376558537,-16.1134538800913,-15.6098793958951,-16.6772641362291,-17.7450575961841,1.22949770587053,-18.9741262950951,-14.3683485880362,-18.8756870182492,-17.0350385879842,-12.9542861529085,-17.1106969499494,-17.8729513506339,-2.09639029852204,-12.4440204571499,-14.7496723691055,-12.1066821531919,-13.1375062122689,-17.4298983777903,7.35768775205052,-18.9156175913494,-12.5964368690418,-19.5328497348615,-21.3542540641275,-17.1938085156693,-9.6329623717877,9.43974543889997,-17.9287331070032,-17.2201516170362,-4.84176821865319,-16.7953189616291,-11.0827962482735,6.13341827723552,-17.5751442795026,-18.1707284336896,-17.9125342962248,-19.9074083254161,-2.14150761576518,-18.7873616696664,-16.5348266863965,-14.767790923637,-17.5291587337506,8.00790936893595,-16.2228319227006,-20.175934174087,7.4804232904945,-14.4620055879497,-16.113056527219,-13.6759044528117,13.2022293470373,-17.6657982103591],"type":"scatter","mode":"markers","name":"Effector Memory T Cell","marker":{"color":"rgba(102,194,165,1)","line":{"color":"rgba(102,194,165,1)"}},"textfont":{"color":"rgba(102,194,165,1)"},"error_y":{"color":"rgba(102,194,165,1)"},"error_x":{"color":"rgba(102,194,165,1)"},"line":{"color":"rgba(102,194,165,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[9.05700339075373,1.31582770457569,7.71129866826428,9.64284253210722,-2.93054633868947,8.01488449123776,1.73017234445989,8.8512205635191,-10.6085782363227,3.00191378192751,10.2663589499106,9.60754861635615,6.36554730667016,2.78085260040477,1.60518134357344,7.08497284512453,-9.93701471824346,-3.73724777055028,-10.9464800679939,5.83027605241538,9.74802910763305,7.23375632097735,10.3147756943685,2.63537008716921,9.87969503104111,2.9947648967973,0.851096695143485,2.95982330040759,5.5357724584898,-4.36001595253798,-4.5995400684101,10.0048414242583,10.2834341290521,6.64017497081236],"y":[9.42016369662749,16.2598662505788,6.16073009561685,-2.40995216662941,17.0802490611973,6.0884210026157,13.0324851488086,9.63182608415293,-1.28370244684747,8.71977138971156,-16.2026383984467,5.62077466672877,6.37329153855569,-16.3254510507658,16.1117903310734,13.2111800958681,-5.28165490662122,15.5273491786175,0.307001981217754,6.30616274348156,-2.41001757364162,6.83076770789913,7.79857006571157,-20.8580407225331,6.24823848686261,-18.9733055199059,-18.084864146926,-20.2878324195875,9.38916201681057,9.53774164368478,9.02122351543075,11.8694209390078,8.23792530574266,6.67187709635809],"type":"scatter","mode":"markers","name":"Effector T Regulatory Cell","marker":{"color":"rgba(252,141,98,1)","line":{"color":"rgba(252,141,98,1)"}},"textfont":{"color":"rgba(252,141,98,1)"},"error_y":{"color":"rgba(252,141,98,1)"},"error_x":{"color":"rgba(252,141,98,1)"},"line":{"color":"rgba(252,141,98,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[9.82891143936811,-3.00196647225079,-16.2636878973757,5.11026001190674,-7.96327259272952,9.8278810873859,8.23099598692419,-14.1328068830468,7.55202175312812,5.58212148331875,-0.968089213823785,-0.691558447469152,4.22047401496733,9.13347093938178,-13.2389163454206,12.7275993677326,-13.385212780602,-11.6767711905462,3.90796220999321,-10.1058977187402,-15.4974877269412,-11.7262439499746,-8.96469744886844,-3.42860593124477,7.33006942842282,-11.7174038942314,3.93814303620079,7.46224187040622,-2.52271126038625,-15.9687795641159,-8.57235426903485,-9.21562151777873,-8.26416314323915,-4.84601253223187,11.3290849681427,1.95401739442417,-12.0106767257361,-6.99277813008846,10.1993777742303,-0.925864772883279,-0.0294976427393101,-1.62499371485408,-3.89113272565772,-3.37191805685573,-6.06683634915917,7.44489014696503,11.8506979167224,-16.1165585028952,7.75200687218933,9.13599607227886,-3.9052052980904,9.42711905621988,-11.5764456740183,-2.30986632161948,-2.80101344559873,-8.47418423574407,3.60594209112652,-12.818035489478,6.12061932001222,-4.93997138252942,-10.0709655747583,-11.3080798752615,-11.7615114398318,-9.6511354498064,7.87559868152668,-1.59995583539679,-12.1076158938546,10.6185783604724,-4.27072367504397,-6.6289221469232,-3.93109620627119,4.37209016342535,7.78196798797721,-11.9993497926742,-5.40163299673636,0.844356426968267,7.25149832125925,-14.549257814265,-2.84591581148156,7.75011498084225,-2.40760105455274,11.8563972429839,-6.89775579946556,-4.43890490045562,-10.4567779527525,4.1944754345686,-12.2686664262743,-17.9774965905025,3.33098043326833,2.25920179307337,-11.6729610313437,9.81837505506589,-4.51042888834907,-14.460521305944,9.719407141031,1.16614902939685,-2.20399706923565,-9.74109013296507,2.51685241926093,-5.74660583936413,6.03341468458895,-1.48333881074084,5.14639604945107,-12.0479781225904,9.44862150446,-3.02307650368023,9.51015296043139,1.76512017529795,-2.0101180061106,3.11848686773141,-5.17226368049842,3.4528729681505,-11.2053652896679,-3.17230186705166,5.23743507166201,-10.557885415587,5.36297744476268,-2.52871812581956,2.20963427724461,0.312493109546483,6.59956451840895,7.46740832723688,1.91010262260896,-1.98629965354965,-12.9475085909362,-0.787555785873725,6.73115074510438,-2.27636621399645,2.97792546441982,0.960353809208676,8.1304638915541,8.4836950741601,-10.7166417232331,9.60213585682176,4.05664331570355,-12.7630804328154,-0.679192258286923,11.6507763440007,-1.16253798832006,4.43155910718925,-16.6155095179111,7.21174356368151,-2.08320304186372,-12.9901107730417,-3.2495720438173,6.69384435463233,-5.06207605418766,0.749349394627608,-4.04566337259661,-0.809704681726989,-9.59505729460747,10.4501732061606,-16.305104514804,-11.6444120095217,8.81181963063932,-1.77739152145807,-12.6968328803359,-10.7049809306755,-12.4882943828428,10.2884510063357,0.371997644469344,-15.1391489673992,8.81894416698965,-1.46010378954201,-10.1755845010003,-1.35850404432497,-5.14819950373142,-8.41251184644206,3.4914738601816,-9.9394576287063,3.64380835425618,-4.07982903761715,3.19201618620682,-13.0337361922979,7.70603774293683,-0.533564774166623,5.50025867192459,-15.5505999207836,-17.3404843193353,-1.28459518800085,-5.08459841534678,-0.889907541794594,2.89131505992518,9.83111307936576,-13.3934892257984,-8.18246290564943,1.05871145735768,-6.83681856140259,8.72685719935613,6.6019429208772,-3.93993006200646,-7.33109921881688,-13.8246756280819,-4.89727184809966,-16.8383949538528,10.7346360732132,5.18740054083238,4.78186902636418,-10.7771865434061,9.26369926671915,-11.1981176072996,-2.2351493628046,1.54855336118919,4.2102841265481,5.14763157067476,4.89652319571709,7.09691582977028,-7.0005755478137,2.92429166154218,5.75355179555096,-1.94018351402956,9.94334428389341,9.91861370883366,6.4350254255442,-8.31775825570102,-1.10999772469839,6.80639392821921,4.72908710375811,2.39648035112287,-6.47837473384119,8.11723377952883,9.01185872985275,-3.47195785338812,1.42258535179755,-12.8995307978385,12.928799632771,-3.85517979721668,-2.64739120566826,-3.15581452052875,-10.9004231645292,6.84044166266121,-8.60348685857969,1.89243982929869,1.16177013190608,5.16078907281793,-3.34341606459096,7.67993241909487,-11.5993243747319,7.70232886715235,7.21905228203758,-10.8316024258295,1.98342671662567,-5.24395941561369,1.42276917088958,-3.46942869890395,-0.160396590088969,-0.52973552678463,5.67038227928319,7.11647437555116,-13.447417763681,-7.67537730573923,-10.2679583593352,-1.94807156536674,-13.9876122953144,-14.8087686026045,-1.95394043338005,-5.77755944378688,9.64487625869794,-15.0491119991027,-3.23870455928025,7.01981910268452,-15.5246449826532,6.2136957990831,-8.88722627972841,-4.25695055555794,-14.8940624230048,5.04888619943999,9.00091945343663,-0.419742225070688,-9.25907816321094,-0.0389637447942507,-1.35395065230248,9.20105750422091,7.46927957871424,-3.31507008261641,0.572508261887164,2.23124919368925,6.03360917359753,11.2146005924885,-4.66441568165984,-15.2736422733279,-9.15377752041512,5.12904382335145,4.0882266665905,0.461831314799731,3.97482665515735,10.8519664180908,12.3010948523928,-1.50861297602589,9.40180723432858,-10.1267842781459,-14.3254330087046,-11.4493864145288,1.12529842400453,7.52782449746136,8.51973455277696,3.75371496161553,3.28753576627298,3.23157223048133,2.96246901198722,6.80137599040706,7.78501145263659,-12.6570833769504,-3.1063261082921,-11.4411946732459,-12.9362281851671,-12.3555114540239,-0.28333411748761,0.755350387541639,-3.9649969760844,3.76756782740247,6.77036626743522,-11.8324950698558,-8.37539909528126,0.679085042846249,-14.9601217324881,-5.1818116282347,-0.0942536726845936,3.56548195374302,-12.2926679294128,-6.12910432435701,-9.918482383013,7.09903538302189,2.16008718644457,-8.10410403638876,-16.7532427217557,4.41264354137519,6.36600922795468,3.39611828225795,-11.7250008361746,-9.05975337407891,5.0792954776135,-8.26470094678932,-12.3290026309716,-12.3810163075483,-2.37752360481419,-4.45136544309075,8.14355923049767,-10.189806344882,8.35094830303059,9.26690983712412,9.70043224865771,3.83264269064231,-13.6943386504416,5.63531880052481,8.0839711172369,-9.38215899273452,10.7606295555205,10.466995043177,-12.4160028951189,-1.00825621425155,6.55730973409536,11.8634528474964,-7.16235345508303,-9.87857700705185,-11.8643616669115,-0.0655645728983408,3.95909449379202,-6.66110477582011,3.6124314740147,6.06000611014882,-6.26920877808011,8.70414236241388,-12.3708513107692,-3.07606905859856,-11.0075685844169,4.0836704198752,-12.1123681950746,-3.88244070834873,-10.3426964760793,-15.8829612504354,2.54520334250744,4.67677432635109,-3.56405154453272,0.565117438569289,11.9748153156141,-10.6201297025315,-11.5121392495465,9.56972139586113,9.38493519816435,0.0498449281812405,1.98049346977713,3.33354445117847,4.49402228006215,-7.96186224336824,-12.6564613966401,-11.2428640140026,10.71085448709,5.309910826302,-4.1912008612685,-6.00060207984623,11.2320287050076,11.4479800255823,7.90371451438669,1.07785797462281,2.50537639501785,0.72359120657806,-14.4541637533114,1.39585806479195,7.00338781258431,2.50753719573538,4.84587729618694,-3.73320579678222,6.97267260261632,-1.55629665582251,-3.37079325635419,7.27811566716854,-13.6135104878247,-3.4482749963,-2.01995382347535,-2.14611337848434,-4.09072615262254,-6.15456234787786,11.6116225418866,3.56318558404695,-0.327441989918096,-3.3290997782366,3.08861359082291,0.20021924189669,-3.55698669208339,2.0445313741246,4.85878957851101,-10.7368285093069,6.54247510900478,10.3738336790499,0.916300211224866,4.45474755223373,-11.0742695809191,9.6909220236751,-5.14179262305946,-9.13925256040715,2.70019371264956,-13.8722650225966,-3.06348763310777,-2.27084700771599,-0.00480882368855796,2.95134655502424,8.89090405719254,-14.8214364838395,-7.21582354574512,-1.18710806271312,-4.64800975333439,-14.4103146911509,9.18076126642862,-2.38016115252994,-0.446759142144465,-9.55133364613346,-0.0984271813938266,3.19063506532798,-14.3846558359694,3.11361846883912,-9.59222540963671,-14.6879580193745,5.14681015456595,2.85735569271041,-6.7935056372843,-2.40100686752974,3.99181744517366,7.20908365649441,9.65398552292046,0.335664680685584,-12.3288088475584,-13.1940751569452,8.92004905480146,-15.9497816728839,-10.1828289522161,4.89453888547524,-13.4531697559213,-9.80858728741571,8.44650653952169,-2.81820794831571,8.12187068293113,-11.5662906299865,4.7455467329387,7.86012121985834,10.2731104021915,-11.6384853997638,-3.61926738524991,8.75423568781965,8.73415503739917,1.77214723493637,-13.0747956341718,0.0970966036085597,-5.28553097863298,8.72958344929029,5.01480765578096,7.13305932617364,-6.90367885550704,-1.90166662639275,-0.0785130993241402,-2.51546247466487,-11.3664861845708,1.84139901284619,-3.25399720559288,-4.82246099656796,4.26384066272578,3.59839926693447,7.59098262208168,3.22524360992331,10.6337403406005,3.46964466644503,8.20869534662265,-1.85838165159572,-0.471454226324056,-3.56385103464068,-15.6470803289408,5.04855201904907,6.79406144720244,-3.43959168860988,-6.96098368540143,7.04638677988106,-8.36890024646393,-9.94027992213227,-10.0738783419843,10.9618017977764,-11.5686983386592,10.2245565289438,-0.041511442884092,0.0171818410957833,-1.4877167314915,-17.4798777156874,2.5099724083714,0.223102107868022,7.67234711509862,-3.830453305316,-15.0906908203436,-11.0437018224022,8.85760497580631,-13.793000506398,-5.42770628760829,7.58952885387356,-3.68547193382706,-15.5563004823308,-11.221441402074,-7.03066717707251,-0.423337520241322,8.841301385236,-6.90208983428567,-11.3195604595538,7.06830473483865,-10.3539460145979,-11.4771218969258,11.2932943684557,-1.35557713695463,-4.18602739352296,-10.0751465490147,1.8788847399775,-4.10020933548653,8.26116887242672,1.87384805832659,-11.6643611778711,7.49096581669704,6.89886624881443,-14.0038814165246,-3.07178865739004,-7.53119984941236,-9.23371108437883,0.129995946427896,-11.5931867703794,2.48109859300615,0.72066770057411,-4.00261296836749,0.0565345794784102,-4.98689691880957,-1.0440663695784,1.06092720810148,-4.32537949856768,5.3075877512733,-11.4554704790414,7.87989838024182,-13.4777994123988,11.2946535504296,4.41543754996927,-3.48331152847722,-1.48866093092278,-1.3940242218004,11.2775353622639,3.33562273664726,6.30098588723641,8.70658523087676,-13.2705672696049,0.766327923268357,7.55707215819368,5.19587118321291,2.31800317440879,-1.29638517874767,-1.1266492514967,-11.4429159185132,1.06045938930989,-10.5373687527947,-12.4923984070613,11.8117442771488,-0.949649808405418,-3.14918213595859,7.47680074253922,9.47983136223432,-3.51315127892931,-4.61340504968856,-9.42293560363734,3.77537212102496,-10.2695216542302,11.1821224324796,9.62687118450968,-1.17121668558615,-3.93403756478549,-5.06708348145476,-9.80520498638217,6.74911722868136,7.15162221475909,7.27480994335103,10.5147485288601,-12.6425818934727,6.74280671691854,0.831822765126827,8.08729861824141,-3.69389188826327,-12.6707253821162,-16.2972092291776,3.77439236409139,2.7236908730811,10.3593393560673,-12.3229140351064,-12.8855862343476,7.19395444800778,-9.37433759057026,-2.61246268326568,-4.32203717191813,-7.49633456691126,-0.339933868028644,1.18671238330189,-3.8372384469258,-2.16840907514794,-8.78464730418214,10.646569505448,6.23702262641971,-2.46278095277826,-3.32608342500181,-11.9994339708017,4.78595728883319,10.1004626161565,-14.0854228157649,7.11563097701212,-3.35268081168093,11.2651997047495,-5.11895583150842,-10.9498110847956,-6.78685441429928,-13.9667144729033,3.19715576239278,7.66434786434726,-14.0155111023421,-6.23364047608626,-10.5932224233214,-13.2109844382228,7.7868944659012,-10.8365314353441,7.57081927345916,10.72335241332,-3.7232025873966,2.65036000281584,3.05988412708265,-3.41732955929093,9.62208983522594,2.4041545341629,-2.82137499922696,10.8093579261054,-6.06474366432108,7.62195715175079,7.59637948642491,6.1109066943915,-16.4797181727674,-7.01528517137479,-2.15360715931152,-11.7738516352146,-8.83953071922091,3.86737733893302,-1.20069981809382,-2.715492437111,-8.21232322755488,2.13932134820078,-8.76068582757213,2.17047392431433,10.4171908216825,5.40644041323777,10.2313878323923,-1.37376226948617,0.672197857752152,-12.5284917213217,-9.72213533062268,-15.0534595981337,-2.79203323676675,-0.795172957832658,6.22536395110988,-13.622218871723,-10.7960227323767,-10.1821594286835,0.130571295687463,1.21353874847712,4.65471198806589,-17.4813038107817,-8.03636110772787,7.00714741638335,2.61557346020311,-6.49582629789168,9.98447415557156,-12.4319361717574,4.98994819641734,-10.2922069562845,4.15105991583381,-6.13060666993313,9.7459362422876,-1.50030216202154,10.112396741618,-6.60131977511768,0.152023997005444,-8.0878030317311,-14.5141855649691,9.3041642813024,-9.61653552327437,-15.4088469327069,-1.46647589381195,-6.10026229643796,-10.7228584739178,3.35477992160217,-8.64448573233908,0.810216909208728,-8.18835435137973,1.47545522992215,8.17655875575261,6.94351802584242,4.99121655893888,-12.2001833095526,8.23983376841392,5.08508410049415,-0.19144892564878,-2.15080996698666,3.05305762261268,3.27855265597793,2.8255953921857,-0.199305439454856,3.07838114462375,-7.95372136419564,4.48744260557881,-3.00441739893663,-2.08236999039003,-7.41668049262339,4.10275902944458,-3.6697836394086,11.9901917760387,6.56274637827125,6.90033151976202,9.83015451386409,-9.68079557349856,7.4701131100295,-11.8108389200304,-14.8622937010971,-11.0635992246399,8.07968803706528,-13.5747680802732,3.20612307421428,9.64472845844451,-3.35022835109835,9.32532694051381,-1.58215914467997,6.81347418097091,1.06436527474142,3.17120508230376,-3.66711732098798,-2.98244750703895,-11.0645368570523,-9.40135268997227,5.55212476112239,-6.16560493452022,3.81873019124052,-1.27030861456813,-11.1959796197472,2.96144423219601,5.20116990332279,-5.22434283445764,6.83515194188371,-11.6664765045506,-1.78128658383716,6.27810516453127,5.6509142931066,2.78277732669555,7.70356117259486,-14.1027219497052,6.58500357982688,9.65950771855894,-14.5457196925591,-13.4435243927939,6.52260281146573,6.90576457774171,6.23752392817671,6.86910783375688,-14.4532693433643,7.42649178348599,-7.20622652350954,-0.550317965567239,-0.611697750365122,-8.13406693871669,1.325187784636,-9.54157969318261,-7.09189932007124,-1.23712731787688,9.18406881853282,-14.0227312612812,9.31155063438081,-13.0109435734237,6.5098992248851,-16.6779390834791,-7.16494029513774,-5.08923495179521,1.5963118888362,-8.97359580462543,-3.60962659757962,-0.61744387483345,-2.85701156634217,2.56357794696911,2.98747824076449,-12.896102287132,-13.9258736873381,2.98671548638628,8.49153238398527,7.08553815402189,-15.9867652858164,-13.6709538394743,-3.65088878315723,3.11953849342822,-13.1839496917453,8.75766847081078,-3.88420442655663,9.64350275686616,-3.59943552623882,7.89382320624922,-14.9878224817506,-3.95301432540554,7.02835741300142,-9.12678758665595,6.21723769253142,4.08111628731198,10.9990127174105,2.01010014786725,-0.878415037347965,10.3845028865434,3.40155729752614,-12.1418086386029,-1.71673740896422,-2.65388907888998,10.4333713655889,10.1158335067172,3.20112670404568,-1.01033115770864,-3.1756773065346,4.22578116202339,-4.40176738214793,9.12595870466929,7.96244806136765,-12.4770167606737,2.87518122474179,-4.23078435556712,-12.5771694231855,8.37341467262872,5.75138764679351,0.660279006477143,7.3323395626482,6.06139339432244,4.13183137810494,2.19975638204812,-10.5504892958463,-2.82072296151232,-14.1858394386605,9.2789493402567,-16.5799941812136,-8.90464766502288,9.21875256640506,9.20421894411271,2.10932795320815,-3.9976735942059,2.10512781362829,9.33830611652755,2.50730663695904,4.58318311997114,6.80932128478773,3.92607583300854,-1.23540773182524,-7.51064271365029,10.5357372122005,-5.5218875911005,0.00976673020773924,-9.89805005605661,-2.23483889784124,-0.832349705887485,7.88706968943287,0.96543954928539,4.97476417641664,0.244998614532728,-9.10127472597457,4.28473610982041,5.22274727155551,-13.0070939544239,-11.8493001484245,1.42854381602167,-12.7399866144193,7.07007002175112,1.24385949720144,-13.4335350527399,7.35756042659514,-12.1372952236963,-14.2731920527005,-8.58410173741731,-7.35729850580197,-14.2132663520284,7.77145318355852,-2.01806303375981,2.11165533581418,3.0771901181428,-8.90779277253557,-16.9509982281864,-3.55023891533676,-14.0097814184945,7.32963773682705,9.118271566155,10.3223779667283,11.9914218289086,-7.40357547883626,-3.21503236779638,-4.90486210994078,-13.6901823001047,1.40596362647126,6.42591343768377,1.98140451255794,-0.259143915494176,1.22751045600258,-12.5335842699494,-10.1679865991152,5.11157833173426,9.92632957743197,-9.70017982570916,3.22030478152524,6.43676328908583,-0.739480456406954,-0.976495840637493,-0.568353632806302,10.2018934153319,-4.22612674752566,-7.44511756534601,5.5549194184107,0.601330692589122,-0.620886096735016,7.99427464095841,-3.93334559288339,-10.8504684645923,2.84081317661909,6.83395463308155,-3.78073461754307,3.78409018387511,-1.30541425020812,-12.3357203394803,6.38971578616204,9.65684994542042,-2.75438339932516,-14.4803163665767,-2.15501430053395,0.465994702172914,3.2285167988725,4.8884276486059,-2.29902687961026,9.16434995883591,9.40110109511421,1.1837201930044,10.4419501548189,-14.8429672895073,-0.564433723719789,-2.05830651686259,-4.35391670968321,11.414165116044,8.5939396898019,-3.7853091020217,-8.46395013637652,-8.06935307815328,8.8385487744995,3.84313804530459,-3.05304359390583,7.08273347127422,8.40531559988768,3.97870544551576,-16.4670822395134,-7.07753069567801,-14.2384247196042,-0.126657088594808,-5.28809161792464,-12.5874350120199,-10.5913232687267,-2.32011198439004,-10.4884599256429,-4.00501773898503,-1.66675170750261,-1.44492856125172,6.59171897248965,-13.7807630629173,-3.34071056786941,-9.92595191234454,-0.871271531215126,4.84447210190695,-7.1683656823241,0.220125787243174,4.18600892926953,-3.66702122765043,-5.8709026100168,-9.61498001990717,-11.1673490949517,-7.50753628267548,-4.18769204949873,-10.2328153946908,-1.29671711423883,9.25233922256825,11.6739934569865,0.0996158734085468,3.96691514485562,-9.12681537118177,-2.81762341296063,-7.24269479141646,-12.8997467944318,1.26425093913796,-11.9080891106839,-9.60075867733841,4.57872269421269,10.100918332724,2.93114678782632,0.971223867035387,2.88436550294712,-12.8614336085156,-8.12749823276476,-13.75374088855,-12.3035540169631,2.32327514643619,7.46189835034588,4.33386843396052,-9.17945031266364,-5.55991360900063,-4.39404665637217,7.36862763331266,4.21658567315317,8.8030100448261,5.54592921961176,11.4502358224436,-4.94818308226012,10.8379170747792,-7.90282113886091,8.68292250505573,-2.51049372240199,-10.2590582809269,0.414952157253194,5.13842097257281,7.88167915832968,9.33209008128285,9.77688304066469,-3.09939171531768,6.96408772603081,4.3295620612701,-15.332191387827,3.46344178929219,2.33018685450472,4.29597182025093,-1.47876778696363,-13.0381138315334,-3.12570613346694,-2.73059870786524,-8.63795891628221,1.69157077101361,-12.1278478231357,8.80881877372487,4.53269780406071,-2.5785420050338,-3.02785550085931,9.03179903908925,-8.30283467289226,-3.84538757753671,10.4030243228389,-10.8869128348097,-11.0365835810929,-9.80976296813444,0.672843607644832,-11.1581879938641,3.52455081731651,1.24233166113872,-12.5760171546041,10.5665549718346,10.1568243344168,-2.36676340444677,0.86883969650285,-11.6771019361012,-5.67916944004976,-13.114908550882,-16.7699164496718,-4.26972676319082,3.46789248006443,-3.71584694999347,6.49582498715062,8.51423657394286,10.5617539414593,11.7707552627807,-4.52441167404108,5.25270585839545,0.789519676614693,-6.46361634924249,-0.561773394102143,7.10202796386736,-15.165033503964,-3.14938720237319,-3.10205777486574,-12.3142986537011,3.92461712474966,-4.35439725062635,7.27740366109258,-2.66093884507689,-2.69122531011063,1.16265003024978,2.54606165930274,-5.33757755794002,10.4707150312347,5.30053421230119,-10.6125172138372,-8.29851165258077,6.6694061446567,-3.46212287384644,-10.8633090396677,-1.98841667427302,-7.84047504370984,0.934151589473817,-3.3548279262894,-11.8910231717265,4.87870037834936,7.64330829935797,3.29250827097822,-14.3121121453113,-2.45711302943458,-1.46618076547506,4.32191941222138,-2.36503119749703,-16.8222330026763,-1.51368749211814,-9.79123124114864,-9.51879561798784,-13.9987247818322,1.27830781675901,2.48925240981731,5.01034774217534,-8.3226926630704,9.65498894519923,-12.4866849083789,10.3136304346567,-4.30743442868922,-11.2332928482887,6.94753258461691],"y":[5.66975964902548,-8.77818950821644,-2.13961178993674,-19.6984183705609,-0.196920975539678,-2.42580987010058,-19.2343069267998,-2.99109571081673,9.52752074000473,17.1364426797165,-15.6402161539033,9.78267413804593,9.43263556966627,5.06757808058878,-4.47926752294106,-17.3691746558434,3.99760455244532,1.4151029400085,8.89224469808186,-2.83992481423707,-0.492336956733319,3.28290991087231,3.85990793553297,-8.90705092873849,-17.7902663841582,1.74661216774987,-19.6411813602325,19.3742870135737,18.9811333759661,0.765645695938311,-4.73163549436878,-0.855021596928189,-2.48653374675299,-12.6792841846805,-17.4875155763953,17.1627125138593,-2.16264078943451,2.15754423793356,-2.80383147240977,-10.7142090887233,17.9389565154667,18.8395192760639,13.7658831814431,-11.1422906848907,7.91357465907774,5.02020416822446,12.3781104233532,1.20421776039783,7.84583189043147,6.3837089411763,13.8932999344707,-2.35928201920734,0.927266973523188,19.5254608802006,-13.198523180346,3.12796600297153,-20.5434998098939,3.52704695759401,10.3820067840877,-9.02868546572871,1.67331190700338,-0.228720331597061,-1.39100558198826,-0.122412142921674,16.5671611812012,-14.4204757055845,1.81665722687687,-16.8537163561203,-11.9264435258769,2.68955297255271,-13.817486831441,7.87312132250302,8.78501077174469,-2.41065861023483,7.74209477699574,-17.875029477808,-20.3382525365537,-0.156889809164829,-8.66988290236121,7.20077879554241,-16.5236246366409,-17.5383460930866,5.4825751966243,-13.2242586383442,2.04709014195926,-14.1287282778212,3.18596852675273,-0.0870944565277938,-21.4542252074024,21.0191128581882,1.60385474551617,-16.964333485658,7.53925839322867,0.806244469881588,-2.8033779330827,-17.3861728309692,17.494924040051,-4.02931917499083,20.413866737375,6.68150994018296,-10.4460495406797,20.1855077863121,15.3628640532909,5.70775721881636,12.6828090398904,15.4353103914631,11.0990568738171,17.0636916704775,-10.3337881049011,11.8316255098366,-6.68512894750448,9.20532089676397,1.18244122540038,15.3853592458602,-20.5062040646386,-3.85040652471447,17.5423193218868,17.8169330823617,15.8112019462963,21.3387802744448,15.8797497722738,5.41224439717998,20.6249474983657,-8.98080173055824,-0.903288417948641,19.6846563527395,19.7363430603188,-9.34848351542492,19.2118956522669,19.8645388208863,-17.7399069662793,13.7281148329862,-2.73031759082039,-2.81041070006992,8.1957884771828,1.03636643502591,-12.2927630262754,12.4362631504786,-14.4579071556128,17.900888083007,-3.98292286863356,19.4754664953722,-14.5664465530997,2.31148961428687,16.6299468011584,6.27441369187808,8.01395323549763,-12.6676261670212,8.19553323534016,-8.3074271563913,3.84597510272244,11.6956907659842,-2.76544888628933,2.29912225727876,15.5046538207351,17.1014019053288,-2.00312377145404,4.92284235823714,-0.695932334115392,7.76873083402207,9.75931456833239,0.953908160832141,18.0787377677213,-12.1647309934251,-2.42537512431964,-8.10786758225995,9.38031138299535,5.42039734168583,18.89016137245,3.75812702143235,18.1215022624889,-12.4198631850512,-9.68705242633241,-0.462287420125959,16.5428625688585,17.2963388559552,19.5845643045607,1.99839679438684,1.05203043609798,-9.44022344452382,-12.5098966893475,19.9883653260893,-11.4668381940495,-2.51920279720007,-1.94362589422666,1.30041141159447,15.4712630727988,-7.97105542424684,9.0885550700181,-16.0708143259203,13.651996886796,4.11706185657858,1.21517331414382,10.8910790040793,-2.59064794134656,11.5909231940057,17.6927126224925,19.7962819482183,1.31752426671125,-2.6266716631909,-2.9914589958855,-15.4066676597385,21.2287038403327,9.17465382615656,6.0829249596365,-21.4100486909801,15.8998983618823,0.121534486468602,16.6542024316003,-18.7837797889461,18.2713082869209,15.7583643487246,-2.47482895690965,15.6281478652042,-4.89756518440682,-13.0722727188178,-9.24452462160515,-18.549461938582,14.5802721093655,4.04565429687868,7.79383009582645,5.46632268170458,-12.5048047973784,-20.5155466098898,1.54284473769175,-17.8643543468521,14.9389946351783,17.8152148860166,15.0874650458087,4.12253813767634,6.79536626466934,-3.16933669862998,19.2736094883604,-21.2466435802668,7.69934125545749,15.1299277220383,6.46163062266483,-0.515585277589916,7.22444121228801,15.6311472667667,1.57422509992914,15.5932848851086,7.82841272501197,20.4393105600063,14.7704126306148,10.0610786618068,-10.0934141484493,-19.3735458271777,13.3824888085295,-3.91201560727871,-6.17802969035687,-3.0782854722129,18.5672730871287,-0.213116993316094,-3.92745993225775,-12.1339659087895,6.24338226102606,-2.81738502775102,-2.08228378769089,17.8136610420233,13.3151896067032,-2.6487991668168,18.2526662464085,-3.2693215964209,8.72366486429651,2.70168099469592,13.2827947914069,16.572979953092,-14.3405385514874,2.70772730819146,-13.3323497518113,16.4590349549654,-2.24230932647231,17.1724444086406,-10.994384941468,-8.60463894901688,8.96317709717228,-19.3901642733629,14.5168847896221,8.69169027305702,-1.53771585385232,-0.984246464803202,13.6827072255216,-17.9832895498747,20.8282787744234,-17.9281699803905,13.8382915891702,13.9770802016478,19.9704223768714,-3.34561909460168,3.34767777678513,4.43979394280692,-5.74764595566222,-14.2799481430226,8.26328633034095,12.8295408942651,-19.869082967735,-9.48802658676562,16.003876005742,16.0020751450403,16.5498827131526,10.3223313012824,0.207876009194289,7.66663009032203,-3.63817244182129,0.528683552321408,4.48202850203215,19.9210461358447,20.389113945851,14.7172745793256,-18.6724228331844,-20.5754501842944,-5.45855740488608,-6.57079439419139,16.2457160493122,-3.53396956780216,-13.2141498278453,10.3884748029506,8.72763756535123,1.38116179980807,3.52742574951163,-2.04680639205122,-14.3704430620874,-14.7898605268285,2.07214115574314,-1.24163081787943,-20.1604873612004,6.41051763781159,-17.3849461031013,0.518411375416865,2.54199777444288,-10.402762631616,3.16938865516432,2.92341685921823,-0.205227723697595,-17.9325411357957,17.5345607378754,9.89385110045275,-0.767757027691558,17.0659670848245,5.05277920748838,11.0046103480639,19.8772905951954,2.4920515631586,7.18944438465895,5.23412624050007,2.71874127630484,12.4916241216863,11.5684429432305,0.422853835792536,18.1503072466146,-7.44621265828443,14.5780235338787,5.58378617164743,1.20760050517734,3.84678867018867,-10.3670858521612,19.6165993844413,5.71517593971534,15.6485901211095,14.8901614364899,-7.82933930400907,14.2692900046865,-3.18561111362506,10.4087048554138,3.77554647180408,15.8170369844064,2.22511779786275,14.6738227638015,-0.901465467116397,-0.651668428603639,14.4059680995789,9.2294033772844,-12.5101697823064,17.4184586800377,-17.426280146391,5.05283338120363,1.67258858208148,-2.2352070495508,-2.49439076976666,-17.9761502107718,16.7376775395646,8.79808686415993,9.58896387521299,0.614259375566449,1.29394408542557,-1.32625445163692,12.4464633902524,16.048594334689,9.55711796359211,-1.24845211538619,14.2485122683077,14.1230741649444,8.83774829576955,15.8414998896177,18.1796254252336,-12.0142657680521,-5.85742795066542,19.8039687681025,14.2356381590713,-12.3597768573599,-19.5095828495316,11.370482328829,16.7097120707782,-10.3481467503858,17.5443569903107,16.7838508213956,-1.27673547941285,16.4034788498926,-9.71142959076393,-8.25137359370254,-14.8338455410491,8.4099998140794,14.9766472579633,-17.9690356760235,16.6089944916805,-15.1259477866307,-18.4553106420749,20.138651478641,15.8731321364437,-18.7289301428151,18.6225119994419,-1.60319356130381,19.0058014396552,12.6696519109037,-19.3353047884268,7.90722260339127,-1.64623142233308,13.3162789105471,-11.5582763606447,-3.05660753524773,8.6383702549955,-3.19176762395208,-11.1129436308296,16.7462900604341,15.2031185312269,13.5354455967921,14.5437388126716,0.360911927572727,-1.38157104266409,18.392931863343,14.5735525859071,-1.01138202350945,5.13578129156389,15.7360762391944,17.267381228412,-3.17754490026488,-9.19494389485269,8.40233366159696,-4.58927213645346,20.8856941629589,6.25946881583429,0.762827368134978,19.6809767643837,18.0899668568342,4.72906544750177,-10.993287349382,-13.9248106263865,6.96041971373793,15.1737701506904,-14.1175660737513,0.891720192916639,0.515578437033615,12.8907920132349,2.40182642167669,-4.26564994701717,13.5109105393715,1.68161762133074,-0.102606395166109,15.7416421600608,17.0735040655409,5.37112495912424,-2.06419947190856,9.50598069363063,7.8591359210575,8.16061450027677,2.5389377327347,10.0588314859113,16.3241975179064,14.1493181368791,8.9689187411779,4.84202946466112,-8.99565672630343,6.64267618933627,11.305339668824,16.8164087320148,10.2335370628633,1.18051060483815,-8.91174989068111,9.279526665975,-14.9626391890019,-3.54743925160024,-15.5825932740027,9.80249637160744,11.536734485222,-18.158195997143,8.65504572529829,14.5665427077952,16.9890147914619,14.7465324701929,9.46403491185392,-18.6675922727859,-8.69350076898722,19.0108274353206,11.2801540572842,-3.04146782020303,7.60749853195372,7.02783548494267,-11.8650018589081,1.10819839262213,13.1853502798213,-0.81137001408533,-0.666704095578504,-1.17910134422225,-16.5468362602718,3.60282661385323,14.2132971369446,-15.3978288470131,19.4702393886969,-10.4696816694967,1.13322845796472,15.0638196519836,15.3837084614033,9.11368801560278,10.9886534986002,3.66309055326045,-1.66960029582586,-2.29142586827462,-3.76804304592304,8.20451639434588,15.2264758981547,16.6244402265399,3.87545633225048,-1.10973349003385,5.39761494610955,17.1692346844561,-1.17549876120324,4.96799123093356,-2.22814040129757,-20.3913156583966,-2.77443811816909,-3.97399573621483,15.4875875599323,19.5626047152569,-7.97550114439175,1.93839121081026,19.4011905787773,14.4293790519012,16.4376369157374,-13.7781594894936,-3.26005927841387,8.35391776239691,16.6080078493928,-2.28365750710105,-15.8687204609801,-3.00515673429595,-2.17465273551824,-13.0225396773461,2.29235424629154,15.8058080252767,19.9674226884741,-12.1839988688698,-17.1261885940117,8.82661517280576,-8.5550739907339,17.296493292508,-11.3559590810084,-20.7727192209101,-0.429788572160804,14.217692337805,-5.19240840581989,13.0449999445833,-18.8321260165657,17.0389882456598,20.4094249330553,15.3371273469079,11.1986964073298,-17.2008678425196,19.0797795082565,10.8855094569571,2.40168209896513,19.9168014105335,17.2553117514865,17.8522171780002,-20.2222723962623,-9.53046464406806,-16.4295558483845,-1.15391471582,9.34542881783022,3.26312313326077,-2.63711089177538,14.5961155339532,-11.8841138816897,-9.90787752099693,15.3221247031161,7.31537632606183,10.2752625177923,-11.9159128882196,0.449192868289589,17.0352734523319,2.90349567418807,14.6684309469985,8.22109747786773,13.8811626967834,13.6188142248342,-11.7549944345284,4.41789300211265,16.1180147701378,14.1288159298647,14.3379385868962,11.4336450879225,-3.81625715860506,-9.12035471944441,8.84925969324289,18.2390604000628,15.1147434875783,3.53066784936364,-2.55730794272695,19.3574124561834,8.69597319814983,15.8969573833719,-4.43751700619594,-3.49781214597606,7.13597358392323,4.17892692005263,-9.83006458092438,-14.236024260158,2.40572730383288,20.0030711636918,16.2443050712752,13.5894964716118,-12.8385728515398,-4.57778255173067,-18.2581302472305,-21.7217644117845,17.3272879784995,18.4834366614735,-4.5668838163869,18.9812725004085,16.5279336265263,3.1536463570625,7.27988690003319,-10.9992408413667,15.1156199886862,6.34218200942106,-1.74759771989044,3.82200928693169,3.32164201229202,-9.6629663018304,15.1507528644543,2.07699543867406,7.17151566870513,2.65972548383651,-2.55222020811325,18.2618196571966,-7.06106249781447,7.36671558412857,11.5612359416949,-12.8941489491538,-16.8951081129048,19.0252425970689,13.3705204996846,-1.97990681578481,-21.4403644740376,-14.5115651918593,-17.1829381801626,5.30579219982831,5.80812981762389,16.2899141269879,20.4593958498046,3.30394972968619,3.72562259841015,-8.33421125611347,0.424972557420911,1.85736699791352,-21.1430359438421,19.333101311897,15.7473602132762,3.29249246741572,17.8121463279215,0.900072044157013,-20.7552528920612,10.5209344628855,-21.5553368657892,14.2816457511048,-10.8321394728403,9.55760865254239,0.729087732189361,1.26883155784994,-0.59765103640581,-8.05603760278139,19.4229032231311,19.6025498787028,2.57474606672288,-4.53930499474173,0.388110576501327,20.6560793611693,-10.4353293429493,7.65317442890324,1.43240249699367,3.27336207931473,14.2276587589568,17.3190151236262,-2.07468389769746,6.60447381677595,2.16800203691403,18.8098324126826,-2.15843873761192,-22.2326126603977,0.174082379051333,-2.96894671984182,19.6535331394537,-2.42572579917431,-3.69369933084895,15.2822632686775,2.19443563309576,-0.00916421248447897,15.0330020505199,-1.71532678013858,-1.7064764118561,-11.6270438601738,-9.26454032177326,-0.480694220675785,8.23138940430177,-1.77228094960533,-19.8024799628438,-2.23457686316135,-12.0589796861826,14.4567313462072,10.1100635264806,18.705083449952,-1.8053801197738,14.1580909399068,12.643752485126,19.9873244032206,-16.9324782556308,10.4545531705857,-17.9187179635911,-17.3454899940387,15.2696328957819,-19.5107780160869,0.0751085382410047,20.8623405564885,-14.621980999277,-13.5699642051154,1.96253890591315,15.8398235424118,-9.10123590695494,-9.73111640168443,8.38491059501091,15.2469748983534,12.1729888611388,-2.19371414743663,5.938134576989,-0.796182141560785,2.69190214953539,-0.161806611707093,19.6900701487769,0.57290472349682,11.8143424471561,-2.7505665508003,-13.3960900834486,15.7170308985772,-13.6680830285086,16.506365408677,6.74678884926463,20.4588207905883,9.87846066161024,9.96469233767841,-2.51598940114582,-0.979431479890049,19.2478578288578,-10.9087608052963,19.3903194869668,-9.63609447203665,2.09166768531411,19.8408129340323,9.69023832779171,7.05596320797161,13.8621824686197,-2.28899041931767,14.9904019068918,-19.3186243401973,9.19114174801527,-20.1638671218294,13.9522958245874,-3.12013199725636,12.9979427134782,-2.55646438981654,1.5771938578123,1.92678821446162,-7.35741387977289,-9.33568939974354,19.6389851756646,18.4335016408565,-2.02843902686987,18.7170765163384,4.34718714154463,-14.565310362859,18.711219812327,6.20476625038319,8.97853137326519,4.48777763912365,-1.75226295272448,17.0472460909636,16.6180218811418,-1.10887933425826,14.5593694794081,-0.352991532538151,17.4120864341467,-0.041668469963273,-0.621690368820052,8.12929424612562,-10.6215285316348,0.353813545611738,16.2256897254895,19.5961743419478,17.6263672434147,14.0128251449006,8.70010978178972,-1.61829291509501,-0.988706838742032,-10.4673041301781,16.6716515651436,13.2365923454686,0.430013306399646,0.27495061153459,13.5491427369848,-9.40492819783824,3.03575397766656,8.1308741111949,17.9385420330635,13.3393570161314,-13.5762293354778,9.89487617730381,-0.153984270042586,-8.96438283106515,16.6675799038749,1.47270080371427,-20.9785748877748,-17.5296719644304,15.2483850673073,19.9798562679143,-9.5237715986463,14.4700228293009,19.9161407097141,0.959397204237898,18.8625751413655,14.7695880933912,5.63845411616412,7.74390247570908,13.2540381426418,15.3659450420157,9.4065629238653,19.7161887993213,11.6769491717758,-1.60534684618136,18.5505570505702,-1.93733020555345,11.8261834124673,14.9931209836848,-1.45495527104816,9.93978974911447,14.9217098534065,-19.7394980397688,7.8549693892823,7.75663425744448,20.9422658748666,20.8019699424477,0.85893574860937,10.3764259158204,-1.78500155199551,15.4485988274013,-3.97561722291679,-1.87500774338994,5.09878374110306,5.2179290157884,-19.1254647914821,7.46864481741188,-11.1960305529515,8.4935616554094,-14.2523828714132,-19.9761762745815,-8.60175717835101,-17.1630731728224,-16.7201291384846,-6.91464845314963,10.9192029630302,8.72416905753062,-9.38851296960255,3.2192728128208,-12.3710648628355,-13.5123755779893,8.19577274135588,-10.1895330746395,18.5368910430836,18.4951535750688,-4.05391199324103,19.4818918813363,-20.0692261361627,0.706705250713441,-0.704858315696957,-18.6350291618185,-5.54863086101619,13.8673574724741,15.0816942906005,1.23253475253951,13.3774146163719,-0.255135073963282,4.08779531461649,0.629557551379091,6.31263917320316,0.375907984189696,17.3177989775091,-9.39244761227013,19.5266656043227,-22.2877749722028,2.9931788056219,2.49161762964059,-12.5111852310286,1.42014864874895,9.02740869929596,14.3018852512707,7.87077030986005,10.9729051559684,5.86241827080051,-13.5215021294372,13.0355403767206,-3.80915641184492,18.0108214051236,18.9157277221578,13.6831366928957,20.7278460559288,-17.2361633156262,1.84826284862035,0.432035640654324,8.59737999131398,-18.1455846950282,2.45760126095541,-15.7674047738025,18.5702610298533,-15.9598464307772,-8.75022627142851,-18.1853231679671,7.88150534900053,-11.576337948329,3.40512300979407,18.5448930772796,-12.4866655906516,-17.0130616029137,8.17315836275974,10.3860841964863,1.9142555341536,-19.4830231764535,18.0886180483037,15.2140264091526,-19.1755068282761,17.9325299986165,2.21423069629377,-14.3491769022764,-2.47638031500891,-9.10398558954643,-5.95706679238743,-10.1744001789963,-15.4702347090409,-10.4323103092065,16.4843193962816,17.6047038748644,13.4644105585632,-2.19592475909374,-18.8947934457866,11.8381243113958,1.7151862558127,-8.67570683478086,-11.534044300752,9.08556760702293,14.9377623573207,18.4683314773426,9.63513491095147,1.31401915324597,-1.271225224239,15.397118521621,16.1527441846277,15.4629089159033,-10.0877367905372,8.15232419539049,8.12238751713012,-0.205833274392683,-4.84138526244564,-2.77894224395762,-13.9448332729583,10.1871378895179,-5.61887869764122,-4.8646317704411,10.0490569905279,-3.78895015051356,-10.4082211497873,16.1497195428697,-17.7882167027861,7.54066807039838,-0.408065406356117,-8.40547288859975,2.64177130706061,-7.97186409424854,13.359357562061,-3.93616855691398,-15.8960856219169,20.577686134897,13.9427721977091,6.15354020451476,-5.09228915686128,1.16633154919136,-2.74347047940198,11.0529867657383,0.803549222062129,-13.6239142359406,13.8178430857289,-17.4693556163821,-9.03507213606613,-15.4892150185873,-6.53126428249449,16.34413616629,-10.0671349499422,-5.87789549981324,9.02832461999755,5.64875186058826,-3.92103396030346,-19.9757099484824,11.5119662435307,-20.4291824270241,16.5062183118915,-11.7302757271135,-0.578599471146989,-5.05305668594502,5.81965401449357,-2.67231847003062,-14.2825822772759,7.10698253044451,-17.8438621769918,0.247612054058772,-10.8695012639889,7.84571768660693,15.0760936521291,-20.9896925695329,11.8489908805816,-20.9762946461049,11.3572062431146,6.91518816788796,13.7032299605268,-3.33281966814274,11.3488702383265,-12.1253005943129,-1.03986794129841,19.2839456431107,17.6928660283789,-14.1593242890703,-2.29093196191276,-2.93440806063435,-11.9658361061073,6.50239323778827,18.9046768716738,-0.105132073970499,18.7564823711172,13.9330544270098,-14.8785073119708,17.987779310703,-2.69247151637152,-8.41990274471359,17.7478670478794,-0.72170821221358,17.3930206723904,0.839712519444847,12.1770556828123,-14.5088592050095,-9.41808655186277,15.7704811143105,17.0577410238385,-4.31006002406029,11.3179990400539,15.0064872579722,-7.00072914645733,0.671524174425753,-5.62430408468967,15.8281324761621,-2.54609926546576,-18.9987209218952,17.559650939614,0.0391167638735266,-17.0451463956825,7.17743331285697,18.7830821145138,21.1871887619488,-1.37783388861623,4.39999947594913,-1.06905161418157,0.00537903355749417,9.83466669189674,19.7337454827457,8.78045774583373,-19.7092401521396,16.9509332803418,15.2192815999903,-16.116060876017,7.55614040969096,13.5453902298381,15.4296801221977,4.55327606829284,17.1853626664548,-10.2110878965165,-1.1557859858247,-11.1687145813132,16.8557798760247,-1.1893675458757,-18.3156070345885,6.46004018607397,6.57408912951245,-8.63285797086896,16.6770500324721,-15.3413750858785,16.3645765587824,5.54272545528201,-17.1014383596195,20.5682045889648,3.50373413625571,5.15603422399348,-18.9273024416562,11.8649366522599,5.83506239276883,-9.46712003146065,4.8708977670112,-13.3430000735059,-16.6237041717262,0.510803400526167,19.8292226059539,-21.9601089146488,-16.4192070967549,0.715291138696762,-9.26215837807002,-16.9938542258804,-17.1268705713455,15.6054871165768,-0.986028339001036,10.0658485168423,-3.30425427896296,-3.80908961717448,-0.577962518282289,18.6995134848471,13.8014950290531,-20.4561685707944,3.9512791602266,5.41351864330652,-3.47403412902778,-2.60907998598067,15.8630352320114,3.06069548794654,-20.1797265139826],"type":"scatter","mode":"markers","name":"T cell","marker":{"color":"rgba(141,160,203,1)","line":{"color":"rgba(141,160,203,1)"}},"textfont":{"color":"rgba(141,160,203,1)"},"error_y":{"color":"rgba(141,160,203,1)"},"error_x":{"color":"rgba(141,160,203,1)"},"line":{"color":"rgba(141,160,203,1)"},"xaxis":"x","yaxis":"y","frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="sessioninfo" class="section level3">
<h3><span class="header-section-number">8.9.7</span> sessionInfo()</h3>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb204-1" title="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 7 x64 (build 7601) Service Pack 1
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=Chinese (Simplified)_People&#39;s Republic of China.936 
## [2] LC_CTYPE=Chinese (Simplified)_People&#39;s Republic of China.936   
## [3] LC_MONETARY=Chinese (Simplified)_People&#39;s Republic of China.936
## [4] LC_NUMERIC=C                                                   
## [5] LC_TIME=Chinese (Simplified)_People&#39;s Republic of China.936    
## 
## attached base packages:
##  [1] splines   parallel  stats4    stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] plotly_4.9.0                scfind_3.2.0               
##  [3] scmap_1.7.0                 googleVis_0.6.4            
##  [5] corrplot_0.84               ggbeeswarm_0.6.0           
##  [7] ggthemes_4.2.0              scater_1.13.9              
##  [9] ouija_0.99.0                Rcpp_1.0.1                 
## [11] SLICER_0.2.0                destiny_2.15.0             
## [13] monocle_2.13.0              DDRTree_0.1.5              
## [15] irlba_2.3.3                 VGAM_1.1-1                 
## [17] ggplot2_3.2.0               Matrix_1.2-17              
## [19] TSCAN_1.23.0                SingleCellExperiment_1.7.0 
## [21] SummarizedExperiment_1.15.5 DelayedArray_0.11.4        
## [23] BiocParallel_1.19.0         Biobase_2.45.0             
## [25] GenomicRanges_1.37.14       GenomeInfoDb_1.21.1        
## [27] IRanges_2.19.10             S4Vectors_0.23.17          
## [29] BiocGenerics_0.31.5         RColorBrewer_1.1-2         
## [31] M3Drop_1.11.4               numDeriv_2016.8-1.1        
## [33] matrixStats_0.54.0          scRNA.seq.funcs_0.1.0      
## [35] knitr_1.23                 
## 
## loaded via a namespace (and not attached):
##   [1] R.methodsS3_1.7.1        coda_0.19-3             
##   [3] tidyr_0.8.3              acepack_1.4.1           
##   [5] R.utils_2.9.0            data.table_1.12.2       
##   [7] rpart_4.1-15             inline_0.3.15           
##   [9] RCurl_1.95-4.12          callr_3.3.0             
##  [11] cowplot_1.0.0            RANN_2.6.1              
##  [13] combinat_0.0-8           proxy_0.4-23            
##  [15] spatstat.data_1.4-0      httpuv_1.5.1            
##  [17] StanHeaders_2.18.1-10    assertthat_0.2.1        
##  [19] viridis_0.5.1            lle_1.1                 
##  [21] xfun_0.8                 hms_0.5.0               
##  [23] evaluate_0.14            promises_1.0.1          
##  [25] DEoptimR_1.0-8           caTools_1.17.1.2        
##  [27] readxl_1.3.1             igraph_1.2.4.1          
##  [29] htmlwidgets_1.3          sparsesvd_0.1-4         
##  [31] tensorA_0.36.1           hash_2.2.6.1            
##  [33] purrr_0.3.2              crosstalk_1.0.0         
##  [35] dplyr_0.8.3              backports_1.1.4         
##  [37] bookdown_0.12            moments_0.14            
##  [39] deldir_0.1-22            vctrs_0.2.0             
##  [41] TTR_0.23-4               abind_1.4-5             
##  [43] RcppEigen_0.3.3.5.0      withr_2.1.2             
##  [45] robustbase_0.93-5        checkmate_1.9.4         
##  [47] vcd_1.4-4                xts_0.11-2              
##  [49] prettyunits_1.0.2        mclust_5.4.5            
##  [51] goftest_1.1-1            cluster_2.1.0           
##  [53] splancs_2.01-40          ape_5.3                 
##  [55] lazyeval_0.2.2           laeken_0.5.0            
##  [57] crayon_1.3.4             elliptic_1.4-0          
##  [59] labeling_0.3             pkgconfig_2.0.2         
##  [61] slam_0.1-45              nlme_3.1-140            
##  [63] vipor_0.4.5              nnet_7.3-12             
##  [65] rlang_0.4.0              sgeostat_1.0-27         
##  [67] rsvd_1.0.1               randomForest_4.6-14     
##  [69] cellranger_1.1.0         polyclip_1.10-0         
##  [71] lmtest_0.9-37            loo_2.1.0               
##  [73] carData_3.0-2            boot_1.3-23             
##  [75] zoo_1.8-6                base64enc_0.1-3         
##  [77] beeswarm_0.2.3           processx_3.4.0          
##  [79] pheatmap_1.0.12          png_0.1-7               
##  [81] viridisLite_0.3.0        bitops_1.0-6            
##  [83] R.oo_1.22.0              KernSmooth_2.23-15      
##  [85] DelayedMatrixStats_1.7.1 stringr_1.4.0           
##  [87] tripack_1.3-8            scales_1.0.0            
##  [89] magrittr_1.5             plyr_1.8.4              
##  [91] gplots_3.0.1.1           gdata_2.18.0            
##  [93] zlibbioc_1.31.0          compiler_3.6.0          
##  [95] HSMMSingleCell_1.5.0     bbmle_1.0.20            
##  [97] cli_1.1.0                XVector_0.25.0          
##  [99] ps_1.3.0                 htmlTable_1.13.1        
## [101] Formula_1.2-3            MASS_7.3-51.4           
## [103] mgcv_1.8-28              tidyselect_0.2.5        
## [105] stringi_1.4.3            forcats_0.4.0           
## [107] highr_0.8                densityClust_0.3        
## [109] yaml_2.2.0               BiocSingular_1.1.5      
## [111] latticeExtra_0.6-28      ggrepel_0.8.1           
## [113] grid_3.6.0               orthopolynom_1.0-5      
## [115] tools_3.6.0              rio_0.5.16              
## [117] rstudioapi_0.10          foreign_0.8-71          
## [119] snowfall_1.84-6.1        gridExtra_2.3           
## [121] cubature_2.0.3           smoother_1.1            
## [123] alphahull_2.2            scatterplot3d_0.3-41    
## [125] Rtsne_0.15               digest_0.6.20           
## [127] FNN_1.1.3                shiny_1.3.2             
## [129] qlcMatrix_0.9.7          car_3.0-3               
## [131] later_0.8.0              httr_1.4.0              
## [133] contfrac_1.1-12          hypergeo_1.2-13         
## [135] colorspace_1.4-1         tensor_1.5              
## [137] ranger_0.11.2            reldist_1.6-6           
## [139] MCMCglmm_2.29            statmod_1.4.32          
## [141] spatstat.utils_1.13-0    sp_1.3-1                
## [143] xtable_1.8-4             jsonlite_1.6            
## [145] spatstat_1.60-1          corpcor_1.6.9           
## [147] rstan_2.19.2             zeallot_0.1.0           
## [149] R6_2.4.0                 Hmisc_4.2-0             
## [151] pillar_1.4.2             htmltools_0.3.6         
## [153] mime_0.7                 DT_0.7                  
## [155] glue_1.3.1               VIM_4.8.0               
## [157] BiocNeighbors_1.3.2      deSolve_1.23            
## [159] class_7.3-15             codetools_0.2-16        
## [161] pkgbuild_1.0.3           lattice_0.20-38         
## [163] tibble_2.1.3             curl_3.3                
## [165] gtools_3.8.1             zip_2.0.3               
## [167] openxlsx_4.1.0.1         survival_2.44-1.1       
## [169] limma_3.41.6             rmarkdown_1.14          
## [171] docopt_0.6.1             fastICA_1.2-2           
## [173] munsell_0.5.0            e1071_1.7-2             
## [175] GenomeInfoDbData_1.2.1   haven_2.1.1             
## [177] reshape2_1.4.3           gtable_0.3.0</code></pre>

</div>
</div>
</div>
<h3> References</h3>
<div id="refs" class="references">
<div id="ref-Altschul1990-ts">
<p>Altschul, Stephen F., Warren Gish, Webb Miller, Eugene W. Myers, and David J. Lipman. 1990. “Basic Local Alignment Search Tool.” <em>Journal of Molecular Biology</em> 215 (3): 403–10. <a href="https://doi.org/10.1016/s0022-2836(05)80360-2">https://doi.org/10.1016/s0022-2836(05)80360-2</a>.</p>
</div>
<div id="ref-Cannoodt2016-uj">
<p>Cannoodt, Robrecht, Wouter Saelens, and Yvan Saeys. 2016. “Computational Methods for Trajectory Inference from Single-Cell Transcriptomics.” <em>Eur. J. Immunol.</em> 46 (11): 2496–2506. <a href="https://doi.org/10.1002/eji.201646347">https://doi.org/10.1002/eji.201646347</a>.</p>
</div>
<div id="ref-tabula2018single">
<p>Consortium, Tabula Muris, and others. 2018. “Single-Cell Transcriptomics of 20 Mouse Organs Creates a Tabula Muris.” <em>Nature</em> 562 (7727): 367.</p>
</div>
<div id="ref-Deng2014-mx">
<p>Deng, Q., D. Ramskold, B. Reinius, and R. Sandberg. 2014. “Single-Cell RNA-Seq Reveals Dynamic, Random Monoallelic Gene Expression in Mammalian Cells.” <em>Science</em> 343 (6167): 193–96. <a href="https://doi.org/10.1126/science.1245316">https://doi.org/10.1126/science.1245316</a>.</p>
</div>
<div id="ref-Van_Dijk2017-bh">
<p>Dijk, David van, Juozas Nainys, Roshan Sharma, Pooja Kathail, Ambrose J Carr, Kevin R Moon, Linas Mazutis, Guy Wolf, Smita Krishnaswamy, and Dana Pe’er. 2017. “MAGIC: A Diffusion-Based Imputation Method Reveals Gene-Gene Interactions in Single-Cell RNA-sequencing Data.” <em>bioRxiv</em>, February, 111591.</p>
</div>
<div id="ref-Guo2015-ok">
<p>Guo, Minzhe, Hui Wang, S. Steven Potter, Jeffrey A. Whitsett, and Yan Xu. 2015. “SINCERA: A Pipeline for Single-Cell RNA-Seq Profiling Analysis.” <em>PLoS Comput Biol</em> 11 (11): e1004575. <a href="https://doi.org/10.1371/journal.pcbi.1004575">https://doi.org/10.1371/journal.pcbi.1004575</a>.</p>
</div>
<div id="ref-Kiselev2017-nb">
<p>Kiselev, Vladimir Yu, and Martin Hemberg. 2017. “Scmap - a Tool for Unsupervised Projection of Single Cell RNA-seq Data.” <em>bioRxiv</em>, July, 150292.</p>
</div>
<div id="ref-Kiselev2016-bq">
<p>Kiselev, Vladimir Yu, Kristina Kirschner, Michael T Schaub, Tallulah Andrews, Andrew Yiu, Tamir Chandra, Kedar N Natarajan, et al. 2017. “SC3: Consensus Clustering of Single-Cell RNA-Seq Data.” <em>Nat Meth</em> 14 (5): 483–86. <a href="https://doi.org/10.1038/nmeth.4236">https://doi.org/10.1038/nmeth.4236</a>.</p>
</div>
<div id="ref-Levine2015-fk">
<p>Levine, Jacob H., Erin F. Simonds, Sean C. Bendall, Kara L. Davis, El-ad D. Amir, Michelle D. Tadmor, Oren Litvin, et al. 2015. “Data-Driven Phenotypic Dissection of AML Reveals Progenitor-Like Cells That Correlate with Prognosis.” <em>Cell</em> 162 (1): 184–97. <a href="https://doi.org/10.1016/j.cell.2015.05.047">https://doi.org/10.1016/j.cell.2015.05.047</a>.</p>
</div>
<div id="ref-Li2017-tz">
<p>Li, Wei Vivian, and Jingyi Jessica Li. 2017. “scImpute: Accurate and Robust Imputation for Single Cell RNA-Seq Data.” <em>bioRxiv</em>, June, 141598.</p>
</div>
<div id="ref-Muraro2016-yk">
<p>Muraro, Mauro J., Gitanjali Dharmadhikari, Dominic Grün, Nathalie Groen, Tim Dielen, Erik Jansen, Leon van Gurp, et al. 2016. “A Single-Cell Transcriptome Atlas of the Human Pancreas.” <em>Cell Systems</em> 3 (4): 385–394.e3. <a href="https://doi.org/10.1016/j.cels.2016.09.002">https://doi.org/10.1016/j.cels.2016.09.002</a>.</p>
</div>
<div id="ref-Regev2017-mw">
<p>Regev, Aviv, Sarah Teichmann, Eric S Lander, Ido Amit, Christophe Benoist, Ewan Birney, Bernd Bodenmiller, et al. 2017. “The Human Cell Atlas.” <em>bioRxiv</em>, May, 121202.</p>
</div>
<div id="ref-Segerstolpe2016-wc">
<p>Segerstolpe, Åsa, Athanasia Palasantza, Pernilla Eliasson, Eva-Marie Andersson, Anne-Christine Andréasson, Xiaoyan Sun, Simone Picelli, et al. 2016. “Single-Cell Transcriptome Profiling of Human Pancreatic Islets in Health and Type 2 Diabetes.” <em>Cell Metabolism</em> 24 (4): 593–607. <a href="https://doi.org/10.1016/j.cmet.2016.08.020">https://doi.org/10.1016/j.cmet.2016.08.020</a>.</p>
</div>
<div id="ref-Welch2016-jr">
<p>Welch, Joshua D., Alexander J. Hartemink, and Jan F. Prins. 2016. “SLICER: Inferring Branched, Nonlinear Cellular Trajectories from Single Cell RNA-Seq Data.” <em>Genome Biol</em> 17 (1). <a href="https://doi.org/10.1186/s13059-016-0975-3">https://doi.org/10.1186/s13059-016-0975-3</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="cleaning-the-expression-matrix.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="seurat-chapter.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": true,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/lemonbases/scRNA.seq.course.CN/edit/master/08biological-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
