<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 scRNA-seq原始数据处理 | scRNA-seq数据分析</title>
  <meta name="description" content="3 scRNA-seq原始数据处理 | scRNA-seq数据分析" />
  <meta name="generator" content="bookdown 0.11 and GitBook 2.6.7" />

  <meta property="og:title" content="3 scRNA-seq原始数据处理 | scRNA-seq数据分析" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 scRNA-seq原始数据处理 | scRNA-seq数据分析" />
  
  
  

<meta name="author" content="碱基吃柠檬" />


<meta name="date" content="2019-07-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction-to-single-cell-rna-seq.html">
<link rel="next" href="construction-of-expression-matrix.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<!-- for Facebook -->  
<meta property="og:url" content="https://github.com/sdbiodog" />
<meta property="og:description" content="本项目是将github上经典的scRNA-seq分析教程进行汉化，源项目参见hemberg-lab/scRNA.seq.course" />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="3 scRNA-seq原始数据处理 | scRNA-seq数据分析" />
<meta name="twitter:description" content="本项目是将github上经典的scRNA-seq分析教程进行汉化，源项目参见hemberg-lab/scRNA.seq.course" />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-71525309-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> 说在前面的话</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#section-1"><i class="fa fa-check"></i><b>1.1</b> 关于本项目</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#section-2"><i class="fa fa-check"></i><b>1.2</b> 视频</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#docker-"><i class="fa fa-check"></i><b>1.4</b> Docker 镜像</a><ul>
<li class="chapter" data-level="1.4.1" data-path="index.html"><a href="index.html#section-3"><i class="fa fa-check"></i><b>1.4.1</b> 运行镜像</a></li>
<li class="chapter" data-level="1.4.2" data-path="index.html"><a href="index.html#section-4"><i class="fa fa-check"></i><b>1.4.2</b> 下载数据或其它文件</a></li>
<li class="chapter" data-level="1.4.3" data-path="index.html"><a href="index.html#rstudio"><i class="fa fa-check"></i><b>1.4.3</b> RStudio</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#section-5"><i class="fa fa-check"></i><b>1.5</b> 手动安装</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#section-6"><i class="fa fa-check"></i><b>1.6</b> 许可</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#section-7"><i class="fa fa-check"></i><b>1.7</b> 准备知识</a></li>
<li class="chapter" data-level="1.8" data-path="index.html"><a href="index.html#section-8"><i class="fa fa-check"></i><b>1.8</b> 联系我们</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> scRNA-seq介绍</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-9"><i class="fa fa-check"></i><b>2.3</b> 工作流程</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-10"><i class="fa fa-check"></i><b>2.4</b> 计算分析</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-11"><i class="fa fa-check"></i><b>2.5</b> 挑战</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-12"><i class="fa fa-check"></i><b>2.6</b> 实验方法</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-13"><i class="fa fa-check"></i><b>2.7</b> 如何选择合适的平台</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html"><i class="fa fa-check"></i><b>3</b> scRNA-seq原始数据处理</a><ul>
<li class="chapter" data-level="3.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#fastqc"><i class="fa fa-check"></i><b>3.1</b> FastQC</a><ul>
<li class="chapter" data-level="3.1.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-14"><i class="fa fa-check"></i><b>3.1.1</b> 解决方案并下载报告</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-15"><i class="fa fa-check"></i><b>3.2</b> 移除接头和低质量碱基</a><ul>
<li class="chapter" data-level="3.2.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#solution"><i class="fa fa-check"></i><b>3.2.1</b> Solution</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-16"><i class="fa fa-check"></i><b>3.3</b> 文件格式</a><ul>
<li class="chapter" data-level="3.3.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#fastq"><i class="fa fa-check"></i><b>3.3.1</b> FastQ</a></li>
<li class="chapter" data-level="3.3.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#bam"><i class="fa fa-check"></i><b>3.3.2</b> BAM</a></li>
<li class="chapter" data-level="3.3.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#cram"><i class="fa fa-check"></i><b>3.3.3</b> CRAM</a></li>
<li class="chapter" data-level="3.3.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-17"><i class="fa fa-check"></i><b>3.3.4</b> 手动查看文件</a></li>
<li class="chapter" data-level="3.3.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#genome-fasta-gtf"><i class="fa fa-check"></i><b>3.3.5</b> Genome (FASTA, GTF)</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-18"><i class="fa fa-check"></i><b>3.4</b> 测序文库拆分</a><ul>
<li class="chapter" data-level="3.4.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-19"><i class="fa fa-check"></i><b>3.4.1</b> 鉴定含有细胞的液滴/微孔</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#starread"><i class="fa fa-check"></i><b>3.5</b> 使用STAR比对read</a><ul>
<li class="chapter" data-level="3.5.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#star"><i class="fa fa-check"></i><b>3.5.1</b> STAR比对命令</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallistopseudo-alignment"><i class="fa fa-check"></i><b>3.6</b> Kallisto和Pseudo-Alignment</a><ul>
<li class="chapter" data-level="3.6.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#k-mer"><i class="fa fa-check"></i><b>3.6.1</b> k-mer是什么?</a></li>
<li class="chapter" data-level="3.6.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#k-merreads"><i class="fa fa-check"></i><b>3.6.2</b> 为什么比对k-mer而不是reads?</a></li>
<li class="chapter" data-level="3.6.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto"><i class="fa fa-check"></i><b>3.6.3</b> Kallisto伪比对模式</a></li>
<li class="chapter" data-level="3.6.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto-pseudo-alignment"><i class="fa fa-check"></i><b>3.6.4</b> Kallisto Pseudo-Alignment命令</a></li>
<li class="chapter" data-level="3.6.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto-1"><i class="fa fa-check"></i><b>3.6.5</b> 理解Kallisto输出结果</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>4</b> 构建表达矩阵</a><ul>
<li class="chapter" data-level="4.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-20"><i class="fa fa-check"></i><b>4.1</b> 质量控制</a></li>
<li class="chapter" data-level="4.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-"><i class="fa fa-check"></i><b>4.2</b> Reads 比对</a></li>
<li class="chapter" data-level="4.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-21"><i class="fa fa-check"></i><b>4.3</b> 比对示例</a></li>
<li class="chapter" data-level="4.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-qc"><i class="fa fa-check"></i><b>4.4</b> Mapping QC</a></li>
<li class="chapter" data-level="4.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads"><i class="fa fa-check"></i><b>4.5</b> Reads定量</a></li>
<li class="chapter" data-level="4.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-22"><i class="fa fa-check"></i><b>4.6</b> 唯一分子标识符</a><ul>
<li class="chapter" data-level="4.6.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umi"><i class="fa fa-check"></i><b>4.6.1</b> UMI介绍</a></li>
<li class="chapter" data-level="4.6.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-23"><i class="fa fa-check"></i><b>4.6.2</b> 比对条形码序列</a></li>
<li class="chapter" data-level="4.6.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#barcodes"><i class="fa fa-check"></i><b>4.6.3</b> Barcodes计数</a></li>
<li class="chapter" data-level="4.6.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-24"><i class="fa fa-check"></i><b>4.6.4</b> 错误校正</a></li>
<li class="chapter" data-level="4.6.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-25"><i class="fa fa-check"></i><b>4.6.5</b> 下游分析</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html"><i class="fa fa-check"></i><b>5</b> R/Bioconductor介绍</a><ul>
<li class="chapter" data-level="5.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#r"><i class="fa fa-check"></i><b>5.1</b> 安装R包</a><ul>
<li class="chapter" data-level="5.1.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#cran"><i class="fa fa-check"></i><b>5.1.1</b> CRAN</a></li>
<li class="chapter" data-level="5.1.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#github-1"><i class="fa fa-check"></i><b>5.1.2</b> Github</a></li>
<li class="chapter" data-level="5.1.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor"><i class="fa fa-check"></i><b>5.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="5.1.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-26"><i class="fa fa-check"></i><b>5.1.4</b> 源码安装</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-27"><i class="fa fa-check"></i><b>5.2</b> 安装说明:</a></li>
<li class="chapter" data-level="5.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-28"><i class="fa fa-check"></i><b>5.3</b> 数据类型/类</a><ul>
<li class="chapter" data-level="5.3.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-29"><i class="fa fa-check"></i><b>5.3.1</b> 数值</a></li>
<li class="chapter" data-level="5.3.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-30"><i class="fa fa-check"></i><b>5.3.2</b> 字符/字符串</a></li>
<li class="chapter" data-level="5.3.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-31"><i class="fa fa-check"></i><b>5.3.3</b> 逻辑值</a></li>
<li class="chapter" data-level="5.3.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-32"><i class="fa fa-check"></i><b>5.3.4</b> 因子</a></li>
<li class="chapter" data-level="5.3.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-33"><i class="fa fa-check"></i><b>5.3.5</b> 检查数据类型</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-34"><i class="fa fa-check"></i><b>5.4</b> 基本数据结构</a></li>
<li class="chapter" data-level="5.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-35"><i class="fa fa-check"></i><b>5.5</b> 更多信息</a></li>
<li class="chapter" data-level="5.6" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-36"><i class="fa fa-check"></i><b>5.6</b> 数据类型</a><ul>
<li class="chapter" data-level="5.6.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-37"><i class="fa fa-check"></i><b>5.6.1</b> 什么是整洁的数据</a></li>
<li class="chapter" data-level="5.6.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#rich-data"><i class="fa fa-check"></i><b>5.6.2</b> 什么是Rich data?</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor-singlecellexperiment--scater"><i class="fa fa-check"></i><b>5.7</b> Bioconductor, <code>SingleCellExperiment</code> 和 <code>scater</code></a><ul>
<li class="chapter" data-level="5.7.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor-1"><i class="fa fa-check"></i><b>5.7.1</b> Bioconductor</a></li>
<li class="chapter" data-level="5.7.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#singlecellexperiment-"><i class="fa fa-check"></i><b>5.7.2</b> <code>SingleCellExperiment</code> 类</a></li>
<li class="chapter" data-level="5.7.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#scater-"><i class="fa fa-check"></i><b>5.7.3</b> <code>scater</code> 包</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2"><i class="fa fa-check"></i><b>5.8</b> ggplot2介绍</a><ul>
<li class="chapter" data-level="5.8.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2-1"><i class="fa fa-check"></i><b>5.8.1</b> ggplot2是什么?</a></li>
<li class="chapter" data-level="5.8.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2j"><i class="fa fa-check"></i><b>5.8.2</b> ggplot2j原则</a></li>
<li class="chapter" data-level="5.8.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#aes"><i class="fa fa-check"></i><b>5.8.3</b> 使用<code>aes</code>映射函数</a></li>
<li class="chapter" data-level="5.8.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#geoms"><i class="fa fa-check"></i><b>5.8.4</b> Geoms</a></li>
<li class="chapter" data-level="5.8.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-38"><i class="fa fa-check"></i><b>5.8.5</b> 绘制2个以上细胞的数据</a></li>
<li class="chapter" data-level="5.8.6" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-39"><i class="fa fa-check"></i><b>5.8.6</b> 绘制热图</a></li>
<li class="chapter" data-level="5.8.7" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-40"><i class="fa fa-check"></i><b>5.8.7</b> 主成分分析</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tabula-muris.html"><a href="tabula-muris.html"><i class="fa fa-check"></i><b>6</b> Tabula Muris</a><ul>
<li class="chapter" data-level="6.1" data-path="tabula-muris.html"><a href="tabula-muris.html#section-41"><i class="fa fa-check"></i><b>6.1</b> 介绍</a></li>
<li class="chapter" data-level="6.2" data-path="tabula-muris.html"><a href="tabula-muris.html#section-42"><i class="fa fa-check"></i><b>6.2</b> 下载数据</a></li>
<li class="chapter" data-level="6.3" data-path="tabula-muris.html"><a href="tabula-muris.html#smartseq2"><i class="fa fa-check"></i><b>6.3</b> 读取Smartseq2数据</a></li>
<li class="chapter" data-level="6.4" data-path="tabula-muris.html"><a href="tabula-muris.html#scater"><i class="fa fa-check"></i><b>6.4</b> 构建scater对象</a></li>
<li class="chapter" data-level="6.5" data-path="tabula-muris.html"><a href="tabula-muris.html#x"><i class="fa fa-check"></i><b>6.5</b> 读取10X数据</a></li>
<li class="chapter" data-level="6.6" data-path="tabula-muris.html"><a href="tabula-muris.html#scater-1"><i class="fa fa-check"></i><b>6.6</b> 创建scater对象</a></li>
<li class="chapter" data-level="6.7" data-path="tabula-muris.html"><a href="tabula-muris.html#section-43"><i class="fa fa-check"></i><b>6.7</b> 进阶练习</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html"><i class="fa fa-check"></i><b>7</b> 表达矩阵质控</a><ul>
<li class="chapter" data-level="7.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#exprs-qc"><i class="fa fa-check"></i><b>7.1</b> UMI表达质控</a><ul>
<li class="chapter" data-level="7.1.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-44"><i class="fa fa-check"></i><b>7.1.1</b> 介绍</a></li>
<li class="chapter" data-level="7.1.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tung"><i class="fa fa-check"></i><b>7.1.2</b> Tung数据集</a></li>
<li class="chapter" data-level="7.1.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-45"><i class="fa fa-check"></i><b>7.1.3</b> 细胞质控</a></li>
<li class="chapter" data-level="7.1.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-48"><i class="fa fa-check"></i><b>7.1.4</b> 细胞过滤</a></li>
<li class="chapter" data-level="7.1.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-51"><i class="fa fa-check"></i><b>7.1.5</b> 手工过滤和自动过滤比较</a></li>
<li class="chapter" data-level="7.1.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-52"><i class="fa fa-check"></i><b>7.1.6</b> 基因分析</a></li>
<li class="chapter" data-level="7.1.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-55"><i class="fa fa-check"></i><b>7.1.7</b> 保存数据</a></li>
<li class="chapter" data-level="7.1.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-56"><i class="fa fa-check"></i><b>7.1.8</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-1"><i class="fa fa-check"></i><b>7.2</b> Reads表达质控</a></li>
<li class="chapter" data-level="7.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-57"><i class="fa fa-check"></i><b>7.3</b> 数据可视化</a><ul>
<li class="chapter" data-level="7.3.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-58"><i class="fa fa-check"></i><b>7.3.1</b> 介绍</a></li>
<li class="chapter" data-level="7.3.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#visual-pca"><i class="fa fa-check"></i><b>7.3.2</b> PCA plot</a></li>
<li class="chapter" data-level="7.3.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#visual-tsne"><i class="fa fa-check"></i><b>7.3.3</b> tSNE可视化</a></li>
<li class="chapter" data-level="7.3.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-62"><i class="fa fa-check"></i><b>7.3.4</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-2"><i class="fa fa-check"></i><b>7.4</b> Reads数据可视化</a></li>
<li class="chapter" data-level="7.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>7.5</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="7.5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>7.5.1</b> Introduction</a></li>
<li class="chapter" data-level="7.5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#correlations-with-pcs"><i class="fa fa-check"></i><b>7.5.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="7.5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#explanatory-variables"><i class="fa fa-check"></i><b>7.5.3</b> Explanatory variables</a></li>
<li class="chapter" data-level="7.5.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#other-confounders"><i class="fa fa-check"></i><b>7.5.4</b> Other confounders</a></li>
<li class="chapter" data-level="7.5.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#exercise"><i class="fa fa-check"></i><b>7.5.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#identifying-confounding-factors-reads"><i class="fa fa-check"></i><b>7.6</b> Identifying confounding factors (Reads)</a></li>
<li class="chapter" data-level="7.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-theory"><i class="fa fa-check"></i><b>7.7</b> Normalization theory</a><ul>
<li class="chapter" data-level="7.7.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-1"><i class="fa fa-check"></i><b>7.7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.7.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#library-size"><i class="fa fa-check"></i><b>7.7.2</b> Library size</a></li>
<li class="chapter" data-level="7.7.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalisations"><i class="fa fa-check"></i><b>7.7.3</b> Normalisations</a></li>
<li class="chapter" data-level="7.7.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#effectiveness"><i class="fa fa-check"></i><b>7.7.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-practice-umi"><i class="fa fa-check"></i><b>7.8</b> Normalization practice (UMI)</a><ul>
<li class="chapter" data-level="7.8.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#raw"><i class="fa fa-check"></i><b>7.8.1</b> Raw</a></li>
<li class="chapter" data-level="7.8.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cpm-1"><i class="fa fa-check"></i><b>7.8.2</b> CPM</a></li>
<li class="chapter" data-level="7.8.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#scran-1"><i class="fa fa-check"></i><b>7.8.3</b> scran</a></li>
<li class="chapter" data-level="7.8.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#downsampling-1"><i class="fa fa-check"></i><b>7.8.4</b> Downsampling</a></li>
<li class="chapter" data-level="7.8.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalisation-for-genetranscript-length"><i class="fa fa-check"></i><b>7.8.5</b> Normalisation for gene/transcript length</a></li>
<li class="chapter" data-level="7.8.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#exercise-1"><i class="fa fa-check"></i><b>7.8.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-practice-reads"><i class="fa fa-check"></i><b>7.9</b> Normalization practice (Reads)</a></li>
<li class="chapter" data-level="7.10" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>7.10</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="7.10.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-2"><i class="fa fa-check"></i><b>7.10.1</b> Introduction</a></li>
<li class="chapter" data-level="7.10.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#remove-unwanted-variation"><i class="fa fa-check"></i><b>7.10.2</b> Remove Unwanted Variation</a></li>
<li class="chapter" data-level="7.10.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#combat"><i class="fa fa-check"></i><b>7.10.3</b> Combat</a></li>
<li class="chapter" data-level="7.10.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>7.10.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="7.10.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#glm"><i class="fa fa-check"></i><b>7.10.5</b> GLM</a></li>
<li class="chapter" data-level="7.10.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#harmony"><i class="fa fa-check"></i><b>7.10.6</b> Harmony</a></li>
<li class="chapter" data-level="7.10.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#how-to-evaluate-and-compare-confounder-removal-strategies"><i class="fa fa-check"></i><b>7.10.7</b> How to evaluate and compare confounder removal strategies</a></li>
<li class="chapter" data-level="7.10.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#big-exercise"><i class="fa fa-check"></i><b>7.10.8</b> Big Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7.11" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#dealing-with-confounders-reads"><i class="fa fa-check"></i><b>7.11</b> Dealing with confounders (Reads)</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>8</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/lemonbases/scRNA.seq.course.CN" target="blank">碱基吃柠檬</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNA-seq数据分析</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="processing_raw_scrna_seq_data" class="section level1">
<h1><span class="header-section-number">3</span> scRNA-seq原始数据处理</h1>
<div id="fastqc" class="section level2">
<h2><span class="header-section-number">3.1</span> FastQC</h2>
<p>获得单细胞RNA-seq数据后，首先要做的就是检查测序质量，我们使用FastQC来完成此任务。 FastQC是用于测序数据的质量控制工具，既可以用于bulk RNA-seq，也可以用于scRNA-seq。 FastQC将测序数据作为输入，并返回测序质量的报告。访问以下链接获取更多关于FastQC的信息：</p>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p>
<p>该网站包含下载和安装FastQC的链接以及生成报告的文档。幸运的是，我们已经为您安装了FastQC<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>，因此这里我们将查看文档。将网页向下滚动到“示例报告”，然后单击“Good Illumina Data”。 这里给出了高质量Illumina测序数据立项报告的示例。</p>
<p>现在让我们自己生成一份FastQC报告。</p>
<p>今天，我们将使用<span class="citation">(Kolodziejczyk et al. <a href="#ref-Kolodziejczyk2015-xy">2015</a>)</span>生成的mESC单细胞数据集进行分析。细胞使用SMART-seq2 protocol构建测序文库并进行双端测序。这些文件位于<code>Share</code>文件夹中。</p>
<p><strong>注意</strong> 本课程的当前文本是为参加我们课程的人员编写的。您必须下载文件（<code>ERR522959_1.fastq</code>和<code>ERR522959_2.fastq</code>）并创建<code>Share</code>目录才能运行命令。你可以在这里找到这些文件：</p>
<p><a href="https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2600/samples/">https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2600/samples/</a></p>
<p>现在让我们来看看文件:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">less</span> Share/ERR522959_1.fastq
<span class="fu">less</span> Share/ERR522959_2.fastq</code></pre>
<p><strong>任务1</strong>: 尝试找出生成FastQC报告的命令，提示: 尝试执行</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">fastqc</span> -h</code></pre>
<p>该命令将告诉您可以传递给FastQC的参数。如果您遇到困难，请随时寻求帮助！ 如果运行成功，则应为正向和反向reads 都会生成.zip和.html文件。运行成功后，请跳到下一节。</p>
<div id="section-14" class="section level3">
<h3><span class="header-section-number">3.1.1</span> 解决方案并下载报告</h3>
<p>如果还没有成功，请使用以下命令生成FastQC报告：</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> fastqc_results
<span class="ex">fastqc</span> -o fastqc_results Share/ERR522959_1.fastq Share/ERR522959_2.fastq</code></pre>
<p>一旦命令执行完毕，您应该总共有四个文件 - 每个read对应一个zip和html文件，报告位于html文件中。 要查看它，我们需要使用filezilla或scp将它从AWS下载到您的计算机上。</p>
<p>下载到本地后，单击打开您的FastQC报告。记得要查看正向和反向read的质量报告！测序的质量如何？有什么我们应该关注的问题吗？ 我们如何解决这些问题呢？</p>
</div>
</div>
<div id="section-15" class="section level2">
<h2><span class="header-section-number">3.2</span> 移除接头和低质量碱基</h2>
<p>Trim Galore是一个cutadapt的封装，用于移除测序接头序列和测序末端的低质量碱基。</p>
<p>鉴于FastQC报告中存在一些接头污染，最好从数据中移除接头序列。</p>
<p><strong>任务2</strong>：数据中使用了哪种类型的接头序列？提示：查看FastQC报告“Adapter Contern”图。</p>
<p>现在让我们使用Trim Galore移除那些有问题的接头序列，修剪后再次检查读取质量，使用FastQC生成另一个报告。</p>
<p><strong>任务3</strong>：找出移除adapter的命令。提示1：您可以使用</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">trim_galore</span> -h</code></pre>
<p>查看Trim Galore的参数描述。
提示2：仔细阅读上述命令的输出。本实验中使用的接头序列非常常见。您是否需要知道接头的实际序列才能将其删除？</p>
<p><strong>任务4</strong>: 为清洗后的数据生成FastQC报告。接头序列污染消失了吗？</p>
<p>一旦您认为您已成功去除接头序列并通过FastQC确认，请使用下一部分核验您的结果。</p>
<div id="solution" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Solution</h3>
<p>您可以使用以下命令去除Nextera测序接头序列：</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> fastqc_trimmed_results
<span class="ex">trim_galore</span> --nextera -o fastqc_trimmed_results Share/ERR522959_1.fastq Share/ERR522959_2.fastq</code></pre>
<p>请记住为清洗后的数据文件重新生成FastQC报告！FastQC现在应该显示’Adaptor Content’为’pass’了。</p>
<p>祝贺您现在已生成测序质量报告并移除接头序列。在下一步中，我们将使用STAR和Kallisto将清洗后的reads比对到参考转录组上。</p>
</div>
</div>
<div id="section-16" class="section level2">
<h2><span class="header-section-number">3.3</span> 文件格式</h2>
<div id="fastq" class="section level3">
<h3><span class="header-section-number">3.3.1</span> FastQ</h3>
<p>FastQ是scRNA-seq数据中最原始数据格式。所有scRNA-seq protocol都使用双端测序，根据使用的protocol，条形码序列(barcode)可能出现在paired-reads中一条或两条上。但是使用唯一分子标识符(UMIs) 的protocol会生成包含细胞和UMI条形码加上接头序列但是没有转录本序列的read。因此虽然是双端测序，但比对时按照单端测序对待。</p>
<p>FastQ文件格式如下:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="op">&gt;</span><span class="ex">ReadID</span>
<span class="ex">READ</span> SEQUENCE
<span class="ex">+</span>
<span class="ex">SEQUENCING</span> QUALITY SCORES</code></pre>
</div>
<div id="bam" class="section level3">
<h3><span class="header-section-number">3.3.2</span> BAM</h3>
<p>BAM文件以标准且高效的方式存储比对结果。SAM文件为直接可读的，而BAM文件是高度压缩的版本。BAM / SAM文件包含头部信息，通常包括样本制备，测序和比对的信息;后面为每个read的比对结果，以tab作为分隔符。</p>
<p>比对行标准格式如下:</p>
<ol style="list-style-type: decimal">
<li><p>QNAME : read名称(如果为UMI文库，则包括UMI条形码)</p></li>
<li><p>FLAG : 数字指示reads比对的类型, <a href="https://broadinstitute.github.io/picard/explain-flags.html">link</a>该网站有所有可能的类型</p></li>
<li><p>RNAME : 参考序列名称 (比如比对到的染色体名称).</p></li>
<li><p>POS : 最左边比对位置</p></li>
<li><p>MAPQ : 比对质量</p></li>
<li><p>CIGAR : 表示reads中匹配/不匹配部分 (可能包括soft-clipping).</p></li>
<li><p>RNEXT : mate/next reads比对到的参考序列名称</p></li>
<li><p>PNEXT : mate/next reads比对到的第一个碱基位置</p></li>
<li><p>TLEN : 模板长度（read比对到的参考区域的长度）</p></li>
<li><p>SEQ : read序列</p></li>
<li><p>QUAL : read质量</p></li>
</ol>
<p>BAM/SAM 文件可通过’samtools’互相转换:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">samtools</span> view -S -b file.sam <span class="op">&gt;</span> file.bam
<span class="ex">samtools</span> view -h file.bam <span class="op">&gt;</span> file.sam</code></pre>
<p>一些测序设备会自动将测序reads比对到标准基因组上，并提供BAM或CRAM格式文件。通常基因组中不包含ERCC序列，因此不会又ERCCs reads比对到在BAM / CRAM文件中。 要量化ERCC（或任何其他遗传变异），或者如果您只想使用不同于标准流程的比对算法（通常过时），那么您将需要将BAM / CRAM文件转换回FastQs:</p>
<p>BAM文件可以使用bedtools转为FastQ。为避免比对到多个基因组位置的一个read转换为FastQ多条read，首先将BAM文件按读取名称排序，并使用samtools删除次级比对。<a href="https://broadinstitute.github.io/picard/index.html">Picard</a>也包含将BAM转换为FastQ文件的方法。</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># sort reads by name</span>
<span class="ex">samtools</span> sort -n original.bam -o sorted_by_name.bam
<span class="co"># remove secondary alignments</span>
<span class="ex">samtools</span> view -b -F 256 sorted_by_name.bam -o primary_alignment_only.bam
<span class="co"># convert to fastq</span>
<span class="ex">bedtools</span> bamtofastq -i primary_alignment_only.bam -fq read1.fq -fq2 read2.fq</code></pre>
</div>
<div id="cram" class="section level3">
<h3><span class="header-section-number">3.3.3</span> CRAM</h3>
<p><a href="https://www.ebi.ac.uk/ena/software/cram-usage">CRAM</a>文件类似SAM文件，其头部信息包括比对使用的参考基因组信息，这使得read中和参考基因组一样的碱基可以进一步压缩。与BAM相比，CRAM还支持有损数据压缩方法以进一步优化存储。CRAM主要由Sanger/EBI测序机构使用。</p>
<p>CRAM和BAM文件可以使用最新版本的samtools（&gt; = v1.0）进行格式转换。但是，这种转换可能需要将参考基因组下载到缓存中。 或者，您可以从CRAM文件的头部元数据预先下载参考基因组，或者询问生成CRAM文件的人获得参考基因组，并使用’-T’指定该文件。因此我们建议在执行此操作之前设置特定的缓存位置：</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">export</span> <span class="va">REF_CACHE=</span>/path_to/cache_directory_for_reference_genome
<span class="ex">samtools</span> view -b -h -T reference_genome.fasta file.cram -o file.bam
<span class="ex">samtools</span> view -C -h -T reference_genome.fasta file.bam -o file.cram</code></pre>
</div>
<div id="section-17" class="section level3">
<h3><span class="header-section-number">3.3.4</span> 手动查看文件</h3>
<p>有时，手动检查文件可能很有用，例如检查文件的头部信息。’less’和’more’可在命令行查看任何文本文件。管道符<code>|</code>可以在多个命令之间传输数据，省却把中间数据存储多个拷贝的过程。</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">less</span> file.txt
<span class="fu">more</span> file.txt
<span class="co"># counts the number of lines in file.txt</span>
<span class="fu">wc</span> -l file.txt
<span class="ex">samtools</span> view -h file.[cram/bam] <span class="kw">|</span> <span class="fu">more</span>
<span class="co"># counts the number of lines in the samtools output</span>
<span class="ex">samtools</span> view -h file.[cram/bam] <span class="kw">|</span> <span class="fu">wc</span> -l</code></pre>
<p><strong>练习</strong></p>
<p>现提供cram示例文件: EXAMPLE.cram</p>
<p><strong>任务1</strong>: 这个文件是怎么生成的？使用了什么软件？参考基因组时什么？(提示: 检查头部信息)</p>
<p><strong>任务2</strong>: 有多少reads比对上/没有比对上？总共有多少reads？secondary alignments有多少? (提示: 使用FLAG)</p>
<p><strong>任务3</strong>: 将CRAM转为Fastq文件。转换后的read只有一个拷贝吗？(将转换后的Fastq文件命名为“10cells_read1.fastq” “10cells_read2.fastq”)</p>
<p>如果您遇到问题，可以通过输入命令来显示每个软件的帮助信息 - 例如 ‘samtools view’，‘bedtools’</p>
<p><strong>答案</strong></p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">samtools</span> view -T data/2000_reference.transcripts.fa -H data/EXAMPLE.cram <span class="kw">|</span> <span class="fu">more</span>
<span class="ex">samtools</span> view -T data/2000_reference.transcripts.fa -f 4 data/EXAMPLE.cram <span class="kw">|</span> <span class="fu">wc</span> -l <span class="co"># unmapped</span>
<span class="ex">samtools</span> view -T data/2000_reference.transcripts.fa -F 260 data/EXAMPLE.cram <span class="kw">|</span> <span class="fu">wc</span> -l <span class="co"># mapped</span>
<span class="ex">samtools</span> view -T data/2000_reference.transcripts.fa -F 256 data/EXAMPLE.cram <span class="kw">|</span> <span class="fu">wc</span> -l <span class="co"># total</span>
<span class="ex">samtools</span> view -T data/2000_reference.transcripts.fa -f 256 data/EXAMPLE.cram <span class="kw">|</span> <span class="fu">wc</span> -l <span class="co"># secondary alignments</span>

<span class="ex">samtools</span> view -b -h -T data/2000_reference.transcripts.fa data/EXAMPLE.cram -o data/EXAMPLE.bam
<span class="ex">samtools</span> sort -n data/EXAMPLE.bam -o data/sorted_EXAMPLE.bam
<span class="ex">samtools</span> view -b -F 256 data/sorted_EXAMPLE.bam -o data/primary_EXAMPLE.bam 
<span class="co"># convert to fastq</span>
<span class="ex">bedtools</span> bamtofastq -i data/primary_EXAMPLE.bam -fq data/10cells_read1.fq -fq2 data/10cells_read2.fq</code></pre>
</div>
<div id="genome-fasta-gtf" class="section level3">
<h3><span class="header-section-number">3.3.5</span> Genome (FASTA, GTF)</h3>
<p>为了比对序列，需要参考基因组和基因组注释文件(GTF或者GFF格式)。模式生物的基因组和注释文件可以从目前主流的基因组数据库下载:
<a href="http://www.ensembl.org/info/data/ftp/index.html">Ensembl</a>,
<a href="ftp://ftp.ncbi.nih.gov/genomes/">NCBI</a>, 或者 <a href="http://hgdownload.soe.ucsc.edu/downloads.html">UCSC Genome Browser</a>.</p>
<p>GTF文件包括基因，转录本和外显子的注释，格式如下：
(1) seqname : 染色体/scaffold编号
(2) source : 注释来源
(3) feature : 注释信息类型(比如基因，转录本，外显子)
(4) start : 起始位置 (bp)
(5) end : 终止 (bp)
(6) score : 得分
(7) strand : + (正链) or - (负链)
(8) frame : 仅对CDS有效，起始编码位置，或者到达下一个密码子需要跳过的碱基个数 (0 = first base, 1 = second base, etc..)
(9) attribute : <code>;</code>分割的键值对来显示其它信息 (比如 names/IDs, biotype)</p>
<p>空字段用“.”填充</p>
<p>根据我们的经验，Ensembl是最容易使用的，并且具有最大的注释集。NCBI往往更严格，仅包括置信度高的基因注释。 而UCSC包含多个使用不同标准的基因组注释。</p>
<p>如果您的实验系统包含非标准序列，则必须将这些序列添加到基因组fasta和gtf中来定量它们的表达。 最常见的是针对ERCC spike-ins，CRISPR相关的序列或其他过表达/报告载体。</p>
<p>为了获得最大的可用性/灵活性，我们建议为添加的任何非标准序列创建完整和详细的fasta序列和gtf序列。</p>
<p>目前没有标准化的方法来做到这一点 以下是我们的自定义perl脚本，用于为ERCC创建一个gtf和fasta文件，可以将其附加到基因组中。如果要量化内含子读数时，您可能还需要更改gtf文件以处理内含子中的重复元素。任何脚本语言甚至“awk”或一些文本编辑器都可以用来相对有效地完成这项任务，但它们超出了本课程的范围。</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># Converts the Annotation file from </span>
<span class="co"># https://www.thermofisher.com/order/catalog/product/4456740 to </span>
<span class="co"># gtf and fasta files that can be added to existing genome fasta &amp; gtf files.</span>

<span class="ex">my</span> @FASTAlines = ();
<span class="ex">my</span> @GTFlines = ();
<span class="ex">open</span> (my <span class="va">$ifh</span>, <span class="st">&quot;ERCC_Controls_Annotation.txt&quot;</span>) <span class="ex">or</span> die <span class="va">$!</span><span class="kw">;</span>
<span class="op">&lt;</span><span class="va">$ifh</span><span class="op">&gt;</span>; <span class="co">#header</span>
<span class="kw">while</span> <span class="kw">(</span><span class="op">&lt;</span><span class="va">$ifh</span><span class="op">&gt;</span><span class="kw">)</span> <span class="kw">{</span>
    <span class="co"># Do all the important stuff</span>
    <span class="ex">chomp</span><span class="kw">;</span>
    <span class="ex">my</span> @record = split(/\t/);
    <span class="ex">my</span> <span class="va">$sequence</span> = <span class="va">$record[4]</span><span class="kw">;</span>
    <span class="va">$sequence</span> =<span class="ex">~</span> <span class="ex">s</span>/\<span class="ex">s+//g</span>; <span class="co"># get rid of any preceeding/tailing white space</span>
    <span class="va">$sequence</span> = <span class="va">$sequence</span><span class="ex">.</span><span class="st">&quot;NNNN&quot;</span><span class="kw">;</span>
    <span class="ex">my</span> <span class="va">$name</span> = <span class="va">$record[0]</span><span class="kw">;</span>
    <span class="ex">my</span> <span class="va">$genbank</span> = <span class="va">$record[1]</span><span class="kw">;</span>
    <span class="ex">push</span>(@FASTAlines, <span class="st">&quot;&gt;</span><span class="va">$name</span><span class="st">\n</span><span class="va">$sequence</span><span class="st">\n&quot;</span>);
<span class="co"># is GTF 1 indexed or 0 indexed? -&gt; it is 1 indexed</span>
<span class="co"># + or - strand?</span>
    <span class="ex">push</span>(@GTFlines, <span class="st">&quot;</span><span class="va">$name</span><span class="st">\tERCC\tgene\t1\t&quot;</span>.(length(<span class="va">$sequence</span>)<span class="ex">-2</span>)<span class="ex">.</span><span class="st">&quot;\t.\t+\t.\tgene_id </span><span class="dt">\&quot;</span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">; transcript_id </span><span class="dt">\&quot;</span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">; exon_number </span><span class="dt">\&quot;</span><span class="st">1</span><span class="dt">\&quot;</span><span class="st">; gene_name </span><span class="dt">\&quot;</span><span class="st">ERCC </span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">\n&quot;</span>);
    <span class="ex">push</span>(@GTFlines, <span class="st">&quot;</span><span class="va">$name</span><span class="st">\tERCC\ttranscript\t1\t&quot;</span>.(length(<span class="va">$sequence</span>)<span class="ex">-2</span>)<span class="ex">.</span><span class="st">&quot;\t.\t+\t.\tgene_id </span><span class="dt">\&quot;</span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">; transcript_id </span><span class="dt">\&quot;</span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">; exon_number </span><span class="dt">\&quot;</span><span class="st">1</span><span class="dt">\&quot;</span><span class="st">; gene_name </span><span class="dt">\&quot;</span><span class="st">ERCC </span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">\n&quot;</span>);
    <span class="ex">push</span>(@GTFlines, <span class="st">&quot;</span><span class="va">$name</span><span class="st">\tERCC\texon\t1\t&quot;</span>.(length(<span class="va">$sequence</span>)<span class="ex">-2</span>)<span class="ex">.</span><span class="st">&quot;\t.\t+\t.\tgene_id </span><span class="dt">\&quot;</span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">; transcript_id </span><span class="dt">\&quot;</span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">; exon_number </span><span class="dt">\&quot;</span><span class="st">1</span><span class="dt">\&quot;</span><span class="st">; gene_name </span><span class="dt">\&quot;</span><span class="st">ERCC </span><span class="va">$name</span><span class="st">-</span><span class="va">$genbank</span><span class="dt">\&quot;</span><span class="st">\n&quot;</span>);
<span class="kw">}</span> <span class="ex">close</span>(<span class="va">$ifh</span>);

<span class="co"># Write output</span>
<span class="ex">open</span>(my <span class="va">$ofh</span>, <span class="st">&quot;&gt;&quot;</span>, <span class="st">&quot;ERCC_Controls.fa&quot;</span>) <span class="ex">or</span> die <span class="va">$!</span><span class="kw">;</span>
<span class="ex">foreach</span> my <span class="va">$line</span> (@FASTAlines) <span class="kw">{</span>
    <span class="ex">print</span> <span class="va">$ofh</span> <span class="va">$line</span><span class="kw">;</span>
<span class="kw">}</span> <span class="ex">close</span> (<span class="va">$ofh</span>);

<span class="ex">open</span>(<span class="va">$ofh</span>, <span class="st">&quot;&gt;&quot;</span>, <span class="st">&quot;ERCC_Controls.gtf&quot;</span>) <span class="ex">or</span> die <span class="va">$!</span><span class="kw">;</span>
<span class="ex">foreach</span> my <span class="va">$line</span> (@GTFlines) <span class="kw">{</span>
    <span class="ex">print</span> <span class="va">$ofh</span> <span class="va">$line</span><span class="kw">;</span>
<span class="kw">}</span> <span class="ex">close</span> (<span class="va">$ofh</span>);</code></pre>
</div>
</div>
<div id="section-18" class="section level2">
<h2><span class="header-section-number">3.4</span> 测序文库拆分</h2>
<p>文库拆分根据Protocol不同或构建的流程不同需要有对应的处理方式。我们所知道的最灵活的文库拆分工具是<a href="https://github.com/sdparekh/zUMIs/wiki/Usage">zUMIs</a>，可用于拆分和比对大多数基于UMI的protocol。对于Smartseq2或其他全长转录本双端测序protocol，数据通常已经被拆分好。诸如GEO或ArrayExpress之类的公共存储库需要在上传之前对基于小规模/基于板的scRNASeq数据拆分，并且许多测序公司在数据返回给您之前自动拆分数据。如果您没有使用已发表的流程，并且数据未被测序公司拆分，则您必须自己对其进行文库拆分。因为不同的建库方案引入的barcode序列的长度和位置不同，通常都需要自己写脚本解决。</p>
<p>对于所有数据类型，文库拆分涉及从一端或双端短序列中识别并移除细胞条形码序列(cell-barcode)。如果提前知道加入的cell-barcodes，比如数据来自基于PCR板的protocol，需要将每个cell-barcode与预期的cell-barcode进行比对，并将其归类于最相近的cell-barcode(根据cell-barcode的设计，一般允许最多1-2错配)。这些数据通常在比对之前进行拆分，从而可以并行比对。</p>
<p>我们提供公开可用<a href="https://github.com/tallulandrews/scRNASeqPipeline">perl</a>脚本，可以拆分任何有或没有UMI 的plate-based的建库方案生成的数据，用法如下：</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">perl</span> utils/1_Flexible_UMI_Demultiplexing.pl data/10cells_read1.fq data/10cells_read2.fq <span class="st">&quot;C12U8&quot;</span> data/10cells_barcodes.txt 2 Ex</code></pre>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">Barcode</span> Structure: 12 bp CellID followed by 8 bp UMI

        <span class="ex">Doesn</span><span class="st">&#39;t match any cell: 0</span>
<span class="st">        Ambiguous: 0</span>
<span class="st">        Exact Matches: 400</span>
<span class="st">        Contain mismatches: 0</span>
<span class="st">        Input Reads: 400</span>
<span class="st">        Output Reads: 400</span></code></pre>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">perl</span> utils/1_Flexible_FullTranscript_Demultiplexing.pl data/10cells_read1.fq data/10cells_read2.fq <span class="st">&quot;start&quot;</span> 12 data/10cells_barcodes.txt 2 Ex</code></pre>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">Doesn</span><span class="st">&#39;t match any cell: 0</span>
<span class="st">Ambiguous: 0</span>
<span class="st">Exact Matches: 400</span>
<span class="st">Contain Mismatches: 0</span>
<span class="st">Input Reads: 400</span>
<span class="st">Output Reads: 400</span></code></pre>
<p>对于包含UMI的数据，文库拆分包括将UMI code附加到包含基因区的序列read名字上。如果数据来自droplet-based protocol或SeqWell，其中预期条形码的数量远远高于预期的细胞数量，为避免生成才能大量的文件么cell-barcode也加到测序read的名字后面。在这些情况下，在量化步骤期间进行文库拆分，以便于识别来源于完整细胞而不是背景噪声的cell-barcode。</p>
<div id="section-19" class="section level3">
<h3><span class="header-section-number">3.4.1</span> 鉴定含有细胞的液滴/微孔</h3>
<p>基于液滴的scRNA-seq方法，只有一部分液滴包含bead和一个完整的细胞。然而生物实验可能不理想，一些RNA会从死细胞/受伤细胞中泄露出去。因此，没有完整细胞的液滴可能捕获少量环境游离RNA，这些RNA将进入测序文库，出现在最终测序结果中。液滴大小，扩增效率和测序的变化将导致“背景”和真实细胞文库大小区别很大。目前已有很多方法用来区分对应真实细胞的cell-barcode 已经使用各种方法来试图区分对应于真实细胞的那些细胞条形码。</p>
<p>大多数方法使用每个barcode的总分子数（可以应用于总reads）并尝试寻找“break point”，区分来自真实细胞较大的文库和来自背景较小的文库。下面加载包含大小文库细胞的示例模拟数据：</p>
<pre class="sourceCode r"><code class="sourceCode r">umi_per_barcode &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/droplet_id_example_per_barcode.txt.gz&quot;</span>)
truth &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/droplet_id_example_truth.gz&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;,&quot;</span>)</code></pre>
<p><strong>练习</strong></p>
<p>多少唯一的barcode被检测到？
数据中多少来自真实的细胞？
为简化计算，去除所有少于10个分子的barcode。</p>
<p><strong>答案</strong></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(umi_per_barcode)[<span class="dv">1</span>]
<span class="kw">dim</span>(truth)[<span class="dv">1</span>]</code></pre>
<p>一种方法是寻找每个条形码对应总分子突然下降的拐点。
One approach is to look for the inflection point where the total molecules per
barcode suddenly drops:</p>
<pre class="sourceCode r"><code class="sourceCode r">barcode_rank &lt;-<span class="st"> </span><span class="kw">rank</span>(<span class="op">-</span>umi_per_barcode[,<span class="dv">2</span>])
<span class="kw">plot</span>(barcode_rank, umi_per_barcode[,<span class="dv">2</span>], <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">8000</span>))</code></pre>
<p><img src="03process-raw_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p>可以看出文库大小近似指数分布，简单起见，对数据进行log转换。</p>
<pre class="sourceCode r"><code class="sourceCode r">log_lib_size &lt;-<span class="st"> </span><span class="kw">log10</span>(umi_per_barcode[,<span class="dv">2</span>])
<span class="kw">plot</span>(barcode_rank, log_lib_size, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">8000</span>))</code></pre>
<p><img src="03process-raw_files/figure-html/unnamed-chunk-21-1.png" width="672" /></p>
<p>从上图可以看出，拐点更加明显了。我们可以手动估计拐点的位置，但是用算法估计更加精确，以及可重复。</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># inflection point</span>
o &lt;-<span class="st"> </span><span class="kw">order</span>(barcode_rank)
log_lib_size &lt;-<span class="st"> </span>log_lib_size[o]
barcode_rank &lt;-<span class="st"> </span>barcode_rank[o]

rawdiff &lt;-<span class="st"> </span><span class="kw">diff</span>(log_lib_size)<span class="op">/</span><span class="kw">diff</span>(barcode_rank)
inflection &lt;-<span class="st"> </span><span class="kw">which</span>(rawdiff <span class="op">==</span><span class="st"> </span><span class="kw">min</span>(rawdiff[<span class="dv">100</span><span class="op">:</span><span class="kw">length</span>(rawdiff)], <span class="dt">na.rm=</span><span class="ot">TRUE</span>))

<span class="kw">plot</span>(barcode_rank, log_lib_size, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">8000</span>))
<span class="kw">abline</span>(<span class="dt">v=</span>inflection, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</code></pre>
<p><img src="03process-raw_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">threshold &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span>log_lib_size[inflection]

cells &lt;-<span class="st"> </span>umi_per_barcode[umi_per_barcode[,<span class="dv">2</span>] <span class="op">&gt;</span><span class="st"> </span>threshold,<span class="dv">1</span>]
TPR &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(cells)
Recall &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(truth[,<span class="dv">1</span>])
<span class="kw">c</span>(TPR, Recall)</code></pre>
<pre><code>## [1] 1.0000000 0.7831707</code></pre>
<p>另外一种方法是构建混合模型，找出重叠区域的最高和最低分布。然而数据可能并不满足假定的分布。</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="op">-</span><span class="dv">92497</span>)
<span class="co"># mixture model</span>
<span class="kw">require</span>(<span class="st">&quot;mixtools&quot;</span>)</code></pre>
<pre><code>## 载入需要的程辑包：mixtools</code></pre>
<pre><code>## mixtools package, version 1.1.0, Released 2017-03-10
## This package is based upon work supported by the National Science Foundation under Grant No. SES-0518772.</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">mix &lt;-<span class="st"> </span><span class="kw">normalmixEM</span>(log_lib_size)</code></pre>
<pre><code>## number of iterations= 43</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(mix, <span class="dt">which=</span><span class="dv">2</span>, <span class="dt">xlab2=</span><span class="st">&quot;log(mol per cell)&quot;</span>)</code></pre>
<p><img src="03process-raw_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw">dnorm</span>(log_lib_size, <span class="dt">mean=</span>mix<span class="op">$</span>mu[<span class="dv">1</span>], <span class="dt">sd=</span>mix<span class="op">$</span>sigma[<span class="dv">1</span>])
p2 &lt;-<span class="st"> </span><span class="kw">dnorm</span>(log_lib_size, <span class="dt">mean=</span>mix<span class="op">$</span>mu[<span class="dv">2</span>], <span class="dt">sd=</span>mix<span class="op">$</span>sigma[<span class="dv">2</span>])
<span class="cf">if</span> (mix<span class="op">$</span>mu[<span class="dv">1</span>] <span class="op">&lt;</span><span class="st"> </span>mix<span class="op">$</span>mu[<span class="dv">2</span>]) {
    split &lt;-<span class="st"> </span><span class="kw">min</span>(log_lib_size[p2 <span class="op">&gt;</span><span class="st"> </span>p1])
} <span class="cf">else</span> {
    split &lt;-<span class="st"> </span><span class="kw">min</span>(log_lib_size[p1 <span class="op">&gt;</span><span class="st"> </span>p2])
}</code></pre>
<p><strong>练习</strong></p>
<p>使用分割点识别细胞并计算TPR和Recall</p>
<p><strong>答案</strong></p>
<pre class="sourceCode r"><code class="sourceCode r">cells &lt;-<span class="st"> </span>umi_per_barcode[umi_per_barcode[,<span class="dv">2</span>] <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span><span class="op">^</span>split,<span class="dv">1</span>]

TPR &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(cells)
Recall &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(truth[,<span class="dv">1</span>])
<span class="kw">c</span>(TPR, Recall)</code></pre>
<p>第三种方法，CellRanger假设真实细胞文库大小变化范围在10倍以内，然后用哪个期望的细胞数目估计区间的分布。</p>
<pre class="sourceCode r"><code class="sourceCode r">n_cells &lt;-<span class="st"> </span><span class="kw">length</span>(truth[,<span class="dv">1</span>])
<span class="co"># CellRanger</span>
totals &lt;-<span class="st"> </span>umi_per_barcode[,<span class="dv">2</span>]
totals &lt;-<span class="st"> </span><span class="kw">sort</span>(totals, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)
<span class="co"># 99th percentile of top n_cells divided by 10</span>
thresh =<span class="st"> </span>totals[<span class="kw">round</span>(<span class="fl">0.01</span><span class="op">*</span>n_cells)]<span class="op">/</span><span class="dv">10</span>
<span class="kw">plot</span>(totals, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">8000</span>))
<span class="kw">abline</span>(<span class="dt">h=</span>thresh, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</code></pre>
<p><img src="03process-raw_files/figure-html/unnamed-chunk-25-1.png" width="672" />
<strong>练习</strong></p>
<p>用该阈值识别细胞并计算TPR和Recall。</p>
<p><strong>答案</strong></p>
<pre class="sourceCode r"><code class="sourceCode r">cells &lt;-<span class="st"> </span>umi_per_barcode[umi_per_barcode[,<span class="dv">2</span>] <span class="op">&gt;</span><span class="st"> </span>thresh,<span class="dv">1</span>]

TPR &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(cells)
Recall &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(truth[,<span class="dv">1</span>])
<span class="kw">c</span>(TPR, Recall)</code></pre>
<p>最后<a href="https://github.com/MarioniLab/DropletUtils">EmptyDrops</a>，目前处于Beta测试阶段，其实用所有液滴的基因 × 细胞分子计数矩阵，从具有极低counts的液滴估计背景RNA的分布，然后寻找与背景基因表达模式不同的细胞。背景RNA通常和群体中大部分细胞的表达模式相似，该方法与拐点方法相结合。因此，EmptyDrops是唯一能够识别高度多样化样本中鉴定小群体细胞条形码的方法。</p>
<p>下面提供了该方法的运行代码：（我们会根据官方发布及时更新代码）</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(<span class="st">&quot;Matrix&quot;</span>)
raw.counts &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/droplet_id_example.rds&quot;</span>)

<span class="kw">require</span>(<span class="st">&quot;DropletUtils&quot;</span>)
<span class="co"># emptyDrops</span>
<span class="kw">set.seed</span>(<span class="dv">100</span>)
e.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(my.counts)
is.cell &lt;-<span class="st"> </span>e.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.01</span>
<span class="kw">sum</span>(is.cell, <span class="dt">na.rm=</span><span class="ot">TRUE</span>)
<span class="kw">plot</span>(e.out<span class="op">$</span>Total, <span class="op">-</span>e.out<span class="op">$</span>LogProb, <span class="dt">col=</span><span class="kw">ifelse</span>(is.cell, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;black&quot;</span>),
    <span class="dt">xlab=</span><span class="st">&quot;Total UMI count&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;-Log Probability&quot;</span>)

cells &lt;-<span class="st"> </span><span class="kw">colnames</span>(raw.counts)[is.cell]

TPR &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(cells)
Recall &lt;-<span class="st"> </span><span class="kw">sum</span>(cells <span class="op">%in%</span><span class="st"> </span>truth[,<span class="dv">1</span>])<span class="op">/</span><span class="kw">length</span>(truth[,<span class="dv">1</span>])
<span class="kw">c</span>(TPR, Recall)</code></pre>
</div>
</div>
<div id="starread" class="section level2">
<h2><span class="header-section-number">3.5</span> 使用STAR比对read</h2>
<p>现在我们已经对测序数据进行质控，并且获得了高质量的清洗后数据，下一步将其比对到参考基因组上。如果我们想要定量基因表达或筛选样品间差异表达的基因，则通常需要某种形式的比对。</p>
<p>短序列比对的工具有很多，目前我们主要关注两个。第一个工具是STAR <span class="citation">(Dobin et al. <a href="#ref-dobin2013star">2013</a>)</span>，对于测序数据中的每一个read，STAR尝试寻找参考基因组中一个或多个最长可能序列。例如，下图所示，read(蓝色)跨越两个外显子和一个可变剪切位点(紫色)。STAR能够找出read的第一部分和序列的第一个外显子仙童，同时第二部分和序列的第二个外显子相同。因为STAR可能够以这种方式识别剪切事件，又被称为“spice aware alinger”。</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-28"></span>
<img src="figures/STAR_explanation.png" alt="STAR比对示意图, 来自Dobin et al." width="80%" />
<p class="caption">
Figure 3.1: STAR比对示意图, 来自Dobin et al.
</p>
</div>
<p>通常，STAR将reads比对到参考基因组是允许其检测新的剪接事件或染色体重排。然而，STAR的一个问题是它需要大量的内存，特别是当参考基因组很大（例如，小鼠和人类）。为了加速我们今天的分析，我们将使用STAR将reads与只包含2000个转录本的参考转录组进行比对。请注意，这不是正常或推荐的做法，我们只是出于时间原因这样做。我们建议比对到参考基因组。</p>
<p>STAR对齐需要两个步骤。第一步，用户向STAR提供参考基因组序列（FASTA）和注释文件（GTF），STAR用它来创建基因组索引。 第二步，STAR将用户的reads比对到基因组索引。</p>
<p>首先现在创建索引。请记住，由于时间的原因，我们比对到转录组而不是基因组，这意味着我们只需要向STAR提供比对的转录本序列。可以从<a href="https://www.ensembl.org/info/data/ftp/index.html">Ensembl</a>获取许多模式生物的转录组数据。</p>
<p>任务 1: 执行以下命令创建索引:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> indices
<span class="fu">mkdir</span> indices/STAR
<span class="ex">STAR</span> --runThreadN 4 --runMode genomeGenerate --genomeDir indices/STAR --genomeFastaFiles Share/2000_reference.transcripts.fa</code></pre>
<p>任务 2: STAR用到的每个参数意义? 提示: 参考<a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR帮助文档</a></p>
<p>任务 3: 如果比对到基因组而不是转录组，命令是什么？</p>
<p>现在索引创建完成，进行比对步骤。</p>
<p>任务 4: 尝试找出应该使用什么命令将我们的清洗后的数据（从ERR522959）比对索引上。参考使用STAR手册， 您认为知道答案，检查它是否与下一节中的解决方案匹配并执行比对。</p>
<p>任务 5: 尝试了解比对结果文件。</p>
<div id="star" class="section level3">
<h3><span class="header-section-number">3.5.1</span> STAR比对命令</h3>
<p>使用以下命令完成比对:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> results
<span class="fu">mkdir</span> results/STAR

<span class="ex">STAR</span> --runThreadN 4 --genomeDir indices/STAR --readFilesIn Share/ERR522959_1.fastq Share/ERR522959_2.fastq --outFileNamePrefix results/STAR/</code></pre>
</div>
</div>
<div id="kallistopseudo-alignment" class="section level2">
<h2><span class="header-section-number">3.6</span> Kallisto和Pseudo-Alignment</h2>
<p>STAR是read比对工具，而Kallisto是伪比对工具<span class="citation">(Bray et al. <a href="#ref-bray_2016">2016</a>)</span>。它们的主要区别是aligner比对到参考基因组或转录组上，而pseudo-aligner将k-mers比对到参考基因组或转录组上。</p>
<div id="k-mer" class="section level3">
<h3><span class="header-section-number">3.6.1</span> k-mer是什么?</h3>
<p>k-mer是来自reads长度为k的序列。例如，假设read为ATCCCGGGTTAT，从中制作7-mer。为此，提取read的前七个碱基找到第一个7-mer。然后向下移动一个碱基得到第二个7-mer，然后计算接下来的七个碱基。图2显示了可以从read得到的所有7-mers：</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-31"></span>
<img src="figures/Kmers.png" alt="示例read产生7-mer"  />
<p class="caption">
Figure 3.2: 示例read产生7-mer
</p>
</div>
</div>
<div id="k-merreads" class="section level3">
<h3><span class="header-section-number">3.6.2</span> 为什么比对k-mer而不是reads?</h3>
<p>主要有两个原因：</p>
<ol style="list-style-type: decimal">
<li><p>psudi-aligners使用计算技巧使得比对k-mers比传统比对reads要快，具体细节参考(Bray et al., 2017)</p></li>
<li><p>在一些情况下，pseudo-aligners比对传统的比对工具更好地处理测序错误。比如，假设序列的第一个碱基存在测序错误，实际为T但是测序为A，这会影响pseduo-aligners比对第一个7-mer，而不影响后面7-mers的比对。</p></li>
</ol>
</div>
<div id="kallisto" class="section level3">
<h3><span class="header-section-number">3.6.3</span> Kallisto伪比对模式</h3>
<p>Kallisto有一个为scRNA-seq实验特别设计的伪比对模式。与STAR不同的是，Kallisto比对到参考转录组而不是参考基因组。 这意味着Kallisto将reads比对到剪接异构体而不是基因，然而对scRNA-seq而言，这很有挑战性： 由于以下原因，对单细胞RNA-seq特异性地将读数映射到同种型而不是基因：</p>
<ul>
<li>scRNA-seq覆盖率比bulk RNA-seq低，意味着从reads获得的信息减少。</li>
<li>许多scRNA-seq protocol 在3’端覆盖存在偏差，意味着如果两个异构体只在5’端不同，则很难确定read来自哪个异构体。</li>
<li>一些scRNA-seq protocol测序读段较短，难以区分read来自哪个异构体。</li>
</ul>
<p>Kallisto的<code>pseudo mode</code>和<code>pseudo alignment</code>略有不同。Kallisto不与异构体比对，而是与等价类(equivalence classes)比对。如果read比对到多个异构体上，Kallisto会记录read比对到包含有所有异构体的等价类，因此可以使用等价类计数而非基因或转录本计数用于下游的聚类等分析。具体见下图：</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-32"></span>
<img src="figures/TCC.jpg" alt="Kallisto等价类示意图, 来自 https://pachterlab.github.io/kallisto/singlecell.html."  />
<p class="caption">
Figure 3.3: Kallisto等价类示意图, 来自 <a href="https://pachterlab.github.io/kallisto/singlecell.html" class="uri">https://pachterlab.github.io/kallisto/singlecell.html</a>.
</p>
</div>
<p>今天我们只用1个细胞进行pseudo-alignment, 但是Kallisto可以同时对大量细胞进行pseudo-alignment以及使用UMI信息，具体参看<a href="https://pachterlab.github.io/kallisto/manual">官方文档</a></p>
<p>类似STAR，pseudo-alignment之前需要用Kallisto生成索引。</p>
<p>任务 6: 使用以下命令构建Kallisto索引, 使用<a href="https://pachterlab.github.io/kallisto/manual">官方手册</a>理解每个参数的意义</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> indices/Kallisto
<span class="ex">kallisto</span> index -i indices/Kallisto/transcripts.idx Share/2000_reference.transcripts.fa</code></pre>
<p>任务 7: 使用Kallisto手册找出运行pseudo-alignment的命令，如果您认为知道答案，可以去下一节核验并运行pseudo-alignment。</p>
</div>
<div id="kallisto-pseudo-alignment" class="section level3">
<h3><span class="header-section-number">3.6.4</span> Kallisto Pseudo-Alignment命令</h3>
<p>使用以下命令执行pseudo-alignment</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> results/Kallisto
<span class="ex">kallisto</span> pseudo -i indices/Kallisto/transcripts.idx -o results/Kallisto -b batch.txt </code></pre>
<p>参考<a href="https://pachterlab.github.io/kallisto/manual">官方手册</a>构建batch.txt。</p>
</div>
<div id="kallisto-1" class="section level3">
<h3><span class="header-section-number">3.6.5</span> 理解Kallisto输出结果</h3>
<p>上述命令会生成4个文件 - matrix.cells, matrix.ec, matrix.tsv and run_info.json.</p>
<ul>
<li>matrix.cells 细胞ID列表. 因为我们只使用一个细胞，该文件应该只包含“ERR522959”</li>
<li>matrix.ec 等价类信息。每一行的第一个数字为等价类ID，第二个数字对应等价类的转录本ID。比如，“10 1,2,3”表示等价类10包括3个转录本，分别是1,2,3. ID顺序对应转录本在参考转录本的顺序。采用0-base技术，转录本ID为1,2,3分别对应2000_reference.transcripts.fa中第2,3,4个转录本。</li>
<li>matrix.tsv 每个等价类在不同cell的read count信息。第一个数字为等价类ID，如matrix.ec一样。第二个数字为cell ID，和matrix.cells文件中细胞名字对应。第三个数字等价类在该细胞的reads count。比如，“5 1 3”表示来自细胞1的3个reads比对到等价类5上。这里的细胞ID同样是0-based，所以细胞1对应matrix.cells的第二行。
Note that zero indexing is used, so cell 1 corresponds to the second line of matrix.cells.</li>
<li>run_info.json Kallisto执行信息，一般可忽略。</li>
</ul>

</div>
</div>
</div>
<h3> References</h3>
<div id="refs" class="references">
<div id="ref-bray_2016">
<p>Bray, Nicolas L, Harold Pimentel, Páll Melsted, and Lior Pachter. 2016. “Near-Optimal Probabilistic Rna-Seq Quantification.” <em>Nat Biotechnol</em> 34 (5): 525–27. <a href="https://doi.org/10.1038/nbt.3519">https://doi.org/10.1038/nbt.3519</a>.</p>
</div>
<div id="ref-dobin2013star">
<p>Dobin, Alexander, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. 2013. “STAR: Ultrafast Universal Rna-Seq Aligner.” <em>Bioinformatics</em> 29 (1): 15–21.</p>
</div>
<div id="ref-Kolodziejczyk2015-xy">
<p>Kolodziejczyk, Aleksandra A., Jong Kyoung Kim, Valentine Svensson, John C. Marioni, and Sarah A. Teichmann. 2015. “The Technology and Biology of Single-Cell RNA Sequencing.” <em>Molecular Cell</em> 58 (4): 610–20. <a href="https://doi.org/10.1016/j.molcel.2015.04.005">https://doi.org/10.1016/j.molcel.2015.04.005</a>.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="1">
<li id="fn1"><p><em>译者注:</em> Docker镜像中已经安装FastQC<a href="processing-raw-scrna-seq-data.html#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-single-cell-rna-seq.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="construction-of-expression-matrix.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": true,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/lemonbases/scRNA.seq.course.CN/edit/master/03process-raw.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
