<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Seurat | scRNA-seq数据分析</title>
  <meta name="description" content="9 Seurat | scRNA-seq数据分析" />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Seurat | scRNA-seq数据分析" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Seurat | scRNA-seq数据分析" />
  
  
  

<meta name="author" content="碱基吃柠檬" />


<meta name="date" content="2019-07-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="biological-analysis.html">
<link rel="next" href="Ideal-scRNA-seq-pipeline.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.46.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.46.1/plotly-latest.min.js"></script>
<!-- for Facebook -->  
<meta property="og:url" content="https://github.com/sdbiodog" />
<meta property="og:description" content="本项目是将github上经典的scRNA-seq分析教程进行汉化，源项目参见hemberg-lab/scRNA.seq.course" />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="9 Seurat | scRNA-seq数据分析" />
<meta name="twitter:description" content="本项目是将github上经典的scRNA-seq分析教程进行汉化，源项目参见hemberg-lab/scRNA.seq.course" />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-71525309-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> 说在前面的话</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#section-1"><i class="fa fa-check"></i><b>1.1</b> 关于本项目</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#section-2"><i class="fa fa-check"></i><b>1.2</b> 视频</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#docker-"><i class="fa fa-check"></i><b>1.4</b> Docker 镜像</a><ul>
<li class="chapter" data-level="1.4.1" data-path="index.html"><a href="index.html#section-3"><i class="fa fa-check"></i><b>1.4.1</b> 运行镜像</a></li>
<li class="chapter" data-level="1.4.2" data-path="index.html"><a href="index.html#section-4"><i class="fa fa-check"></i><b>1.4.2</b> 下载数据或其它文件</a></li>
<li class="chapter" data-level="1.4.3" data-path="index.html"><a href="index.html#rstudio"><i class="fa fa-check"></i><b>1.4.3</b> RStudio</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#section-5"><i class="fa fa-check"></i><b>1.5</b> 手动安装</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#section-6"><i class="fa fa-check"></i><b>1.6</b> 许可</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#section-7"><i class="fa fa-check"></i><b>1.7</b> 准备知识</a></li>
<li class="chapter" data-level="1.8" data-path="index.html"><a href="index.html#section-8"><i class="fa fa-check"></i><b>1.8</b> 联系我们</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> scRNA-seq介绍</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-9"><i class="fa fa-check"></i><b>2.3</b> 工作流程</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-10"><i class="fa fa-check"></i><b>2.4</b> 计算分析</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-11"><i class="fa fa-check"></i><b>2.5</b> 挑战</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-12"><i class="fa fa-check"></i><b>2.6</b> 实验方法</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#section-13"><i class="fa fa-check"></i><b>2.7</b> 如何选择合适的平台</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html"><i class="fa fa-check"></i><b>3</b> scRNA-seq原始数据处理</a><ul>
<li class="chapter" data-level="3.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#fastqc"><i class="fa fa-check"></i><b>3.1</b> FastQC</a><ul>
<li class="chapter" data-level="3.1.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-14"><i class="fa fa-check"></i><b>3.1.1</b> 解决方案并下载报告</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-15"><i class="fa fa-check"></i><b>3.2</b> 移除接头和低质量碱基</a><ul>
<li class="chapter" data-level="3.2.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#solution"><i class="fa fa-check"></i><b>3.2.1</b> Solution</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-16"><i class="fa fa-check"></i><b>3.3</b> 文件格式</a><ul>
<li class="chapter" data-level="3.3.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#fastq"><i class="fa fa-check"></i><b>3.3.1</b> FastQ</a></li>
<li class="chapter" data-level="3.3.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#bam"><i class="fa fa-check"></i><b>3.3.2</b> BAM</a></li>
<li class="chapter" data-level="3.3.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#cram"><i class="fa fa-check"></i><b>3.3.3</b> CRAM</a></li>
<li class="chapter" data-level="3.3.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-17"><i class="fa fa-check"></i><b>3.3.4</b> 手动查看文件</a></li>
<li class="chapter" data-level="3.3.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#genome-fasta-gtf"><i class="fa fa-check"></i><b>3.3.5</b> Genome (FASTA, GTF)</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-18"><i class="fa fa-check"></i><b>3.4</b> 测序文库拆分</a><ul>
<li class="chapter" data-level="3.4.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#section-19"><i class="fa fa-check"></i><b>3.4.1</b> 鉴定含有细胞的液滴/微孔</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#starread"><i class="fa fa-check"></i><b>3.5</b> 使用STAR比对read</a><ul>
<li class="chapter" data-level="3.5.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#star"><i class="fa fa-check"></i><b>3.5.1</b> STAR比对命令</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallistopseudo-alignment"><i class="fa fa-check"></i><b>3.6</b> Kallisto和Pseudo-Alignment</a><ul>
<li class="chapter" data-level="3.6.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#k-mer"><i class="fa fa-check"></i><b>3.6.1</b> k-mer是什么?</a></li>
<li class="chapter" data-level="3.6.2" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#k-merreads"><i class="fa fa-check"></i><b>3.6.2</b> 为什么比对k-mer而不是reads?</a></li>
<li class="chapter" data-level="3.6.3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto"><i class="fa fa-check"></i><b>3.6.3</b> Kallisto伪比对模式</a></li>
<li class="chapter" data-level="3.6.4" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto-pseudo-alignment"><i class="fa fa-check"></i><b>3.6.4</b> Kallisto Pseudo-Alignment命令</a></li>
<li class="chapter" data-level="3.6.5" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#kallisto-1"><i class="fa fa-check"></i><b>3.6.5</b> 理解Kallisto输出结果</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>4</b> 构建表达矩阵</a><ul>
<li class="chapter" data-level="4.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-20"><i class="fa fa-check"></i><b>4.1</b> 质量控制</a></li>
<li class="chapter" data-level="4.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-"><i class="fa fa-check"></i><b>4.2</b> Reads 比对</a></li>
<li class="chapter" data-level="4.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-21"><i class="fa fa-check"></i><b>4.3</b> 比对示例</a></li>
<li class="chapter" data-level="4.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-qc"><i class="fa fa-check"></i><b>4.4</b> Mapping QC</a></li>
<li class="chapter" data-level="4.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads"><i class="fa fa-check"></i><b>4.5</b> Reads定量</a></li>
<li class="chapter" data-level="4.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umichapter"><i class="fa fa-check"></i><b>4.6</b> 唯一分子标识符</a><ul>
<li class="chapter" data-level="4.6.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umi"><i class="fa fa-check"></i><b>4.6.1</b> UMI介绍</a></li>
<li class="chapter" data-level="4.6.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-22"><i class="fa fa-check"></i><b>4.6.2</b> 比对条形码序列</a></li>
<li class="chapter" data-level="4.6.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#barcodes"><i class="fa fa-check"></i><b>4.6.3</b> Barcodes计数</a></li>
<li class="chapter" data-level="4.6.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-23"><i class="fa fa-check"></i><b>4.6.4</b> 错误校正</a></li>
<li class="chapter" data-level="4.6.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#section-24"><i class="fa fa-check"></i><b>4.6.5</b> 下游分析</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html"><i class="fa fa-check"></i><b>5</b> R/Bioconductor介绍</a><ul>
<li class="chapter" data-level="5.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#r"><i class="fa fa-check"></i><b>5.1</b> 安装R包</a><ul>
<li class="chapter" data-level="5.1.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#cran"><i class="fa fa-check"></i><b>5.1.1</b> CRAN</a></li>
<li class="chapter" data-level="5.1.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#github-1"><i class="fa fa-check"></i><b>5.1.2</b> Github</a></li>
<li class="chapter" data-level="5.1.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor"><i class="fa fa-check"></i><b>5.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="5.1.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-25"><i class="fa fa-check"></i><b>5.1.4</b> 源码安装</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-26"><i class="fa fa-check"></i><b>5.2</b> 安装说明:</a></li>
<li class="chapter" data-level="5.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-27"><i class="fa fa-check"></i><b>5.3</b> 数据类型/类</a><ul>
<li class="chapter" data-level="5.3.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-28"><i class="fa fa-check"></i><b>5.3.1</b> 数值</a></li>
<li class="chapter" data-level="5.3.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-29"><i class="fa fa-check"></i><b>5.3.2</b> 字符/字符串</a></li>
<li class="chapter" data-level="5.3.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-30"><i class="fa fa-check"></i><b>5.3.3</b> 逻辑值</a></li>
<li class="chapter" data-level="5.3.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-31"><i class="fa fa-check"></i><b>5.3.4</b> 因子</a></li>
<li class="chapter" data-level="5.3.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-32"><i class="fa fa-check"></i><b>5.3.5</b> 检查数据类型</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-33"><i class="fa fa-check"></i><b>5.4</b> 基本数据结构</a></li>
<li class="chapter" data-level="5.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-34"><i class="fa fa-check"></i><b>5.5</b> 更多信息</a></li>
<li class="chapter" data-level="5.6" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-35"><i class="fa fa-check"></i><b>5.6</b> 数据类型</a><ul>
<li class="chapter" data-level="5.6.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-36"><i class="fa fa-check"></i><b>5.6.1</b> 什么是整洁的数据</a></li>
<li class="chapter" data-level="5.6.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#rich-data"><i class="fa fa-check"></i><b>5.6.2</b> 什么是Rich data?</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor-singlecellexperiment--scater"><i class="fa fa-check"></i><b>5.7</b> Bioconductor, <code>SingleCellExperiment</code> 和 <code>scater</code></a><ul>
<li class="chapter" data-level="5.7.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#bioconductor-1"><i class="fa fa-check"></i><b>5.7.1</b> Bioconductor</a></li>
<li class="chapter" data-level="5.7.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#singlecellexperiment-"><i class="fa fa-check"></i><b>5.7.2</b> <code>SingleCellExperiment</code> 类</a></li>
<li class="chapter" data-level="5.7.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#scater-"><i class="fa fa-check"></i><b>5.7.3</b> <code>scater</code> 包</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2"><i class="fa fa-check"></i><b>5.8</b> ggplot2介绍</a><ul>
<li class="chapter" data-level="5.8.1" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2-1"><i class="fa fa-check"></i><b>5.8.1</b> ggplot2是什么?</a></li>
<li class="chapter" data-level="5.8.2" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#ggplot2j"><i class="fa fa-check"></i><b>5.8.2</b> ggplot2j原则</a></li>
<li class="chapter" data-level="5.8.3" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#aes"><i class="fa fa-check"></i><b>5.8.3</b> 使用<code>aes</code>映射函数</a></li>
<li class="chapter" data-level="5.8.4" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#geoms"><i class="fa fa-check"></i><b>5.8.4</b> Geoms</a></li>
<li class="chapter" data-level="5.8.5" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-37"><i class="fa fa-check"></i><b>5.8.5</b> 绘制2个以上细胞的数据</a></li>
<li class="chapter" data-level="5.8.6" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-38"><i class="fa fa-check"></i><b>5.8.6</b> 绘制热图</a></li>
<li class="chapter" data-level="5.8.7" data-path="intro-r-bioc.html"><a href="intro-r-bioc.html#section-39"><i class="fa fa-check"></i><b>5.8.7</b> 主成分分析</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tabula-muris.html"><a href="tabula-muris.html"><i class="fa fa-check"></i><b>6</b> Tabula Muris</a><ul>
<li class="chapter" data-level="6.1" data-path="tabula-muris.html"><a href="tabula-muris.html#section-40"><i class="fa fa-check"></i><b>6.1</b> 介绍</a></li>
<li class="chapter" data-level="6.2" data-path="tabula-muris.html"><a href="tabula-muris.html#section-41"><i class="fa fa-check"></i><b>6.2</b> 下载数据</a></li>
<li class="chapter" data-level="6.3" data-path="tabula-muris.html"><a href="tabula-muris.html#smartseq2"><i class="fa fa-check"></i><b>6.3</b> 读取Smartseq2数据</a></li>
<li class="chapter" data-level="6.4" data-path="tabula-muris.html"><a href="tabula-muris.html#scater"><i class="fa fa-check"></i><b>6.4</b> 构建scater对象</a></li>
<li class="chapter" data-level="6.5" data-path="tabula-muris.html"><a href="tabula-muris.html#x"><i class="fa fa-check"></i><b>6.5</b> 读取10X数据</a></li>
<li class="chapter" data-level="6.6" data-path="tabula-muris.html"><a href="tabula-muris.html#scater-1"><i class="fa fa-check"></i><b>6.6</b> 创建scater对象</a></li>
<li class="chapter" data-level="6.7" data-path="tabula-muris.html"><a href="tabula-muris.html#section-42"><i class="fa fa-check"></i><b>6.7</b> 进阶练习</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html"><i class="fa fa-check"></i><b>7</b> 表达矩阵质控</a><ul>
<li class="chapter" data-level="7.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#exprs-qc"><i class="fa fa-check"></i><b>7.1</b> UMI表达质控</a><ul>
<li class="chapter" data-level="7.1.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-43"><i class="fa fa-check"></i><b>7.1.1</b> 介绍</a></li>
<li class="chapter" data-level="7.1.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tung"><i class="fa fa-check"></i><b>7.1.2</b> Tung数据集</a></li>
<li class="chapter" data-level="7.1.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-44"><i class="fa fa-check"></i><b>7.1.3</b> 细胞质控</a></li>
<li class="chapter" data-level="7.1.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-47"><i class="fa fa-check"></i><b>7.1.4</b> 细胞过滤</a></li>
<li class="chapter" data-level="7.1.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-50"><i class="fa fa-check"></i><b>7.1.5</b> 手工过滤和自动过滤比较</a></li>
<li class="chapter" data-level="7.1.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-51"><i class="fa fa-check"></i><b>7.1.6</b> 基因分析</a></li>
<li class="chapter" data-level="7.1.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-54"><i class="fa fa-check"></i><b>7.1.7</b> 保存数据</a></li>
<li class="chapter" data-level="7.1.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-55"><i class="fa fa-check"></i><b>7.1.8</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-1"><i class="fa fa-check"></i><b>7.2</b> Reads表达质控</a></li>
<li class="chapter" data-level="7.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-56"><i class="fa fa-check"></i><b>7.3</b> 数据可视化</a><ul>
<li class="chapter" data-level="7.3.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-57"><i class="fa fa-check"></i><b>7.3.1</b> 介绍</a></li>
<li class="chapter" data-level="7.3.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#visual-pca"><i class="fa fa-check"></i><b>7.3.2</b> PCA plot</a></li>
<li class="chapter" data-level="7.3.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#visual-tsne"><i class="fa fa-check"></i><b>7.3.3</b> tSNE可视化</a></li>
<li class="chapter" data-level="7.3.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-61"><i class="fa fa-check"></i><b>7.3.4</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-2"><i class="fa fa-check"></i><b>7.4</b> Reads数据可视化</a></li>
<li class="chapter" data-level="7.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-62"><i class="fa fa-check"></i><b>7.5</b> 识别混淆因子</a><ul>
<li class="chapter" data-level="7.5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-63"><i class="fa fa-check"></i><b>7.5.1</b> 介绍</a></li>
<li class="chapter" data-level="7.5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-64"><i class="fa fa-check"></i><b>7.5.2</b> 和组成分的相关性</a></li>
<li class="chapter" data-level="7.5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-66"><i class="fa fa-check"></i><b>7.5.3</b> 解释变量</a></li>
<li class="chapter" data-level="7.5.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-67"><i class="fa fa-check"></i><b>7.5.4</b> 其它混淆因子</a></li>
<li class="chapter" data-level="7.5.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-68"><i class="fa fa-check"></i><b>7.5.5</b> 练习</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-3"><i class="fa fa-check"></i><b>7.6</b> 识别READS中混淆因子</a></li>
<li class="chapter" data-level="7.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-69"><i class="fa fa-check"></i><b>7.7</b> 标准化理论</a><ul>
<li class="chapter" data-level="7.7.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-70"><i class="fa fa-check"></i><b>7.7.1</b> 介绍</a></li>
<li class="chapter" data-level="7.7.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-71"><i class="fa fa-check"></i><b>7.7.2</b> 文库大小</a></li>
<li class="chapter" data-level="7.7.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-72"><i class="fa fa-check"></i><b>7.7.3</b> 标准化</a></li>
<li class="chapter" data-level="7.7.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#effectiveness"><i class="fa fa-check"></i><b>7.7.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#umi-1"><i class="fa fa-check"></i><b>7.8</b> UMI标准化练习</a><ul>
<li class="chapter" data-level="7.8.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#raw"><i class="fa fa-check"></i><b>7.8.1</b> Raw</a></li>
<li class="chapter" data-level="7.8.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cpm-1"><i class="fa fa-check"></i><b>7.8.2</b> CPM</a></li>
<li class="chapter" data-level="7.8.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#scran-1"><i class="fa fa-check"></i><b>7.8.3</b> scran</a></li>
<li class="chapter" data-level="7.8.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-73"><i class="fa fa-check"></i><b>7.8.4</b> 下采样</a></li>
<li class="chapter" data-level="7.8.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-74"><i class="fa fa-check"></i><b>7.8.5</b> 基因/转录本长度标准化</a></li>
<li class="chapter" data-level="7.8.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-75"><i class="fa fa-check"></i><b>7.8.6</b> 练习</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-4"><i class="fa fa-check"></i><b>7.9</b> Reads标准化练习</a></li>
<li class="chapter" data-level="7.10" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>7.10</b> 处理混淆因子</a><ul>
<li class="chapter" data-level="7.10.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-76"><i class="fa fa-check"></i><b>7.10.1</b> 介绍</a></li>
<li class="chapter" data-level="7.10.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-77"><i class="fa fa-check"></i><b>7.10.2</b> 移除不需要的变异</a></li>
<li class="chapter" data-level="7.10.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#combat"><i class="fa fa-check"></i><b>7.10.3</b> Combat</a></li>
<li class="chapter" data-level="7.10.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>7.10.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="7.10.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#glm"><i class="fa fa-check"></i><b>7.10.5</b> GLM</a></li>
<li class="chapter" data-level="7.10.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#harmony"><i class="fa fa-check"></i><b>7.10.6</b> Harmony</a></li>
<li class="chapter" data-level="7.10.7" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-78"><i class="fa fa-check"></i><b>7.10.7</b> 评价和比较混淆因子去除策略</a></li>
<li class="chapter" data-level="7.10.8" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#section-81"><i class="fa fa-check"></i><b>7.10.8</b> 大作业</a></li>
</ul></li>
<li class="chapter" data-level="7.11" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#reads-5"><i class="fa fa-check"></i><b>7.11</b> 处理Reads中混淆因子</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>8</b> 生物学分析</a><ul>
<li class="chapter" data-level="8.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-82"><i class="fa fa-check"></i><b>8.1</b> 聚类介绍</a><ul>
<li class="chapter" data-level="8.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-83"><i class="fa fa-check"></i><b>8.1.1</b> 介绍</a></li>
<li class="chapter" data-level="8.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#section-84"><i class="fa fa-check"></i><b>8.1.2</b> 降维</a></li>
<li class="chapter" data-level="8.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-85"><i class="fa fa-check"></i><b>8.1.3</b> 聚类方法</a></li>
<li class="chapter" data-level="8.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#section-88"><i class="fa fa-check"></i><b>8.1.4</b> 聚类的挑战</a></li>
<li class="chapter" data-level="8.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#scrna-seq-1"><i class="fa fa-check"></i><b>8.1.5</b> scRNA-seq数据的工具</a></li>
<li class="chapter" data-level="8.1.6" data-path="biological-analysis.html"><a href="biological-analysis.html#section-89"><i class="fa fa-check"></i><b>8.1.6</b> 比较聚类结果</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clust-methods"><i class="fa fa-check"></i><b>8.2</b> 聚类示例</a><ul>
<li class="chapter" data-level="8.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-"><i class="fa fa-check"></i><b>8.2.1</b> Deng 数据集</a></li>
<li class="chapter" data-level="8.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>8.2.2</b> SC3</a></li>
<li class="chapter" data-level="8.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>8.2.3</b> tSNE + kmeans</a></li>
<li class="chapter" data-level="8.2.4" data-path="biological-analysis.html"><a href="biological-analysis.html#sincera-1"><i class="fa fa-check"></i><b>8.2.4</b> SINCERA</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-90"><i class="fa fa-check"></i><b>8.3</b> 特征选择</a><ul>
<li class="chapter" data-level="8.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#null"><i class="fa fa-check"></i><b>8.3.1</b> 识别基因与null模型</a></li>
<li class="chapter" data-level="8.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>8.3.2</b> Correlated Expression</a></li>
<li class="chapter" data-level="8.3.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-91"><i class="fa fa-check"></i><b>8.3.3</b> 比较不同的方法</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="biological-analysis.html"><a href="biological-analysis.html#section-92"><i class="fa fa-check"></i><b>8.4</b> 伪时间分析</a><ul>
<li class="chapter" data-level="8.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng"><i class="fa fa-check"></i><b>8.4.1</b> Deng数据初探</a></li>
<li class="chapter" data-level="8.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#tscan"><i class="fa fa-check"></i><b>8.4.2</b> TSCAN</a></li>
<li class="chapter" data-level="8.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>8.4.3</b> monocle</a></li>
<li class="chapter" data-level="8.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#diffusion-maps"><i class="fa fa-check"></i><b>8.4.4</b> Diffusion maps</a></li>
<li class="chapter" data-level="8.4.5" data-path="biological-analysis.html"><a href="biological-analysis.html#slicer"><i class="fa fa-check"></i><b>8.4.5</b> SLICER</a></li>
<li class="chapter" data-level="8.4.6" data-path="biological-analysis.html"><a href="biological-analysis.html#ouija"><i class="fa fa-check"></i><b>8.4.6</b> Ouija</a></li>
<li class="chapter" data-level="8.4.7" data-path="biological-analysis.html"><a href="biological-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>8.4.7</b> 方法比较</a></li>
<li class="chapter" data-level="8.4.8" data-path="biological-analysis.html"><a href="biological-analysis.html#expression-of-genes-through-time"><i class="fa fa-check"></i><b>8.4.8</b> Expression of genes through time</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="biological-analysis.html"><a href="biological-analysis.html#imputation"><i class="fa fa-check"></i><b>8.5</b> Imputation</a><ul>
<li class="chapter" data-level="8.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#scimpute"><i class="fa fa-check"></i><b>8.5.1</b> scImpute</a></li>
<li class="chapter" data-level="8.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#drimpute"><i class="fa fa-check"></i><b>8.5.2</b> DrImpute</a></li>
<li class="chapter" data-level="8.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#magic"><i class="fa fa-check"></i><b>8.5.3</b> MAGIC</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="biological-analysis.html"><a href="biological-analysis.html#dechapter"><i class="fa fa-check"></i><b>8.6</b> 差异表达分析</a><ul>
<li class="chapter" data-level="8.6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>8.6.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="8.6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>8.6.2</b> Single cell RNA-seq</a></li>
<li class="chapter" data-level="8.6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#Differences-in-Distribution"><i class="fa fa-check"></i><b>8.6.3</b> 分布的差异</a></li>
<li class="chapter" data-level="8.6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#Models-of-single-cell-RNASeq-data"><i class="fa fa-check"></i><b>8.6.4</b> scRNA-seq数据模型</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="biological-analysis.html"><a href="biological-analysis.html#DE-in-a-real-dataset"><i class="fa fa-check"></i><b>8.7</b> 差异表达分析实战</a><ul>
<li class="chapter" data-level="8.7.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-93"><i class="fa fa-check"></i><b>8.7.1</b> 介绍</a></li>
<li class="chapter" data-level="8.7.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>8.7.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="8.7.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>8.7.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="8.7.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger"><i class="fa fa-check"></i><b>8.7.4</b> edgeR</a></li>
<li class="chapter" data-level="8.7.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle-1"><i class="fa fa-check"></i><b>8.7.5</b> Monocle</a></li>
<li class="chapter" data-level="8.7.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>8.7.6</b> MAST</a></li>
<li class="chapter" data-level="8.7.7" data-path="biological-analysis.html"><a href="biological-analysis.html#slow-method1h"><i class="fa fa-check"></i><b>8.7.7</b> Slow method(&gt;1h)</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="biological-analysis.html"><a href="biological-analysis.html#scrna-seq-2"><i class="fa fa-check"></i><b>8.8</b> 比较/组合scRNA-seq数据集</a><ul>
<li class="chapter" data-level="8.8.1" data-path="biological-analysis.html"><a href="biological-analysis.html#section-95"><i class="fa fa-check"></i><b>8.8.1</b> 介绍</a></li>
<li class="chapter" data-level="8.8.2" data-path="biological-analysis.html"><a href="biological-analysis.html#datasets"><i class="fa fa-check"></i><b>8.8.2</b> Datasets</a></li>
<li class="chapter" data-level="8.8.3" data-path="biological-analysis.html"><a href="biological-analysis.html#scmap"><i class="fa fa-check"></i><b>8.8.3</b> scmap</a></li>
<li class="chapter" data-level="8.8.4" data-path="biological-analysis.html"><a href="biological-analysis.html#section-98"><i class="fa fa-check"></i><b>8.8.4</b> 细胞间映射</a></li>
<li class="chapter" data-level="8.8.5" data-path="biological-analysis.html"><a href="biological-analysis.html#metaneighbour"><i class="fa fa-check"></i><b>8.8.5</b> Metaneighbour</a></li>
<li class="chapter" data-level="8.8.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mnncorrect-1"><i class="fa fa-check"></i><b>8.8.6</b> mnnCorrect</a></li>
<li class="chapter" data-level="8.8.7" data-path="biological-analysis.html"><a href="biological-analysis.html#seurat"><i class="fa fa-check"></i><b>8.8.7</b> 典型相关分析 (Seurat)</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="biological-analysis.html"><a href="biological-analysis.html#scrna-seq-3"><i class="fa fa-check"></i><b>8.9</b> 检索scRNA-seq数据</a><ul>
<li class="chapter" data-level="8.9.1" data-path="biological-analysis.html"><a href="biological-analysis.html#about"><i class="fa fa-check"></i><b>8.9.1</b> About</a></li>
<li class="chapter" data-level="8.9.2" data-path="biological-analysis.html"><a href="biological-analysis.html#section-100"><i class="fa fa-check"></i><b>8.9.2</b> 数据集</a></li>
<li class="chapter" data-level="8.9.3" data-path="biological-analysis.html"><a href="biological-analysis.html#section-101"><i class="fa fa-check"></i><b>8.9.3</b> 基因索引</a></li>
<li class="chapter" data-level="8.9.4" data-path="biological-analysis.html"><a href="biological-analysis.html#marker"><i class="fa fa-check"></i><b>8.9.4</b> Marker基因</a></li>
<li class="chapter" data-level="8.9.5" data-path="biological-analysis.html"><a href="biological-analysis.html#section-102"><i class="fa fa-check"></i><b>8.9.5</b> 搜索基因列表</a></li>
<li class="chapter" data-level="8.9.6" data-path="biological-analysis.html"><a href="biological-analysis.html#in-silico-gating"><i class="fa fa-check"></i><b>8.9.6</b> In-silico gating</a></li>
<li class="chapter" data-level="8.9.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo"><i class="fa fa-check"></i><b>8.9.7</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="seurat-chapter.html"><a href="seurat-chapter.html"><i class="fa fa-check"></i><b>9</b> Seurat</a><ul>
<li class="chapter" data-level="9.1" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Setup-the-Seurat-Object"><i class="fa fa-check"></i><b>9.1</b> 构建Seurat对象</a></li>
<li class="chapter" data-level="9.2" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Standard-pre-processing-workflow"><i class="fa fa-check"></i><b>9.2</b> 标准预处理流程</a><ul>
<li class="chapter" data-level="9.2.1" data-path="seurat-chapter.html"><a href="seurat-chapter.html#QC-and-selecting-cells-for-further-analysis"><i class="fa fa-check"></i><b>9.2.1</b> 质控和选择细胞</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Normalizing-the-data"><i class="fa fa-check"></i><b>9.3</b> 数据标准化</a></li>
<li class="chapter" data-level="9.4" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>9.4</b> 检测高可变基因</a></li>
<li class="chapter" data-level="9.5" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Scaling-the-data-and-removing-unwanted-sources-of-variation"><i class="fa fa-check"></i><b>9.5</b> 缩放数据并删除不需要的变异来源</a></li>
<li class="chapter" data-level="9.6" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Perform-linear-dimensional-reduction"><i class="fa fa-check"></i><b>9.6</b> 线性降维</a></li>
<li class="chapter" data-level="9.7" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Determine-statistically-significant-principal-components"><i class="fa fa-check"></i><b>9.7</b> 统计显著主成分</a></li>
<li class="chapter" data-level="9.8" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Cluster-the-cells"><i class="fa fa-check"></i><b>9.8</b> 聚类</a></li>
<li class="chapter" data-level="9.9" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Run-Non-linear-dimensional-reduction-tSNE"><i class="fa fa-check"></i><b>9.9</b> 非线性降维</a></li>
<li class="chapter" data-level="9.10" data-path="seurat-chapter.html"><a href="seurat-chapter.html#umap"><i class="fa fa-check"></i><b>9.10</b> 运行UMAP</a></li>
<li class="chapter" data-level="9.11" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Finding-differentially-expressed-genes-cluster-biomarkers"><i class="fa fa-check"></i><b>9.11</b> 差异表达分析</a></li>
<li class="chapter" data-level="9.12" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Assigning-cell-type-identity-to-clusters"><i class="fa fa-check"></i><b>9.12</b> 确定cluster细胞类型</a></li>
<li class="chapter" data-level="9.13" data-path="seurat-chapter.html"><a href="seurat-chapter.html#Further-subdivisions-within-cell-types"><i class="fa fa-check"></i><b>9.13</b> 细分细胞类型</a></li>
<li class="chapter" data-level="9.14" data-path="seurat-chapter.html"><a href="seurat-chapter.html#sessioninfo-1"><i class="fa fa-check"></i><b>9.14</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="Ideal-scRNA-seq-pipeline.html"><a href="Ideal-scRNA-seq-pipeline.html"><i class="fa fa-check"></i><b>10</b> “理想”scRANA-seq流程</a><ul>
<li class="chapter" data-level="10.1" data-path="Ideal-scRNA-seq-pipeline.html"><a href="Ideal-scRNA-seq-pipeline.html#Experimental-Design"><i class="fa fa-check"></i><b>10.1</b> 实验设计</a></li>
<li class="chapter" data-level="10.2" data-path="Ideal-scRNA-seq-pipeline.html"><a href="Ideal-scRNA-seq-pipeline.html#Processing-Reads"><i class="fa fa-check"></i><b>10.2</b> reads处理</a></li>
<li class="chapter" data-level="10.3" data-path="Ideal-scRNA-seq-pipeline.html"><a href="Ideal-scRNA-seq-pipeline.html#Preparing-Expression-Matrix"><i class="fa fa-check"></i><b>10.3</b> 准备表达矩阵</a></li>
<li class="chapter" data-level="10.4" data-path="Ideal-scRNA-seq-pipeline.html"><a href="Ideal-scRNA-seq-pipeline.html#Biological-Interpretation"><i class="fa fa-check"></i><b>10.4</b> 生物学解释</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="Advanced-exercises.html"><a href="Advanced-exercises.html"><i class="fa fa-check"></i><b>11</b> 进阶练习</a></li>
<li class="chapter" data-level="12" data-path="Resources.html"><a href="Resources.html"><i class="fa fa-check"></i><b>12</b> 相关资源</a><ul>
<li class="chapter" data-level="12.1" data-path="Resources.html"><a href="Resources.html#scRNA-seq-protocols"><i class="fa fa-check"></i><b>12.1</b> scRNA-seq protocols</a></li>
<li class="chapter" data-level="12.2" data-path="Resources.html"><a href="Resources.html#external-rna-control-consortium-ercc"><i class="fa fa-check"></i><b>12.2</b> External RNA Control Consortium (ERCC)</a></li>
<li class="chapter" data-level="12.3" data-path="Resources.html"><a href="Resources.html#scRNA-seq-analysis-tools"><i class="fa fa-check"></i><b>12.3</b> scRNA-seq分析工具</a></li>
<li class="chapter" data-level="12.4" data-path="Resources.html"><a href="Resources.html#scrna-seq-4"><i class="fa fa-check"></i><b>12.4</b> scRNA-seq公共数据集</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>13</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/lemonbases/scRNA.seq.course.CN" target="blank">碱基吃柠檬</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNA-seq数据分析</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="seurat-chapter" class="section level1">
<h1><span class="header-section-number">9</span> Seurat</h1>
<p><a href="http://satijalab.org/seurat/">Seurat</a>最初是作为scRNA-seq数据的聚类工具，但是在过去的几年中，该软件包包含越来越多的功能。目前Seurat是一个流行的R软件包，可以对scRNA-seq数进行质量控制，分析和探索，涵盖本课程的许多任务。</p>
<p><strong>注意</strong> 超过<span class="math inline">\(5000\)</span>个细胞的数据集推荐使用<code>Seurat</code>。对于较小的数据集，可以使用<code>SC3</code>。
We recommend using <code>Seurat</code> for datasets with more than <span class="math inline">\(5000\)</span> cells. For smaller dataset a good alternative will be <code>SC3</code>.</p>
<p><strong>注意</strong> 本章使用是 <a href="https://satijalab.org/seurat/pbmc3k_tutorial.html">该教程</a>.</p>
<div id="Setup-the-Seurat-Object" class="section level2">
<h2><span class="header-section-number">9.1</span> 构建Seurat对象</h2>
<p>我们将分析10X Genomics免费提供的Peripheral Blood Mononuclear Cells (PBMC)数据集。用Illumina NextSeq 500对2,700个细胞进行了测序。原始数据在 <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">这里</a>。</p>
<p>从读入数据开始，Seurat中的所有功能都已配置为与稀疏矩阵配合使用，从而为Drop-seq/inDrop/10x数据节省大量内存和速度。</p>
<div class="sourceCode" id="cb616"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb616-1" title="1"><span class="kw">library</span>(Seurat)</a>
<a class="sourceLine" id="cb616-2" title="2"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb616-3" title="3"><span class="kw">library</span>(cowplot)</a>
<a class="sourceLine" id="cb616-4" title="4"></a>
<a class="sourceLine" id="cb616-5" title="5"><span class="co"># Load the PBMC dataset</span></a>
<a class="sourceLine" id="cb616-6" title="6">pbmc.data &lt;-<span class="st"> </span><span class="kw">Read10X</span>(<span class="dt">data.dir =</span> <span class="st">&quot;data/pbmc3k_filtered_gene_bc_matrices/hg19/&quot;</span>)</a>
<a class="sourceLine" id="cb616-7" title="7"></a>
<a class="sourceLine" id="cb616-8" title="8"><span class="co"># Examine the memory savings between regular and sparse matrices</span></a>
<a class="sourceLine" id="cb616-9" title="9">dense.size &lt;-<span class="st"> </span><span class="kw">object.size</span>(<span class="dt">x =</span> <span class="kw">as.matrix</span>(<span class="dt">x =</span> pbmc.data))</a>
<a class="sourceLine" id="cb616-10" title="10">dense.size</a>
<a class="sourceLine" id="cb616-11" title="11"></a>
<a class="sourceLine" id="cb616-12" title="12">sparse.size &lt;-<span class="st"> </span><span class="kw">object.size</span>(<span class="dt">x =</span> pbmc.data)</a>
<a class="sourceLine" id="cb616-13" title="13">sparse.size</a>
<a class="sourceLine" id="cb616-14" title="14"></a>
<a class="sourceLine" id="cb616-15" title="15">dense.size<span class="op">/</span>sparse.size</a>
<a class="sourceLine" id="cb616-16" title="16"></a>
<a class="sourceLine" id="cb616-17" title="17"><span class="co"># Initialize the Seurat object with the raw (non-normalized data).  Keep all</span></a>
<a class="sourceLine" id="cb616-18" title="18"><span class="co"># genes expressed in &gt;= 3 cells (~0.1% of the data). Keep all cells with at</span></a>
<a class="sourceLine" id="cb616-19" title="19"><span class="co"># least 200 detected genes</span></a>
<a class="sourceLine" id="cb616-20" title="20">pbmc &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(<span class="dt">counts =</span> pbmc.data, <span class="dt">min.cells =</span> <span class="dv">3</span>, <span class="dt">min.features  =</span> <span class="dv">200</span>, <span class="dt">project =</span> <span class="st">&quot;10X_PBMC&quot;</span>, <span class="dt">assay =</span> <span class="st">&quot;RNA&quot;</span>)</a></code></pre></div>
</div>
<div id="Standard-pre-processing-workflow" class="section level2">
<h2><span class="header-section-number">9.2</span> 标准预处理流程</h2>
<p>以下步骤包括Seurat中scRNA-seq数据的标准预处理流程。包括创建Seurat对象，基于QC指标的细胞选择和过滤，数据标准化和缩放，以及高可变基因的检测。</p>
<div id="QC-and-selecting-cells-for-further-analysis" class="section level3">
<h3><span class="header-section-number">9.2.1</span> 质控和选择细胞</h3>
<p>虽然<code>CreateSeuratObject</code>使用基本的minimum gene-cutoff，但可以根据技术或生物参数在此阶段过滤掉细胞。 Seurat允许用户定义的标准探索QC指标和过滤细胞。在下面的例子中，对基因和分子counts进行可视化，绘制它们之间的关系，并排除multiplets具有异常数量的基因的细胞。当然，这不是排除细胞doublets的保证方法，但将其作为用户定义过滤异常细胞的示例。 我们还根据存在的线粒体基因的百分比过滤细胞。</p>
<div class="sourceCode" id="cb617"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb617-1" title="1"><span class="co"># The number of genes and UMIs (nGene and nUMI) are automatically calculated</span></a>
<a class="sourceLine" id="cb617-2" title="2"><span class="co"># for every object by Seurat.  For non-UMI data, nUMI represents the sum of</span></a>
<a class="sourceLine" id="cb617-3" title="3"><span class="co"># the non-normalized values within a cell We calculate the percentage of</span></a>
<a class="sourceLine" id="cb617-4" title="4"><span class="co"># mitochondrial genes here and store it in percent.mito using AddMetaData.</span></a>
<a class="sourceLine" id="cb617-5" title="5"><span class="co"># We use object@raw.data since this represents non-transformed and</span></a>
<a class="sourceLine" id="cb617-6" title="6"><span class="co"># non-log-normalized counts The % of UMI mapping to MT-genes is a common</span></a>
<a class="sourceLine" id="cb617-7" title="7"><span class="co"># scRNA-seq QC metric.</span></a>
<a class="sourceLine" id="cb617-8" title="8">mito.genes &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="dt">pattern =</span> <span class="st">&quot;^MT-&quot;</span>, <span class="dt">x =</span> <span class="kw">rownames</span>(pbmc<span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]), <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb617-9" title="9"></a>
<a class="sourceLine" id="cb617-10" title="10">percent.mito &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(pbmc<span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]][mito.genes, ])<span class="op">/</span>Matrix<span class="op">::</span><span class="kw">colSums</span>(pbmc<span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]])</a>
<a class="sourceLine" id="cb617-11" title="11"></a>
<a class="sourceLine" id="cb617-12" title="12"><span class="co"># AddMetaData adds columns to object@meta.data, and is a great place to</span></a>
<a class="sourceLine" id="cb617-13" title="13"><span class="co"># stash QC stats</span></a>
<a class="sourceLine" id="cb617-14" title="14"></a>
<a class="sourceLine" id="cb617-15" title="15"><span class="co">#Seurat v2 function, but shows compatibility in Seurat v3</span></a>
<a class="sourceLine" id="cb617-16" title="16">pbmc &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> pbmc, <span class="dt">metadata =</span> percent.mito, <span class="dt">col.name =</span> <span class="st">&quot;percent.mito&quot;</span>) </a>
<a class="sourceLine" id="cb617-17" title="17"><span class="co">#in case the above function does not work simply do:</span></a>
<a class="sourceLine" id="cb617-18" title="18">pbmc<span class="op">$</span>percent.mito &lt;-<span class="st"> </span>percent.mito</a>
<a class="sourceLine" id="cb617-19" title="19"></a>
<a class="sourceLine" id="cb617-20" title="20"><span class="kw">VlnPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;percent.mito&quot;</span>), <span class="dt">ncol =</span> <span class="dv">3</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-4-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb618"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb618-1" title="1"><span class="co"># GenePlot is typically used to visualize gene-gene relationships, but can</span></a>
<a class="sourceLine" id="cb618-2" title="2"><span class="co"># be used for anything calculated by the object, i.e. columns in</span></a>
<a class="sourceLine" id="cb618-3" title="3"><span class="co"># object@meta.data, PC scores etc.  Since there is a rare subset of cells</span></a>
<a class="sourceLine" id="cb618-4" title="4"><span class="co"># with an outlier level of high mitochondrial percentage and also low UMI</span></a>
<a class="sourceLine" id="cb618-5" title="5"><span class="co"># content, we filter these as well</span></a>
<a class="sourceLine" id="cb618-6" title="6"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb618-7" title="7"><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> pbmc, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;percent.mito&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-5-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb619"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb619-1" title="1"><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> pbmc, <span class="dt">feature1 =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;nFeature_RNA&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-5-2.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb620"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb620-1" title="1"><span class="co"># We filter out cells that have unique gene counts (nFeature_RNA) over 2,500 or less than</span></a>
<a class="sourceLine" id="cb620-2" title="2"><span class="co"># 200 Note that &gt; and &lt; are used to define a&#39;gate&#39;.  </span></a>
<a class="sourceLine" id="cb620-3" title="3"><span class="co">#-Inf and Inf should be used if you don&#39;t want a lower or upper threshold.</span></a>
<a class="sourceLine" id="cb620-4" title="4">pbmc &lt;-<span class="st"> </span><span class="kw">subset</span>(<span class="dt">x =</span> pbmc, <span class="dt">subset =</span> nFeature_RNA <span class="op">&gt;</span><span class="st"> </span><span class="dv">200</span> <span class="op">&amp;</span><span class="st"> </span>nFeature_RNA <span class="op">&lt;</span><span class="st"> </span><span class="dv">2500</span> <span class="op">&amp;</span><span class="st"> </span>percent.mito <span class="op">&gt;</span><span class="st">  </span><span class="op">-</span><span class="ot">Inf</span> <span class="op">&amp;</span><span class="st"> </span>percent.mito <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span> )</a></code></pre></div>
</div>
</div>
<div id="Normalizing-the-data" class="section level2">
<h2><span class="header-section-number">9.3</span> 数据标准化</h2>
<p>从数据集中删除不需要的细胞后，下一步是数据标准化。 默认情况下，采用全局缩放归一化方法“LogNormalize”，通过总表达对每个细胞的基因表达值进行标准化，乘以比例因子（默认为10,000），并对结果进行对数转换。</p>
<div class="sourceCode" id="cb621"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb621-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> pbmc, <span class="dt">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="dt">scale.factor =</span> <span class="dv">10000</span>)</a></code></pre></div>
</div>
<div id="Detection-of-variable-genes-across-the-single-cells" class="section level2">
<h2><span class="header-section-number">9.4</span> 检测高可变基因</h2>
<p>Seurat计算高度可变的基因，并关注下游分析。<code>FindVariableGenes</code>计算每个基因的表达均值和离散程度，将这些基因放入bin中，然后计算每个bin离散程度的z-score。这有助于控制方差和表达均值之间的关系。(Macosko et al.)没有更改该函数，但很快就会出现用于HVG鉴定的新方法。 建议用户设置这些参数以在dispersion plot标记异常值，但具体的参数设置可能会根据数据类型，样本中的异质性和标准化策略而有所不同。 这里的参数识别~2000个可变基因，并且代表UMI数据的典型参数设置，其归一化为总共1e4个分子。</p>
<div class="sourceCode" id="cb622"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb622-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(<span class="dt">object =</span> pbmc, <span class="dt">mean.function =</span> ExpMean, <span class="dt">dispersion.function =</span> LogVMR, <span class="dt">x.low.cutoff =</span> <span class="fl">0.0125</span>, <span class="dt">x.high.cutoff =</span> <span class="dv">3</span>, <span class="dt">y.cutoff =</span> <span class="fl">0.5</span>, <span class="dt">nfeatures =</span> <span class="dv">2000</span>)</a></code></pre></div>
<p>查看FindVariableFeatures输出的输出。基因不存储在对象中，但可以通过这种方式访问。</p>
<div class="sourceCode" id="cb623"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb623-1" title="1"><span class="kw">head</span>(<span class="dt">x =</span> <span class="kw">HVFInfo</span>(<span class="dt">object =</span> pbmc))</a>
<a class="sourceLine" id="cb623-2" title="2"><span class="co">##                      mean    variance variance.standardized</span></a>
<a class="sourceLine" id="cb623-3" title="3"><span class="co">## AL627309.1    0.003411676 0.003401325             0.9330441</span></a>
<a class="sourceLine" id="cb623-4" title="4"><span class="co">## AP006222.2    0.001137225 0.001136363             0.9924937</span></a>
<a class="sourceLine" id="cb623-5" title="5"><span class="co">## RP11-206L10.2 0.001895375 0.001892500             0.9627290</span></a>
<a class="sourceLine" id="cb623-6" title="6"><span class="co">## RP11-206L10.9 0.001137225 0.001136363             0.9924937</span></a>
<a class="sourceLine" id="cb623-7" title="7"><span class="co">## LINC00115     0.006823351 0.006779363             0.9062135</span></a>
<a class="sourceLine" id="cb623-8" title="8"><span class="co">## NOC2L         0.107278241 0.159514698             0.7849309</span></a></code></pre></div>
</div>
<div id="Scaling-the-data-and-removing-unwanted-sources-of-variation" class="section level2">
<h2><span class="header-section-number">9.5</span> 缩放数据并删除不需要的变异来源</h2>
<p>单细胞数据集可能包含“不感兴趣”的变异来源。这不仅包括技术噪音，还包括批次效应，甚至包括生物变异来源（细胞周期）。 正如 <a href="https://www.nature.com/articles/nbt.3102">Buettner et al, NBT, 2015</a>所建议的那样，从分析中回归这些信号可以改善下游降维和聚类结果。 为了减轻这些信号的影响，Seurat基于用户定义的变量构建线性模型预测基因表达。这些模型的缩放z-score残差存储在scale.data中，用于降维和聚类。</p>
<p>我们可以建立由批次效应驱动的基因表达中的细胞-细胞变异，细胞比对率（由Drop-seq数据的Drop-seq工具提供），检测到的分子数量和线粒体基因表达回归模型。对于循环细胞，我们还可以学习‘cell-cycle’ score(<a href="http://satijalab.org/seurat/cell_cycle_vignette.html">示例</a>) 并对其进行回归。 在这个有丝分裂后血细胞的简单例子中，我们对每个细胞检测到的分子数量以及线粒体基因含量百分比建立回归模型。</p>
<p>Seurat v2.0将此回归实现为数据扩展过程的一部分。<code>RegressOut</code>函数已经被弃用，并将其替换为<code>ScaleData</code>中的vars.to.regress参数。</p>
<div class="sourceCode" id="cb624"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb624-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(<span class="dt">object =</span> pbmc, <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">&quot;nCounts_RNA&quot;</span>, <span class="st">&quot;percent.mito&quot;</span>))</a></code></pre></div>
</div>
<div id="Perform-linear-dimensional-reduction" class="section level2">
<h2><span class="header-section-number">9.6</span> 线性降维</h2>
<p>–&gt; Seurat v2: 接下来对缩放后的数据执行PCA。 默认情况下，<code>object@var.genes</code>中的基因用作输入，但可以使用pc.genes定义。 通常对高度可变基因运行降维可以提高性能。对于UMI数据，特别是在消除技术因素之后，PCA在更大的基因集（包括整个转录组）上运行时返回相似（尽管更慢）的结果。</p>
<p>–&gt; refered to Seurat v3 (latest): 高可变基因可通过函数HVFInfo(object)访问。尽管RunPCA有一个features参数，用于指定计算PCA的features，但测试修改该参数值，PCA结果相同，表明features参数中提供的基因并不完全是用于计算PCA。不知道该函数是直接获取HVG还是不考虑它们。</p>
<div class="sourceCode" id="cb625"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb625-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(<span class="dt">object =</span> pbmc,  <span class="dt">npcs =</span> <span class="dv">30</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p>–&gt; refered to Seurat v2: Seurat提供了几种可视化PCA中细胞和基因的方法，包括<code>PrintPCA</code>, <code>VizPCA</code>, <code>PCAPlot</code>, 和 <code>PCHeatmap</code></p>
<p>–&gt; Seurat v3 (latest):
Seurat v3 提供以下可视化的方法:
- PCA
- 定量特征着色的PCA图
- 单个细胞散点图
- 各基因的散点图
- 小提琴和ridge图
- Violin and Ridge plots
- 热图</p>
<div class="sourceCode" id="cb626"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb626-1" title="1"><span class="co"># Examine and visualize PCA results a few different ways</span></a>
<a class="sourceLine" id="cb626-2" title="2"><span class="kw">DimPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-12-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb627"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb627-1" title="1"><span class="co"># Dimensional reduction plot, with cells colored by a quantitative feature</span></a>
<a class="sourceLine" id="cb627-2" title="2"><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features =</span> <span class="st">&quot;MS4A1&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-13-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb628"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb628-1" title="1"><span class="co"># Scatter plot across single cells, replaces GenePlot</span></a>
<a class="sourceLine" id="cb628-2" title="2"><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> pbmc, <span class="dt">feature1 =</span> <span class="st">&quot;MS4A1&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;PC_1&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-14-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb629"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb629-1" title="1"><span class="kw">FeatureScatter</span>(<span class="dt">object =</span> pbmc, <span class="dt">feature1 =</span> <span class="st">&quot;MS4A1&quot;</span>, <span class="dt">feature2 =</span> <span class="st">&quot;CD3D&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-14-2.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb630"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb630-1" title="1"><span class="co"># Scatter plot across individual features, repleaces CellPlot</span></a>
<a class="sourceLine" id="cb630-2" title="2"><span class="kw">CellScatter</span>(<span class="dt">object =</span> pbmc, <span class="dt">cell1 =</span> <span class="st">&quot;AGTCTACTAGGGTG&quot;</span>, <span class="dt">cell2 =</span> <span class="st">&quot;CACAGATGGTTTCT&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-15-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb631"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb631-1" title="1"><span class="kw">VariableFeaturePlot</span>(<span class="dt">object =</span> pbmc)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-16-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb632"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb632-1" title="1"><span class="co"># Violin and Ridge plots</span></a>
<a class="sourceLine" id="cb632-2" title="2"><span class="kw">VlnPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;CCL5&quot;</span>, <span class="st">&quot;IL32&quot;</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-17-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb633"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb633-1" title="1"><span class="kw">RidgePlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">feature =</span> <span class="kw">c</span>(<span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;CCL5&quot;</span>, <span class="st">&quot;IL32&quot;</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-17-2.png" style="display: block; margin: auto;" /></p>
<p><code>DimHeatmap</code>允许探索数据集中异质性的主要来源，并且确定下游分析包含的PCs。细胞和基因根据其PCA分数排序。将cells.use设置为数字可以绘制两端的“extreme”细胞，从而大大加快了大型数据集的绘图速度。虽然一种监督分析，但是为探索相关基因集的有用工具。</p>
<div class="sourceCode" id="cb634"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb634-1" title="1"><span class="co"># Heatmaps</span></a>
<a class="sourceLine" id="cb634-2" title="2"><span class="kw">DimHeatmap</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">cells =</span> <span class="dv">200</span>, <span class="dt">balanced =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-18-1.png" style="display: block; margin: auto;" /></p>
<p>Seurat 3.0中不再支持ProjectPCA函数。</p>
</div>
<div id="Determine-statistically-significant-principal-components" class="section level2">
<h2><span class="header-section-number">9.7</span> 统计显著主成分</h2>
<p>为克服scRNA-seq数据广泛技术噪音，Seurat根据其PCA评分对细胞进行聚类，每个PC基本上代表相关基因集中信息组合的“metagene”。 因此，确定下游分析使用多少PC是重要的一步。</p>
<p><a href="http://www.cell.com/abstract/S0092-8674(15)00549-8">Macosko et al</a>受jackStraw程序启发实现重采样检验。 随机扰动数据的一个子集（默认为1％）并重新运行PCA，构建基因得分的“零分布”，并重复此过程。富集了低P值基因集的PC为显著PC。</p>
<div class="sourceCode" id="cb635"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb635-1" title="1"><span class="co"># </span><span class="al">NOTE</span><span class="co">: This process can take a long time for big datasets, comment out for</span></a>
<a class="sourceLine" id="cb635-2" title="2"><span class="co"># expediency.  More approximate techniques such as those implemented in</span></a>
<a class="sourceLine" id="cb635-3" title="3"><span class="co"># PCElbowPlot() can be used to reduce computation time</span></a>
<a class="sourceLine" id="cb635-4" title="4">pbmc &lt;-<span class="st"> </span><span class="kw">JackStraw</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims =</span> <span class="dv">20</span>, <span class="dt">num.replicate =</span> <span class="dv">100</span>,  <span class="dt">prop.freq =</span> <span class="fl">0.1</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><code>JackStrawPlot</code>函数提供了一个可视化工具，用于比较PC的p值分布和均匀分布（虚线）。显著PC将显示强烈富集低p值的基因（虚线上方的实线）。 在这种情况下，PC 1-10都很重要。</p>
<div class="sourceCode" id="cb636"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb636-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">ScoreJackStraw</span>(<span class="dt">object =</span> pbmc, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb637"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb637-1" title="1"><span class="kw">JackStrawPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-21-1.png" style="display: block; margin: auto;" /></p>
<p>确定使用哪些PC的更特别的方法是查看主成分的标准偏差图，并在图中拐点地方绘制cutoff。这可以通过<code>ElbowPlot</code>来完成。 在这个例子中，看起来拐点落在PC5周围。</p>
<div class="sourceCode" id="cb638"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb638-1" title="1"><span class="kw">ElbowPlot</span>(<span class="dt">object =</span> pbmc)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-22-1.png" style="display: block; margin: auto;" /></p>
<p>选择PC,识别数据集的真实维度，是Seurat的重要一步，但对用户来说可能具有挑战性/不确定性。因此，我们建议考虑这三种方法。第一个是更多的监督分析，探索PC以确定异质性的相关来源，并且可以与GSEA一起使用。第二个基于随机空模型的统计检验，但对于大型数据集来说耗时的，并且可能不会返回明确的PC cutoff。第三种是常用的启发式算法，可以立即计算。在这个例子中，所有三种方法产生了类似的结果，可以在PC 7-10之间选择cutoff。这里采用jackStraw结果，看到PCHeatmap在这些PC中返回可解释的信号（包括canonical dendritic cell markers）。虽然结果只是受到cutoff的微小变化的影响（可以在下面进行测试），但强烈建议您始终探索选择进行下游分析的PCs。</p>
</div>
<div id="Cluster-the-cells" class="section level2">
<h2><span class="header-section-number">9.8</span> 聚类</h2>
<p>Seurat现在包括基于图形的聚类方法，虽然聚类分析的距离度量（基于先前识别的PC）保持不变，但将细胞距离矩阵分成cluster的方法已经大大改进。我们的方法受到最近scRNA-seq数据基于图聚类方法 <a href="http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract">SNN-Cliq, Xu and Su, Bioinformatics, 2015</a>和CyTOF data <a href="http://www.ncbi.nlm.nih.gov/pubmed/26095251">PhenoGraph, Levine et al., Cell, 2015</a>的启发。简而言之，这些方法将细胞嵌入到图结构中 - 例如K-最近邻（KNN），相邻的细胞具有相似的基因表达模式，然后将该图划分为高度互连的“quasi-cliques”或“quasi-cliques”。与PhenoGraph一样，我们首先根据PCA空间中的欧氏距离构建KNN图，并根据其局部邻域中的重叠部分（Jaccard相似性）优化任意两个细胞之间的边权。为了聚类细胞，应用模块化优化技术，例如Louvain算法（默认）或SLM <a href="http://dx.doi.org/10.1088/1742-5468/2008/10/P10008">SLM, Blondel et al., Journal of Statistical Mechanics</a>以迭代方式将细胞分组在一起来优化标准模块化函数。</p>
<p><code>FindClusters</code>函数实现上述过程，并包含一个分辨率参数，用于设置下游聚类的“granularity”，值越大cluster越多。 将此参数设置在0.6-1.2之间对包含3k细胞的单细胞数据集返回不错的记过。较大的数据集通常会增加最佳分辨率。 最新的聚类结果将存储在<code>seurat_clusters</code>的元数据中。</p>
<p>首先计算k-最近邻并构造SNN图(<code>FindNeighbors</code>)，然后运行<code>FindClusters</code>。</p>
<div class="sourceCode" id="cb639"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb639-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(pbmc, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>)</a>
<a class="sourceLine" id="cb639-2" title="2">pbmc &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(pbmc, <span class="dt">resolution =</span> <span class="fl">0.5</span>, <span class="dt">algorithm =</span> <span class="dv">1</span>)</a></code></pre></div>
</div>
<div id="Run-Non-linear-dimensional-reduction-tSNE" class="section level2">
<h2><span class="header-section-number">9.9</span> 非线性降维</h2>
<p>Seurat继续使用tSNE作为可视化和探索这些数据集的强大工具。虽然不再建议直接在tSNE上进行聚类，上述基于图聚类cluster中的细胞在tSNE图中应该定位在一起。这是因为tSNE旨在将高维空间中具有相似局部邻域的细胞在低维空间放在一起。tSNE的输入，建议使用与聚类分析相同的PC，尽管genes.use参数也支持基于缩放基因表达计算tSNE。</p>
<div class="sourceCode" id="cb640"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb640-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(<span class="dt">object =</span> pbmc, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">do.fast =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb640-2" title="2"><span class="co"># note that you can set do.label=T to help label individual clusters</span></a>
<a class="sourceLine" id="cb640-3" title="3"><span class="kw">DimPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-24-1.png" style="display: block; margin: auto;" /></p>
</div>
<div id="umap" class="section level2">
<h2><span class="header-section-number">9.10</span> 运行UMAP</h2>
<p>为了并排显示这两个条件，可以使用split.by参数来显示由cluster着色的每个条件。
To visualize the two conditions side-by-side, we can use the split.by argument to show each condition colored by cluster.</p>
<div class="sourceCode" id="cb641"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb641-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(pbmc, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>)</a>
<a class="sourceLine" id="cb641-2" title="2"><span class="kw">DimPlot</span>(pbmc, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="dt">split.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-25-1.png" style="display: block; margin: auto;" /></p>
<p>您可以在此时保存对象，以便可以轻松地将其加载或者与协作者共享，而无需重新运行上面执行的计算密集型步骤。</p>
<div class="sourceCode" id="cb642"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb642-1" title="1"><span class="kw">saveRDS</span>(pbmc, <span class="dt">file =</span> <span class="st">&quot;data/pbmc_tutorial.rds&quot;</span>)</a></code></pre></div>
</div>
<div id="Finding-differentially-expressed-genes-cluster-biomarkers" class="section level2">
<h2><span class="header-section-number">9.11</span> 差异表达分析</h2>
<p>Seurat可以通过差异表达找到cluster的makers。默认情况下，通过与所有其他细胞进行比较，识别单个cluster的（在ident.1中指定）的正负marker。 <code>FindAllMarkers</code>为所有cluster自动执行此过程，但也可以检验cluster之间或所有细胞间。</p>
<p>min.pct参数要求在两组细胞中的任一组中最小检测到基因的百分比，并且thresh.test参数要求基因在两组之间以一定量差异表达（平均）。 可以将这两者设置为0，但随着时间的显著增加 - 因为这将检验大量不太可能具有高度差异基因。 加速这些计算的另一种选择，可以设置<code>max.cells.per.ident</code>。这将对每个类进行下采样，使其不多于设置的细胞。 虽然这会损失效力，但显著增加速度，并且最高度差异表达的基因可能仍会上升到顶部。</p>
<div class="sourceCode" id="cb643"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb643-1" title="1"><span class="co"># find all markers of cluster 1</span></a>
<a class="sourceLine" id="cb643-2" title="2">cluster1.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">ident.1 =</span> <span class="dv">1</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>)</a>
<a class="sourceLine" id="cb643-3" title="3"><span class="kw">print</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> cluster1.markers, <span class="dt">n =</span> <span class="dv">5</span>))</a></code></pre></div>
<div class="sourceCode" id="cb644"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb644-1" title="1"><span class="co"># find all markers distinguishing cluster 5 from clusters 0 and 3</span></a>
<a class="sourceLine" id="cb644-2" title="2">cluster5.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">ident.1 =</span> <span class="dv">2</span>, <span class="dt">ident.2 =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">3</span>), <span class="dt">min.pct =</span> <span class="fl">0.25</span>)</a>
<a class="sourceLine" id="cb644-3" title="3"><span class="kw">print</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> cluster5.markers, <span class="dt">n =</span> <span class="dv">5</span>))</a></code></pre></div>
<div class="sourceCode" id="cb645"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb645-1" title="1"><span class="co"># find markers for every cluster compared to all remaining cells, report</span></a>
<a class="sourceLine" id="cb645-2" title="2"><span class="co"># only the positive ones</span></a>
<a class="sourceLine" id="cb645-3" title="3">pbmc.markers &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">only.pos =</span> <span class="ot">TRUE</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, <span class="dt">thresh.use =</span> <span class="fl">0.25</span>)</a>
<a class="sourceLine" id="cb645-4" title="4">pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">2</span>, avg_logFC)</a></code></pre></div>
<p>Seurat有多种差异表达检验，可以使用test.use参数设置（详细信息见 <a href="http://satijalab.org/seurat/de_vignette.html">DE vignette</a>）。 例如，ROC检验返回marker的“classification power”（范围从0 - random, to 1 - perfect）。</p>
<div class="sourceCode" id="cb646"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb646-1" title="1">cluster1.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">thresh.use =</span> <span class="fl">0.25</span>, <span class="dt">test.use =</span> <span class="st">&quot;roc&quot;</span>, <span class="dt">only.pos =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>Seurat提供了几种用于可视化marker表达的工具。
* <code>VlnPlot</code> (显示cluster见的表达概率分布),
* <code>FeaturePlot</code> (tSNE或PCA可视化基因表达，也是最常用的),
建议尝试以下可视化方法：
* <code>RidgePlot</code>,
* <code>CellPlot</code>
* <code>DotPlot</code></p>
<div class="sourceCode" id="cb647"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb647-1" title="1"><span class="kw">VlnPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features =</span><span class="kw">c</span>(<span class="st">&quot;NKG7&quot;</span>, <span class="st">&quot;PF4&quot;</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-31-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb648"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb648-1" title="1"><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;GNLY&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;FCER1A&quot;</span>, <span class="st">&quot;FCGR3A&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;PPBP&quot;</span>, <span class="st">&quot;CD8A&quot;</span>), <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;blue&quot;</span>), <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-32-1.png" style="display: block; margin: auto;" />
<code>DoHeatmap</code>为给定的细胞和基因生成表达热图。在这种情况下，为每个cluster绘制前20个markers（如果小于20，绘制所有marker）。</p>
<div class="sourceCode" id="cb649"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb649-1" title="1">top10 &lt;-<span class="st"> </span>pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">10</span>, avg_logFC)</a>
<a class="sourceLine" id="cb649-2" title="2"><span class="co"># setting slim.col.label to TRUE will print just the cluster IDS instead of</span></a>
<a class="sourceLine" id="cb649-3" title="3"><span class="co"># every cell name</span></a>
<a class="sourceLine" id="cb649-4" title="4"><span class="kw">DoHeatmap</span>(<span class="dt">object =</span> pbmc, <span class="dt">features =</span> top10<span class="op">$</span>gene, <span class="dt">label =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-33-1.png" style="display: block; margin: auto;" /></p>
</div>
<div id="Assigning-cell-type-identity-to-clusters" class="section level2">
<h2><span class="header-section-number">9.12</span> 确定cluster细胞类型</h2>
<p>对于该数据集，使用经典marker确定cluster细胞类型。
Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types.</p>
<div class="sourceCode" id="cb650"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb650-1" title="1">current.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb650-2" title="2">new.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CD4 T cells&quot;</span>, <span class="st">&quot;CD14+ Monocytes&quot;</span>, <span class="st">&quot;B cells&quot;</span>, <span class="st">&quot;CD8 T cells&quot;</span>, <span class="st">&quot;FCGR3A+ Monocytes&quot;</span>, <span class="st">&quot;NK cells&quot;</span>, <span class="st">&quot;Dendritic cells&quot;</span>, <span class="st">&quot;Megakaryocytes&quot;</span>)</a>
<a class="sourceLine" id="cb650-3" title="3">pbmc<span class="op">@</span>active.ident &lt;-<span class="st"> </span>plyr<span class="op">::</span><span class="kw">mapvalues</span>(<span class="dt">x =</span> pbmc<span class="op">@</span>active.ident, <span class="dt">from =</span> current.cluster.ids, <span class="dt">to =</span> new.cluster.ids)</a>
<a class="sourceLine" id="cb650-4" title="4"><span class="kw">DimPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">do.label =</span> <span class="ot">TRUE</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-34-1.png" style="display: block; margin: auto;" /></p>
</div>
<div id="Further-subdivisions-within-cell-types" class="section level2">
<h2><span class="header-section-number">9.13</span> 细分细胞类型</h2>
<p>如果扰乱我们上面的一些参数选择（例如，设置<code>resolution=0.8</code>或更改PC的数量），可能会看到CD4 T细胞细分为两个群体。 您可以探索此细分以查找区分两个T细胞marker基因。但是，在重新聚类(覆盖<code>object@ident</code>)之前，可以存储我们重命名的identities，以便以后恢复。</p>
<div class="sourceCode" id="cb651"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb651-1" title="1"><span class="co"># First lets stash our identities for later</span></a>
<a class="sourceLine" id="cb651-2" title="2">pbmc &lt;-<span class="st"> </span><span class="kw">StashIdent</span>(<span class="dt">object =</span> pbmc, <span class="dt">save.name =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>)</a>
<a class="sourceLine" id="cb651-3" title="3"></a>
<a class="sourceLine" id="cb651-4" title="4"><span class="co"># Note that if you set save.snn=T above, you don&#39;t need to recalculate the</span></a>
<a class="sourceLine" id="cb651-5" title="5"><span class="co"># SNN, and can simply put: pbmc &lt;- FindClusters(pbmc,resolution = 0.8)</span></a>
<a class="sourceLine" id="cb651-6" title="6">pbmc &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">resolution =</span> <span class="fl">0.8</span>, <span class="dt">print.output =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<div class="sourceCode" id="cb652"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb652-1" title="1"><span class="co"># Demonstration of how to plot two tSNE plots side by side, and how to color</span></a>
<a class="sourceLine" id="cb652-2" title="2"><span class="co"># points based on different criteria</span></a>
<a class="sourceLine" id="cb652-3" title="3">plot1 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">do.return =</span> <span class="ot">TRUE</span>, <span class="dt">no.legend =</span> <span class="ot">TRUE</span>, <span class="dt">do.label =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb652-4" title="4">plot2 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">reduction =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">do.return =</span> <span class="ot">TRUE</span>, <span class="dt">group.by =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>, <span class="dt">no.legend =</span> <span class="ot">TRUE</span>, <span class="dt">do.label =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb652-5" title="5"><span class="kw">plot_grid</span>(plot1, plot2)</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-36-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb653"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb653-1" title="1"><span class="co"># Find discriminating markers</span></a>
<a class="sourceLine" id="cb653-2" title="2">tcell.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb653-3" title="3"></a>
<a class="sourceLine" id="cb653-4" title="4"><span class="co"># Most of the markers tend to be expressed in C1 (i.e. S100A4). However, we</span></a>
<a class="sourceLine" id="cb653-5" title="5"><span class="co"># can see that CCR7 is upregulated in C0, strongly indicating that we can</span></a>
<a class="sourceLine" id="cb653-6" title="6"><span class="co"># differentiate memory from naive CD4 cells.  cols.use demarcates the color</span></a>
<a class="sourceLine" id="cb653-7" title="7"><span class="co"># palette from low to high expression</span></a>
<a class="sourceLine" id="cb653-8" title="8"><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;S100A4&quot;</span>, <span class="st">&quot;CCR7&quot;</span>), <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;blue&quot;</span>))</a></code></pre></div>
<p><img src="https://scrnaseq-course.cog.sanger.ac.uk/website/seurat_files/figure-html/unnamed-chunk-37-1.png" style="display: block; margin: auto;" /></p>
<p>memory/naive划分有点弱，可以查看更多细胞看看这是否更有说服力。与此同时，我们可以恢复原来cluster identities以进行下游处理。</p>
<div class="sourceCode" id="cb654"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb654-1" title="1">pbmc &lt;-<span class="st"> </span><span class="kw">SetIdent</span>(<span class="dt">object =</span> pbmc, <span class="dt">value =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>)</a>
<a class="sourceLine" id="cb654-2" title="2"><span class="kw">saveRDS</span>(pbmc, <span class="dt">file =</span> <span class="st">&quot;data/pbmc3k_final.rds&quot;</span>)</a></code></pre></div>
</div>
<div id="sessioninfo-1" class="section level2">
<h2><span class="header-section-number">9.14</span> sessionInfo()</h2>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 7 x64 (build 7601) Service Pack 1
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=Chinese (Simplified)_People&#39;s Republic of China.936 
## [2] LC_CTYPE=Chinese (Simplified)_People&#39;s Republic of China.936   
## [3] LC_MONETARY=Chinese (Simplified)_People&#39;s Republic of China.936
## [4] LC_NUMERIC=C                                                   
## [5] LC_TIME=Chinese (Simplified)_People&#39;s Republic of China.936    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] cowplot_1.0.0 dplyr_0.8.3   Seurat_3.0.2  knitr_1.23   
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.4.0          tidyr_0.8.3         viridisLite_0.3.0  
##  [4] jsonlite_1.6        splines_3.6.0       lsei_1.2-0         
##  [7] R.utils_2.9.0       gtools_3.8.1        Rdpack_0.11-0      
## [10] assertthat_0.2.1    yaml_2.2.0          ggrepel_0.8.1      
## [13] globals_0.12.4      pillar_1.4.2        lattice_0.20-38    
## [16] reticulate_1.12     glue_1.3.1          digest_0.6.20      
## [19] RColorBrewer_1.1-2  SDMTools_1.1-221.1  colorspace_1.4-1   
## [22] htmltools_0.3.6     Matrix_1.2-17       R.oo_1.22.0        
## [25] plyr_1.8.4          pkgconfig_2.0.2     bibtex_0.4.2       
## [28] tsne_0.1-3          listenv_0.7.0       bookdown_0.12      
## [31] purrr_0.3.2         scales_1.0.0        RANN_2.6.1         
## [34] gdata_2.18.0        Rtsne_0.15          tibble_2.1.3       
## [37] ggplot2_3.2.0       ROCR_1.0-7          pbapply_1.4-0      
## [40] lazyeval_0.2.2      survival_2.44-1.1   magrittr_1.5       
## [43] crayon_1.3.4        evaluate_0.14       R.methodsS3_1.7.1  
## [46] future_1.14.0       nlme_3.1-140        MASS_7.3-51.4      
## [49] gplots_3.0.1.1      ica_1.0-2           tools_3.6.0        
## [52] fitdistrplus_1.0-14 data.table_1.12.2   gbRd_0.4-11        
## [55] stringr_1.4.0       plotly_4.9.0        munsell_0.5.0      
## [58] cluster_2.1.0       irlba_2.3.3         compiler_3.6.0     
## [61] rsvd_1.0.1          caTools_1.17.1.2    rlang_0.4.0        
## [64] grid_3.6.0          ggridges_0.5.1      htmlwidgets_1.3    
## [67] igraph_1.2.4.1      bitops_1.0-6        rmarkdown_1.14     
## [70] npsurv_0.4-0        gtable_0.3.0        codetools_0.2-16   
## [73] reshape2_1.4.3      R6_2.4.0            gridExtra_2.3      
## [76] zoo_1.8-6           future.apply_1.3.0  KernSmooth_2.23-15 
## [79] metap_1.1           ape_5.3             stringi_1.4.3      
## [82] parallel_3.6.0      Rcpp_1.0.1          sctransform_0.2.0  
## [85] png_0.1-7           tidyselect_0.2.5    xfun_0.8           
## [88] lmtest_0.9-37</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="biological-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="Ideal-scRNA-seq-pipeline.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": true,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/lemonbases/scRNA.seq.course.CN/edit/master/09seurat.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
